---
title: "RNAseq analysis on the Cryostem cohort"
author: "Helena"
date: "09/10/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r imports, echo=FALSE, results='hide', warning = FALSE}
suppressPackageStartupMessages({
  library(dplyr)
  library(openxlsx)
 # library(metabolomics)
  #library(BioGVHD)
  library(tidyverse)
  library(data.table)
  library(dendextend)
  library(ggplot2)
  library(mlbench)
  library(caret)
  library(nnet)
  library(pROC)
  library(igraph)
  #library(edgeR)
  #library(triwise)
  library(htmlwidgets)
  #library(biomaRt)
  #library(org.Hs.eg.db) 
  #library(clusterProfiler)
  #library(VennDiagram)
  #library(qsub)
  #library(kableExtra)
})
options("scipen"=100)
```


```{r, echo = FALSE}
## Liesbet's functions
###Get DE genes
getDEgenes<-function(expMatrix, pValCutOff, logFCcutOff){
  topgenes<-expMatrix[expMatrix$adj.P.Val<pValCutOff,]
  genes_up<-topgenes[topgenes$logFC>logFCcutOff,]
  genes_down<-topgenes[topgenes$logFC< -logFCcutOff,]
  ##Sort genes on logFC
  genes_up<-genes_up[order(genes_up$logFC, decreasing=TRUE),]
  genes_down<-genes_down[order(genes_down$logFC, decreasing=TRUE),]
  genes_de_sorted<-rbind(genes_up, genes_down)

  return(genes_de_sorted)
}

###Normalize per gene
normalizePerGene<-function(expMatrix){
  resultMatrix<-t(apply(expMatrix, 1, function(x)(x-min(x))/(max(x)-min(x))))

  return(resultMatrix)
}

## convert to Entrez Id
convertIDs <- function( ids, fromKey, toKey, db, ifMultiple=c( "putNA", "useFirst" ) ) {
stopifnot( inherits( db, "AnnotationDb" ) )
ifMultiple <- match.arg( ifMultiple )
suppressWarnings( selRes <- AnnotationDbi::select( 
db, keys=ids, keytype=fromKey, columns=c(fromKey,toKey) ) )
if( ifMultiple == "putNA" ) {
duplicatedIds <- selRes[ duplicated( selRes[,1] ), 1 ] 
selRes <- selRes[ ! selRes[,1] %in% duplicatedIds, ] }
return( selRes[ match( ids, selRes[,1] ), 2 ] )
}
```

## RNA seq analysis of the Cryostem cohort

I first read in the raw data:

```{r read_data}
countData<-read.xlsx("~/Documents/VIB/Projects/Integrative_Paris/National_cohort/RNAseq/Global Table Count Cryostem 01.2019.xlsx")
colnames(countData) <- as.character(countData[1,]) # set patient names as colnames
rownames(countData) <- countData$ID # set gene names as rownames
countData_withinfo <- countData[,-1]
```

I can then look at the extra information on the patients' unmapped genes, lib size, ...

```{r, echo = FALSE}
## Look at the info on samples:
plot(as.numeric(countData_withinfo[2,]), ylab = "unmapped genes", xlab = "patients")
plot(as.numeric(countData_withinfo[3,]), ylab = rownames(countData)[3], xlab = "patients")
above <- which(as.numeric(countData_withinfo[3,]) > 7*10^6)
colnames(countData_withinfo)[above]
plot(as.numeric(countData_withinfo[4,]), ylab = rownames(countData)[4], xlab = "patients")
plot(as.numeric(countData_withinfo[5,]), ylab = rownames(countData)[5], xlab = "patients")

countData <- countData[6:nrow(countData), 2:ncol(countData)] # rm the informations on unmapped, ambiguous...
dim(countData)
```

I then load in the clinical data, and reorder the RNAseq and clinical data tables so that they are in the same order:

```{r, echo = FALSE}
colData<-read.xlsx("~/Documents/VIB/Projects/Integrative_Paris/National_cohort/Data synthesis national cohort Cryostem 15042019_FINAL_filtered.xlsx")
colData <- colData[!is.na(colData$RNAseq.ID),] # remove rows with no RNAseq sample
rownames(colData) <- colData$X3
colData <- colData[which(rownames(colData) %in% colnames(countData)),] # remove rows for which I don't have the RNA seq data

### Reorder
countData<-countData[,rownames(colData)]
```

I then identify the donors and recipients, for which we have complete cases:
```{r, echo = FALSE}
complete_couples <- colData$COUPLENUMBER[which(duplicated(colData$COUPLENUMBER))]

print(paste0("We have ", length(complete_couples), " complete couples in the RNA seq data of the Cryostem cohort."))
complete_names <- rownames(colData)[which(colData$COUPLENUMBER %in% complete_couples)]
donor_names <- complete_names[grep("D", complete_names)]
donor_id <- which(colnames(countData) %in% donor_names)

recip_names <- complete_names[grep("R", complete_names)]
recip_id <- which(colnames(countData) %in% recip_names)
```

Finally, I generate the age and gender vectors that will be used for modeling in 
all patients (donors, recipients and couples):
```{r, echo = FALSE}
couple_info <- colData %>% 
  dplyr::filter(COUPLENUMBER %in% complete_couples) %>% 
  mutate(Id.Cryostem.R = X3) %>% 
  arrange(COUPLENUMBER) %>%
  dplyr::select(c("COUPLENUMBER", "Gender", "DOG", "DOB", "GROUP",
           "RIN", "CMVStatus", "Id.Cryostem.R")) %>% 
  mutate(DOB = as.Date(DOB, format = "%d.%m.%Y"),
         DOG = as.Date(DOG, format = "%d.%m.%Y")) 

ifelse(grepl("R", rownames(couple_info)[1]), 
       gender_donors <- couple_info$Gender[c(FALSE,TRUE)],
       gender_donors <- couple_info$Gender[c(TRUE,FALSE)])

ifelse(grepl("R", rownames(couple_info)[1]), 
       gender_recip <- couple_info$Gender[c(TRUE,FALSE)],
       gender_recip <- couple_info$Gender[c(FALSE,TRUE)])

gender_comp <- paste0(gender_donors, gender_recip)

ifelse(grepl("R", rownames(couple_info)[1]), 
       age_recip <- couple_info$DOG[c(TRUE, FALSE)] - couple_info$DOB[c(TRUE, FALSE)],
       age_recip <- couple_info$DOG[c(FALSE, TRUE)] - couple_info$DOB[c(FALSE, TRUE)])

couple_info <- couple_info %>% 
  rownames_to_column("rownames") %>% 
  mutate(gender_comp = rep(gender_comp, each = 2),
         age_recip = rep(age_recip, each = 2)) %>% 
  column_to_rownames("rownames")
head(couple_info)

save(couple_info, file = "../data/RNA_seq_national/couple_info.RData")
```

# Analysis on the recipients

### Preprocessing :

I will first focus on the recipients only, and perform a standard preprocessing 
(keep genes with more than one count per million in at least three patients):

```{r preproc_recip, echo=FALSE, eval = FALSE}
counts <- countData[,c(recip_id)]

count_nb <- apply(counts,2,as.numeric) # turn characters into numerics
rownames(count_nb) <- rownames(counts) # restore the rownmes which were lost in the transfo
counts <- count_nb

### Preprocessing
y <- DGEList(counts = counts)
keep = rowSums(cpm(y)>1) >= 3
y = y[keep,]
dim(y) # from 69000 genes to 18000 genes after this filter

# Reset lib sizes
y$samples$lib.size = colSums(y$counts)
```

The names of the genes are in the ensembl ID format. When possible, I replace them with the associated gene symbol:
```{r, eval = FALSE, echo=FALSE}
##CONVERT ENSEMBL GENE NAME TO ENTREZID & SYMBOL
hgnc_symbol <- convertIDs( rownames(y$counts), "ENSEMBL", "SYMBOL", org.Hs.eg.db )

counts_gene_symbol <- y$counts
names_2change <- hgnc_symbol[-which(is.na(hgnc_symbol))]
names_2keep_idx <- which(is.na(hgnc_symbol))
dupp <- names_2change[which(duplicated(names_2change))]
dupp_idx <- which(hgnc_symbol %in% dupp)[c(T,F)]
final_ids_2keep <- c(names_2keep_idx, dupp_idx)

rownames(counts_gene_symbol)[-final_ids_2keep] <- hgnc_symbol[-final_ids_2keep]

y$counts <- counts_gene_symbol
```

Perform normalisation and initiate the design matrix (per group)

```{r, eval = FALSE, echo=FALSE}
# compute norm_factors
normfac <- calcNormFactors(y, method = "TMM")

#### Create design matrix
# for the three groups :
patient_groups <- tolower(colData$GROUP)
TS <- patient_groups[c(recip_id)]

TS <- factor(TS, levels=unique(TS))
design <- model.matrix(~0+TS)
colnames(design) <- levels(TS)
design

#voom
#In limma, read counts are converted to log2-counts-per-million (logCPM) and
#the mean-variance relationship is modelled either with precision weights or
#with an empirical Bayes prior trend. The precision weights approach is called “voom”
# lib.size <- counts$samples$lib.size * counts$samples$norm.factors
# E <- t(log2(t(counts + 0.5)/(lib.size + 1) * 1e+06))
v <- voom(y, design, plot = TRUE)
expTable<-v$E
```

Save the produced objects:

```{r, eval = FALSE, echo=FALSE}
DE_list <- list(v, y, design)

save(expTable, file = "~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/recip_only/expression_table_recip.RData")
save(DE_list, file = "~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/recip_only/data_for_DE_recip.RData")
```

Load the produced data :
```{r}
load("~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/recip_only/expression_table_recip.RData")
load("~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/recip_only/data_for_DE_recip.RData")
```

### Differential Analysis 

Fit a model to the design:
```{r, echo=FALSE}
names(DE_list) <- c("v", "y", "design")
fit <- lmFit(DE_list$v, DE_list$design)

#### Create contrast matrix
cont.matrix <- makeContrasts(group1=non_tolerant-primary_tolerant,
                             group2=non_tolerant-secondary_tolerant,
                             group3=primary_tolerant-secondary_tolerant,
                             levels=DE_list$design)
cont.matrix
fit2 = contrasts.fit(fit, cont.matrix)

#### Moderate T-test with 'shrunken' se
fit.eb <- eBayes(fit2)
```

After adjusting the pvalues for multiple testing, we can see how many genes were
identified as differentially expressed between the different tolerance groups:
```{r}
#### Adjust P-values via Benjamini-Hochberg
##### 1. non-tol vs tol-1
allGenesGroup1<-topTable(fit.eb, adjust="BH", sort.by="P",number=Inf, coef=1)
DEgenesGroup1<-getDEgenes(allGenesGroup1,0.05,1)
paste0("Non tolerant vs primary: ", dim(DEgenesGroup1)[1])
##84

##### 2. non-tol vs tol-2
allGenesGroup2<-topTable(fit.eb, adjust="BH", sort.by="P",number=Inf, coef=2)
DEgenesGroup2<-getDEgenes(allGenesGroup2,0.05,1)
paste0("Non tolerant vs secondary: ", dim(DEgenesGroup2)[1])
##57

##### 3. tol-1 vs tol-2
allGenesGroup3<-topTable(fit.eb, adjust="BH", sort.by="P",number=Inf, coef=3)
DEgenesGroup3<-getDEgenes(allGenesGroup3,0.05,1)
paste0("Primary vs secondary: ", dim(DEgenesGroup3)[1])
##0
```

No genes were identified as differentially expressed between the different groups,
in the Cryostem cohort.

### Machine learning approach :

filter out lowly variable genes:

```{r, echo = FALSE}
sds <- apply(expTable,1,sd)
plot(sort(sds), type = "l")
abline(h=quantile(sds, 0.5), col="red")
expT <- expTable[which(sds>=quantile(sds, 0.5)),]
rnames <- rownames(expT)
expT <- apply(expT,2,as.numeric)
rownames(expT) <- rnames
```

Prepare the data:
```{r, cache=FALSE, results='hide', warning=FALSE, comment=FALSE, warning=FALSE, message = FALSE, include=FALSE, echo = FALSE}
patient_groups <- tolower(colData$GROUP)
group_tmp <- patient_groups[c(recip_id)]

norm_data <- as.data.frame(t(expT)) %>% 
  rownames_to_column("rn") %>% 
  mutate(group = as.factor(group_tmp)) %>% 
  dplyr::select("rn", "group", rownames(expT)) %>% 
  column_to_rownames("rn")

colnames(norm_data) <- make.names(colnames(norm_data))
norm_data_3gr <- norm_data
```

We perform a PCA on the recipients to see if we already observe some structure 
in the data, and if there isn't too much batch effect:
```{r, echo = FALSE}
colData_recip <- colData[c(recip_id),] %>% 
  mutate(GROUP = tolower(GROUP)) %>% 
  mutate(Id.Cryostem.R = X3)

pca <- prcomp(norm_data %>% dplyr::select(-group))

pca_2plot <- as.data.frame(pca$x) %>% 
  mutate(Id.Cryostem.R = rownames(pca$x)) %>% 
  left_join(colData_recip, by = "Id.Cryostem.R")
ggplot(pca_2plot, aes(x = PC1, y = PC2, col = as.factor(GROUP), label = Id.Cryostem.R)) + 
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0)

ggplot(pca_2plot, aes(x = PC1, y = PC2, col = as.factor(RNAseq.RUN), label = Id.Cryostem.R)) + 
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0)

ggplot(pca_2plot, aes(x = PC1, y = PC2, col = as.numeric(RIN), label = Id.Cryostem.R)) + 
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0)

ggplot(pca_2plot, aes(x = PC1, y = PC2, col = as.factor(GROUP), label = Id.Cryostem.R)) + 
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  theme(axis.ticks=element_blank(),
        axis.text = element_blank(),
        axis.title = element_blank(),
          text = element_text(size=25)) +
  ggtitle("Tolerance group of the Cryostem recipients") +
  theme(
      panel.background = element_rect(fill = "white", colour = "white",
                                      size = 2, linetype = "solid"),
      panel.grid.major = element_line(size = 0.5, linetype = 'solid',
                                      colour = "#99999920")
      # panel.grid.minor = element_line(size = 0.25, linetype = 'solid',
      #                                 colour = "grey")
    ) +
  labs(col="GROUP") 

```
We don't observe much structure, relatively to the tolerance groups, or batches.

We can try to build a random forest model using all the genes of the 
recipients:
```{r, echo = FALSE}
set.seed(1)
rf_recip <- ranger::ranger(group~., norm_data, num.trees = 5000,
                           importance = "impurity")
rf_recip$confusion.matrix
rf_recip$prediction.error*100
```
But it seems like the data doesn't allow us to classify into the three tolerance
groups. We can try to simplify the classification problem:

First, we try to separate tolerant from non tolerant recipients:
```{r, eval = FALSE}
set.seed(1)
rf <- rf_tol_non_tol(df_orig = norm_data, ntree = 5000)
rf
```

Second, we can try to separate primary tolerant from secondary tolerant recipients:
```{r, eval = FALSE}
set.seed(1)
rf <- rf_tol1_tol2(df_orig = norm_data, ntree = 5000)
rf
```
But it seems like we can't classify between these two groups of patients yet.

### Feature selection

### Identifying features that vary between non and tolerant recipients:

Run the permutation distribution, taking demographic variables into account:
I first add the age and gender compatibility information to the data.
```{r, echo = FALSE}
recip_info <- couple_info %>% 
  dplyr::filter(grepl("R", rownames(couple_info))) 

norm_data <- norm_data %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  left_join(recip_info %>% dplyr::select(Id.Cryostem.R, gender_comp, age_recip),
            by = "Id.Cryostem.R") %>% 
  column_to_rownames("Id.Cryostem.R")
```

And then compute the permutation distribution for the non vs tolerance:
```{r, eval=FALSE, echo = FALSE}
norm_data$group <- as.character(norm_data$group)
norm_data$group[which(norm_data$group!="non_tolerant")] <- "tolerant"
norm_data$group <- as.factor(norm_data$group)

perm_vals_PRISM(norm_data = norm_data, 
                PRISM_name = "rnaseq_r_TNT",
                rds_name = "handle_cryo_rna_r_TNT.rds")

handle3 <- readRDS("handle_cryo_rna_r_TNT.rds")
perm_vals <- qsub_retrieve(handle3)

save(perm_vals, file = "~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/recip_only/perm_vals_TNT.RData")
save(norm_data, file= "~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/recip_only/norm_data_TNT.RData")
```

Compute the median per group for every gene, to use it later in the graphs:
```{r}
load("~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/recip_only/perm_vals_TNT.RData")
load("~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/recip_only/norm_data_TNT.RData")

gr_medians <- norm_data %>%
  group_by(group) %>%
  summarize_if(is.numeric, median) %>%
  as.data.frame()
rownames(gr_medians) <- gr_medians$group
```

#### Threshold of 0.95

These genes were selected as informative to classify the non and tolerant 
recipients, with a 0.95 threshold:
```{r, echo = FALSE}
qt <- unlist(map(perm_vals, 1))
densityplot(qt)
selected_genes_qt <- colnames(norm_data)[which(qt>.95)+1]
write.xlsx(selected_genes_qt, file = "../data/RNA_seq_national/recip_only/perm_val_TNT_95_thresh.xlsx")
print(paste0(length(selected_genes_qt),
             " genes were selected with a 0.95 threshold in the Cryostem cohort."))
```

Which of these selected genes are common with the Cryostem cohort?
```{r, echo = FALSE}
sel_louis <- read.xlsx("../data/RNAseq/recip_only/perm_val_TNT_95_thresh.xlsx", colNames = FALSE)
sel_common <- which(selected_genes_qt %in% sel_louis$X1)
print(paste0(length(sel_common), " genes were common between the two cohorts."))
#selected_genes_qt[sel_common]
```

Graph of the 140 selected genes that were found commonly between the two cohorts:
```{r, echo = FALSE}
correlations <- norm_data[,selected_genes_qt[sel_common]] %>% 
  as.matrix %>%
  cor %>%
  as.data.frame %>%
  rownames_to_column(var = 'var1') %>%
  gather(var2, value, -var1) %>% 
  dplyr::filter(var1 < var2)

correlated_features <- correlations %>% 
  dplyr::filter(value > 0.7)
gr <- graph_from_data_frame(correlated_features, directed = FALSE, vertices=selected_genes_qt[sel_common])
nodes <- get.vertex.attribute(gr)
color_nodes <- rep("#FF000080", length(nodes$name))

medians <- gr_medians[,nodes$name]
for (gene in seq_along(colnames(medians))){
  if (medians["tolerant", gene] == max(medians[, gene])){
    color_nodes[gene] <- "#0000FF80"
  }
}

# recip_path <- "~/Desktop/"
# 
# pdf(paste0(recip_path, "cryo_common_metabolites_recip_05.pdf"), width = 12, height = 8)
set.seed(1)
plot(gr, vertex.size = 7, vertex.label.cex = .5, edge.width = (E(gr)$value - .7) * 20,
     vertex.color = color_nodes, layout = layout_nicely)
# dev.off()

```

I save these features:
```{r, echo = FALSE}
rna_recip_cryo_TNT_95 <- norm_data[,c("group", selected_genes_qt[sel_common])]
save(rna_recip_cryo_TNT_95,
     file = "../data/RNA_seq_national/recip_only/rna_recip_cryo_TNT_95.RData")

# Identify connected components in the graph:
gr_components <- decompose(gr, mode = "strong", max.comps = NA,
                 min.vertices = 0)
comp_sizes <- sapply(gr_components, diameter)

# Generate a new annotated excel table
annot_genes <- as.data.frame(selected_genes_qt[sel_common])

# Compute the FC and pvalues:
patient_groups <- tolower(colData$GROUP)
patient_groups[which(patient_groups != "non_tolerant")] <- "tolerant"
TS <- patient_groups[c(recip_id)]

TS <- factor(TS, levels=unique(TS))
design <- model.matrix(~0+TS)
colnames(design) <- levels(TS)
#design
fit <- lmFit(t(norm_data[,selected_genes_qt[sel_common]]), design)
  
#### Create contrast matrix
cont.matrix <- makeContrasts(group1=non_tolerant-tolerant,
                             group2=tolerant-non_tolerant,
                             levels=design)
#cont.matrix
fit2 = contrasts.fit(fit, cont.matrix)

#### Moderate T-test with 'shrunken' se
fit.eb <- eBayes(fit2)


allGenesGroup1<-topTable(fit.eb, adjust="BH", sort.by="P",number=Inf, coef=1)
DEgenesGroup1<-getDEgenes(allGenesGroup1,1,0)

# length(which(selected_genes_qt[sel_common] %in% rownames(DEgenesGroup1)))

# add them to the table

DEgenesGroup1 <- DEgenesGroup1[as.character(annot_genes$`selected_genes_qt[sel_common]`),] 
annot_genes <- cbind(annot_genes, DEgenesGroup1)


for( i in which(comp_sizes != 0)){
  vertex_names <- vertex_attr(gr_components[[i]], index = V(gr_components[[i]]))[[1]]
  v_names <- c(vertex_names, rep(0, nrow(annot_genes) - length(vertex_names)))
  annot_genes <- cbind(annot_genes, v_names)
}
write.xlsx(annot_genes, file = "../data/RNA_seq_national/recip_only/annot_140selected_genes_recip_TNT_95.xlsx")
```


We can see how many of these genes had the same behaviour in the two cohorts (ie, 
were over or underexpressed in the same group of patients)
```{r, echo = FALSE}
dir <- lapply(seq_len(length(sel_common)), function(idx){
  gene <- selected_genes_qt[sel_common[idx]]
  overexp <- rownames(gr_medians)[which.is.max(gr_medians[gene][,1])]
  list(gene, overexp)
})
dir_genes <- do.call(rbind, dir)
write.xlsx(dir_genes, file = "../data/RNA_seq_national/recip_only/perm_val_dir_TNT_95_thresh.xlsx")
dir_genes_cryo <- dir_genes

dir_genes_louis <- read.xlsx("../data/RNAseq/recip_only/perm_val_dir_TNT_95_thresh.xlsx", rowNames = FALSE)
dir_genes_cryo <- as.data.frame(dir_genes_cryo) %>% 
  column_to_rownames("V1")
dir_genes_cryo <- dir_genes_cryo[dir_genes_louis$V1,]
sel_beh <- dir_genes_louis$V1[which(dir_genes_louis$V2 == dir_genes_cryo)]
sel_beh <- as.data.frame(sel_beh)
write.xlsx(sel_beh, file = "../data/RNAseq/recip_only/perm_val_beh_TNT_95_thresh.xlsx")

print(paste0(length(which(dir_genes_louis$V2 == dir_genes_cryo)),
             " genes out of the ", length(sel_common),
             " common selected genes have the same behaviour in both cohorts."))
```


Graph of the 111 genes that were found commonly in the two cohorts AND had the
same behaviour in the two cohorts:
```{r}
selb <- as.character(sel_beh$sel_beh)
correlations <- norm_data[,selb] %>% 
  as.matrix %>%
  cor %>%
  as.data.frame %>%
  rownames_to_column(var = 'var1') %>%
  gather(var2, value, -var1) %>% 
  dplyr::filter(var1 < var2)

correlated_features <- correlations %>% 
  dplyr::filter(value > 0.7)
gr <- graph_from_data_frame(correlated_features, directed = FALSE, vertices=selb)
nodes <- get.vertex.attribute(gr)
color_nodes <- rep("#FF000080", length(nodes$name))

medians <- gr_medians[,nodes$name]
for (gene in seq_along(colnames(medians))){
  if (medians["tolerant", gene] == max(medians[, gene])){
    color_nodes[gene] <- "#0000FF80"
  }
}

# recip_path <- "~/Desktop/"
# 
# pdf(paste0(recip_path, "cryo_common_metabolites_recip_05.pdf"), width = 12, height = 8)
set.seed(1)
plot(gr, vertex.size = 7, vertex.label.cex = .5, edge.width = (E(gr)$value - .7) * 20,
     vertex.color = color_nodes, layout = layout_nicely)
# dev.off()
```


I save the 111 features in an excel file:
The excel file contains information on which features were correlated in the graph above.
It also contains, for every gene, the associated logFC, pvalue..., where the comparison 
is in non tolerant versus tolerant patients (eg, a positive Fold Change for a gene 
therefore means that the gene is overexpressed in non tolerant patients)
```{r, echo = FALSE}
rna_recip_cryo_beh_TNT_95 <- norm_data[,c("group", selb)]
save(rna_recip_cryo_beh_TNT_95,
     file = "../data/RNA_seq_national/recip_only/rna_recip_cryo_beh_TNT_95.RData")

# Identify connected components in the graph:
gr_components <- decompose(gr, mode = "strong", max.comps = NA,
                 min.vertices = 0)
comp_sizes <- sapply(gr_components, diameter)

# Generate a new annotated excel table
annot_genes <- as.data.frame(selb)

DEgenesGroup1b <- DEgenesGroup1[as.character(annot_genes$selb),]
annot_genes <- cbind(annot_genes, DEgenesGroup1b)

# add the information on correlated genes to the table

for( i in which(comp_sizes != 0)){
  vertex_names <- vertex_attr(gr_components[[i]], index = V(gr_components[[i]]))[[1]]
  v_names <- c(vertex_names, rep(0, nrow(annot_genes) - length(vertex_names)))
  annot_genes <- cbind(annot_genes, v_names)
}
write.xlsx(annot_genes, file = "../data/RNA_seq_national/recip_only/annot_111selected_genes_recip_TNT_95.xlsx")
```

Now that we have selected features that should help us classify non and 
tolerant recipients, we can see that a RF model built on these selected
features gives slightly better results than the one built on all features:
```{r, eval = FALSE}
set.seed(1)
rf <- ranger::ranger(group~., data = rna_recip_cryo_TNT_95, num.trees = 5000, importance = "impurity")
rf$confusion.matrix
paste0("prediction error: ", rf$prediction.error*100)
```

We can also try to use only the features that were common and behaved similarly
in the two cohorts, we can also build a classifier on these 111 features. But 
the RF model built on the 278 selected features isn't better than the one built 
on the 140 common genes:
```{r, echo = FALSE, eval = FALSE}
set.seed(1)
rf <- rf_tol_non_tol(df_orig = rna_recip_cryo_beh_TNT_95, ntree = 5000)
rf
```

#### Threshold of 0.90

These genes were selected as informative to classify the non and tolerant 
recipients, with a 0.90 threshold:
```{r, echo = FALSE}
qt <- unlist(map(perm_vals, 1))
densityplot(qt)
selected_genes_qt <- colnames(norm_data)[which(qt>.90)+1]
write.xlsx(selected_genes_qt, file = "../data/RNA_seq_national/recip_only/perm_val_TNT_90_thresh.xlsx")
print(paste0(length(selected_genes_qt),
             " genes were above the 0.90 threshold in the Cryostem cohort."))
```

Which of these selected genes are common with the Cryostem cohort?
```{r}
sel_louis <- read.xlsx("../data/RNAseq/recip_only/perm_val_TNT_90_thresh.xlsx", colNames = FALSE)
sel_common <- which(selected_genes_qt %in% sel_louis$X1)
print(paste0(length(sel_common), " genes were common between the two cohorts."))
# selected_genes_qt[sel_common]
```

Graph of these selected genes that were found commonly between the two cohorts:
```{r, echo = FALSE}
correlations <- norm_data[,selected_genes_qt[sel_common]] %>% 
  as.matrix %>%
  cor %>%
  as.data.frame %>%
  rownames_to_column(var = 'var1') %>%
  gather(var2, value, -var1) %>% 
  dplyr::filter(var1 < var2)

correlated_features <- correlations %>% 
  dplyr::filter(value > 0.7)
gr <- graph_from_data_frame(correlated_features, directed = FALSE, vertices=selected_genes_qt[sel_common])
nodes <- get.vertex.attribute(gr)
color_nodes <- rep("#FF000080", length(nodes$name))

medians <- gr_medians[,nodes$name]
for (gene in seq_along(colnames(medians))){
  if (medians["tolerant", gene] == max(medians[, gene])){
    color_nodes[gene] <- "#0000FF80"
  }
}

# recip_path <- "~/Desktop/"
# 
# pdf(paste0(recip_path, "cryo_common_metabolites_recip_05.pdf"), width = 12, height = 8)
set.seed(1)
plot(gr, vertex.size = 7, vertex.label.cex = .5, edge.width = (E(gr)$value - .7) * 20,
     vertex.color = color_nodes, layout = layout_nicely)
# dev.off()

```

I save the 408 features in an excel file:
The excel file contains information on which features were correlated in the graph above.
It also contains, for every gene, the associated logFC, pvalue..., where the comparison 
is in non tolerant versus tolerant patients (eg, a positive Fold Change for a gene 
therefore means that the gene is overexpressed in non tolrant patients)
```{r, echo = FALSE}
rna_recip_cryo_TNT_90 <- norm_data[,c("group", selected_genes_qt[sel_common])]
save(rna_recip_cryo_TNT_90,
     file = "../data/RNA_seq_national/recip_only/rna_recip_cryo_TNT_90.RData")

# Identify connected components in the graph:
gr_components <- decompose(gr, mode = "strong", max.comps = NA,
                 min.vertices = 0)
comp_sizes <- sapply(gr_components, diameter)

# Generate a new annotated excel table
annot_genes <- as.data.frame(selected_genes_qt[sel_common])

# Compute the FC and pvalues:
patient_groups <- tolower(colData$GROUP)
patient_groups[which(patient_groups != "non_tolerant")] <- "tolerant"
TS <- patient_groups[c(recip_id)]

TS <- factor(TS, levels=unique(TS))
design <- model.matrix(~0+TS)
colnames(design) <- levels(TS)
#design
fit <- lmFit(t(norm_data[,selected_genes_qt[sel_common]]), design)
  
#### Create contrast matrix
cont.matrix <- makeContrasts(group1=non_tolerant-tolerant,
                             group2=tolerant-non_tolerant,
                             levels=design)
#cont.matrix
fit2 = contrasts.fit(fit, cont.matrix)

#### Moderate T-test with 'shrunken' se
fit.eb <- eBayes(fit2)


allGenesGroup1<-topTable(fit.eb, adjust="BH", sort.by="P",number=Inf, coef=1)
DEgenesGroup1<-getDEgenes(allGenesGroup1,1,0)

#length(which(selected_genes_qt[sel_common] %in% rownames(DEgenesGroup1)))

# add them to the table

DEgenesGroup1 <- DEgenesGroup1[as.character(annot_genes$`selected_genes_qt[sel_common]`),] 
annot_genes <- cbind(annot_genes, DEgenesGroup1)

# add the information on correlated genes to the table

for( i in which(comp_sizes != 0)){
  vertex_names <- vertex_attr(gr_components[[i]], index = V(gr_components[[i]]))[[1]]
  v_names <- c(vertex_names, rep(0, nrow(annot_genes) - length(vertex_names)))
  annot_genes <- cbind(annot_genes, v_names)
}
write.xlsx(annot_genes, file = "../data/RNA_seq_national/recip_only/annot_408selected_genes_recip_TNT_90.xlsx")
```

We can see how many of these genes had the same behaviour in the two cohorts (ie, 
were over or underexpressed in the same group of patients)
```{r, echo = FALSE}
dir <- lapply(seq_len(length(sel_common)), function(idx){
  gene <- selected_genes_qt[sel_common[idx]]
  overexp <- rownames(gr_medians)[which.is.max(gr_medians[gene][,1])]
  list(gene, overexp)
})
dir_genes <- do.call(rbind, dir)
write.xlsx(dir_genes, file = "../data/RNA_seq_national/recip_only/perm_val_dir_TNT_90_thresh.xlsx")
dir_genes_cryo <- dir_genes

dir_genes_louis <- read.xlsx("../data/RNAseq/recip_only/perm_val_dir_TNT_90_thresh.xlsx", rowNames = FALSE)
dir_genes_cryo <- as.data.frame(dir_genes_cryo) %>% 
  column_to_rownames("V1")
dir_genes_cryo <- dir_genes_cryo[dir_genes_louis$V1,]
sel_beh <- dir_genes_louis$V1[which(dir_genes_louis$V2 == dir_genes_cryo)]
sel_beh <- as.data.frame(sel_beh)
write.xlsx(sel_beh, file = "../data/RNAseq/recip_only/perm_val_beh_TNT_90_thresh.xlsx")

print(paste0(length(which(dir_genes_louis$V2 == dir_genes_cryo)),
             " genes out of the ", length(sel_common),
             " common selected genes have the same behaviour in both cohorts."))
```

Graph of the 278 genes that were found commonly in the two cohorts AND had the
same behaviour in the two cohorts:
```{r}
selb <- as.character(sel_beh$sel_beh)
correlations <- norm_data[,selb] %>% 
  as.matrix %>%
  cor %>%
  as.data.frame %>%
  rownames_to_column(var = 'var1') %>%
  gather(var2, value, -var1) %>% 
  dplyr::filter(var1 < var2)

correlated_features <- correlations %>% 
  dplyr::filter(value > 0.7)
gr <- graph_from_data_frame(correlated_features, directed = FALSE, vertices=selb)
nodes <- get.vertex.attribute(gr)
color_nodes <- rep("#FF000080", length(nodes$name))

medians <- gr_medians[,nodes$name]
for (gene in seq_along(colnames(medians))){
  if (medians["tolerant", gene] == max(medians[, gene])){
    color_nodes[gene] <- "#0000FF80"
  }
}

# recip_path <- "~/Desktop/"
# 
# pdf(paste0(recip_path, "cryo_common_metabolites_recip_05.pdf"), width = 12, height = 8)
set.seed(1)
plot(gr, vertex.size = 7, vertex.label.cex = .5, edge.width = (E(gr)$value - .7) * 20,
     vertex.color = color_nodes, layout = layout_nicely)
# dev.off()
```

I save the 278 features in an excel file:
The excel file contains information on which features were correlated in the graph above.
It also contains, for every gene, the associated logFC, pvalue..., where the comparison 
is in non tolerant versus tolerant patients (eg, a positive Fold Change for a gene 
therefore means that the gene is overexpressed in non tolrant patients)
```{r, echo = FALSE}
rna_recip_cryo_beh_TNT_90 <- norm_data[,c("group", selb)]
save(rna_recip_cryo_beh_TNT_90,
     file = "../data/RNA_seq_national/recip_only/rna_recip_cryo_beh_TNT_90.RData")

# Identify connected components in the graph:
gr_components <- decompose(gr, mode = "strong", max.comps = NA,
                 min.vertices = 0)
comp_sizes <- sapply(gr_components, diameter)

# Generate a new annotated excel table
annot_genes <- as.data.frame(selb)

DEgenesGroup1b <- DEgenesGroup1[as.character(annot_genes$selb),]
annot_genes <- cbind(annot_genes, DEgenesGroup1b)

# add the information on correlated genes to the table

for( i in which(comp_sizes != 0)){
  vertex_names <- vertex_attr(gr_components[[i]], index = V(gr_components[[i]]))[[1]]
  v_names <- c(vertex_names, rep(0, nrow(annot_genes) - length(vertex_names)))
  annot_genes <- cbind(annot_genes, v_names)
}
write.xlsx(annot_genes, file = "../data/RNA_seq_national/recip_only/annot_278selected_genes_recip_TNT_90.xlsx")
```

Now that we have selected features that should help us classify tolerant and 
non tolerant recipients, we can see that a RF model built on these selected
features is slightly better than the one built on all genes:
```{r, eval = FALSE}
set.seed(1)
rf <- rf_tol_non_tol(df_orig = rna_recip_cryo_TNT_90, ntree = 5000)
rf
```

We can also try to use only the features that were common and behaved similarly
in the two cohorts, we can also build a classifier on these 278 features. But 
the RF model built on the 278 selected features isn't better than the one built 
on the 408 common genes:
```{r, echo = FALSE, eval = FALSE}
set.seed(1)
rf <- rf_tol_non_tol(df_orig = rna_recip_cryo_beh_TNT_90, ntree = 5000)
rf
```

#### Including gender compatibility and age recip in the feature selection

Now that we have selected metabolites that seemed to play a role in tolerance that 
were common to both cohorts (with a threhold over 0.90), we decide to 
investigate into these metabolites to see if some of them are related to the age
of the recipients, to the gender compatibility between the donor and recipient,
to both or to none of these two causes.

I first build new models using the metabolites that were kept in the 
two cohorts as informative when comparing tolerant and non tolerant patients: 

```{r}
new_models <- lapply(seq_along(selb), function(idx){
  dta_tmp <- norm_data[,c(selb[idx],
                        "group", "age_recip", "gender_comp")]
  colnames(dta_tmp)[1] <- "Var"
  fit1 <- glm(group ~ Var, data=dta_tmp,trace=F, family = "binomial")
  fit2 <- glm(group ~ gender_comp + Var,
              data=dta_tmp,trace=F, family = "binomial")
  fit3 <- glm(group ~ age_recip + Var, 
              data=dta_tmp,trace=F, family = "binomial")
  fit4 <- glm(group ~ gender_comp + age_recip + Var, 
              data=dta_tmp,trace=F, family = "binomial")
  list(fit1, fit2, fit3, fit4)
})
```


```{r, eval = FALSE}
plots_gender_age2(pdf_name = "../plots/rnaseq_national/recip/age_and_gender/tol_nontol_0.90_behavior.pdf",
                 norm_data = norm_data,
                 features = selb, compar = "TNT",
                 excel_name = "../plots/rnaseq_national/recip/age_and_gender/tol_nontol_0.90_behavior.xlsx")

```





### Identifying features that vary between primary and secondary tolerant recipients:

Run the permutation distribution, taking demographic variables into account:
I first add the age and gender compatibility information to the data.
```{r, echo = FALSE}
recip_info <- couple_info %>% 
  dplyr::filter(grepl("R", rownames(couple_info))) 

norm_data <- norm_data_3gr %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  left_join(recip_info %>% dplyr::select(Id.Cryostem.R, gender_comp, age_recip),
            by = "Id.Cryostem.R") %>% 
  column_to_rownames("Id.Cryostem.R")
```

And then compute the permutation distribution for the non vs tolerance:
```{r, eval=FALSE, echo = FALSE}
norm_data <- norm_data %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::filter(group != "non_tolerant") %>% 
  mutate(group = as.factor(as.character(group))) %>% 
  column_to_rownames("Id.Cryostem.R")

perm_vals_PRISM(norm_data = norm_data, 
                PRISM_name = "rnaseq_r_PS",
                rds_name = "handle_cryo_rna_r_PS.rds")

handle3 <- readRDS("handle_cryo_rna_r_PS.rds")
perm_vals <- qsub_retrieve(handle3)

save(perm_vals, file = "~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/recip_only/perm_vals_PS.RData")
save(norm_data, file= "~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/recip_only/norm_data_PS.RData")
```

Compute the median per group for every gene, to use it later in the graphs:
```{r}
load("~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/recip_only/perm_vals_PS.RData")
load("~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/recip_only/norm_data_PS.RData")

gr_medians <- norm_data %>%
  group_by(group) %>%
  summarize_if(is.numeric, median) %>%
  as.data.frame()
rownames(gr_medians) <- gr_medians$group
```

#### Threshold of 0.95

These genes were selected as informative to classify the non and tolerant 
recipients, with a 0.95 threshold:
```{r, echo = FALSE}
qt <- unlist(map(perm_vals, 1))
densityplot(qt)
selected_genes_qt <- colnames(norm_data)[which(qt>.95)+1]
write.xlsx(selected_genes_qt, file = "../data/RNA_seq_national/recip_only/perm_val_PS_95_thresh.xlsx")
print(paste0(length(selected_genes_qt),
             " genes were above the 0.95 threshold in the Cryostem cohort."))
```

Which of these selected genes are common with the Cryostem cohort?
```{r, echo = FALSE}
sel_louis <- read.xlsx("../data/RNAseq/recip_only/perm_val_PS_95_thresh.xlsx", colNames = FALSE)
sel_common <- which(selected_genes_qt %in% sel_louis$X1)
print(paste0(length(sel_common), " genes were common between the two cohorts: "))
print(selected_genes_qt[sel_common])
# selected_genes_qt[sel_common]
```

I save these features:
```{r, echo = FALSE}
rna_recip_cryo_PS_95 <- norm_data[,c("group", selected_genes_qt[sel_common])]
save(rna_recip_cryo_PS_95,
     file = "../data/RNA_seq_national/recip_only/rna_recip_cryo_PS_95.RData")

# Generate a new annotated excel table
annot_genes <- as.data.frame(selected_genes_qt[sel_common])

# Compute the FC and pvalues:
TS <- colData[rownames(norm_data),] %>% 
  dplyr::select(GROUP) %>% 
  mutate(GROUP = tolower(GROUP))
TS <- TS$GROUP

TS <- factor(TS, levels=unique(TS))
design <- model.matrix(~0+TS)
colnames(design) <- levels(TS)
# design
fit <- lmFit(t(norm_data[,selected_genes_qt[sel_common]]), design)
  
#### Create contrast matrix
cont.matrix <- makeContrasts(group1=primary_tolerant-secondary_tolerant,
                             group2=secondary_tolerant-primary_tolerant,
                             levels=design)
#cont.matrix
fit2 = contrasts.fit(fit, cont.matrix)

#### Moderate T-test with 'shrunken' se
fit.eb <- eBayes(fit2)


allGenesGroup1<-topTable(fit.eb, adjust="BH", sort.by="P",number=Inf, coef=1)
DEgenesGroup1<-getDEgenes(allGenesGroup1,1,0)

# length(which(selected_genes_qt[sel_common] %in% rownames(DEgenesGroup1)))

# add them to the table

DEgenesGroup1 <- DEgenesGroup1[as.character(annot_genes$`selected_genes_qt[sel_common]`),] 
annot_genes <- cbind(annot_genes, DEgenesGroup1)

write.xlsx(annot_genes, file = "../data/RNA_seq_national/recip_only/annot_3selected_genes_recip_PS_95.xlsx")
```

Boxplots of these common metabolites:
```{r, echo = FALSE}
load("../data/RNA_seq_national/recip_only/rna_recip_cryo_PS_95.RData")
tmp_cryo <- rna_recip_cryo_PS_95
load("../data/RNAseq/recip_only/rna_recip_louis_PS_95.RData")
tmp_louis <- rna_recip_louis_PS_95

common_genes_95 <- selected_genes_qt[sel_common]

for(i in seq_along(common_genes_95)){
  cryo_tmp <- data.frame(Var = tmp_cryo[, i+1],group = tmp_cryo[, 1])
  g1<- ggplot(cryo_tmp) + geom_boxplot(aes(x = group, y = Var, col = group)) +
    geom_jitter(aes(x = group, y = Var, col = group)) + theme_minimal() +
    scale_color_manual(breaks = c("primary_tolerant", "secondary_tolerant"),
                       values = c("springgreen3", "royalblue1")) +
    ggtitle(paste0(colnames(tmp_cryo)[i+1]," in Cryostem"))+
    theme(plot.title = element_text(size = 8, face = "bold"))
  
  louis_tmp <- data.frame(Var = tmp_louis[, i+1],group = tmp_louis[, 1])
  g2<- ggplot(louis_tmp) + geom_boxplot(aes(x = group, y = Var, col = group)) +
    geom_jitter(aes(x = group, y = Var, col = group)) + theme_minimal() +
    scale_color_manual(breaks = c("primary_tolerant", "secondary_tolerant"),
                       values = c("springgreen3", "royalblue1")) +
    ggtitle(paste0(colnames(tmp_louis)[i+1]," in St Louis"))+
    theme(plot.title = element_text(size = 8, face = "bold"))
  gmix <- patchwork::wrap_plots(g1, g2)
  print(gmix)
}
```

We can see how many of these genes had the same behaviour in the two cohorts (ie, 
were over or underexpressed in the same group of patients)
```{r, echo = FALSE}
dir <- lapply(seq_len(length(sel_common)), function(idx){
  gene <- selected_genes_qt[sel_common[idx]]
  overexp <- rownames(gr_medians)[which.is.max(gr_medians[gene][,1])]
  list(gene, overexp)
})
dir_genes <- do.call(rbind, dir)
write.xlsx(dir_genes, file = "../data/RNA_seq_national/recip_only/perm_val_dir_PS_95_thresh.xlsx")
dir_genes_cryo <- dir_genes

dir_genes_louis <- read.xlsx("../data/RNAseq/recip_only/perm_val_dir_PS_95_thresh.xlsx", rowNames = FALSE)
dir_genes_cryo <- as.data.frame(dir_genes_cryo) %>% 
  column_to_rownames("V1")
dir_genes_cryo <- dir_genes_cryo[dir_genes_louis$V1,]
sel_beh <- dir_genes_louis$V1[which(dir_genes_louis$V2 == dir_genes_cryo)]
sel_beh <- as.data.frame(sel_beh)
write.xlsx(sel_beh, file = "../data/RNAseq/recip_only/perm_val_beh_PS_95_thresh.xlsx")

print(paste0(length(which(dir_genes_louis$V2 == dir_genes_cryo)),
             " genes out of the ", length(sel_common),
             " common selected genes has the same behaviour in both cohorts."))
```

I save the 1 feature in an excel file:
The excel file contains, for every gene, the associated logFC, pvalue..., where the comparison 
is in primary versus secondary tolerant patients (eg, a positive Fold Change for a gene 
therefore means that the gene is overexpressed in non tolerant patients)
```{r, echo = FALSE}
selb <- as.character(sel_beh$sel_beh)
rna_recip_cryo_beh_PS_95 <- norm_data[,c("group", selb)]
save(rna_recip_cryo_beh_PS_95,
     file = "../data/RNA_seq_national/recip_only/rna_recip_cryo_beh_PS_95.RData")

# Generate a new annotated excel table
annot_genes <- as.data.frame(selb)

DEgenesGroup1b <- DEgenesGroup1[as.character(annot_genes$selb),]
annot_genes <- cbind(annot_genes, DEgenesGroup1b)

write.xlsx(annot_genes, file = "../data/RNA_seq_national/recip_only/annot_1selected_gene_recip_PS_95.xlsx")
```

Now that we have selected features that should help us classify non and 
tolerant recipients, we can see that a RF model built on these selected
features gives slightly better results than the one built on all features:
```{r, eval = FALSE}
set.seed(1)
rf <- ranger::ranger(group~., data = rna_recip_cryo_PS_95, num.trees = 5000, importance = "impurity")
rf$confusion.matrix
paste0("prediction error: ", rf$prediction.error*100)
```


#### Threshold of 0.90

These genes were selected as informative to classify the non and tolerant 
recipients, with a 0.90 threshold:
```{r, echo = FALSE}
qt <- unlist(map(perm_vals, 1))
densityplot(qt)
selected_genes_qt <- colnames(norm_data)[which(qt>.90)+1]
write.xlsx(selected_genes_qt, file = "../data/RNA_seq_national/recip_only/perm_val_PS_90_thresh.xlsx")
print(paste0(length(selected_genes_qt),
             " genes were above the 0.95 threshold in the Cryostem cohort."))
```

Which of these selected genes are common with the Cryostem cohort?
```{r, echo = FALSE}
sel_louis <- read.xlsx("../data/RNAseq/recip_only/perm_val_PS_90_thresh.xlsx", colNames = FALSE)
sel_common <- which(selected_genes_qt %in% sel_louis$X1)
print(paste0(length(sel_common), " genes were common between the two cohorts."))
# selected_genes_qt[sel_common]
```

Graph of these selected genes that were found commonly between the two cohorts:
```{r, echo = FALSE}
correlations <- norm_data[,selected_genes_qt[sel_common]] %>% 
  as.matrix %>%
  cor %>%
  as.data.frame %>%
  rownames_to_column(var = 'var1') %>%
  gather(var2, value, -var1) %>% 
  dplyr::filter(var1 < var2)

correlated_features <- correlations %>% 
  dplyr::filter(value > 0.7)
gr <- graph_from_data_frame(correlated_features, directed = FALSE, vertices=selected_genes_qt[sel_common])
nodes <- get.vertex.attribute(gr)
color_nodes <- rep("#0000FF80", length(nodes$name))

medians <- gr_medians[,nodes$name]
for (gene in seq_along(colnames(medians))){
  if (medians["primary_tolerant", gene] == max(medians[, gene])){
    color_nodes[gene] <- "#00FF0080"
  }
}

# recip_path <- "~/Desktop/"
# 
# pdf(paste0(recip_path, "cryo_common_metabolites_recip_05.pdf"), width = 12, height = 8)
set.seed(1)
plot(gr, vertex.size = 7, vertex.label.cex = .5, edge.width = (E(gr)$value - .7) * 20,
     vertex.color = color_nodes, layout = layout_nicely)
# dev.off()

```

I save the 45 features in an excel file:
The excel file contains information on which features were correlated in the graph above.
It also contains, for every gene, the associated logFC, pvalue..., where the comparison 
is in non tolerant versus tolerant patients (eg, a positive Fold Change for a gene 
therefore means that the gene is overexpressed in non tolrant patients)
```{r, echo = FALSE}
rna_recip_cryo_PS_90 <- norm_data[,c("group", selected_genes_qt[sel_common])]
save(rna_recip_cryo_PS_90,
     file = "../data/RNA_seq_national/recip_only/rna_recip_cryo_PS_90.RData")

# Identify connected components in the graph:
gr_components <- decompose(gr, mode = "strong", max.comps = NA,
                 min.vertices = 0)
comp_sizes <- sapply(gr_components, diameter)

# Generate a new annotated excel table
annot_genes <- as.data.frame(selected_genes_qt[sel_common])

# Compute the FC and pvalues:
TS <- colData[rownames(norm_data),] %>% 
  dplyr::select(GROUP) %>% 
  mutate(GROUP = tolower(GROUP))
TS <- TS$GROUP

TS <- factor(TS, levels=unique(TS))
design <- model.matrix(~0+TS)
colnames(design) <- levels(TS)
#design
fit <- lmFit(t(norm_data[,selected_genes_qt[sel_common]]), design)
  
#### Create contrast matrix
cont.matrix <- makeContrasts(group1=primary_tolerant-secondary_tolerant,
                             group2=secondary_tolerant-primary_tolerant,
                             levels=design)
#cont.matrix
fit2 = contrasts.fit(fit, cont.matrix)

#### Moderate T-test with 'shrunken' se
fit.eb <- eBayes(fit2)


allGenesGroup1<-topTable(fit.eb, adjust="BH", sort.by="P",number=Inf, coef=1)
DEgenesGroup1<-getDEgenes(allGenesGroup1,1,0)

#length(which(selected_genes_qt[sel_common] %in% rownames(DEgenesGroup1)))

# add them to the table

DEgenesGroup1 <- DEgenesGroup1[as.character(annot_genes$`selected_genes_qt[sel_common]`),] 
annot_genes <- cbind(annot_genes, DEgenesGroup1)

# add the information on correlated genes to the table

for( i in which(comp_sizes != 0)){
  vertex_names <- vertex_attr(gr_components[[i]], index = V(gr_components[[i]]))[[1]]
  v_names <- c(vertex_names, rep(0, nrow(annot_genes) - length(vertex_names)))
  annot_genes <- cbind(annot_genes, v_names)
}
write.xlsx(annot_genes, file = "../data/RNA_seq_national/recip_only/annot_45selected_genes_recip_PS_90.xlsx")
```

We can see how many of these genes had the same behaviour in the two cohorts (ie, 
were over or underexpressed in the same group of patients)
```{r, echo = FALSE}
dir <- lapply(seq_len(length(sel_common)), function(idx){
  gene <- selected_genes_qt[sel_common[idx]]
  overexp <- rownames(gr_medians)[which.is.max(gr_medians[gene][,1])]
  list(gene, overexp)
})
dir_genes <- do.call(rbind, dir)
write.xlsx(dir_genes, file = "../data/RNA_seq_national/recip_only/perm_val_dir_PS_90_thresh.xlsx")
dir_genes_cryo <- dir_genes

dir_genes_louis <- read.xlsx("../data/RNAseq/recip_only/perm_val_dir_PS_90_thresh.xlsx", rowNames = FALSE)
dir_genes_cryo <- as.data.frame(dir_genes_cryo) %>% 
  column_to_rownames("V1")
dir_genes_cryo <- dir_genes_cryo[dir_genes_louis$V1,]
sel_beh <- dir_genes_louis$V1[which(dir_genes_louis$V2 == dir_genes_cryo)]
sel_beh <- as.data.frame(sel_beh)
write.xlsx(sel_beh, file = "../data/RNAseq/recip_only/perm_val_beh_PS_90_thresh.xlsx")

print(paste0(length(which(dir_genes_louis$V2 == dir_genes_cryo)),
             " genes out of the ", length(sel_common),
             " common selected genes have the same behaviour in both cohorts."))
```

Graph of the 24 genes that were found commonly in the two cohorts AND had the
same behaviour in the two cohorts:
```{r}
selb <- as.character(sel_beh$sel_beh)
correlations <- norm_data[,selb] %>% 
  as.matrix %>%
  cor %>%
  as.data.frame %>%
  rownames_to_column(var = 'var1') %>%
  gather(var2, value, -var1) %>% 
  dplyr::filter(var1 < var2)

correlated_features <- correlations %>% 
  dplyr::filter(value > 0.7)
gr <- graph_from_data_frame(correlated_features, directed = FALSE, vertices=selb)
nodes <- get.vertex.attribute(gr)
color_nodes <- rep("#0000FF80", length(nodes$name))

medians <- gr_medians[,nodes$name]
for (gene in seq_along(colnames(medians))){
  if (medians["primary_tolerant", gene] == max(medians[, gene])){
    color_nodes[gene] <- "#00FF0080"
  }
}

# recip_path <- "~/Desktop/"
# 
# pdf(paste0(recip_path, "cryo_common_metabolites_recip_05.pdf"), width = 12, height = 8)
set.seed(1)
plot(gr, vertex.size = 7, vertex.label.cex = .5, edge.width = (E(gr)$value - .7) * 20,
     vertex.color = color_nodes, layout = layout_nicely)
# dev.off()
```

I save the 24 features in an excel file:
The excel file contains information on which features were correlated in the graph above.
It also contains, for every gene, the associated logFC, pvalue..., where the comparison 
is in non tolerant versus tolerant patients (eg, a positive Fold Change for a gene 
therefore means that the gene is overexpressed in non tolrant patients)
```{r, echo = FALSE}
selb <- as.character(sel_beh$sel_beh)
rna_recip_cryo_beh_PS_90 <- norm_data[,c("group", selb)]
save(rna_recip_cryo_beh_PS_90,
     file = "../data/RNA_seq_national/recip_only/rna_recip_cryo_beh_PS_90.RData")

# Identify connected components in the graph:
gr_components <- decompose(gr, mode = "strong", max.comps = NA,
                 min.vertices = 0)
comp_sizes <- sapply(gr_components, diameter)

# Generate a new annotated excel table
annot_genes <- as.data.frame(selb)

DEgenesGroup1b <- DEgenesGroup1[as.character(annot_genes$selb),]
annot_genes <- cbind(annot_genes, DEgenesGroup1b)

# add the information on correlated genes to the table

for( i in which(comp_sizes != 0)){
  vertex_names <- vertex_attr(gr_components[[i]], index = V(gr_components[[i]]))[[1]]
  v_names <- c(vertex_names, rep(0, nrow(annot_genes) - length(vertex_names)))
  annot_genes <- cbind(annot_genes, v_names)
}
write.xlsx(annot_genes, file = "../data/RNA_seq_national/recip_only/annot_24selected_genes_recip_PS_90.xlsx")
```

Now that we have selected 45 features that should help us classify primary and 
secondary tolerant recipients, we can see that a RF model built on these selected
features is slightly better than the one built on all genes:
```{r, eval = FALSE}
set.seed(1)
rf <- ranger::ranger(group~., data = rna_recip_cryo_PS_90, num.trees = 5000, importance = "impurity")
rf$confusion.matrix
paste0("prediction error: ", rf$prediction.error*100)
```

We can also try to use only the 24 features that were common and behaved 
similarly in the two cohorts. But the RF model built on the 24 selected 
features isn't better than the one built on the 44 common genes:
```{r, echo = FALSE, eval = FALSE}
set.seed(1)
rf <- ranger::ranger(group~., data = rna_recip_cryo_beh_PS_90, num.trees = 5000, importance = "impurity")
rf$confusion.matrix
paste0("prediction error: ", rf$prediction.error*100)
```


#### Including gender compatibility and age recip in the feature selection

Now that we have selected metabolites that seemed to play a role in tolerance that 
were common to both cohorts (with a threhold over 0.90), we decide to 
investigate into these metabolites to see if some of them are related to the age
of the recipients, to the gender compatibility between the donor and recipient,
to both or to none of these two causes.

I first build new models using the metabolites that were kept in the 
two cohorts as informative when comparing tolerant and non tolerant patients: 

```{r}
new_models <- lapply(seq_along(selb), function(idx){
  dta_tmp <- norm_data[,c(selb[idx],
                        "group", "age_recip", "gender_comp")]
  colnames(dta_tmp)[1] <- "Var"
  fit1 <- glm(group ~ Var, data=dta_tmp,trace=F, family = "binomial")
  fit2 <- glm(group ~ gender_comp + Var,
              data=dta_tmp,trace=F, family = "binomial")
  fit3 <- glm(group ~ age_recip + Var, 
              data=dta_tmp,trace=F, family = "binomial")
  fit4 <- glm(group ~ gender_comp + age_recip + Var, 
              data=dta_tmp,trace=F, family = "binomial")
  list(fit1, fit2, fit3, fit4)
})
```


```{r, eval = FALSE}
plots_gender_age2(pdf_name = "../plots/rnaseq_national/recip/age_and_gender/prim_sec_0.90_behavior.pdf",
                 norm_data = norm_data,
                 features = selb,
                 compar = "PTST",
                 excel_name = "../plots/rnaseq_national/recip/age_and_gender/prim_sec_0.90_behavior.xlsx")

```

Now that we have generated forest plots and associated statistics, we can see 
which of the selected features that had the same behaviour in both cohorts were
consistently associated with age or gender. In the tolerant versus non-tolerant
recipients:
```{r}
louis <- read.xlsx("../plots/rnaseq/recip/age_and_gender/tol_nontol_0.90.xlsx", 
                   rowNames = TRUE)
cryo <- read.xlsx("../plots/rnaseq_national/recip/age_and_gender/tol_nontol_0.90_behavior.xlsx",
                  rowNames = TRUE)
gender_ft <- c()
age_ft <- c()
gender_age_ft <- c()

for(ft in rownames(louis)){
  l_ft <- louis[ft,]
  c_ft <- cryo[ft,]
  if(l_ft$pval_var < 0.05 & l_ft$pval_var_gender > 0.05 & 
     c_ft$pval_var < 0.05 & c_ft$pval_var_gender > 0.05 &
     sign(l_ft$beta_var_gender) == sign(c_ft$beta_var_gender)){
    gender_ft <- c(gender_ft, ft)
  }
  
  if(l_ft$pval_var < 0.05 & l_ft$pval_var_age > 0.05 & 
     c_ft$pval_var < 0.05 & c_ft$pval_var_age > 0.05 &
     sign(l_ft$beta_var_age) == sign(c_ft$beta_var_age)){
    age_ft <- c(age_ft, ft)
  }
  
  if(l_ft$pval_var < 0.05 & l_ft$pval_var_gender_age > 0.05 & 
     c_ft$pval_var < 0.05 & c_ft$pval_var_gender_age > 0.05 &
     sign(l_ft$beta_var_age) == sign(c_ft$beta_var_age)){
    gender_age_ft <- c(gender_age_ft, ft)
  }
  
}

print("These features are associated with gender in both cohorts:")
gender_ft
print("These features are associated with age in both cohorts:")
age_ft
print("These features are associated with gender and age in both cohorts:")
gender_age_ft[which(!gender_age_ft %in% unique(age_ft, gender_ft))]
```

In the primary versus secondary-tolerant recipients:
```{r}
louis <- read.xlsx("../plots/rnaseq/recip/age_and_gender/prim_sec_0.90.xlsx", 
                   rowNames = TRUE)
cryo <- read.xlsx("../plots/rnaseq_national/recip/age_and_gender/prim_sec_0.90_behavior.xlsx",
                  rowNames = TRUE)
gender_ft <- c()
age_ft <- c()
gender_age_ft <- c()

for(ft in rownames(louis)){
  l_ft <- louis[ft,]
  c_ft <- cryo[ft,]
  if(l_ft$pval_var < 0.05 & l_ft$pval_var_gender > 0.05 & 
     c_ft$pval_var < 0.05 & c_ft$pval_var_gender > 0.05 &
     sign(l_ft$beta_var_gender) == sign(c_ft$beta_var_gender)){
    gender_ft <- c(gender_ft, ft)
  }
  
  if(l_ft$pval_var < 0.05 & l_ft$pval_var_age > 0.05 & 
     c_ft$pval_var < 0.05 & c_ft$pval_var_age > 0.05 &
     sign(l_ft$beta_var_age) == sign(c_ft$beta_var_age)){
    age_ft <- c(age_ft, ft)
  }
  
  if(l_ft$pval_var < 0.05 & l_ft$pval_var_gender_age > 0.05 & 
     c_ft$pval_var < 0.05 & c_ft$pval_var_gender_age > 0.05 &
     sign(l_ft$beta_var_age) == sign(c_ft$beta_var_age)){
    gender_age_ft <- c(gender_age_ft, ft)
  }
  
}

print("These features are associated with gender in both cohorts:")
gender_ft
print("These features are associated with age in both cohorts:")
age_ft
print("These features are associated with gender and age in both cohorts:")
gender_age_ft
```





# Analysis on the donors only :

```{r preproc_donors, echo=FALSE, eval = FALSE}
counts <- countData[,c(donor_id)]

count_nb <- apply(counts,2,as.numeric) # turn characters into numerics
rownames(count_nb) <- rownames(counts) # restore the rownmes which were lost in the transfo
counts <- count_nb

### Preprocessing
y <- DGEList(counts = counts)
keep = rowSums(cpm(y)>1) >= 3
y = y[keep,]
dim(y) # from 69000 genes to 18000 genes after this filter

# Reset lib sizes
y$samples$lib.size = colSums(y$counts)
```

The names of the genes are in the ensembl ID format. When possible, I replace them with the associated gene symbol:
```{r, eval = FALSE, echo = FALSE}
##CONVERT ENSEMBL GENE NAME TO ENTREZID & SYMBOL
hgnc_symbol <- convertIDs( rownames(y$counts), "ENSEMBL", "SYMBOL", org.Hs.eg.db )

counts_gene_symbol <- y$counts
names_2change <- hgnc_symbol[-which(is.na(hgnc_symbol))]
names_2keep_idx <- which(is.na(hgnc_symbol))
dupp <- names_2change[which(duplicated(names_2change))]
dupp_idx <- which(hgnc_symbol %in% dupp)[c(T,F)]
final_ids_2keep <- c(names_2keep_idx, dupp_idx)

rownames(counts_gene_symbol)[-final_ids_2keep] <- hgnc_symbol[-final_ids_2keep]

y$counts <- counts_gene_symbol
```
Perform normalisation and initiate the design matrix (per group)

```{r, eval = FALSE, echo = FALSE}
# compute norm_factors
normfac <- calcNormFactors(y, method = "TMM")

#### Create design matrix
# for the three groups :
patient_groups <- tolower(colData$GROUP)
TS <- patient_groups[c(donor_id)]

TS <- factor(TS, levels=unique(TS))
design <- model.matrix(~0+TS)
colnames(design) <- levels(TS)
design

#voom
#In limma, read counts are converted to log2-counts-per-million (logCPM) and
#the mean-variance relationship is modelled either with precision weights or
#with an empirical Bayes prior trend. The precision weights approach is called “voom”
# lib.size <- counts$samples$lib.size * counts$samples$norm.factors
# E <- t(log2(t(counts + 0.5)/(lib.size + 1) * 1e+06))
v <- voom(y, design, plot = TRUE)
expTable<-v$E
```

Save the produced objects:

```{r, eval = FALSE, echo = FALSE}
DE_list <- list(v, y, design)

save(expTable, file = "~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/donors_only/expression_table_donors.RData")
save(DE_list, file = "~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/donors_only/data_for_DE_donors.RData")
```

Load the produced data :
```{r}
load("~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/donors_only/expression_table_donors.RData")
load("~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/donors_only/data_for_DE_donors.RData")
```

### Differential Analysis 

Fit a model to the design:
```{r, echo = FALSE}
names(DE_list) <- c("v", "y", "design")
fit <- lmFit(DE_list$v, DE_list$design)

#### Create contrast matrix
cont.matrix <- makeContrasts(group1=non_tolerant-primary_tolerant,
                             group2=non_tolerant-secondary_tolerant,
                             group3=primary_tolerant-secondary_tolerant,
                             levels=DE_list$design)
cont.matrix
fit2 = contrasts.fit(fit, cont.matrix)

#### Moderate T-test with 'shrunken' se
fit.eb <- eBayes(fit2)
```

Adjust the pvalues for multiple testing :
Number of DE genes found for the three groups
```{r, echo = FALSE}
#### Adjust P-values via Benjamini-Hochberg
##### 1. non-tol vs tol-1
allGenesGroup1<-topTable(fit.eb, adjust="BH", sort.by="P",number=Inf, coef=1)
DEgenesGroup1<-getDEgenes(allGenesGroup1,0.05,1)
paste0("Non tolerant vs primary: ", dim(DEgenesGroup1)[1])
##84

##### 2. non-tol vs tol-2
allGenesGroup2<-topTable(fit.eb, adjust="BH", sort.by="P",number=Inf, coef=2)
DEgenesGroup2<-getDEgenes(allGenesGroup2,0.05,1)
paste0("Non tolerant vs secondary: ", dim(DEgenesGroup2)[1])
##57

##### 3. tol-1 vs tol-2
allGenesGroup3<-topTable(fit.eb, adjust="BH", sort.by="P",number=Inf, coef=3)
DEgenesGroup3<-getDEgenes(allGenesGroup3,0.05,1)
paste0("Primary vs secondary: ", dim(DEgenesGroup3)[1])
##0
```

### Machine learning approach :

filter out lowly variable genes:

```{r, echo = FALSE}
sds <- apply(expTable,1,sd)
plot(sort(sds), type = "l")
abline(h=quantile(sds, 0.5), col="red")
expT <- expTable[which(sds>=quantile(sds, 0.5)),]
rnames <- rownames(expT)
expT <- apply(expT,2,as.numeric)
rownames(expT) <- rnames
```

Prepare the data:
```{r, cache=FALSE, results='hide', warning=FALSE, comment=FALSE, warning=FALSE, message = FALSE, include=FALSE, echo = FALSE}

norm_data <- t(expT)
patient_groups <- tolower(colData$GROUP)
group_tmp <- patient_groups[c(donor_id)]
norm_data <- as.data.frame(norm_data) %>% 
  rownames_to_column("rn") %>% 
  mutate(group = as.factor(group_tmp)) %>% 
  dplyr::select(rn, group, rownames(expT)) %>% 
  column_to_rownames("rn")
colnames(norm_data) <- make.names(colnames(norm_data))

norm_data_3gr <- norm_data
```

We perform a PCA on the donors to see if we already observe some structure 
in the data, and if there isn't too much batch effect:
```{r, echo = FALSE}
colData_donor <- colData[c(donor_id),] %>% 
  mutate(GROUP = tolower(GROUP))

pca <- prcomp(norm_data[,-1])

pca_2plot <- as.data.frame(pca$x) %>% 
  mutate(Id.Cryostem.R = rownames(pca$x)) %>% 
  left_join(colData_donor %>% mutate(Id.Cryostem.R = X3), by = "Id.Cryostem.R")
ggplot(pca_2plot, aes(x = PC1, y = PC2, col = as.factor(GROUP), label = Id.Cryostem.R)) + 
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0)

ggplot(pca_2plot, aes(x = PC1, y = PC2, col = as.factor(RNAseq.RUN), label = Id.Cryostem.R)) + 
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0)

ggplot(pca_2plot, aes(x = PC1, y = PC2, col = as.numeric(RIN), label = Id.Cryostem.R)) + 
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0)

```
We don't observe much structure on the PCA.

We can try to build a random forest model using all the genes of the 
donors:
```{r, echo = FALSE}
set.seed(1)
rf_donor <- ranger::ranger(group~., norm_data, mtry = 40, num.trees = 1000,
                           importance = "impurity")
rf_donor$confusion.matrix
rf_donor$prediction.error*100
```

But it seems like the data doesn't allow us to classify into the three tolerance
groups. We can try to simplify the classification problem:

First, we try to separate tolerant from non tolerant donors:
```{r, eval = FALSE}
set.seed(1)
rf <- rf_tol_non_tol(df_orig = norm_data, ntree = 5000)
rf
```

Second, we can try to classify primary tolerant from secondary tolerant donors:
```{r, eval = FALSE}
set.seed(1)
rf <- rf_tol1_tol2(df_orig = norm_data, ntree = 5000)
rf
```
But the error rates are still very high, even after splitting the classification 
problem in two.

## Feature selection

### Identifying features that vary between non and tolerant donors:

Run the permutation distribution, taking demographic variables into account:
I first add the age and gender compatibility information to the data.
```{r, echo = FALSE}
donor_info <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::filter(grepl("D", rownames(couple_info))) 

norm_data <- norm_data_3gr %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  left_join(donor_info %>% dplyr::select(Id.Cryostem.R, gender_comp, age_recip),
            by = "Id.Cryostem.R") %>% 
  column_to_rownames("Id.Cryostem.R")
```

And then compute the permutation distribution for the non vs tolerance:
```{r, eval=FALSE, echo = FALSE}
norm_data$group <- as.character(norm_data$group)
norm_data$group[which(norm_data$group!="non_tolerant")] <- "tolerant"
norm_data$group <- as.factor(norm_data$group)

perm_vals_PRISM(norm_data = norm_data, 
                PRISM_name = "rnaseq_d_TNT",
                rds_name = "handle_cryo_rna_d_TNT.rds")

handle3 <- readRDS("handle_cryo_rna_d_TNT.rds")
perm_vals <- qsub_retrieve(handle3)

save(perm_vals, file = "~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/donors_only/perm_vals_TNT.RData")
save(norm_data, file= "~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/donors_only/norm_data_TNT.RData")
```

Compute the median per group for every gene, to use it later in the graphs:
```{r}
load("~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/donors_only/perm_vals_TNT.RData")
load("~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/donors_only/norm_data_TNT.RData")

gr_medians <- norm_data %>%
  group_by(group) %>%
  summarize_if(is.numeric, median) %>%
  as.data.frame()
rownames(gr_medians) <- gr_medians$group
```

#### Threshold of 0.95

These genes were selected as informative to classify the non and tolerant 
recipients, with a 0.95 threshold:
```{r, echo = FALSE}
qt <- unlist(map(perm_vals, 1))
densityplot(qt)
selected_genes_qt <- colnames(norm_data)[which(qt>.95)+1]
write.xlsx(selected_genes_qt, file = "../data/RNA_seq_national/donors_only/perm_val_TNT_95_thresh.xlsx")
print(paste0(length(selected_genes_qt),
             " genes were above the 0.95 threshold in the Cryostem cohort."))
```

Which of these selected genes are common with the Cryostem cohort?
```{r, echo = FALSE}
sel_louis <- read.xlsx("../data/RNAseq/donors_only/perm_val_TNT_95_thresh.xlsx", colNames = FALSE)
sel_common <- which(selected_genes_qt %in% sel_louis$X1)
print(paste0(length(sel_common), " genes were common between the two cohorts."))
# selected_genes_qt[sel_common]
```

Graph of these selected genes that were found commonly between the two cohorts:
```{r, echo = FALSE}
correlations <- norm_data[,selected_genes_qt[sel_common]] %>% 
  as.matrix %>%
  cor %>%
  as.data.frame %>%
  rownames_to_column(var = 'var1') %>%
  gather(var2, value, -var1) %>% 
  dplyr::filter(var1 < var2)

correlated_features <- correlations %>% 
  dplyr::filter(value > 0.7)
gr <- graph_from_data_frame(correlated_features, directed = FALSE, vertices=selected_genes_qt[sel_common])
nodes <- get.vertex.attribute(gr)
color_nodes <- rep("#FF000080", length(nodes$name))

medians <- gr_medians[,nodes$name]
for (gene in seq_along(colnames(medians))){
  if (medians["tolerant", gene] == max(medians[, gene])){
    color_nodes[gene] <- "#0000FF80"
  }
}

# recip_path <- "~/Desktop/"
# 
# pdf(paste0(recip_path, "cryo_common_metabolites_recip_05.pdf"), width = 12, height = 8)
set.seed(1)
plot(gr, vertex.size = 7, vertex.label.cex = .5, edge.width = (E(gr)$value - .7) * 20,
     vertex.color = color_nodes, layout = layout_nicely)
# dev.off()

```

I save these features:
```{r, echo = FALSE}
rna_donor_cryo_TNT_95 <- norm_data[,c("group", selected_genes_qt[sel_common])]
save(rna_donor_cryo_TNT_95,
     file = "../data/RNA_seq_national/donors_only/rna_donor_cryo_TNT_95.RData")

# Identify connected components in the graph:
gr_components <- decompose(gr, mode = "strong", max.comps = NA,
                 min.vertices = 0)
comp_sizes <- sapply(gr_components, diameter)

# Generate a new annotated excel table
annot_genes <- as.data.frame(selected_genes_qt[sel_common])

# Compute the FC and pvalues:
patient_groups <- tolower(colData$GROUP)
patient_groups[which(patient_groups != "non_tolerant")] <- "tolerant"
TS <- patient_groups[c(donor_id)]

TS <- factor(TS, levels=unique(TS))
design <- model.matrix(~0+TS)
colnames(design) <- levels(TS)
#design
fit <- lmFit(t(norm_data[,selected_genes_qt[sel_common]]), design)
  
#### Create contrast matrix
cont.matrix <- makeContrasts(group1=non_tolerant-tolerant,
                             group2=tolerant-non_tolerant,
                             levels=design)
#cont.matrix
fit2 = contrasts.fit(fit, cont.matrix)

#### Moderate T-test with 'shrunken' se
fit.eb <- eBayes(fit2)


allGenesGroup1<-topTable(fit.eb, adjust="BH", sort.by="P",number=Inf, coef=1)
DEgenesGroup1<-getDEgenes(allGenesGroup1,1,0)

# length(which(selected_genes_qt[sel_common] %in% rownames(DEgenesGroup1)))

# add them to the table

DEgenesGroup1 <- DEgenesGroup1[as.character(annot_genes$`selected_genes_qt[sel_common]`),] 
annot_genes <- cbind(annot_genes, DEgenesGroup1)


for( i in which(comp_sizes != 0)){
  vertex_names <- vertex_attr(gr_components[[i]], index = V(gr_components[[i]]))[[1]]
  v_names <- c(vertex_names, rep(0, nrow(annot_genes) - length(vertex_names)))
  annot_genes <- cbind(annot_genes, v_names)
}
write.xlsx(annot_genes, file = "../data/RNA_seq_national/donors_only/annot_70selected_genes_donor_TNT_95.xlsx")
```


We can see how many of these genes had the same behaviour in the two cohorts (ie, 
were over or underexpressed in the same group of patients)
```{r, echo = FALSE}
dir <- lapply(seq_len(length(sel_common)), function(idx){
  gene <- selected_genes_qt[sel_common[idx]]
  overexp <- rownames(gr_medians)[which.is.max(gr_medians[gene][,1])]
  list(gene, overexp)
})
dir_genes <- do.call(rbind, dir)
write.xlsx(dir_genes, file = "../data/RNA_seq_national/donors_only/perm_val_dir_TNT_95_thresh.xlsx")
dir_genes_cryo <- dir_genes

dir_genes_louis <- read.xlsx("../data/RNAseq/donors_only/perm_val_dir_TNT_95_thresh.xlsx", rowNames = FALSE)
dir_genes_cryo <- as.data.frame(dir_genes_cryo) %>% 
  column_to_rownames("V1")
dir_genes_cryo <- dir_genes_cryo[dir_genes_louis$V1,]
sel_beh <- dir_genes_louis$V1[which(dir_genes_louis$V2 == dir_genes_cryo)]
sel_beh <- as.data.frame(sel_beh)
write.xlsx(sel_beh, file = "../data/RNAseq/donors_only/perm_val_beh_TNT_95_thresh.xlsx")

print(paste0(length(which(dir_genes_louis$V2 == dir_genes_cryo)),
             " genes out of the ", length(sel_common),
             " common selected genes have the same behaviour in both cohorts."))
```

Boxplots of thes 5 common genes:
```{r, echo = FALSE}
load("../data/RNA_seq_national/donors_only/rna_donor_cryo_TNT_95.RData")
tmp_cryo <- rna_donor_cryo_TNT_95
load("../data/RNAseq/donors_only/rna_donor_louis_TNT_95.RData")
tmp_louis <- rna_donor_louis_TNT_95

selb <- as.character(sel_beh$sel_beh)
common_genes_95 <- selb

for(gene in common_genes_95){
  idx <- which(colnames(tmp_cryo)==gene)
  cryo_tmp <- data.frame(Var = tmp_cryo[, idx],group = tmp_cryo[, 1])
  g1<- ggplot(cryo_tmp) + geom_boxplot(aes(x = group, y = Var, col = group)) +
    geom_jitter(aes(x = group, y = Var, col = group)) + theme_minimal() +
    scale_color_manual(breaks = c("non_tolerant", "tolerant"),
                       values = c("red", "royalblue1")) +
    ggtitle(paste0(colnames(tmp_cryo)[idx]," in Cryostem"))+
    theme(plot.title = element_text(size = 8, face = "bold"))
  
  idx <- which(colnames(tmp_louis)==gene)
  louis_tmp <- data.frame(Var = tmp_louis[, i+1],group = tmp_louis[, 1])
  g2<- ggplot(louis_tmp) + geom_boxplot(aes(x = group, y = Var, col = group)) +
    geom_jitter(aes(x = group, y = Var, col = group)) + theme_minimal() +
    scale_color_manual(breaks = c("non_tolerant", "tolerant"),
                       values = c("red", "royalblue1")) +
    ggtitle(paste0(colnames(tmp_louis)[idx]," in St Louis"))+
    theme(plot.title = element_text(size = 8, face = "bold"))
  gmix <- patchwork::wrap_plots(g1, g2)
  print(gmix)
}
```

I save the 5 features in an excel file:
The excel file contains information on which features were correlated in the graph above.
It also contains, for every gene, the associated logFC, pvalue..., where the comparison 
is in non tolerant versus tolerant patients (eg, a positive Fold Change for a gene 
therefore means that the gene is overexpressed in non tolerant patients)
```{r, echo = FALSE}
rna_donor_cryo_beh_TNT_95 <- norm_data[,c("group", selb)]
save(rna_donor_cryo_beh_TNT_95,
     file = "../data/RNA_seq_national/donors_only/rna_donor_cryo_beh_TNT_95.RData")


# Generate a new annotated excel table
annot_genes <- as.data.frame(selb)

DEgenesGroup1b <- DEgenesGroup1[as.character(annot_genes$selb),]
annot_genes <- cbind(annot_genes, DEgenesGroup1b)

write.xlsx(annot_genes, file = "../data/RNA_seq_national/donors_only/annot_5selected_genes_donor_TNT_95.xlsx")
```

Now that we have selected features that should help us classify non and 
tolerant donorients, we can see that a RF model built on these selected
features gives slightly better results than the one built on all features:
```{r, eval = FALSE}
set.seed(1)
rf <- ranger::ranger(group~., data = rna_donor_cryo_TNT_95, num.trees = 5000, importance = "impurity")
rf$confusion.matrix
paste0("prediction error: ", rf$prediction.error*100)
```

We can also try to use only the features that were common and behaved similarly
in the two cohorts, we can also build a classifier on these 111 features. But 
the RF model built on the 5 selected features isn't better than the one built 
on the 70 common genes:
```{r, echo = FALSE, eval = FALSE}
set.seed(1)
rf <- rf_tol_non_tol(df_orig = rna_donor_cryo_beh_TNT_95, ntree = 5000)
rf
```

#### Threshold of 0.90

These genes were selected as informative to classify the non and tolerant 
donorients, with a 0.90 threshold:
```{r, echo = FALSE}
qt <- unlist(map(perm_vals, 1))
densityplot(qt)
selected_genes_qt <- colnames(norm_data)[which(qt>.90)+1]
write.xlsx(selected_genes_qt, file = "../data/RNA_seq_national/donors_only/perm_val_TNT_90_thresh.xlsx")
print(paste0(length(selected_genes_qt),
             " genes were above the 0.90 threshold in the Cryostem cohort."))
```

Which of these selected genes are common with the Cryostem cohort?
```{r, echo = FALSE}
sel_louis <- read.xlsx("../data/RNAseq/donors_only/perm_val_TNT_90_thresh.xlsx", colNames = FALSE)
sel_common <- which(selected_genes_qt %in% sel_louis$X1)
print(paste0(length(sel_common), " genes were common between the two cohorts."))
# selected_genes_qt[sel_common]
```

Graph of these selected genes that were found commonly between the two cohorts:
```{r, echo = FALSE}
correlations <- norm_data[,selected_genes_qt[sel_common]] %>% 
  as.matrix %>%
  cor %>%
  as.data.frame %>%
  rownames_to_column(var = 'var1') %>%
  gather(var2, value, -var1) %>% 
  dplyr::filter(var1 < var2)

correlated_features <- correlations %>% 
  dplyr::filter(value > 0.7)
gr <- graph_from_data_frame(correlated_features, directed = FALSE, vertices=selected_genes_qt[sel_common])
nodes <- get.vertex.attribute(gr)
color_nodes <- rep("#FF000080", length(nodes$name))

medians <- gr_medians[,nodes$name]
for (gene in seq_along(colnames(medians))){
  if (medians["tolerant", gene] == max(medians[, gene])){
    color_nodes[gene] <- "#0000FF80"
  }
}

# recip_path <- "~/Desktop/"
# 
# pdf(paste0(recip_path, "cryo_common_metabolites_recip_05.pdf"), width = 12, height = 8)
set.seed(1)
plot(gr, vertex.size = 7, vertex.label.cex = .5, edge.width = (E(gr)$value - .7) * 20,
     vertex.color = color_nodes, layout = layout_nicely)
# dev.off()

```

I save these features:
```{r, echo = FALSE}
rna_donor_cryo_TNT_90 <- norm_data[,c("group", selected_genes_qt[sel_common])]
save(rna_donor_cryo_TNT_90,
     file = "../data/RNA_seq_national/donors_only/rna_donor_cryo_TNT_90.RData")

# Identify connected components in the graph:
gr_components <- decompose(gr, mode = "strong", max.comps = NA,
                 min.vertices = 0)
comp_sizes <- sapply(gr_components, diameter)

# Generate a new annotated excel table
annot_genes <- as.data.frame(selected_genes_qt[sel_common])

# Compute the FC and pvalues:
patient_groups <- tolower(colData$GROUP)
patient_groups[which(patient_groups != "non_tolerant")] <- "tolerant"
TS <- patient_groups[c(donor_id)]

TS <- factor(TS, levels=unique(TS))
design <- model.matrix(~0+TS)
colnames(design) <- levels(TS)
#design
fit <- lmFit(t(norm_data[,selected_genes_qt[sel_common]]), design)
  
#### Create contrast matrix
cont.matrix <- makeContrasts(group1=non_tolerant-tolerant,
                             group2=tolerant-non_tolerant,
                             levels=design)
#cont.matrix
fit2 = contrasts.fit(fit, cont.matrix)

#### Moderate T-test with 'shrunken' se
fit.eb <- eBayes(fit2)


allGenesGroup1<-topTable(fit.eb, adjust="BH", sort.by="P",number=Inf, coef=1)
DEgenesGroup1<-getDEgenes(allGenesGroup1,1,0)

# length(which(selected_genes_qt[sel_common] %in% rownames(DEgenesGroup1)))

# add them to the table

DEgenesGroup1 <- DEgenesGroup1[as.character(annot_genes$`selected_genes_qt[sel_common]`),] 
annot_genes <- cbind(annot_genes, DEgenesGroup1)


for( i in which(comp_sizes != 0)){
  vertex_names <- vertex_attr(gr_components[[i]], index = V(gr_components[[i]]))[[1]]
  v_names <- c(vertex_names, rep(0, nrow(annot_genes) - length(vertex_names)))
  annot_genes <- cbind(annot_genes, v_names)
}
write.xlsx(annot_genes, file = "../data/RNA_seq_national/donors_only/annot_310selected_genes_donor_TNT_90.xlsx")
```


We can see how many of these genes had the same behaviour in the two cohorts (ie, 
were over or underexpressed in the same group of patients)
```{r, echo = FALSE}
dir <- lapply(seq_len(length(sel_common)), function(idx){
  gene <- selected_genes_qt[sel_common[idx]]
  overexp <- rownames(gr_medians)[which.is.max(gr_medians[gene][,1])]
  list(gene, overexp)
})
dir_genes <- do.call(rbind, dir)
write.xlsx(dir_genes, file = "../data/RNA_seq_national/donors_only/perm_val_dir_TNT_90_thresh.xlsx")
dir_genes_cryo <- dir_genes

dir_genes_louis <- read.xlsx("../data/RNAseq/donors_only/perm_val_dir_TNT_90_thresh.xlsx", rowNames = FALSE)
dir_genes_cryo <- as.data.frame(dir_genes_cryo) %>% 
  column_to_rownames("V1")
dir_genes_cryo <- dir_genes_cryo[dir_genes_louis$V1,]
sel_beh <- dir_genes_louis$V1[which(dir_genes_louis$V2 == dir_genes_cryo)]
sel_beh <- as.data.frame(sel_beh)
write.xlsx(sel_beh, file = "../data/RNAseq/donors_only/perm_val_beh_TNT_90_thresh.xlsx")

print(paste0(length(which(dir_genes_louis$V2 == dir_genes_cryo)),
             " genes out of the ", length(sel_common),
             " common selected genes have the same behaviour in both cohorts."))
```

Graph of these 22 selected genes that were found commonly between the two cohorts:
```{r, echo = FALSE}
selb <- as.character(sel_beh$sel_beh)
correlations <- norm_data[,selb] %>% 
  as.matrix %>%
  cor %>%
  as.data.frame %>%
  rownames_to_column(var = 'var1') %>%
  gather(var2, value, -var1) %>% 
  dplyr::filter(var1 < var2)

correlated_features <- correlations %>% 
  dplyr::filter(value > 0.7)
gr <- graph_from_data_frame(correlated_features, directed = FALSE, vertices=selb)
nodes <- get.vertex.attribute(gr)
color_nodes <- rep("#FF000080", length(nodes$name))

medians <- gr_medians[,nodes$name]
for (gene in seq_along(colnames(medians))){
  if (medians["tolerant", gene] == max(medians[, gene])){
    color_nodes[gene] <- "#0000FF80"
  }
}

# recip_path <- "~/Desktop/"
# 
# pdf(paste0(recip_path, "cryo_common_metabolites_recip_05.pdf"), width = 12, height = 8)
set.seed(1)
plot(gr, vertex.size = 7, vertex.label.cex = .5, edge.width = (E(gr)$value - .7) * 20,
     vertex.color = color_nodes, layout = layout_nicely)
# dev.off()

```

I save these features:
```{r, echo = FALSE}
rna_donor_cryo_beh_TNT_90 <- norm_data[,c("group", selb)]
save(rna_donor_cryo_beh_TNT_90,
     file = "../data/RNA_seq_national/donors_only/rna_donor_cryo_beh_TNT_90.RData")

# Identify connected components in the graph:
gr_components <- decompose(gr, mode = "strong", max.comps = NA,
                 min.vertices = 0)
comp_sizes <- sapply(gr_components, diameter)

# Generate a new annotated excel table
annot_genes <- as.data.frame(selb)

# Compute the FC and pvalues:
patient_groups <- tolower(colData$GROUP)
patient_groups[which(patient_groups != "non_tolerant")] <- "tolerant"
TS <- patient_groups[c(donor_id)]

TS <- factor(TS, levels=unique(TS))
design <- model.matrix(~0+TS)
colnames(design) <- levels(TS)
#design
fit <- lmFit(t(norm_data[,selb]), design)
  
#### Create contrast matrix
cont.matrix <- makeContrasts(group1=non_tolerant-tolerant,
                             group2=tolerant-non_tolerant,
                             levels=design)
#cont.matrix
fit2 = contrasts.fit(fit, cont.matrix)

#### Moderate T-test with 'shrunken' se
fit.eb <- eBayes(fit2)


allGenesGroup1<-topTable(fit.eb, adjust="BH", sort.by="P",number=Inf, coef=1)
DEgenesGroup1<-getDEgenes(allGenesGroup1,1,0)

# length(which(selected_genes_qt[sel_common] %in% rownames(DEgenesGroup1)))

# add them to the table

DEgenesGroup1 <- DEgenesGroup1[as.character(annot_genes$selb),] 
annot_genes <- cbind(annot_genes, DEgenesGroup1)


for( i in which(comp_sizes != 0)){
  vertex_names <- vertex_attr(gr_components[[i]], index = V(gr_components[[i]]))[[1]]
  v_names <- c(vertex_names, rep(0, nrow(annot_genes) - length(vertex_names)))
  annot_genes <- cbind(annot_genes, v_names)
}
write.xlsx(annot_genes, file = "../data/RNA_seq_national/donors_only/annot_22selected_genes_donor_TNT_90.xlsx")
```


Now that we have selected features that should help us classify primary and 
secondary tolerant recipients, we can see that a RF model built on these selected
features is better than the one built on all genes:
```{r, echo = FALSE, eval = FALSE}
set.seed(1)
rf <- ranger::ranger(group~., data = rna_donor_cryo_TNT_90, num.trees = 5000, importance = "impurity")
rf$confusion.matrix
paste0("prediction error: ", rf$prediction.error*100)
```

We can do the same on the 22 genes that had the same behaviour in both cohorts,
but it doesn't seem to improve:
```{r, echo = FALSE, eval = FALSE}
set.seed(1)
rf <- ranger::ranger(group~., data = rna_donor_cryo_beh_TNT_90, num.trees = 5000, importance = "impurity")
rf$confusion.matrix
paste0("prediction error: ", rf$prediction.error*100)
```

#### Including gender compatibility and age recip in the feature selection

Now that we have selected metabolites that seemed to play a role in tolerance that 
were common to both cohorts (with a threhold over 0.90), we decide to 
investigate into these metabolites to see if some of them are related to the age
of the recipients, to the gender compatibility between the donor and recipient,
to both or to none of these two causes.

I first build new models using the metabolites that were kept in the 
two cohorts as informative when comparing tolerant and non tolerant patients: 

```{r}
new_models <- lapply(seq_along(selb), function(idx){
  dta_tmp <- norm_data[,c(selb[idx],
                        "group", "age_recip", "gender_comp")]
  colnames(dta_tmp)[1] <- "Var"
  fit1 <- glm(group ~ Var, data=dta_tmp,trace=F, family = "binomial")
  fit2 <- glm(group ~ gender_comp + Var,
              data=dta_tmp,trace=F, family = "binomial")
  fit3 <- glm(group ~ age_recip + Var, 
              data=dta_tmp,trace=F, family = "binomial")
  fit4 <- glm(group ~ gender_comp + age_recip + Var, 
              data=dta_tmp,trace=F, family = "binomial")
  list(fit1, fit2, fit3, fit4)
})
```


```{r, eval = FALSE}
plots_gender_age(pdf_name = "../plots/rnaseq_national/donors/age_and_gender/tol_nontol_0.90_behavior.pdf",
                 norm_data = norm_data,
                 features = selb)

```


### Identifying features that vary between primary and secondary tolerant donors:

Run the permutation distribution, taking demographic variables into account:
I first add the age and gender compatibility information to the data.
```{r, echo = FALSE}
donor_info <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::filter(grepl("D", rownames(couple_info))) 

norm_data <- norm_data_3gr %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  left_join(donor_info %>% dplyr::select(Id.Cryostem.R, gender_comp, age_recip),
            by = "Id.Cryostem.R") %>% 
  column_to_rownames("Id.Cryostem.R")
```

And then compute the permutation distribution for the non vs tolerance:
```{r, eval=FALSE, echo = FALSE}
norm_data <- norm_data %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::filter(group != "non_tolerant") %>% 
  mutate(group = as.factor(as.character(group))) %>% 
  column_to_rownames("Id.Cryostem.R")

perm_vals_PRISM(norm_data = norm_data, 
                PRISM_name = "rnaseq_d_PS",
                rds_name = "handle_cryo_rna_d_PS.rds")

handle3 <- readRDS("handle_cryo_rna_d_PS.rds")
perm_vals <- qsub_retrieve(handle3)

save(perm_vals, file = "~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/donors_only/perm_vals_PS.RData")
save(norm_data, file= "~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/donors_only/norm_data_PS.RData")
```

Compute the median per group for every gene, to use it later in the graphs:
```{r}
load("~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/donors_only/perm_vals_PS.RData")
load("~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/donors_only/norm_data_PS.RData")

gr_medians <- norm_data %>%
  group_by(group) %>%
  summarize_if(is.numeric, median) %>%
  as.data.frame()
rownames(gr_medians) <- gr_medians$group
```

#### Threshold of 0.95

These genes were selected as informative to classify the non and tolerant 
recipients, with a 0.95 threshold:
```{r, echo = FALSE}
qt <- unlist(map(perm_vals, 1))
densityplot(qt)
selected_genes_qt <- colnames(norm_data)[which(qt>.95)+1]
write.xlsx(selected_genes_qt, file = "../data/RNA_seq_national/donors_only/perm_val_PS_95_thresh.xlsx")
print(paste0(length(selected_genes_qt),
             " genes were above the 0.90 threshold in the Cryostem cohort."))
```

Which of these selected genes are common with the Cryostem cohort?
```{r, echo = FALSE}
sel_louis <- read.xlsx("../data/RNAseq/donors_only/perm_val_PS_95_thresh.xlsx", colNames = FALSE)
sel_common <- which(selected_genes_qt %in% sel_louis$X1)
print(paste0(length(sel_common), " genes were common between the two cohorts."))
# selected_genes_qt[sel_common]
```

Graph of these selected genes that were found commonly between the two cohorts:
```{r, echo = FALSE}
correlations <- norm_data[,selected_genes_qt[sel_common]] %>% 
  as.matrix %>%
  cor %>%
  as.data.frame %>%
  rownames_to_column(var = 'var1') %>%
  gather(var2, value, -var1) %>% 
  dplyr::filter(var1 < var2)

correlated_features <- correlations %>% 
  dplyr::filter(value > 0.7)
gr <- graph_from_data_frame(correlated_features, directed = FALSE, vertices=selected_genes_qt[sel_common])
nodes <- get.vertex.attribute(gr)
color_nodes <- rep("#0000FF80", length(nodes$name))

medians <- gr_medians[,nodes$name]
for (gene in seq_along(colnames(medians))){
  if (medians["primary_tolerant", gene] == max(medians[, gene])){
    color_nodes[gene] <- "#00FF0080"
  }
}

# recip_path <- "~/Desktop/"
# 
# pdf(paste0(recip_path, "cryo_common_metabolites_recip_05.pdf"), width = 12, height = 8)
set.seed(1)
plot(gr, vertex.size = 7, vertex.label.cex = .5, edge.width = (E(gr)$value - .7) * 20,
     vertex.color = color_nodes, layout = layout_nicely)
# dev.off()

```

I save these features:
```{r, echo = FALSE}
save_genes_annot(norm_data = norm_data,
                 selected_genes = selected_genes_qt[sel_common],
                 path_obj = "../data/RNA_seq_national/donors_only/rna_donor_cryo_PS_95.RDS",
                 compar_design = "PS",
                 graph = gr,
                 path_annot_obj = "../data/RNA_seq_national/donors_only/annot_26selected_genes_donor_PS_95.xlsx")
```


We can see how many of these genes had the same behaviour in the two cohorts (ie, 
were over or underexpressed in the same group of patients)
```{r, echo = FALSE}
dir <- lapply(seq_len(length(sel_common)), function(idx){
  gene <- selected_genes_qt[sel_common[idx]]
  overexp <- rownames(gr_medians)[which.is.max(gr_medians[gene][,1])]
  list(gene, overexp)
})
dir_genes <- do.call(rbind, dir)
write.xlsx(dir_genes, file = "../data/RNA_seq_national/donors_only/perm_val_dir_PS_95_thresh.xlsx")
dir_genes_cryo <- dir_genes

dir_genes_louis <- read.xlsx("../data/RNAseq/donors_only/perm_val_dir_PS_95_thresh.xlsx", rowNames = FALSE)
dir_genes_cryo <- as.data.frame(dir_genes_cryo) %>% 
  column_to_rownames("V1")
dir_genes_cryo <- dir_genes_cryo[dir_genes_louis$V1,]
sel_beh <- dir_genes_louis$V1[which(dir_genes_louis$V2 == dir_genes_cryo)]
sel_beh <- as.data.frame(sel_beh)
write.xlsx(sel_beh, file = "../data/RNAseq/donors_only/perm_val_beh_PS_95_thresh.xlsx")

print(paste0(length(which(dir_genes_louis$V2 == dir_genes_cryo)),
             " genes out of the ", length(sel_common),
             " common selected genes have the same behaviour in both cohorts."))
```

Graph of the 6 genes that had the same behaviour in the two cohorts:
```{r, echo = FALSE}
selb <- as.character(sel_beh$sel_beh)
correlations <- norm_data[,selb] %>% 
  as.matrix %>%
  cor %>%
  as.data.frame %>%
  rownames_to_column(var = 'var1') %>%
  gather(var2, value, -var1) %>% 
  dplyr::filter(var1 < var2)

correlated_features <- correlations %>% 
  dplyr::filter(value > 0.7)
gr <- graph_from_data_frame(correlated_features, directed = FALSE, vertices=selb)
nodes <- get.vertex.attribute(gr)
color_nodes <- rep("#0000FF80", length(nodes$name))

medians <- gr_medians[,nodes$name]
for (gene in seq_along(colnames(medians))){
  if (medians["primary_tolerant", gene] == max(medians[, gene])){
    color_nodes[gene] <- "#00FF0080"
  }
}

# recip_path <- "~/Desktop/"
# 
# pdf(paste0(recip_path, "cryo_common_metabolites_recip_05.pdf"), width = 12, height = 8)
set.seed(1)
plot(gr, vertex.size = 7, vertex.label.cex = .5, edge.width = (E(gr)$value - .7) * 20,
     vertex.color = color_nodes, layout = layout_nicely)
# dev.off()

```

```{r, echo = FALSE}
save_genes_annot(norm_data = norm_data,
                 selected_genes = selb,
                 path_obj = "../data/RNA_seq_national/donors_only/rna_donor_cryo_beh_PS_95.RDS",
                 compar_design = "PS",
                 graph = gr,
                 path_annot_obj = "../data/RNA_seq_national/donors_only/annot_6selected_genes_donor_PS_95.xlsx")
```

Now that we have selected features that should help us classify primary and 
secondary tolerant recipients, we can see that a RF model built on these selected
features is better than the one built on all genes:
```{r, echo = FALSE}
rna_donor_cryo_PS_95 <- readRDS("../data/RNA_seq_national/donors_only/rna_donor_cryo_PS_95.RDS")
set.seed(1)
rf <- ranger::ranger(group~., data = rna_donor_cryo_PS_95, num.trees = 5000, importance = "impurity")
rf$confusion.matrix
paste0("prediction error: ", rf$prediction.error*100)
```

We can also build a RF model using only the 6 genes that had the same behaviour
in both cohorts, but the classification resutls are slightly worse: 
```{r, echo = FALSE}
rna_donor_cryo_beh_PS_95 <- readRDS("../data/RNA_seq_national/donors_only/rna_donor_cryo_beh_PS_95.RDS")
set.seed(1)
rf <- ranger::ranger(group~., data = rna_donor_cryo_beh_PS_95, num.trees = 5000, importance = "impurity")
rf$confusion.matrix
paste0("prediction error: ", rf$prediction.error*100)
```


#### Threshold of 0.90

These genes were selected as informative to classify the non and tolerant 
recipients, with a 0.90 threshold:
```{r, echo = FALSE}
qt <- unlist(map(perm_vals, 1))
densityplot(qt)
selected_genes_qt <- colnames(norm_data)[which(qt>.90)+1]
write.xlsx(selected_genes_qt, file = "../data/RNA_seq_national/donors_only/perm_val_PS_90_thresh.xlsx")
print(paste0(length(selected_genes_qt),
             " genes were above the 0.90 threshold in the Cryostem cohort."))
```

Which of these selected genes are common with the Cryostem cohort?
```{r, echo = FALSE}
sel_louis <- read.xlsx("../data/RNAseq/donors_only/perm_val_PS_90_thresh.xlsx", colNames = FALSE)
sel_common <- which(selected_genes_qt %in% sel_louis$X1)
print(paste0(length(sel_common), " genes were common between the two cohorts."))
# selected_genes_qt[sel_common]
```

Graph of these selected genes that were found commonly between the two cohorts:
```{r, echo = FALSE}
correlations <- norm_data[,selected_genes_qt[sel_common]] %>% 
  as.matrix %>%
  cor %>%
  as.data.frame %>%
  rownames_to_column(var = 'var1') %>%
  gather(var2, value, -var1) %>% 
  dplyr::filter(var1 < var2)

correlated_features <- correlations %>% 
  dplyr::filter(value > 0.7)
gr <- graph_from_data_frame(correlated_features, directed = FALSE, vertices=selected_genes_qt[sel_common])
nodes <- get.vertex.attribute(gr)
color_nodes <- rep("#0000FF80", length(nodes$name))

medians <- gr_medians[,nodes$name]
for (gene in seq_along(colnames(medians))){
  if (medians["primary_tolerant", gene] == max(medians[, gene])){
    color_nodes[gene] <- "#00FF0080"
  }
}

# recip_path <- "~/Desktop/"
# 
# pdf(paste0(recip_path, "cryo_common_metabolites_recip_05.pdf"), width = 12, height = 8)
set.seed(1)
plot(gr, vertex.size = 7, vertex.label.cex = .5, edge.width = (E(gr)$value - .7) * 20,
     vertex.color = color_nodes, layout = layout_nicely)
# dev.off()

```

```{r, echo = FALSE}
save_genes_annot(norm_data = norm_data,
                 selected_genes = selected_genes_qt[sel_common],
                 path_obj = "../data/RNA_seq_national/donors_only/rna_donor_cryo_PS_90.RDS",
                 compar_design = "PS",
                 graph = gr,
                 path_annot_obj = "../data/RNA_seq_national/donors_only/annot_104selected_genes_donor_PS_90.xlsx")
```

We can see how many of these genes had the same behaviour in the two cohorts (ie, 
were over or underexpressed in the same group of patients)
```{r, echo = FALSE}
dir <- lapply(seq_len(length(sel_common)), function(idx){
  gene <- selected_genes_qt[sel_common[idx]]
  overexp <- rownames(gr_medians)[which.is.max(gr_medians[gene][,1])]
  list(gene, overexp)
})
dir_genes <- do.call(rbind, dir)
write.xlsx(dir_genes, file = "../data/RNA_seq_national/donors_only/perm_val_dir_PS_90_thresh.xlsx")
dir_genes_cryo <- dir_genes

dir_genes_louis <- read.xlsx("../data/RNAseq/donors_only/perm_val_dir_PS_90_thresh.xlsx", rowNames = FALSE)
dir_genes_cryo <- as.data.frame(dir_genes_cryo) %>% 
  column_to_rownames("V1")
dir_genes_cryo <- dir_genes_cryo[dir_genes_louis$V1,]
sel_beh <- dir_genes_louis$V1[which(dir_genes_louis$V2 == dir_genes_cryo)]
sel_beh <- as.data.frame(sel_beh)
write.xlsx(sel_beh, file = "../data/RNAseq/donors_only/perm_val_beh_PS_90_thresh.xlsx")

print(paste0(length(which(dir_genes_louis$V2 == dir_genes_cryo)),
             " genes out of the ", length(sel_common),
             " common selected genes have the same behaviour in both cohorts."))
```

Graph of the 16 genes that had the same behaviour in the two cohorts:
```{r, echo = FALSE}
selb <- as.character(sel_beh$sel_beh)
correlations <- norm_data[,selb] %>% 
  as.matrix %>%
  cor %>%
  as.data.frame %>%
  rownames_to_column(var = 'var1') %>%
  gather(var2, value, -var1) %>% 
  dplyr::filter(var1 < var2)

correlated_features <- correlations %>% 
  dplyr::filter(value > 0.7)
gr <- graph_from_data_frame(correlated_features, directed = FALSE, vertices=selb)
nodes <- get.vertex.attribute(gr)
color_nodes <- rep("#0000FF80", length(nodes$name))

medians <- gr_medians[,nodes$name]
for (gene in seq_along(colnames(medians))){
  if (medians["primary_tolerant", gene] == max(medians[, gene])){
    color_nodes[gene] <- "#00FF0080"
  }
}

# recip_path <- "~/Desktop/"
# 
# pdf(paste0(recip_path, "cryo_common_metabolites_recip_05.pdf"), width = 12, height = 8)
set.seed(1)
plot(gr, vertex.size = 7, vertex.label.cex = .5, edge.width = (E(gr)$value - .7) * 20,
     vertex.color = color_nodes, layout = layout_nicely)
# dev.off()

```

```{r, echo = FALSE}
save_genes_annot(norm_data = norm_data,
                 selected_genes = selb,
                 path_obj = "../data/RNA_seq_national/donors_only/rna_donor_cryo_beh_PS_90.RDS",
                 compar_design = "PS",
                 graph = gr,
                 path_annot_obj = "../data/RNA_seq_national/donors_only/annot_16selected_genes_donor_PS_90.xlsx")
```


Now that we have selected features that should help us classify primary and 
secondary tolerant recipients, we can see that a RF model built on these selected
features is better than the one built on all genes:
```{r, echo = FALSE}
rna_donors_cryo_PS_90 <- readRDS("../data/RNA_seq_national/donors_only/rna_donor_cryo_PS_90.RDS")
set.seed(1)
rf <- ranger::ranger(group~., data = rna_donors_cryo_PS_90, num.trees = 5000, importance = "impurity")
rf$confusion.matrix
paste0("prediction error: ", rf$prediction.error*100)
```

We can also build a RF only on the 16 genes that had the same behaviour in both
cohorts, but this doesn't improve the results:
```{r}
rna_donors_cryo_beh_PS_90 <- readRDS("../data/RNA_seq_national/donors_only/rna_donor_cryo_beh_PS_90.RDS")
set.seed(1)
rf <- ranger::ranger(group~., data = rna_donors_cryo_beh_PS_90, num.trees = 5000, importance = "impurity")
rf$confusion.matrix
paste0("prediction error: ", rf$prediction.error*100)
```


#### Including gender compatibility and age recip in the feature selection

Now that we have selected metabolites that seemed to play a role in tolerance that 
were common to both cohorts (with a threhold over 0.90), we decide to 
investigate into these metabolites to see if some of them are related to the age
of the recipients, to the gender compatibility between the donor and recipient,
to both or to none of these two causes.

I first build new models using the metabolites that were kept in the 
two cohorts as informative when comparing tolerant and non tolerant patients: 

```{r}
new_models <- lapply(seq_along(selb), function(idx){
  dta_tmp <- norm_data[,c(selb[idx],
                        "group", "age_recip", "gender_comp")]
  colnames(dta_tmp)[1] <- "Var"
  fit1 <- glm(group ~ Var, data=dta_tmp,trace=F, family = "binomial")
  fit2 <- glm(group ~ gender_comp + Var,
              data=dta_tmp,trace=F, family = "binomial")
  fit3 <- glm(group ~ age_recip + Var, 
              data=dta_tmp,trace=F, family = "binomial")
  fit4 <- glm(group ~ gender_comp + age_recip + Var, 
              data=dta_tmp,trace=F, family = "binomial")
  list(fit1, fit2, fit3, fit4)
})
```


```{r, eval = FALSE}
plots_gender_age(pdf_name = "../plots/rnaseq_national/donors/age_and_gender/prim_sec_0.90_behavior.pdf",
                 norm_data = norm_data,
                 features = selb,
                 compar = "PTST")

```






# Analysis on the donors - recipients

```{r preproc_d_r, echo=FALSE}
counts <- countData[,c(donor_id, recip_id)]

count_nb <- apply(counts,2,as.numeric) # turn characters into numerics
rownames(count_nb) <- rownames(counts) # restore the rownmes which were lost in the transfo
counts <- count_nb

# Keep only the patients that form complete cases (donor + recipient)

colData_rd <- colData[colnames(counts),]
complete_couples <- colData_rd$COUPLENUMBER[which(duplicated(colData_rd$COUPLENUMBER))]
counts <- counts[,which(colData_rd$COUPLENUMBER %in% complete_couples)]

colData_rd <- colData_rd[colnames(counts),]

### Preprocessing
y <- DGEList(counts = counts)
keep = rowSums(cpm(y)>1) >= 3
y = y[keep,]
dim(y) # from 69000 genes to 18000 genes after this filter

# Reset lib sizes
y$samples$lib.size = colSums(y$counts)
```

The names of the genes are in the ensembl ID format. When possible, I replace them with the associated gene symbol:
```{r, eval = FALSE, echo = FALSE}
##CONVERT ENSEMBL GENE NAME TO ENTREZID & SYMBOL
hgnc_symbol <- convertIDs( rownames(y$counts), "ENSEMBL", "SYMBOL", org.Hs.eg.db )

counts_gene_symbol <- y$counts
names_2change <- hgnc_symbol[-which(is.na(hgnc_symbol))]
names_2keep_idx <- which(is.na(hgnc_symbol))
dupp <- names_2change[which(duplicated(names_2change))]
dupp_idx <- which(hgnc_symbol %in% dupp)[c(T,F)]
final_ids_2keep <- c(names_2keep_idx, dupp_idx)

rownames(counts_gene_symbol)[-final_ids_2keep] <- hgnc_symbol[-final_ids_2keep]

y$counts <- counts_gene_symbol
```

Perform normalisation and initiate the design matrix (per group)

```{r, eval = FALSE, echo = FALSE}
# compute norm_factors
normfac <- calcNormFactors(y, method = "TMM")

#### Create design matrix
# for the three groups :
TS <- tolower(colData_rd$GROUP)

TS <- factor(TS, levels=unique(TS))
design <- model.matrix(~0+TS)
colnames(design) <- levels(TS)
design

#voom
#In limma, read counts are converted to log2-counts-per-million (logCPM) and
#the mean-variance relationship is modelled either with precision weights or
#with an empirical Bayes prior trend. The precision weights approach is called “voom”
# lib.size <- counts$samples$lib.size * counts$samples$norm.factors
# E <- t(log2(t(counts + 0.5)/(lib.size + 1) * 1e+06))
v <- voom(y, design, plot = TRUE)
expTable<-v$E
```

Save the produced objects:

```{r, eval = FALSE, echo = FALSE}
DE_list <- list(v, y, design)

save(expTable, file = "~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/donors_and_recip/expression_table_dr.RData")
save(DE_list, file = "~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/donors_and_recip/data_for_DE_dr.RData")
```

Load the data :
```{r, echo = FALSE}
load("~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/donors_and_recip/expression_table_dr.RData")
load("~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/donors_and_recip/data_for_DE_dr.RData")

big_mat <- as.data.frame(t(expTable)) %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  left_join(colData %>% mutate(Id.Cryostem.R = X3), by = "Id.Cryostem.R")

#big_mat <- big_mat[which(duplicated(big_mat$COUPLENUMBER)),]

vars2rm <- colnames(colData)[-c(3, 4)]
big_mat2 <- big_mat %>%
  group_by(COUPLENUMBER) %>%
  dplyr::select(-vars2rm)

subst <- big_mat2 %>%
  summarise_if(is.numeric, ~.[1]-.[2])

colData_tmp <- colData[big_mat2$X3,]
colData_tmp$GROUP <- as.factor(tolower(colData_tmp$GROUP))
colData_subst <- colData_tmp %>% 
  arrange(COUPLENUMBER) %>% 
  dplyr::select(c("GROUP", "COUPLENUMBER")) %>% 
  unique()

subst <- subst %>% 
  mutate("couple" = paste0("couple_",COUPLENUMBER),
         "group" = colData_subst$GROUP) %>% 
  column_to_rownames("couple") %>% 
  dplyr::select (-"COUPLENUMBER") 

subst <- subst[,c(ncol(subst), 1:(ncol(subst)-1))]

colnames(subst) <- make.names(colnames(subst))
colnames(subst)[-1] <- paste0("subst_", colnames(subst))[-1]
```


### Machine learning approach :

filter out lowly variable genes:

```{r, echo = FALSE}
sds <- apply(subst[,-1],2,sd)
plot(sort(sds), type = "l")
abline(h=quantile(sds, 0.5), col="red")
expT <- subst[,which(sds>=quantile(sds, 0.5))+1]
rnames <- rownames(expT)
expT <- apply(expT,2,as.numeric)
rownames(expT) <- rnames
```

Prepare the data:
```{r, cache=FALSE, results='hide', warning=FALSE, comment=FALSE, warning=FALSE, message = FALSE, include=FALSE, echo = FALSE}
norm_data <- expT
group_tmp <- subst$group
norm_data <- as.data.frame(norm_data) %>% 
  rownames_to_column("rn") %>% 
  mutate(group = as.factor(group_tmp)) %>% 
  dplyr::select(rn, group, colnames(expT)) %>% 
  column_to_rownames("rn")
colnames(norm_data) <- make.names(colnames(norm_data))
norm_data_3gr_dr <- norm_data
```

We can apply a principal components analysis to see if there is some structure
in the data that can already be seen:
```{r, echo = FALSE}
genes_in_norm_data <- colnames(norm_data %>% dplyr::select(-group))
colnames(big_mat) <- make.names(colnames(big_mat))
bmat <- big_mat %>% 
  dplyr::select(c(gsub("subst_", "", genes_in_norm_data),
                  "GROUP", "Id.Cryostem.R", "COUPLENUMBER")) %>% 
  mutate(GROUP = as.factor(tolower(GROUP)))

tosave <- bmat %>% column_to_rownames("Id.Cryostem.R")
saveRDS(tosave, "../data/RNA_seq_national/compar_donors_recipients/mat_pheno_funct_donors_and_recipients.RDS")
saveRDS(norm_data, "../data/RNA_seq_national/compar_donors_recipients/mat_all_patients.RDS")

toplot <- bmat %>% 
  dplyr::select(-GROUP, -COUPLENUMBER) %>% 
  column_to_rownames("Id.Cryostem.R")
pca <- prcomp(as.matrix(toplot))

pca_2plot <- as.data.frame(pca$x, stringsAsFactors = F) %>% 
  dplyr::select(PC1, PC2) %>% 
  mutate(Id.Cryostem.R = rownames(pca$x)) %>% 
  left_join(bmat, by = "Id.Cryostem.R")
ggplot(pca_2plot, aes(x = PC1, y = PC2, col = as.factor(GROUP), label = Id.Cryostem.R)) + 
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0)
```

In the following PCA plot, we plot the recipients as black dots, the donors as 
white dots, and we link donors and recipients that belong to the same couple by
a line of the color of their group.
```{r, echo = FALSE}
plot(pca$x)
distances = c()
for (i in names(table(bmat$COUPLENUMBER))){
  group_status <- bmat$GROUP[which(bmat$COUPLENUMBER==i)]
  pt_coord <- pca$x[as.numeric(rownames(bmat)[which(bmat$COUPLENUMBER==i)]),c(1:2)]
  if(group_status[1]=="non_tolerant"){
    lines(pt_coord,col="red")
    points(pt_coord[1,1], pt_coord[1,2], pch = 19)
  } else if(group_status[1]=="primary_tolerant"){
    lines(pt_coord,col="green")
    points(pt_coord[1,1], pt_coord[1,2], pch = 19)
  } else {
    lines(pt_coord,col="blue")
    points(pt_coord[1,1], pt_coord[1,2], pch = 19)
  }
  distances = c(distances, sqrt((pt_coord[2,1]-pt_coord[1,1])^2 + (pt_coord[2,2]-pt_coord[1,2])^2))
}
```

It looks like, in the RNAseq data of the Cryostem cohort, the distances 
between non tolerant donors and recipients are not really bigger than the 
distances between primary or secondary tolerant donors and recipients. 
```{r, echo = FALSE}

list_couples <- names(table(bmat$COUPLENUMBER))
names(distances) <- paste0("couple_",list_couples)
couple_dist_matrix <- bmat[,c("COUPLENUMBER", "GROUP")] %>% arrange(COUPLENUMBER) %>% unique()
couple_dist_matrix$GROUP <- as.factor(couple_dist_matrix$GROUP)
dis_colors <- c("red","green","blue")[couple_dist_matrix$GROUP]
order_dis <- order(distances)

plot(distances[order_dis], col = dis_colors[order_dis], pch=19,
     main = "Distance between donor and recipient from same couple",
     ylab = "Distance D/R", xaxt = "n", xlab = "")
axis(1, at=seq_along(list_couples), labels=FALSE)
text(x=seq_along(list_couples), y=par()$usr[3]-0.1*(par()$usr[4]-par()$usr[3]),
     labels=names(distances[order_dis]), srt=45, adj=1, xpd=TRUE, cex = .7)
legend("topleft", c("non_tolerant","primary_tolerant","secondary_tolerant"),
       col = c("red","green","blue"), pch = 19)
```

We can already try to classify the couples based on all the genes:
```{r, echo = FALSE, eval = FALSE}
set.seed(1)
rf_dr <- ranger::ranger(group~., norm_data, num.trees = 1000,
                           importance = "impurity")
rf_dr$confusion.matrix
rf_dr$prediction.error
```

Trying to classify between non tolerant and tolerant couples:

```{r, echo=FALSE, eval = FALSE}
set.seed(1)
rf1_subst <- rf_tol_non_tol(df_orig = norm_data, ntree = 5000)
rf1_subst
```

Trying to classify only primary and secondary tolerant couples:
```{r, echo = FALSE, eval = FALSE}
set.seed(1)
rf_subst_tol <- rf_tol1_tol2(df_orig = norm_data, ntree = 5000)
rf_subst_tol
```


## Feature selection

### Finding features that mainly differ between tolerant and non tolerant couples:

Run the permutation distribution, taking demographic variables into account:
I first add the age and gender compatibility information to the data.
```{r, echo = FALSE}
dr_info <- couple_info %>% 
  mutate(COUPLENUMBER = paste0("couple_", COUPLENUMBER)) %>% 
  dplyr::filter(grepl("R", rownames(couple_info))) 

norm_data <- norm_data_3gr_dr %>% 
  rownames_to_column("COUPLENUMBER") %>% 
  left_join(dr_info %>% dplyr::select(COUPLENUMBER, gender_comp, age_recip),
            by = "COUPLENUMBER") %>% 
  column_to_rownames("COUPLENUMBER")
```

And then compute the permutation distribution for the non vs tolerance:
```{r, eval=FALSE, echo = FALSE}
norm_data$group <- as.character(norm_data$group)
norm_data$group[which(norm_data$group!="non_tolerant")] <- "tolerant"
norm_data$group <- as.factor(norm_data$group)

perm_vals_PRISM(norm_data = norm_data, PRISM_name = "rnaseq_dr_cryo_TNT",
                rds_name = "handle_cryo_rna_dr_TNT.rds")

handle3 <- readRDS("handle_cryo_rna_dr_TNT.rds")
perm_vals <- qsub_retrieve(handle3)

save(perm_vals, file = "~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/donors_and_recip/perm_vals_TNT.RData")
save(norm_data, file= "~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/donors_and_recip/norm_data_TNT.RData")
```

Compute the median per group for every gene, to use it later in the graphs:
```{r}
load("~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/donors_and_recip/perm_vals_TNT.RData")
load("~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/donors_and_recip/norm_data_TNT.RData")

gr_medians <- norm_data %>%
  group_by(group) %>%
  summarize_if(is.numeric, median) %>%
  as.data.frame()
rownames(gr_medians) <- gr_medians$group
```

#### Threshold of 0.95

These genes were selected as informative to classify the non and tolerant 
couples, with a 0.95 threshold:
```{r, echo = FALSE}
qt <- unlist(map(perm_vals, 1))
densityplot(qt)
selected_genes_qt <- colnames(norm_data)[which(qt>.95)+1]
write.xlsx(selected_genes_qt, file = "../data/RNA_seq_national/donors_and_recip/perm_val_TNT_95_thresh.xlsx")
print(paste0(length(selected_genes_qt),
             " genes were above the 0.90 threshold in the Cryostem cohort."))
```

Which of these selected genes are common with the Cryostem cohort?
```{r, echo = FALSE}
sel_louis <- read.xlsx("../data/RNAseq/donors_and_recip/perm_val_TNT_95_thresh.xlsx", colNames = FALSE)
sel_common <- which(selected_genes_qt %in% sel_louis$X1)
print(paste0(length(sel_common), " genes were common between the two cohorts."))
# selected_genes_qt[sel_common]
```

Graph of these selected genes that were found commonly between the two cohorts:
```{r, echo = FALSE}
correlations <- norm_data[,selected_genes_qt[sel_common]] %>% 
  as.matrix %>%
  cor %>%
  as.data.frame %>%
  rownames_to_column(var = 'var1') %>%
  gather(var2, value, -var1) %>% 
  dplyr::filter(var1 < var2)

correlated_features <- correlations %>% 
  dplyr::filter(value > 0.7)
gr <- graph_from_data_frame(correlated_features, directed = FALSE, vertices=selected_genes_qt[sel_common])
nodes <- get.vertex.attribute(gr)
color_nodes <- rep("#FF000080", length(nodes$name))

medians <- gr_medians[,nodes$name]
for (gene in seq_along(colnames(medians))){
  if (medians["tolerant", gene] == max(medians[, gene])){
    color_nodes[gene] <- "#0000FF80"
  }
}

# recip_path <- "~/Desktop/"
# 
# pdf(paste0(recip_path, "cryo_common_metabolites_recip_05.pdf"), width = 12, height = 8)
set.seed(1)
plot(gr, vertex.size = 7, vertex.label.cex = .5, edge.width = (E(gr)$value - .7) * 20,
     vertex.color = color_nodes, layout = layout_nicely)
# dev.off()

```

```{r, echo = FALSE}
save_genes_annot(norm_data = norm_data,
                 selected_genes = selected_genes_qt[sel_common],
                 path_obj = "../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_TNT_95.RDS",
                 compar_design = "TNT",
                 graph = gr,
                 path_annot_obj = "../data/RNA_seq_national/donors_and_recip/annot_68selected_genes_dr_TNT_95.xlsx")
```

We can see how many of these genes had the same behaviour in the two cohorts (ie, 
were over or underexpressed in the same group of patients)
```{r, echo = FALSE}
dir <- lapply(seq_len(length(sel_common)), function(idx){
  gene <- selected_genes_qt[sel_common[idx]]
  overexp <- rownames(gr_medians)[which.is.max(gr_medians[gene][,1])]
  list(gene, overexp)
})
dir_genes <- do.call(rbind, dir)
write.xlsx(dir_genes, file = "../data/RNA_seq_national/donors_and_recip/perm_val_dir_TNT_95_thresh.xlsx")
dir_genes_cryo <- dir_genes

dir_genes_louis <- read.xlsx("../data/RNAseq/donors_and_recip/perm_val_dir_TNT_95_thresh.xlsx", rowNames = FALSE)
dir_genes_cryo <- as.data.frame(dir_genes_cryo) %>% 
  column_to_rownames("V1")
dir_genes_cryo <- dir_genes_cryo[dir_genes_louis$V1,]
sel_beh <- dir_genes_louis$V1[which(dir_genes_louis$V2 == dir_genes_cryo)]
sel_beh <- as.data.frame(sel_beh)
write.xlsx(sel_beh, file = "../data/RNAseq/donors_and_recip/perm_val_beh_TNT_95_thresh.xlsx")

print(paste0(length(which(dir_genes_louis$V2 == dir_genes_cryo)),
             " genes out of the ", length(sel_common),
             " common selected genes have the same behaviour in both cohorts."))
```

Graph of the 28 genes that had the same behaviour in the two cohorts:
```{r, echo = FALSE}
selb <- as.character(sel_beh$sel_beh)
correlations <- norm_data[,selb] %>% 
  as.matrix %>%
  cor %>%
  as.data.frame %>%
  rownames_to_column(var = 'var1') %>%
  gather(var2, value, -var1) %>% 
  dplyr::filter(var1 < var2)

correlated_features <- correlations %>% 
  dplyr::filter(value > 0.7)
gr <- graph_from_data_frame(correlated_features, directed = FALSE, vertices=selb)
nodes <- get.vertex.attribute(gr)
color_nodes <- rep("#0000FF80", length(nodes$name))

medians <- gr_medians[,nodes$name]
for (gene in seq_along(colnames(medians))){
  if (medians["non_tolerant", gene] == max(medians[, gene])){
    color_nodes[gene] <- "#FF000080"
  }
}

# recip_path <- "~/Desktop/"
# 
# pdf(paste0(recip_path, "cryo_common_metabolites_recip_05.pdf"), width = 12, height = 8)
set.seed(1)
plot(gr, vertex.size = 7, vertex.label.cex = .5, edge.width = (E(gr)$value - .7) * 20,
     vertex.color = color_nodes, layout = layout_nicely)
# dev.off()

```

```{r, echo = FALSE}
save_genes_annot(norm_data = norm_data,
                 selected_genes = selb,
                 path_obj = "../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_beh_TNT_95.RDS",
                 compar_design = "TNT",
                 graph = gr,
                 path_annot_obj = "../data/RNA_seq_national/donors_and_recip/annot_28selected_genes_dr_TNT_95.xlsx")
```

Now that we have selected features that should help us classify primary and 
secondary tolerant recipients, we can see that a RF model built on these selected
features is better than the one built on all genes:
```{r, echo = FALSE}
rna_dr_cryo_TNT_95 <- readRDS("../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_TNT_95.RDS")
set.seed(1)
rf <- ranger::ranger(group~., data = rna_dr_cryo_TNT_95, num.trees = 5000, importance = "impurity")
rf$confusion.matrix
paste0("prediction error: ", rf$prediction.error*100)
```

We can also build a RF only on the 28 genes that had the same behaviour in 
the two cohorts, which doesn't improve the classification:
```{r, echo = FALSE}
rna_dr_cryo_beh_TNT_95 <- readRDS("../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_beh_TNT_95.RDS")
set.seed(1)
rf <- ranger::ranger(group~., data = rna_dr_cryo_beh_TNT_95, num.trees = 5000, importance = "impurity")
rf$confusion.matrix
paste0("prediction error: ", rf$prediction.error*100)
```

#### Threshold of 0.90

These genes were selected as informative to classify the non and tolerant 
recipients, with a 0.90 threshold:
```{r, echo = FALSE}
qt <- unlist(map(perm_vals, 1))
densityplot(qt)
selected_genes_qt <- colnames(norm_data)[which(qt>.90)+1]
write.xlsx(selected_genes_qt, file = "../data/RNA_seq_national/donors_and_recip/perm_val_TNT_90_thresh.xlsx")
print(paste0(length(selected_genes_qt),
             " genes were above the 0.90 threshold in the Cryostem cohort."))
```

Which of these selected genes are common with the Cryostem cohort?
```{r, echo = FALSE}
sel_louis <- read.xlsx("../data/RNAseq/donors_and_recip/perm_val_TNT_90_thresh.xlsx", colNames = FALSE)
sel_common <- which(selected_genes_qt %in% sel_louis$X1)
print(paste0(length(sel_common), " genes were common between the two cohorts."))
# selected_genes_qt[sel_common]
```

Graph of these selected genes that were found commonly between the two cohorts:
```{r, echo = FALSE}
correlations <- norm_data[,selected_genes_qt[sel_common]] %>% 
  as.matrix %>%
  cor %>%
  as.data.frame %>%
  rownames_to_column(var = 'var1') %>%
  gather(var2, value, -var1) %>% 
  dplyr::filter(var1 < var2)

correlated_features <- correlations %>% 
  dplyr::filter(value > 0.7)
gr <- graph_from_data_frame(correlated_features, directed = FALSE, vertices=selected_genes_qt[sel_common])
nodes <- get.vertex.attribute(gr)
color_nodes <- rep("#FF000080", length(nodes$name))

medians <- gr_medians[,nodes$name]
for (gene in seq_along(colnames(medians))){
  if (medians["tolerant", gene] == max(medians[, gene])){
    color_nodes[gene] <- "#0000FF80"
  }
}

# recip_path <- "~/Desktop/"
# 
# pdf(paste0(recip_path, "cryo_common_metabolites_recip_05.pdf"), width = 12, height = 8)
set.seed(1)
plot(gr, vertex.size = 7, vertex.label.cex = .5, edge.width = (E(gr)$value - .7) * 20,
     vertex.color = color_nodes, layout = layout_nicely)
# dev.off()

```

```{r, echo = FALSE}
save_genes_annot(norm_data = norm_data,
                 selected_genes = selected_genes_qt[sel_common],
                 path_obj = "../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_TNT_90.RDS",
                 compar_design = "TNT",
                 graph = gr,
                 path_annot_obj = "../data/RNA_seq_national/donors_and_recip/annot_196selected_genes_dr_TNT_90.xlsx")
```

We can see how many of these genes had the same behaviour in the two cohorts (ie, 
were over or underexpressed in the same group of patients)
```{r, echo = FALSE}
dir <- lapply(seq_len(length(sel_common)), function(idx){
  gene <- selected_genes_qt[sel_common[idx]]
  overexp <- rownames(gr_medians)[which.is.max(gr_medians[gene][,1])]
  list(gene, overexp)
})
dir_genes <- do.call(rbind, dir)
write.xlsx(dir_genes, file = "../data/RNA_seq_national/donors_and_recip/perm_val_dir_TNT_90_thresh.xlsx")
dir_genes_cryo <- dir_genes

dir_genes_louis <- read.xlsx("../data/RNAseq/donors_and_recip/perm_val_dir_TNT_90_thresh.xlsx", rowNames = FALSE)
dir_genes_cryo <- as.data.frame(dir_genes_cryo) %>% 
  column_to_rownames("V1")
dir_genes_cryo <- dir_genes_cryo[dir_genes_louis$V1,]
sel_beh <- dir_genes_louis$V1[which(dir_genes_louis$V2 == dir_genes_cryo)]
sel_beh <- as.data.frame(sel_beh)
write.xlsx(sel_beh, file = "../data/RNAseq/donors_and_recip/perm_val_beh_TNT_90_thresh.xlsx")

print(paste0(length(which(dir_genes_louis$V2 == dir_genes_cryo)),
             " genes out of the ", length(sel_common),
             " common selected genes have the same behaviour in both cohorts."))
```

Graph of the 87 genes that had the same behaviour in the two cohorts:
```{r, echo = FALSE}
selb <- as.character(sel_beh$sel_beh)
correlations <- norm_data[,selb] %>% 
  as.matrix %>%
  cor %>%
  as.data.frame %>%
  rownames_to_column(var = 'var1') %>%
  gather(var2, value, -var1) %>% 
  dplyr::filter(var1 < var2)

correlated_features <- correlations %>% 
  dplyr::filter(value > 0.7)
gr <- graph_from_data_frame(correlated_features, directed = FALSE, vertices=selb)
nodes <- get.vertex.attribute(gr)
color_nodes <- rep("#0000FF80", length(nodes$name))

medians <- gr_medians[,nodes$name]
for (gene in seq_along(colnames(medians))){
  if (medians["non_tolerant", gene] == max(medians[, gene])){
    color_nodes[gene] <- "#FF000080"
  }
}

# recip_path <- "~/Desktop/"
# 
# pdf(paste0(recip_path, "cryo_common_metabolites_recip_05.pdf"), width = 12, height = 8)
set.seed(1)
plot(gr, vertex.size = 7, vertex.label.cex = .5, edge.width = (E(gr)$value - .7) * 20,
     vertex.color = color_nodes, layout = layout_nicely)
# dev.off()

```

```{r, echo = FALSE}
save_genes_annot(norm_data = norm_data,
                 selected_genes = selb,
                 path_obj = "../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_beh_TNT_90.RDS",
                 compar_design = "TNT",
                 graph = gr,
                 path_annot_obj = "../data/RNA_seq_national/donors_and_recip/annot_87selected_genes_dr_TNT_90.xlsx")
```

Now that we have selected features that should help us classify primary and 
secondary tolerant recipients, we can see that a RF model built on these selected
features is better than the one built on all genes:
```{r, echo = FALSE}
rna_dr_cryo_TNT_90 <- readRDS("../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_TNT_90.RDS")
set.seed(1)
rf <- ranger::ranger(group~., data = rna_dr_cryo_TNT_90, num.trees = 5000, importance = "impurity")
rf$confusion.matrix
paste0("prediction error: ", rf$prediction.error*100)
```

We can also build a RF only on the 28 genes that had the same behaviour in 
the two cohorts, which doesn't improve the classification:
```{r, echo = FALSE}
rna_dr_cryo_beh_TNT_90 <- readRDS("../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_beh_TNT_90.RDS")
set.seed(1)
rf <- ranger::ranger(group~., data = rna_dr_cryo_beh_TNT_90, num.trees = 5000, importance = "impurity")
rf$confusion.matrix
paste0("prediction error: ", rf$prediction.error*100)
```


### Including age and gender in the analysis 

Of the 196 genes that were selected as informative on tolerance in both cohorts, 
which ones are still high after correcting for age and gender?
```{r, echo = FALSE}
sel_com_names <- selected_genes_qt[sel_common]

qt_demo <- unlist(map(perm_vals, 2))
qt_sel <- qt_demo[which(qt>.90)+1]
names(qt_sel) <- colnames(norm_data)[which(qt>.90)+1]
qt_sel <- qt_sel[sel_com_names]

densityplot(qt_sel)
```

A lot of genes (third "bump" in the density plot) still add information on the 
tolerance even after correction for gender and age:
```{r, echo = FALSE}
names(qt_sel)[which(qt_sel>0.9)]
order_demo_corr <- sel_com_names[order(qt_sel, decreasing=T)]
```

Which of the 196 selected genes were most linked to gender?
```{r, message = FALSE, error = FALSE, echo = FALSE}
auc_g <- lapply(sel_com_names, function(gene_g){
  fit <- nnet::multinom(gender_comp ~ eval(parse(text=gene_g)), data=norm_data,trace=F)
  AUC <- pROC::multiclass.roc(norm_data$gender_comp,as.numeric(fitted(fit)[,1]))$auc[[1]]
})

```

```{r, echo = FALSE}
order_gender <- sel_com_names[order(unlist(auc_g), decreasing = T)]
densityplot(unlist(auc_g))
# sel_com_names[which(auc_g > .65)]
```

Boxplots of these most gender related genes:
```{r, echo = FALSE}
toplot <- norm_data[,c("gender_comp", sel_com_names[which(auc_g > .66)])]
for(i in seq_along(sel_com_names[which(auc_g > .66)])){
  data_tmp <- data.frame(Var = toplot[, i+1],group = toplot[, 1])
  g<- ggplot(data_tmp) + geom_boxplot(aes(x = group, y = Var, col = group)) + 
    geom_jitter(aes(x = group, y = Var, col = group), shape=16, position=position_jitter(0.2)) +
    theme_minimal()
  print(g+ggtitle(colnames(toplot)[i+1]))
}

```

Which of the 196 selected genes were most linked to age?
```{r, echo = FALSE}
# I'll rather try with a linear model:
auc_a <- lapply (sel_com_names, function(gene_a){
  fit <- lm(age_recip ~ eval(parse(text=gene_a)), data=norm_data)
  stat_val <- summary(fit)$sigma^2
})

```

```{r, echo = FALSE}
order_age <- sel_com_names[order(unlist(auc_a))]
densityplot(unlist(auc_a))
# which(auc_a<29000000)
```

We can plot the genes that seemed to be most linked to the age of the recipients:
```{r, echo = FALSE}
toplot <- norm_data[,c("age_recip", "group", sel_com_names[which(auc_a < 28000000)])]
toplot$age_recip <- round(toplot$age_recip/365)
for(i in seq_along(sel_com_names[which(auc_a < 28000000)])){
  data_tmp <- data.frame(Var = toplot[, i+2],age_recip = toplot[, 1], group = toplot[,2])
  g<- ggplot(data_tmp, aes(x = age_recip, y = Var)) + 
    geom_point(aes(x = age_recip, y = Var, col = as.factor(group))) + geom_smooth(method = "loess") +
    theme_minimal()
  print(g+ ggtitle(colnames(toplot)[i+2]))
}
```

### Investigating into metabolites that mainly differ between tolerant and non tolerant couples:

Run the permutation distribution, taking demographic variables into account:
I first add the age and gender compatibility information to the data.
```{r, echo = FALSE}
dr_info <- couple_info %>% 
  mutate(COUPLENUMBER = paste0("couple_", COUPLENUMBER)) %>% 
  dplyr::filter(grepl("R", rownames(couple_info))) 

norm_data <- norm_data_3gr_dr %>% 
  rownames_to_column("COUPLENUMBER") %>% 
  left_join(dr_info %>% dplyr::select(COUPLENUMBER, gender_comp, age_recip),
            by = "COUPLENUMBER") %>% 
  column_to_rownames("COUPLENUMBER")
```

And then compute the permutation distribution for the non vs tolerance:
```{r, eval=FALSE, echo = FALSE}
norm_data <- norm_data %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::filter(group != "non_tolerant") %>% 
  mutate(group = as.factor(as.character(group))) %>% 
  column_to_rownames("Id.Cryostem.R")

perm_vals_PRISM(norm_data = norm_data, PRISM_name = "rnaseq_dr_cryo_PS",
                rds_name = "handle_cryo_rna_dr_PS.rds")

handle3 <- readRDS("handle_cryo_rna_dr_PS.rds")
perm_vals <- qsub_retrieve(handle3)

save(perm_vals, file = "~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/donors_and_recip/perm_vals_PS.RData")
save(norm_data, file= "~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/donors_and_recip/norm_data_PS.RData")
```

Compute the median per group for every gene, to use it later in the graphs:
```{r}
load("~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/donors_and_recip/perm_vals_PS.RData")
load("~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/RNA_seq_national/donors_and_recip/norm_data_PS.RData")

gr_medians <- norm_data %>%
  group_by(group) %>%
  summarize_if(is.numeric, median) %>%
  as.data.frame()
rownames(gr_medians) <- gr_medians$group
```

#### Threshold of 0.95

These genes were selected as informative to classify the non and tolerant 
recipients, with a 0.95 threshold:
```{r, echo = FALSE}
qt <- unlist(map(perm_vals, 1))
densityplot(qt)
selected_genes_qt <- colnames(norm_data)[which(qt>.95)+1]
write.xlsx(selected_genes_qt, file = "../data/RNA_seq_national/donors_and_recip/perm_val_PS_95_thresh.xlsx")
print(paste0(length(selected_genes_qt),
             " genes were above the 0.90 threshold in the Cryostem cohort."))
```

Which of these selected genes are common with the Cryostem cohort?
```{r, echo = FALSE}
sel_louis <- read.xlsx("../data/RNAseq/donors_and_recip/perm_val_PS_95_thresh.xlsx", colNames = FALSE)
sel_common <- which(selected_genes_qt %in% sel_louis$X1)
print(paste0(length(sel_common), " genes were common between the two cohorts."))
# selected_genes_qt[sel_common]
```

Graph of these selected genes that were found commonly between the two cohorts:
```{r, echo = FALSE}
correlations <- norm_data[,selected_genes_qt[sel_common]] %>% 
  as.matrix %>%
  cor %>%
  as.data.frame %>%
  rownames_to_column(var = 'var1') %>%
  gather(var2, value, -var1) %>% 
  dplyr::filter(var1 < var2)

correlated_features <- correlations %>% 
  dplyr::filter(value > 0.7)
gr <- graph_from_data_frame(correlated_features, directed = FALSE, vertices=selected_genes_qt[sel_common])
nodes <- get.vertex.attribute(gr)
color_nodes <- rep("#0000FF80", length(nodes$name))

medians <- gr_medians[,nodes$name]
for (gene in seq_along(colnames(medians))){
  if (medians["primary_tolerant", gene] == max(medians[, gene])){
    color_nodes[gene] <- "#00FF0080"
  }
}

# recip_path <- "~/Desktop/"
# 
# pdf(paste0(recip_path, "cryo_common_metabolites_recip_05.pdf"), width = 12, height = 8)
set.seed(1)
plot(gr, vertex.size = 7, vertex.label.cex = .5, edge.width = (E(gr)$value - .7) * 20,
     vertex.color = color_nodes, layout = layout_nicely)
# dev.off()

```

```{r}
colData_orig <- colData
```


```{r}
# rownames(norm_data) are couple numbers -> I change the rownames of colData 
# to couples as well so that I can keep using my functions
colData <- colData_orig
colData <- colData %>% 
  dplyr::filter(grepl("R", rownames(colData))) %>% 
  mutate(COUPLE = paste0("couple_", COUPLENUMBER))
rownames(colData) <- colData$COUPLE
```

```{r, echo = FALSE}
save_genes_annot(norm_data = norm_data,
                 selected_genes = selected_genes_qt[sel_common],
                 path_obj = "../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_PS_95.RDS",
                 compar_design = "PS",
                 graph = gr,
                 path_annot_obj = "../data/RNA_seq_national/donors_and_recip/annot_111selected_genes_dr_PS_95.xlsx")
```

We can see how many of these genes had the same behaviour in the two cohorts (ie, 
were over or underexpressed in the same group of patients)
```{r, echo = FALSE}
dir <- lapply(seq_len(length(sel_common)), function(idx){
  gene <- selected_genes_qt[sel_common[idx]]
  overexp <- rownames(gr_medians)[which.is.max(gr_medians[gene][,1])]
  list(gene, overexp)
})
dir_genes <- do.call(rbind, dir)
write.xlsx(dir_genes, file = "../data/RNA_seq_national/donors_and_recip/perm_val_dir_PS_95_thresh.xlsx")
dir_genes_cryo <- dir_genes

dir_genes_louis <- read.xlsx("../data/RNAseq/donors_and_recip/perm_val_dir_PS_95_thresh.xlsx", rowNames = FALSE)
dir_genes_cryo <- as.data.frame(dir_genes_cryo) %>% 
  column_to_rownames("V1")
dir_genes_cryo <- dir_genes_cryo[dir_genes_louis$V1,]
sel_beh <- dir_genes_louis$V1[which(dir_genes_louis$V2 == dir_genes_cryo)]
sel_beh <- as.data.frame(sel_beh)
write.xlsx(sel_beh, file = "../data/RNAseq/donors_and_recip/perm_val_beh_PS_95_thresh.xlsx")

print(paste0(length(which(dir_genes_louis$V2 == dir_genes_cryo)),
             " genes out of the ", length(sel_common),
             " common selected genes have the same behaviour in both cohorts."))
```

Graph of the 6 genes that had the same behaviour in the two cohorts:
```{r, echo = FALSE}
selb <- as.character(sel_beh$sel_beh)
correlations <- norm_data[,selb] %>% 
  as.matrix %>%
  cor %>%
  as.data.frame %>%
  rownames_to_column(var = 'var1') %>%
  gather(var2, value, -var1) %>% 
  dplyr::filter(var1 < var2)

correlated_features <- correlations %>% 
  dplyr::filter(value > 0.7)
gr <- graph_from_data_frame(correlated_features, directed = FALSE, vertices=selb)
nodes <- get.vertex.attribute(gr)
color_nodes <- rep("#0000FF80", length(nodes$name))

medians <- gr_medians[,nodes$name]
for (gene in seq_along(colnames(medians))){
  if (medians["primary_tolerant", gene] == max(medians[, gene])){
    color_nodes[gene] <- "#00FF0080"
  }
}

# recip_path <- "~/Desktop/"
# 
# pdf(paste0(recip_path, "cryo_common_metabolites_recip_05.pdf"), width = 12, height = 8)
set.seed(1)
plot(gr, vertex.size = 7, vertex.label.cex = .5, edge.width = (E(gr)$value - .7) * 20,
     vertex.color = color_nodes, layout = layout_nicely)
# dev.off()

```

```{r, echo = FALSE}
save_genes_annot(norm_data = norm_data,
                 selected_genes = selb,
                 path_obj = "../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_beh_PS_95.RDS",
                 compar_design = "PS",
                 graph = gr,
                 path_annot_obj = "../data/RNA_seq_national/donors_and_recip/annot_6selected_genes_dr_PS_95.xlsx")
```


Now that we have selected features that should help us classify primary and 
secondary tolerant couples, we can see that a RF model built on these selected
features is better than the one built on all genes:
```{r, echo = FALSE}
rna_dr_cryo_PS_95 <- readRDS("../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_PS_95.RDS")
set.seed(1)
rf <- ranger::ranger(group~., data = rna_dr_cryo_PS_95, num.trees = 5000, importance = "impurity")
rf$confusion.matrix
paste0("prediction error: ", rf$prediction.error*100)
```

We can also build a RF only on the 6 genes that had the same behaviour in both
cohorts, but this doesn't improve the results:
```{r}
rna_dr_cryo_beh_PS_95 <- readRDS("../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_beh_PS_95.RDS")
set.seed(1)
rf <- ranger::ranger(group~., data = rna_dr_cryo_beh_PS_95, num.trees = 5000, importance = "impurity")
rf$confusion.matrix
paste0("prediction error: ", rf$prediction.error*100)
```

#### Threshold of 0.90

These genes were selected as informative to classify the non and tolerant 
recipients, with a 0.90 threshold:
```{r, echo = FALSE}
qt <- unlist(map(perm_vals, 1))
densityplot(qt)
selected_genes_qt <- colnames(norm_data)[which(qt>.90)+1]
write.xlsx(selected_genes_qt, file = "../data/RNA_seq_national/donors_and_recip/perm_val_PS_90_thresh.xlsx")
print(paste0(length(selected_genes_qt),
             " genes were above the 0.90 threshold in the Cryostem cohort."))
```

Which of these selected genes are common with the Cryostem cohort?
```{r, echo = FALSE}
sel_louis <- read.xlsx("../data/RNAseq/donors_and_recip/perm_val_PS_90_thresh.xlsx", colNames = FALSE)
sel_common <- which(selected_genes_qt %in% sel_louis$X1)
print(paste0(length(sel_common), " genes were common between the two cohorts."))
# selected_genes_qt[sel_common]
```

Graph of these selected genes that were found commonly between the two cohorts:
```{r, echo = FALSE}
correlations <- norm_data[,selected_genes_qt[sel_common]] %>% 
  as.matrix %>%
  cor %>%
  as.data.frame %>%
  rownames_to_column(var = 'var1') %>%
  gather(var2, value, -var1) %>% 
  dplyr::filter(var1 < var2)

correlated_features <- correlations %>% 
  dplyr::filter(value > 0.7)
gr <- graph_from_data_frame(correlated_features, directed = FALSE, vertices=selected_genes_qt[sel_common])
nodes <- get.vertex.attribute(gr)
color_nodes <- rep("#0000FF80", length(nodes$name))

medians <- gr_medians[,nodes$name]
for (gene in seq_along(colnames(medians))){
  if (medians["primary_tolerant", gene] == max(medians[, gene])){
    color_nodes[gene] <- "#00FF0080"
  }
}

# recip_path <- "~/Desktop/"
# 
# pdf(paste0(recip_path, "cryo_common_metabolites_recip_05.pdf"), width = 12, height = 8)
set.seed(1)
plot(gr, vertex.size = 7, vertex.label.cex = .5, edge.width = (E(gr)$value - .7) * 20,
     vertex.color = color_nodes, layout = layout_nicely)
# dev.off()

```

```{r, echo = FALSE}
save_genes_annot(norm_data = norm_data,
                 selected_genes = selected_genes_qt[sel_common],
                 path_obj = "../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_PS_90.RDS",
                 compar_design = "PS",
                 graph = gr,
                 path_annot_obj = "../data/RNA_seq_national/donors_and_recip/annot_337selected_genes_dr_PS_90.xlsx")
```

We can see how many of these genes had the same behaviour in the two cohorts (ie, 
were over or underexpressed in the same group of patients)
```{r, echo = FALSE}
dir <- lapply(seq_len(length(sel_common)), function(idx){
  gene <- selected_genes_qt[sel_common[idx]]
  overexp <- rownames(gr_medians)[which.is.max(gr_medians[gene][,1])]
  list(gene, overexp)
})
dir_genes <- do.call(rbind, dir)
write.xlsx(dir_genes, file = "../data/RNA_seq_national/donors_and_recip/perm_val_dir_PS_90_thresh.xlsx")
dir_genes_cryo <- dir_genes

dir_genes_louis <- read.xlsx("../data/RNAseq/donors_and_recip/perm_val_dir_PS_90_thresh.xlsx", rowNames = FALSE)
dir_genes_cryo <- as.data.frame(dir_genes_cryo) %>% 
  column_to_rownames("V1")
dir_genes_cryo <- dir_genes_cryo[dir_genes_louis$V1,]
sel_beh <- dir_genes_louis$V1[which(dir_genes_louis$V2 == dir_genes_cryo)]
sel_beh <- as.data.frame(sel_beh)
write.xlsx(sel_beh, file = "../data/RNAseq/donors_and_recip/perm_val_beh_PS_90_thresh.xlsx")

print(paste0(length(which(dir_genes_louis$V2 == dir_genes_cryo)),
             " genes out of the ", length(sel_common),
             " common selected genes have the same behaviour in both cohorts."))
```

Graph of the 15 genes that had the same behaviour in the two cohorts:
```{r, echo = FALSE}
selb <- as.character(sel_beh$sel_beh)
correlations <- norm_data[,selb] %>% 
  as.matrix %>%
  cor %>%
  as.data.frame %>%
  rownames_to_column(var = 'var1') %>%
  gather(var2, value, -var1) %>% 
  dplyr::filter(var1 < var2)

correlated_features <- correlations %>% 
  dplyr::filter(value > 0.7)
gr <- graph_from_data_frame(correlated_features, directed = FALSE, vertices=selb)
nodes <- get.vertex.attribute(gr)
color_nodes <- rep("#0000FF80", length(nodes$name))

medians <- gr_medians[,nodes$name]
for (gene in seq_along(colnames(medians))){
  if (medians["primary_tolerant", gene] == max(medians[, gene])){
    color_nodes[gene] <- "#00FF0080"
  }
}

# recip_path <- "~/Desktop/"
# 
# pdf(paste0(recip_path, "cryo_common_metabolites_recip_05.pdf"), width = 12, height = 8)
set.seed(1)
plot(gr, vertex.size = 7, vertex.label.cex = .5, edge.width = (E(gr)$value - .7) * 20,
     vertex.color = color_nodes, layout = layout_nicely)
# dev.off()

```

```{r, echo = FALSE}
save_genes_annot(norm_data = norm_data,
                 selected_genes = selb,
                 path_obj = "../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_beh_PS_90.RDS",
                 compar_design = "PS",
                 graph = gr,
                 path_annot_obj = "../data/RNA_seq_national/donors_and_recip/annot_15selected_genes_dr_PS_90.xlsx")
```


Now that we have selected features that should help us classify primary and 
secondary tolerant couples, we can see that a RF model built on these selected
features is better than the one built on all genes:
```{r, echo = FALSE}
rna_dr_cryo_PS_90 <- readRDS("../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_PS_90.RDS")
set.seed(1)
rf <- ranger::ranger(group~., data = rna_dr_cryo_PS_90, num.trees = 5000, importance = "impurity")
rf$confusion.matrix
paste0("prediction error: ", rf$prediction.error*100)
```

We can also build a RF only on the 15 genes that had the same behaviour in both
cohorts, but this doesn't improve the results:
```{r}
rna_dr_cryo_beh_PS_90 <- readRDS("../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_beh_PS_90.RDS")
set.seed(1)
rf <- ranger::ranger(group~., data = rna_dr_cryo_beh_PS_90, num.trees = 5000, importance = "impurity")
rf$confusion.matrix
paste0("prediction error: ", rf$prediction.error*100)
```

### Including age and gender in the analysis 

Of the 337 genes that were selected as informative on tolerance in both cohorts, 
which ones are still high after correcting for age and gender?
```{r, echo = FALSE}
sel_com_names <- selected_genes_qt[sel_common]

qt_demo <- unlist(map(perm_vals, 2))
qt_sel <- qt_demo[which(qt>.90)+1]
names(qt_sel) <- colnames(norm_data)[which(qt>.90)+1]
qt_sel <- qt_sel[sel_com_names]

densityplot(qt_sel)
```

A lot of genes (third "bump" in the density plot) still add information on the 
tolerance even after correction for gender and age:
```{r}
names(qt_sel)[which(qt_sel>0.90)]
order_demo_corr <- sel_com_names[order(qt_sel, decreasing=T)]
```

Which of the 337 selected genes were most linked to gender?
```{r, message = FALSE, error = FALSE, echo = FALSE}
auc_g <- lapply(sel_com_names, function(gene_g){
  fit <- nnet::multinom(gender_comp ~ eval(parse(text=gene_g)), data=norm_data,trace=F)
  AUC <- pROC::multiclass.roc(norm_data$gender_comp,as.numeric(fitted(fit)[,1]))$auc[[1]]
})

```

```{r, echo = FALSE}
order_gender <- sel_com_names[order(unlist(auc_g), decreasing = T)]
densityplot(unlist(auc_g))
# sel_com_names[which(auc_g > .68)]
```

Boxplots of these most gender related genes:
```{r, echo = FALSE}
toplot <- norm_data[,c("gender_comp", sel_com_names[which(auc_g > .69)])]
for(i in seq_along(sel_com_names[which(auc_g > .69)])){
  data_tmp <- data.frame(Var = toplot[, i+1],group = toplot[, 1])
  g<- ggplot(data_tmp) + geom_boxplot(aes(x = group, y = Var, col = group)) + 
    geom_jitter(aes(x = group, y = Var, col = group), shape=16, position=position_jitter(0.2)) +
    theme_minimal()
  print(g+ggtitle(colnames(toplot)[i+1]))
}

```

Which of the 337 selected genes were most linked to age?
```{r, echo = FALSE}
# I'll rather try with a linear model:
auc_a <- lapply (sel_com_names, function(gene_a){
  fit <- lm(age_recip ~ eval(parse(text=gene_a)), data=norm_data)
  stat_val <- summary(fit)$sigma^2
})

```

```{r, echo = FALSE}
order_age <- sel_com_names[order(unlist(auc_a))]
densityplot(unlist(auc_a))
# which(auc_a<31000000)
```

We can plot the genes that seemed to be most linked to the age of the recipients:
```{r, echo = FALSE}
toplot <- norm_data[,c("age_recip", "group", sel_com_names[which(auc_a < 31000000)])]
toplot$age_recip <- round(toplot$age_recip/365)
for(i in seq_along(sel_com_names[which(auc_a < 31000000)])){
  data_tmp <- data.frame(Var = toplot[, i+2],age_recip = toplot[, 1], group = toplot[,2])
  g<- ggplot(data_tmp, aes(x = age_recip, y = Var)) + 
    geom_point(aes(x = age_recip, y = Var, col = as.factor(group))) + geom_smooth(method = "loess") +
    theme_minimal()
  print(g+ ggtitle(colnames(toplot)[i+2]))
}
```





# Combining the information on couples, recipients and donors:

Now that we have selected features at the couple level, at the recipient level 
and at the donor level, we can try to combine these features:

## Classifying non versus tolerant patients based on all the information:

#### Threshold of 0.95

We first combine the features that were selected as highly informative to 
classify between tolerant and non tolerant patients from the couples, recipients 
and donors into one big table:
```{r, echo = FALSE}
# couple info:
load("../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_TNT_95.RData")
# # add "couple_" in front of the feature names 
# colnames2keep <- which(colnames(pctgs_sel_ft_couples_90) %in% c("group", "COUPLENUMBER"))
# colnames(pctgs_sel_ft_couples_90)[-colnames2keep] <-
#   gsub("X", "couple_", colnames(pctgs_sel_ft_couples_90)[-colnames2keep])
# combine feature info with couple info about gender_comp...
pctgs_couples <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::select("Id.Cryostem.R", "COUPLENUMBER", "gender_comp", "age_recip") %>% 
  dplyr::filter(grepl("R", Id.Cryostem.R)) %>% 
  inner_join(rna_dr_cryo_TNT_95 %>% 
               rownames_to_column("COUPLENUMBER") %>% 
               mutate(COUPLENUMBER = gsub("couple_","", COUPLENUMBER)), by = "COUPLENUMBER")

# recip info:
load("../data/RNA_seq_national/recip_only/rna_recip_cryo_TNT_95.RData")
colnames2keep <- which(colnames(rna_recip_cryo_TNT_95) %in% c("group"))
colnames(rna_recip_cryo_TNT_95)[-colnames2keep] <-
  paste0("R_", colnames(rna_recip_cryo_TNT_95)[-colnames2keep])
pctgs_recip <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::select("Id.Cryostem.R", "COUPLENUMBER") %>% 
  right_join(rna_recip_cryo_TNT_95 %>%
               rownames_to_column("Id.Cryostem.R"), by = "Id.Cryostem.R")

# donor info :
load("../data//RNA_seq_national/donors_only/rna_donors_Cryo_TNT_95.RData")
colnames2keep <- which(colnames(rna_donors_cryo_TNT_95) %in% c("group"))
colnames(rna_donors_cryo_TNT_95)[-colnames2keep] <-
  paste0("D_", colnames(rna_donors_cryo_TNT_95)[-colnames2keep])
pctgs_donor <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::select("Id.Cryostem.R", "COUPLENUMBER") %>% 
  right_join(rna_donors_cryo_TNT_95 %>%
               rownames_to_column("Id.Cryostem.R"), by = "Id.Cryostem.R") 

# combine :
pctgs_all <- pctgs_couples %>% 
  left_join(pctgs_recip, by = c("Id.Cryostem.R", "COUPLENUMBER", "group")) %>% 
  left_join(pctgs_donor, by = c("COUPLENUMBER", "group"))
```

We can generate a PCA and a tSNE using all these features to see how the tolerant 
and non tolerant patients are represented in the reduced dimensions:
```{r, echo = FALSE}
pctgs_all_only <- pctgs_all %>% 
  column_to_rownames("Id.Cryostem.R.x") %>% 
  select_if(is.numeric) %>% 
  dplyr::select(-"age_recip")

pca_all <- prcomp(as.matrix(pctgs_all_only))
```

The 2 first components of the PCA seem to hold quite a lot of the variability in the data:
```{r}
plot(pca_all)
```

```{r, echo = FALSE}
pca_2plot <- as.data.frame(pca_all$x) %>% 
  mutate(Id.Cryostem.R.x = rownames(pca_all$x)) %>% 
  left_join(pctgs_all, by = "Id.Cryostem.R.x")
ggplot(pca_2plot, aes(x = PC1, y = PC2, col = as.factor(group), label = Id.Cryostem.R.x)) + 
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R.x),hjust=0, vjust=0)
```
The tolerant patients seem to have higher PC1 values on the PCA (they are 
situated more on the right), even though the two groups are quite mixed, which 
suggests that it might still be difficult to classify the patients into
tolerant and non tolerant groups based on the features we selected.

We can also generate a tSNE on the same data:
```{r, echo = FALSE}
tsne_all <- Rtsne::Rtsne(pctgs_all_only, perplexity = 10)
tsne_2plot <- as.data.frame(tsne_all$Y) %>% 
  magrittr::set_colnames(c("X1", "X2")) %>% 
  mutate(Id.Cryostem.R.x = rownames(pctgs_all_only)) %>% 
  left_join(pctgs_all, by = "Id.Cryostem.R.x")

ggplot(tsne_2plot, aes(x = X1, y = X2, col = as.factor(group), 
                       label = Id.Cryostem.R.x)) + 
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R.x),hjust=0, vjust=0)
```


We can see if building a model using all these features combined leads to good 
classification results between tolerant and non tolerant patients:
```{r, echo = FALSE}
pctgs_all_RF <- pctgs_all %>% 
  dplyr::select(-"COUPLENUMBER", -"gender_comp", -"age_recip") %>% 
  mutate(group = as.factor(group)) %>% 
  column_to_rownames("Id.Cryostem.R.x")

set.seed(1)
nb_features = ncol(pctgs_all_RF)-1

rf <- rf_tol_non_tol(df_orig = pctgs_all_RF,
                      mtry = round(sqrt(nb_features)), ntree = 1000, package_2use = "ranger")
rf$confusion.matrix
paste0("prediction error: ", rf$prediction.error*100)
```

The classification error is quite high, but we can still see which features were 
selected in the model to classify tolerant from non tolerant patients in the 
Cryostem cohort:
```{r, echo = FALSE}
sort(rf$variable.importance, decreasing = TRUE)[1:20]
```

The variables that were mainly selected as informative in the model are the ones
from recipients and couples.

We can visualise the top variables in a heatmap:
In the first heatmap, the patients (in rows) are ordered by group and the 20 top genes
(in columns) are ordered by importance in the random forest model.
In the second heatmap, the patients and 20 top genes are clustered hierarchically:
the patients that are most similar are closer to each other, same for the genes.
```{r, echo = FALSE}
mat2plot <- pctgs_all_RF[,c("group", names(sort(rf$variable.importance, decreasing = TRUE))[1:20])] %>% 
  rownames_to_column("rownames") %>% 
  arrange(group) %>% 
  column_to_rownames("rownames")

annot_patients <- as.data.frame(mat2plot$group)
rownames(annot_patients) <- rownames(mat2plot)
colnames(annot_patients) <- "group"
ann_colors <- list(group = c(non_tolerant = "indianred1", 
                             tolerant = "royalblue1"))

pheatmap::pheatmap(mat2plot[,-which(colnames(mat2plot)=="group")], 
                   annotation_row = annot_patients,
                   annotation_colors = ann_colors[1],
                   cluster_rows = FALSE,
                   cluster_cols = FALSE,
                   main = "Heatmap with patients ordered by group",
                   scale = "column")

pheatmap::pheatmap(mat2plot[,-which(colnames(mat2plot)=="group")], 
                   annotation_row = annot_patients,
                   annotation_colors = ann_colors[1],
                   cluster_rows = TRUE,
                   main = "Heatmap with patients clustered",
                   scale = "column")
```

Reminder: high values for the variables that start with "subst_" mean that the
difference of expression between the donor and the recipient of the same couple 
was high for this specific feature.

#### Threshold of 0.90

We first combine the features that were selected as highly informative to 
classify between tolerant and non tolerant patients from the couples, recipients 
and donors into one big table:
```{r, echo = FALSE}
# couple info:
load("../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_TNT_90.RData")
# # add "couple_" in front of the feature names 
# colnames2keep <- which(colnames(pctgs_sel_ft_couples_90) %in% c("group", "COUPLENUMBER"))
# colnames(pctgs_sel_ft_couples_90)[-colnames2keep] <-
#   gsub("X", "couple_", colnames(pctgs_sel_ft_couples_90)[-colnames2keep])
# combine feature info with couple info about gender_comp...
pctgs_couples <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::select("Id.Cryostem.R", "COUPLENUMBER", "gender_comp", "age_recip") %>% 
  dplyr::filter(grepl("R", Id.Cryostem.R)) %>% 
  inner_join(rna_dr_cryo_TNT_90 %>% 
               rownames_to_column("COUPLENUMBER") %>% 
               mutate(COUPLENUMBER = gsub("couple_","", COUPLENUMBER)), by = "COUPLENUMBER")

# recip info:
load("../data/RNA_seq_national/recip_only/rna_recip_cryo_TNT_90.RData")
colnames2keep <- which(colnames(rna_recip_cryo_TNT_90) %in% c("group"))
colnames(rna_recip_cryo_TNT_90)[-colnames2keep] <-
  paste0("R_", colnames(rna_recip_cryo_TNT_90)[-colnames2keep])
pctgs_recip <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::select("Id.Cryostem.R", "COUPLENUMBER") %>% 
  right_join(rna_recip_cryo_TNT_90 %>%
               rownames_to_column("Id.Cryostem.R"), by = "Id.Cryostem.R")

# donor info :
load("../data//RNA_seq_national/donors_only/rna_donors_Cryo_TNT_90.RData")
colnames2keep <- which(colnames(rna_donors_cryo_TNT_90) %in% c("group"))
colnames(rna_donors_cryo_TNT_90)[-colnames2keep] <-
  paste0("D_", colnames(rna_donors_cryo_TNT_90)[-colnames2keep])
pctgs_donor <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::select("Id.Cryostem.R", "COUPLENUMBER") %>% 
  right_join(rna_donors_cryo_TNT_90 %>%
               rownames_to_column("Id.Cryostem.R"), by = "Id.Cryostem.R") 

# combine :
pctgs_all <- pctgs_couples %>% 
  left_join(pctgs_recip, by = c("Id.Cryostem.R", "COUPLENUMBER", "group")) %>% 
  left_join(pctgs_donor, by = c("COUPLENUMBER", "group"))
```

We can generate a PCA and a tSNE using all these features to see how the tolerant 
and non tolerant patients are represented in the reduced dimensions:
```{r, echo = FALSE}
pctgs_all_only <- pctgs_all %>% 
  column_to_rownames("Id.Cryostem.R.x") %>% 
  select_if(is.numeric) %>% 
  dplyr::select(-"age_recip")

pca_all <- prcomp(as.matrix(pctgs_all_only))
```

The 2 first components of the PCA seem to hold quite a lot of the variability in the data:
```{r}
plot(pca_all)
```

```{r, echo = FALSE}
pca_2plot <- as.data.frame(pca_all$x) %>% 
  mutate(Id.Cryostem.R.x = rownames(pca_all$x)) %>% 
  left_join(pctgs_all, by = "Id.Cryostem.R.x")
ggplot(pca_2plot, aes(x = PC1, y = PC2, col = as.factor(group), label = Id.Cryostem.R.x)) + 
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R.x),hjust=0, vjust=0)
```
The tolerant patients seem to have higher PC1 values on the PCA (they are 
situated more on the right), even though the two groups are quite mixed, which 
suggests that it might still be difficult to classify the patients into
tolerant and non tolerant groups based on the features we selected.

We can also generate a tSNE on the same data:
```{r, echo = FALSE}
tsne_all <- Rtsne::Rtsne(pctgs_all_only, perplexity = 10)
tsne_2plot <- as.data.frame(tsne_all$Y) %>% 
  magrittr::set_colnames(c("X1", "X2")) %>% 
  mutate(Id.Cryostem.R.x = rownames(pctgs_all_only)) %>% 
  left_join(pctgs_all, by = "Id.Cryostem.R.x")

ggplot(tsne_2plot, aes(x = X1, y = X2, col = as.factor(group), 
                       label = Id.Cryostem.R.x)) + 
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R.x),hjust=0, vjust=0)
```


We can see if building a model using all these features combined leads to good 
classification results between tolerant and non tolerant patients:
```{r, echo = FALSE}
pctgs_all_RF <- pctgs_all %>% 
  dplyr::select(-"COUPLENUMBER", -"gender_comp", -"age_recip") %>% 
  mutate(group = as.factor(group)) %>% 
  column_to_rownames("Id.Cryostem.R.x")

set.seed(1)
nb_features = ncol(pctgs_all_RF)-1

rf <- rf_tol_non_tol(df_orig = pctgs_all_RF,
                      mtry = round(sqrt(nb_features)), ntree = 1000, package_2use = "ranger")
rf$confusion.matrix
paste0("prediction error: ", rf$prediction.error*100)
```

The classification error is quite high, but we can still see which features were 
selected in the model to classify tolerant from non tolerant patients in the 
Cryostem cohort:
```{r, echo = FALSE}
sort(rf$variable.importance, decreasing = TRUE)[1:20]
```

The variables that were mainly selected as informative in the model are the ones
from recipients and couples.

We can visualise the top variables in a heatmap:
In the first heatmap, the patients (in rows) are ordered by group and the 20 top genes
(in columns) are ordered by importance in the random forest model.
In the second heatmap, the patients and 20 top genes are clustered hierarchically:
the patients that are most similar are closer to each other, same for the genes.
```{r, echo = FALSE}
mat2plot <- pctgs_all_RF[,c("group", names(sort(rf$variable.importance, decreasing = TRUE))[1:20])] %>% 
  rownames_to_column("rownames") %>% 
  arrange(group) %>% 
  column_to_rownames("rownames")

annot_patients <- as.data.frame(mat2plot$group)
rownames(annot_patients) <- rownames(mat2plot)
colnames(annot_patients) <- "group"
ann_colors <- list(group = c(non_tolerant = "indianred1", 
                             tolerant = "royalblue1"))

pheatmap::pheatmap(mat2plot[,-which(colnames(mat2plot)=="group")], 
                   annotation_row = annot_patients,
                   annotation_colors = ann_colors[1],
                   cluster_rows = FALSE,
                   cluster_cols = FALSE,
                   main = "Heatmap with patients ordered by group",
                   scale = "column")

pheatmap::pheatmap(mat2plot[,-which(colnames(mat2plot)=="group")], 
                   annotation_row = annot_patients,
                   annotation_colors = ann_colors[1],
                   cluster_rows = TRUE,
                   main = "Heatmap with patients clustered",
                   scale = "column")
```

Reminder: high values for the variables that start with "subst_" mean that the
difference of expression between the donor and the recipient of the same couple 
was high for this specific feature.

#### Focusing only on the genes that had a similar behaviour in the two cohorts:

We first combine the features that were selected as highly informative to 
classify between tolerant and non tolerant patients from the couples, recipients 
and donors into one big table:
```{r, echo = FALSE}
# couple info:
rna_dr_cryo_TNT_90 <- readRDS("../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_beh_TNT_90.RDS")
pctgs_couples <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::select("Id.Cryostem.R", "COUPLENUMBER", "gender_comp", "age_recip") %>% 
  dplyr::filter(grepl("R", Id.Cryostem.R)) %>% 
  inner_join(rna_dr_cryo_TNT_90 %>% 
               rownames_to_column("COUPLENUMBER") %>% 
               mutate(COUPLENUMBER = gsub("couple_","", COUPLENUMBER)), by = "COUPLENUMBER")

# recip info:
load("../data/RNA_seq_national/recip_only/rna_recip_cryo_beh_TNT_90.RData")
colnames2keep <- which(colnames(rna_recip_cryo_beh_TNT_90) %in% c("group"))
colnames(rna_recip_cryo_beh_TNT_90)[-colnames2keep] <-
  paste0("R_", colnames(rna_recip_cryo_beh_TNT_90)[-colnames2keep])
pctgs_recip <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::select("Id.Cryostem.R", "COUPLENUMBER") %>% 
  right_join(rna_recip_cryo_beh_TNT_90 %>%
               rownames_to_column("Id.Cryostem.R"), by = "Id.Cryostem.R")

# donor info :
load("../data//RNA_seq_national/donors_only/rna_donor_cryo_beh_TNT_90.RData")
colnames2keep <- which(colnames(rna_donor_cryo_beh_TNT_90) %in% c("group"))
colnames(rna_donor_cryo_beh_TNT_90)[-colnames2keep] <-
  paste0("D_", colnames(rna_donor_cryo_beh_TNT_90)[-colnames2keep])
pctgs_donor <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::select("Id.Cryostem.R", "COUPLENUMBER") %>% 
  right_join(rna_donor_cryo_beh_TNT_90 %>%
               rownames_to_column("Id.Cryostem.R"), by = "Id.Cryostem.R") 

# combine :
pctgs_all <- pctgs_couples %>% 
  left_join(pctgs_recip, by = c("Id.Cryostem.R", "COUPLENUMBER", "group")) %>% 
  left_join(pctgs_donor, by = c("COUPLENUMBER", "group"))
```

We can generate a PCA and a tSNE using all these features to see how the tolerant 
and non tolerant patients are represented in the reduced dimensions:
```{r, echo = FALSE}
pctgs_all_only <- pctgs_all %>% 
  column_to_rownames("Id.Cryostem.R.x") %>% 
  select_if(is.numeric) %>% 
  dplyr::select(-"age_recip")

pca_all <- prcomp(as.matrix(pctgs_all_only))
```

The 2 first components of the PCA seem to hold quite a lot of the variability in the data:
```{r}
plot(pca_all)
```

```{r, echo = FALSE}
pca_2plot <- as.data.frame(pca_all$x) %>% 
  mutate(Id.Cryostem.R.x = rownames(pca_all$x)) %>% 
  left_join(pctgs_all, by = "Id.Cryostem.R.x")
ggplot(pca_2plot, aes(x = PC1, y = PC2, col = as.factor(group), label = Id.Cryostem.R.x)) + 
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R.x),hjust=0, vjust=0)
```
The tolerant patients seem to have higher PC1 values on the PCA (they are 
situated more on the right), even though the two groups are quite mixed, which 
suggests that it might still be difficult to classify the patients into
tolerant and non tolerant groups based on the features we selected.

We can also generate a tSNE on the same data:
```{r, echo = FALSE}
tsne_all <- Rtsne::Rtsne(pctgs_all_only, perplexity = 10)
tsne_2plot <- as.data.frame(tsne_all$Y) %>% 
  magrittr::set_colnames(c("X1", "X2")) %>% 
  mutate(Id.Cryostem.R.x = rownames(pctgs_all_only)) %>% 
  left_join(pctgs_all, by = "Id.Cryostem.R.x")

ggplot(tsne_2plot, aes(x = X1, y = X2, col = as.factor(group), 
                       label = Id.Cryostem.R.x)) + 
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R.x),hjust=0, vjust=0)
```


We can see if building a model using all these features combined leads to good 
classification results between tolerant and non tolerant patients:
```{r, echo = FALSE}
pctgs_all_RF <- pctgs_all %>% 
  dplyr::select(-"COUPLENUMBER", -"gender_comp", -"age_recip") %>% 
  mutate(group = as.factor(group)) %>% 
  column_to_rownames("Id.Cryostem.R.x")

set.seed(1)
nb_features = ncol(pctgs_all_RF)-1

rf <- rf_tol_non_tol(df_orig = pctgs_all_RF,
                      mtry = round(sqrt(nb_features)), ntree = 1000, package_2use = "ranger")
rf$confusion.matrix
paste0("prediction error: ", rf$prediction.error*100)
```

The classification error is quite high, but we can still see which features were 
selected in the model to classify tolerant from non tolerant patients in the 
Cryostem cohort:
```{r, echo = FALSE}
sort(rf$variable.importance, decreasing = TRUE)[1:20]
```

The variables that were mainly selected as informative in the model are the ones
from recipients and couples.

We can visualise the top variables in a heatmap:
In the first heatmap, the patients (in rows) are ordered by group and the 20 top genes
(in columns) are ordered by importance in the random forest model.
In the second heatmap, the patients and 20 top genes are clustered hierarchically:
the patients that are most similar are closer to each other, same for the genes.
```{r, echo = FALSE}
mat2plot <- pctgs_all_RF[,c("group", names(sort(rf$variable.importance, decreasing = TRUE))[1:20])] %>% 
  rownames_to_column("rownames") %>% 
  arrange(group) %>% 
  column_to_rownames("rownames")

annot_patients <- as.data.frame(mat2plot$group)
rownames(annot_patients) <- rownames(mat2plot)
colnames(annot_patients) <- "group"
ann_colors <- list(group = c(non_tolerant = "indianred1", 
                             tolerant = "royalblue1"))

pheatmap::pheatmap(mat2plot[,-which(colnames(mat2plot)=="group")], 
                   annotation_row = annot_patients,
                   annotation_colors = ann_colors[1],
                   cluster_rows = FALSE,
                   cluster_cols = FALSE,
                   main = "Heatmap with patients ordered by group",
                   scale = "column")

pheatmap::pheatmap(mat2plot[,-which(colnames(mat2plot)=="group")], 
                   annotation_row = annot_patients,
                   annotation_colors = ann_colors[1],
                   cluster_rows = TRUE,
                   main = "Heatmap with patients clustered",
                   scale = "column")
```

Reminder: high values for the variables that start with "subst_" mean that the
difference of expression between the donor and the recipient of the same couple 
was high for this specific feature.


## Classifying primary versus secondary tolerant patients based on all the information:

#### Threshold of 0.95

We first combine the features that were selected as highly informative to 
classify between primary and secondary tolerant patients from the couples, recipients 
and donors into one big table:
```{r, echo = FALSE}
# couple info:
load("../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_PS_95.RData")
pctgs_couples <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::select("Id.Cryostem.R", "COUPLENUMBER", "gender_comp", "age_recip") %>% 
  dplyr::filter(grepl("R", Id.Cryostem.R)) %>% 
  inner_join(rna_dr_cryo_PS_95 %>% 
               rownames_to_column("COUPLENUMBER") %>% 
               mutate(COUPLENUMBER = gsub("couple_","", COUPLENUMBER)), by = "COUPLENUMBER")

# recip info:
load("../data/RNA_seq_national/recip_only/rna_recip_cryo_PS_95.RData")
colnames2keep <- which(colnames(rna_recip_cryo_PS_95) %in% c("group"))
colnames(rna_recip_cryo_PS_95)[-colnames2keep] <-
  paste0("R_", colnames(rna_recip_cryo_PS_95)[-colnames2keep])
pctgs_recip <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::select("Id.Cryostem.R", "COUPLENUMBER") %>% 
  right_join(rna_recip_cryo_PS_95 %>%
               rownames_to_column("Id.Cryostem.R"), by = "Id.Cryostem.R")

# donor info :
load("../data//RNA_seq_national/donors_only/rna_donors_Cryo_PS_95.RData")
colnames2keep <- which(colnames(rna_donors_cryo_PS_95) %in% c("group"))
colnames(rna_donors_cryo_PS_95)[-colnames2keep] <-
  paste0("D_", colnames(rna_donors_cryo_PS_95)[-colnames2keep])
pctgs_donor <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::select("Id.Cryostem.R", "COUPLENUMBER") %>% 
  right_join(rna_donors_cryo_PS_95 %>%
               rownames_to_column("Id.Cryostem.R"), by = "Id.Cryostem.R") 

# combine :
pctgs_all <- pctgs_couples %>% 
  left_join(pctgs_recip, by = c("Id.Cryostem.R", "COUPLENUMBER", "group")) %>% 
  left_join(pctgs_donor, by = c("COUPLENUMBER", "group"))
```

We can generate a PCA and a tSNE using all these features to see how the tolerant 
and non tolerant patients are represented in the reduced dimensions:
```{r, echo = FALSE}
pctgs_all_only <- pctgs_all %>% 
  column_to_rownames("COUPLENUMBER") %>% 
  select_if(is.numeric) %>% 
  dplyr::select(-"age_recip")

pca_all <- prcomp(pctgs_all_only)
```

The 2 first components of the PCA seem to hold quite a lot of the variability in the data:
```{r}
plot(pca_all)
```

```{r, echo = FALSE}
pca_2plot <- as.data.frame(pca_all$x) %>% 
  mutate(COUPLENUMBER = rownames(pca_all$x)) %>% 
  left_join(pctgs_all, by = "COUPLENUMBER")
ggplot(pca_2plot, aes(x = PC1, y = PC2, col = as.factor(group), label = Id.Cryostem.R.x)) + 
  geom_point() +
  scale_color_manual(breaks = c("primary_tolerant", "secondary_tolerant"),
                     values = c("springgreen3", "royalblue1")) +
  geom_text(aes(label=Id.Cryostem.R.x),hjust=0, vjust=0)
```
The primary tolerant patients seem to have lower PC1 values on the PCA 
(they are situated more on the left), but the two groups are still quite mixed.
We can still see which genes are driving the first PC:
```{r}
pca_all$rotation[which(abs(pca_all$rotation[,"PC1"])>0.15),c(1)]
```

It seems like the differences between donors and recipients in these 5 genes 
mainly drive the 1st PC.

We also see a separation of the patients in a top and a low group. We can see
which genes are mainly driving the second axis of the PCA:
```{r}
pca_all$rotation[which(abs(pca_all$rotation[,"PC2"])>0.1),c(1,2)]
```

It seems like the expression of the gene UTY in the donors is mainly driving 
the separation into two groups in the PCA:
```{r, echo = FALSE}
ggplot(pca_2plot, aes(x = PC1, y = PC2, col = D_UTY, label = Id.Cryostem.R.x)) + 
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R.x),hjust=0, vjust=0)
```


We can also generate a tSNE on the same data:
```{r, echo = FALSE}
set.seed(1)
tsne_all <- Rtsne::Rtsne(pctgs_all_only, perplexity = 3)
tsne_2plot <- as.data.frame(tsne_all$Y) %>% 
  magrittr::set_colnames(c("X1", "X2")) %>% 
  mutate(COUPLENUMBER = rownames(pctgs_all_only)) %>% 
  left_join(pctgs_all, by = "COUPLENUMBER")

ggplot(tsne_2plot, aes(x = X1, y = X2, col = as.factor(group), 
                       label = Id.Cryostem.R.x)) + 
  geom_point() +
  scale_color_manual(breaks = c("primary_tolerant", "secondary_tolerant"),
                     values = c("springgreen3", "royalblue1")) +
  geom_text(aes(label=Id.Cryostem.R.x),hjust=0, vjust=0)
```


We can see if building a model using all these features combined leads to good 
classification results between tolerant and non tolerant patients:
```{r, echo = FALSE}
pctgs_all_RF <- pctgs_all %>% 
  dplyr::select(-"COUPLENUMBER", -"gender_comp", -"age_recip") %>% 
  mutate(group = as.factor(group)) %>% 
  column_to_rownames("Id.Cryostem.R.x")

set.seed(1)
rf <- ranger::ranger(group~., data = pctgs_all_RF, num.trees = 5000, importance = "impurity")
rf$confusion.matrix
paste0("prediction error: ", rf$prediction.error*100)
```

We can see which features were selected in the model to classify tolerant from 
non tolerant patients in the St Louis cohort:
```{r, echo = FALSE}
sort(rf$variable.importance, decreasing = TRUE)[1:20]
```

The variables that were mainly selected as informative in the model are the ones
from recipients and couples.

We can visualise the top variables in a heatmap:
```{r, echo = FALSE}
mat2plot <- pctgs_all_RF[,c("group", names(sort(rf$variable.importance, decreasing = TRUE))[1:20])] %>% 
  rownames_to_column("rownames") %>% 
  arrange(group) %>% 
  column_to_rownames("rownames")

annot_patients <- as.data.frame(mat2plot$group)
rownames(annot_patients) <- rownames(mat2plot)
colnames(annot_patients) <- "group"
ann_colors <- list(group = c(primary_tolerant = "springgreen3", 
                             secondary_tolerant = "royalblue1"))

pheatmap::pheatmap(mat2plot[,-which(colnames(mat2plot)=="group")], 
                   annotation_row = annot_patients,
                   annotation_colors = ann_colors[1],
                   cluster_rows = FALSE,
                   cluster_cols = FALSE,
                   main = "Heatmap with patients ordered by group",
                   scale = "column")

pheatmap::pheatmap(mat2plot[,-which(colnames(mat2plot)=="group")], 
                   annotation_row = annot_patients,
                   annotation_colors = ann_colors[1],
                   cluster_rows = TRUE,
                   main = "Heatmap with patients clustered",
                   scale = "column")
```

Reminder: high values for the variables that start with "subst_" mean that the
difference of expression between the donor and the recipient of the same couple 
was high for this specific feature.

#### Threshold of 0.90

We first combine the features that were selected as highly informative to 
classify between primary and secondary tolerant patients from the couples, recipients 
and donors into one big table:
```{r, echo = FALSE}
# couple info:
load("../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_PS_90.RData")
# # add "couple_" in front of the feature names 
# colnames2keep <- which(colnames(pctgs_sel_ft_couples_90) %in% c("group", "COUPLENUMBER"))
# colnames(pctgs_sel_ft_couples_90)[-colnames2keep] <-
#   gsub("X", "couple_", colnames(pctgs_sel_ft_couples_90)[-colnames2keep])
# combine feature info with couple info about gender_comp...
pctgs_couples <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::select("Id.Cryostem.R", "COUPLENUMBER", "gender_comp", "age_recip") %>% 
  dplyr::filter(grepl("R", Id.Cryostem.R)) %>% 
  inner_join(rna_dr_cryo_PS_90 %>% 
               rownames_to_column("COUPLENUMBER") %>% 
               mutate(COUPLENUMBER = gsub("couple_","", COUPLENUMBER)), by = "COUPLENUMBER")

# recip info:
load("../data/RNA_seq_national/recip_only/rna_recip_cryo_PS_90.RData")
colnames2keep <- which(colnames(rna_recip_cryo_PS_90) %in% c("group"))
colnames(rna_recip_cryo_PS_90)[-colnames2keep] <-
  paste0("R_", colnames(rna_recip_cryo_PS_90)[-colnames2keep])
pctgs_recip <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::select("Id.Cryostem.R", "COUPLENUMBER") %>% 
  right_join(rna_recip_cryo_PS_90 %>%
               rownames_to_column("Id.Cryostem.R"), by = "Id.Cryostem.R")

# donor info :
load("../data//RNA_seq_national/donors_only/rna_donors_Cryo_PS_90.RData")
colnames2keep <- which(colnames(rna_donors_cryo_PS_90) %in% c("group"))
colnames(rna_donors_cryo_PS_90)[-colnames2keep] <-
  paste0("D_", colnames(rna_donors_cryo_PS_90)[-colnames2keep])
pctgs_donor <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::select("Id.Cryostem.R", "COUPLENUMBER") %>% 
  right_join(rna_donors_cryo_PS_90 %>%
               rownames_to_column("Id.Cryostem.R"), by = "Id.Cryostem.R") 

# combine :
pctgs_all <- pctgs_couples %>% 
  left_join(pctgs_recip, by = c("Id.Cryostem.R", "COUPLENUMBER", "group")) %>% 
  left_join(pctgs_donor, by = c("COUPLENUMBER", "group"))
```

We can generate a PCA and a tSNE using all these features to see how the tolerant 
and non tolerant patients are represented in the reduced dimensions:
```{r, echo = FALSE}
pctgs_all_only <- pctgs_all %>% 
  column_to_rownames("COUPLENUMBER") %>% 
  select_if(is.numeric) %>% 
  dplyr::select(-"age_recip")

pca_all <- prcomp(pctgs_all_only)
```

The 2 first components of the PCA seem to hold quite a lot of the variability in the data:
```{r}
plot(pca_all)
```

```{r, echo = FALSE}
pca_2plot <- as.data.frame(pca_all$x) %>% 
  mutate(COUPLENUMBER = rownames(pca_all$x)) %>% 
  left_join(pctgs_all, by = "COUPLENUMBER")
ggplot(pca_2plot, aes(x = PC1, y = PC2, col = as.factor(group), label = Id.Cryostem.R.x)) + 
  geom_point() +
  scale_color_manual(breaks = c("primary_tolerant", "secondary_tolerant"),
                     values = c("springgreen3", "royalblue1")) +
  geom_text(aes(label=Id.Cryostem.R.x),hjust=0, vjust=0)
```
The primary tolerant patients seem to have higher PC1 values on the PCA 
(they are situated more on the right), which suggests that we might be 
able to classify primary from secondary tolerant patients based on the features 
we selected.

We can also generate a tSNE on the same data:
```{r, echo = FALSE}
tsne_all <- Rtsne::Rtsne(pctgs_all_only, perplexity = 3)
tsne_2plot <- as.data.frame(tsne_all$Y) %>% 
  magrittr::set_colnames(c("X1", "X2")) %>% 
  mutate(COUPLENUMBER = rownames(pctgs_all_only)) %>% 
  left_join(pctgs_all, by = "COUPLENUMBER")

ggplot(tsne_2plot, aes(x = X1, y = X2, col = as.factor(group), 
                       label = Id.Cryostem.R.x)) + 
  geom_point() +
  scale_color_manual(breaks = c("primary_tolerant", "secondary_tolerant"),
                     values = c("springgreen3", "royalblue1")) +
  geom_text(aes(label=Id.Cryostem.R.x),hjust=0, vjust=0)
```


We can see if building a model using all these features combined leads to good 
classification results between tolerant and non tolerant patients:
```{r, echo = FALSE}
pctgs_all_RF <- pctgs_all %>% 
  dplyr::select(-"COUPLENUMBER", -"gender_comp", -"age_recip") %>% 
  mutate(group = as.factor(group)) %>% 
  column_to_rownames("Id.Cryostem.R.x")

set.seed(1)
rf <- ranger::ranger(group~., data = pctgs_all_RF, num.trees = 5000, importance = "impurity")
rf$confusion.matrix
paste0("prediction error: ", rf$prediction.error*100)
```

We can see which features were selected in the model to classify tolerant from 
non tolerant patients in the St Louis cohort:
```{r, echo = FALSE}
sort(rf$variable.importance, decreasing = TRUE)[1:20]
```

The variables that were mainly selected as informative in the model are the ones
from recipients and couples.

We can visualise the top variables in a heatmap:
```{r, echo = FALSE}
mat2plot <- pctgs_all_RF[,c("group", names(sort(rf$variable.importance, decreasing = TRUE))[1:20])] %>% 
  rownames_to_column("rownames") %>% 
  arrange(group) %>% 
  column_to_rownames("rownames")

annot_patients <- as.data.frame(mat2plot$group)
rownames(annot_patients) <- rownames(mat2plot)
colnames(annot_patients) <- "group"
ann_colors <- list(group = c(primary_tolerant = "springgreen3", 
                             secondary_tolerant = "royalblue1"))

pheatmap::pheatmap(mat2plot[,-which(colnames(mat2plot)=="group")], 
                   annotation_row = annot_patients,
                   annotation_colors = ann_colors[1],
                   cluster_rows = FALSE,
                   cluster_cols = FALSE,
                   main = "Heatmap with patients ordered by group",
                   scale = "column")

pheatmap::pheatmap(mat2plot[,-which(colnames(mat2plot)=="group")], 
                   annotation_row = annot_patients,
                   annotation_colors = ann_colors[1],
                   cluster_rows = TRUE,
                   main = "Heatmap with patients clustered",
                   scale = "column")
```

Reminder: high values for the variables that start with "subst_" mean that the
difference of expression between the donor and the recipient of the same couple 
was high for this specific feature.

#### Focusing only on the genes that had a common behaviour in the two cohorts:

We first combine the features that were selected as highly informative to 
classify between primary and secondary tolerant patients from the couples, recipients 
and donors into one big table:
```{r, echo = FALSE}
# couple info:
rna_dr_cryo_PS_90 <- readRDS("../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_beh_PS_90.RDS")
pctgs_couples <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::select("Id.Cryostem.R", "COUPLENUMBER", "gender_comp", "age_recip") %>% 
  dplyr::filter(grepl("R", Id.Cryostem.R)) %>% 
  inner_join(rna_dr_cryo_PS_90 %>% 
               rownames_to_column("COUPLENUMBER") %>% 
               mutate(COUPLENUMBER = gsub("couple_","", COUPLENUMBER)), by = "COUPLENUMBER")

# recip info:
load("../data/RNA_seq_national/recip_only/rna_recip_cryo_beh_PS_90.RData")
colnames2keep <- which(colnames(rna_recip_cryo_beh_PS_90) %in% c("group"))
colnames(rna_recip_cryo_beh_PS_90)[-colnames2keep] <-
  paste0("R_", colnames(rna_recip_cryo_beh_PS_90)[-colnames2keep])
pctgs_recip <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::select("Id.Cryostem.R", "COUPLENUMBER") %>% 
  right_join(rna_recip_cryo_beh_PS_90 %>%
               rownames_to_column("Id.Cryostem.R"), by = "Id.Cryostem.R")

# donor info :
rna_donors_cryo_PS_90 <- readRDS("../data//RNA_seq_national/donors_only/rna_donor_cryo_beh_PS_90.RDS")
colnames2keep <- which(colnames(rna_donors_cryo_PS_90) %in% c("group"))
colnames(rna_donors_cryo_PS_90)[-colnames2keep] <-
  paste0("D_", colnames(rna_donors_cryo_PS_90)[-colnames2keep])
pctgs_donor <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::select("Id.Cryostem.R", "COUPLENUMBER") %>% 
  right_join(rna_donors_cryo_PS_90 %>%
               rownames_to_column("Id.Cryostem.R"), by = "Id.Cryostem.R") 

# combine :
pctgs_all <- pctgs_couples %>% 
  left_join(pctgs_recip, by = c("Id.Cryostem.R", "COUPLENUMBER", "group")) %>% 
  left_join(pctgs_donor, by = c("COUPLENUMBER", "group"))
```

We can generate a PCA and a tSNE using all these features to see how the tolerant 
and non tolerant patients are represented in the reduced dimensions:
```{r, echo = FALSE}
pctgs_all_only <- pctgs_all %>% 
  column_to_rownames("COUPLENUMBER") %>% 
  select_if(is.numeric) %>% 
  dplyr::select(-"age_recip")

pca_all <- prcomp(pctgs_all_only)
```

The 2 first components of the PCA seem to hold quite a lot of the variability in the data:
```{r}
plot(pca_all)
```

```{r, echo = FALSE}
pca_2plot <- as.data.frame(pca_all$x) %>% 
  mutate(COUPLENUMBER = rownames(pca_all$x)) %>% 
  left_join(pctgs_all, by = "COUPLENUMBER")
ggplot(pca_2plot, aes(x = PC1, y = PC2, col = as.factor(group), label = Id.Cryostem.R.x)) + 
  geom_point() +
  scale_color_manual(breaks = c("primary_tolerant", "secondary_tolerant"),
                     values = c("springgreen3", "royalblue1")) +
  geom_text(aes(label=Id.Cryostem.R.x),hjust=0, vjust=0)
```
The primary tolerant patients seem to have higher PC1 values on the PCA 
(they are situated more on the right), which suggests that we might be 
able to classify primary from secondary tolerant patients based on the features 
we selected.

We can also generate a tSNE on the same data:
```{r, echo = FALSE}
tsne_all <- Rtsne::Rtsne(pctgs_all_only, perplexity = 3)
tsne_2plot <- as.data.frame(tsne_all$Y) %>% 
  magrittr::set_colnames(c("X1", "X2")) %>% 
  mutate(COUPLENUMBER = rownames(pctgs_all_only)) %>% 
  left_join(pctgs_all, by = "COUPLENUMBER")

ggplot(tsne_2plot, aes(x = X1, y = X2, col = as.factor(group), 
                       label = Id.Cryostem.R.x)) + 
  geom_point() +
  scale_color_manual(breaks = c("primary_tolerant", "secondary_tolerant"),
                     values = c("springgreen3", "royalblue1")) +
  geom_text(aes(label=Id.Cryostem.R.x),hjust=0, vjust=0)
```


We can see if building a model using all these features combined leads to good 
classification results between tolerant and non tolerant patients:
```{r, echo = FALSE}
pctgs_all_RF <- pctgs_all %>% 
  dplyr::select(-"COUPLENUMBER", -"gender_comp", -"age_recip") %>% 
  mutate(group = as.factor(group)) %>% 
  column_to_rownames("Id.Cryostem.R.x")

set.seed(1)
rf <- ranger::ranger(group~., data = pctgs_all_RF, num.trees = 5000, importance = "impurity")
rf$confusion.matrix
paste0("prediction error: ", rf$prediction.error*100)
```

We can see which features were selected in the model to classify tolerant from 
non tolerant patients in the St Louis cohort:
```{r, echo = FALSE}
sort(rf$variable.importance, decreasing = TRUE)[1:20]
```

The variables that were mainly selected as informative in the model are the ones
from recipients and couples.

We can visualise the top variables in a heatmap:
```{r, echo = FALSE}
mat2plot <- pctgs_all_RF[,c("group", names(sort(rf$variable.importance, decreasing = TRUE))[1:20])] %>% 
  rownames_to_column("rownames") %>% 
  arrange(group) %>% 
  column_to_rownames("rownames")

annot_patients <- as.data.frame(mat2plot$group)
rownames(annot_patients) <- rownames(mat2plot)
colnames(annot_patients) <- "group"
ann_colors <- list(group = c(primary_tolerant = "springgreen3", 
                             secondary_tolerant = "royalblue1"))

pheatmap::pheatmap(mat2plot[,-which(colnames(mat2plot)=="group")], 
                   annotation_row = annot_patients,
                   annotation_colors = ann_colors[1],
                   cluster_rows = FALSE,
                   cluster_cols = FALSE,
                   main = "Heatmap with patients ordered by group",
                   scale = "column")

pheatmap::pheatmap(mat2plot[,-which(colnames(mat2plot)=="group")], 
                   annotation_row = annot_patients,
                   annotation_colors = ann_colors[1],
                   cluster_rows = TRUE,
                   main = "Heatmap with patients clustered",
                   scale = "column")
```

Reminder: high values for the variables that start with "subst_" mean that the
difference of expression between the donor and the recipient of the same couple 
was high for this specific feature.

Summary of the number of significative genes found overall, between non tolerant
and tolerant patients:
```{r, message = FALSE}
r_tnt_95_louis <- read.xlsx("../data/RNAseq/recip_only/perm_val_TNT_95_thresh.xlsx", colNames = FALSE)
r_tnt_95_cryo <- read.xlsx("../data/RNA_seq_national/recip_only/perm_val_TNT_95_thresh.xlsx", colNames = FALSE)
load("../data/RNA_seq_national/recip_only/rna_recip_cryo_TNT_95.RData")
r_tnt_95_beh <- read.xlsx("../data/RNAseq/recip_only/perm_val_beh_TNT_95_thresh.xlsx")

summary_genes <- as.data.frame(t(c("Recip_TNT", 0.95, length(r_tnt_95_louis$X1),
                                   length(r_tnt_95_cryo$X1), (ncol(rna_recip_cryo_TNT_95)-1),
                                   length(r_tnt_95_beh$sel_beh))))

r_tnt_90_louis <- read.xlsx("../data/RNAseq/recip_only/perm_val_TNT_90_thresh.xlsx", colNames = FALSE)
r_tnt_90_cryo <- read.xlsx("../data/RNA_seq_national/recip_only/perm_val_TNT_90_thresh.xlsx", colNames = FALSE)
load("../data/RNA_seq_national/recip_only/rna_recip_cryo_TNT_90.RData")
r_tnt_90_beh <- read.xlsx("../data/RNAseq/recip_only/perm_val_beh_TNT_90_thresh.xlsx")

summary_genes <- rbind(summary_genes, 
                       as.data.frame(t(c("Recip_TNT", 0.90, length(r_tnt_90_louis$X1),
                                   length(r_tnt_90_cryo$X1), (ncol(rna_recip_cryo_TNT_90)-1),
                                   length(r_tnt_90_beh$sel_beh)))))

d_tnt_95_louis <- read.xlsx("../data/RNAseq/donors_only/perm_val_TNT_95_thresh.xlsx", colNames = FALSE)
d_tnt_95_cryo <- read.xlsx("../data/RNA_seq_national/donors_only/perm_val_TNT_95_thresh.xlsx", colNames = FALSE)
load("../data/RNA_seq_national/donors_only/rna_donors_cryo_TNT_95.RData")
d_tnt_95_beh <- read.xlsx("../data/RNAseq/donors_only/perm_val_beh_TNT_95_thresh.xlsx")

summary_genes <- rbind(summary_genes, 
                       as.data.frame(t(c("Donors_TNT", 0.95, length(d_tnt_95_louis$X1),
                                   length(d_tnt_95_cryo$X1), ncol(rna_donors_cryo_TNT_95)-1,
                                   length(d_tnt_95_beh$sel_beh)))))

d_tnt_90_louis <- read.xlsx("../data/RNAseq/donors_only/perm_val_TNT_90_thresh.xlsx", colNames = FALSE)
d_tnt_90_cryo <- read.xlsx("../data/RNA_seq_national/donors_only/perm_val_TNT_90_thresh.xlsx", colNames = FALSE)
load("../data/RNA_seq_national/donors_only/rna_donors_cryo_TNT_90.RData")
d_tnt_90_beh <- read.xlsx("../data/RNAseq/donors_only/perm_val_beh_TNT_90_thresh.xlsx")

summary_genes <- rbind(summary_genes, 
                       as.data.frame(t(c("Donors_TNT", 0.90, length(d_tnt_90_louis$X1),
                                   length(d_tnt_90_cryo$X1), ncol(rna_donors_cryo_TNT_90)-1,
                                   length(d_tnt_90_beh$sel_beh)))))

dr_tnt_95_louis <- read.xlsx("../data/RNAseq/donors_and_recip/perm_val_TNT_95_thresh.xlsx", colNames = FALSE)
dr_tnt_95_cryo <- read.xlsx("../data/RNA_seq_national/donors_and_recip/perm_val_TNT_95_thresh.xlsx", colNames = FALSE)
load("../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_TNT_95.RData")
dr_tnt_95_beh <- read.xlsx("../data/RNAseq/donors_and_recip/perm_val_beh_TNT_95_thresh.xlsx")

summary_genes <- rbind(summary_genes, 
                       as.data.frame(t(c("D-R_TNT", 0.95, length(dr_tnt_95_louis$X1),
                                   length(dr_tnt_95_cryo$X1), ncol(rna_dr_cryo_TNT_95)-1,
                                   length(dr_tnt_95_beh$sel_beh)))))

dr_tnt_90_louis <- read.xlsx("../data/RNAseq/donors_and_recip/perm_val_TNT_90_thresh.xlsx", colNames = FALSE)
dr_tnt_90_cryo <- read.xlsx("../data/RNA_seq_national/donors_and_recip/perm_val_TNT_90_thresh.xlsx", colNames = FALSE)
load("../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_TNT_90.RData")
dr_tnt_90_beh <- read.xlsx("../data/RNAseq/donors_and_recip/perm_val_beh_TNT_90_thresh.xlsx")

summary_genes <- rbind(summary_genes, 
                       as.data.frame(t(c("D-R_TNT", 0.90, length(dr_tnt_90_louis$X1),
                                   length(dr_tnt_90_cryo$X1), ncol(rna_dr_cryo_TNT_90)-1,
                                   length(dr_tnt_90_beh$sel_beh)))))

colnames(summary_genes) <- c("type", "threshold", "St Louis", "Cryostem", "both", "behaviour")

summary_genes %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover"))
```

Summary of the number of significative genes found overall, between primary
and secondary tolerant patients:
```{r, message = FALSE}
r_ps_95_louis <- read.xlsx("../data/RNAseq/recip_only/perm_val_PS_95_thresh.xlsx", colNames = FALSE)
r_ps_95_cryo <- read.xlsx("../data/RNA_seq_national/recip_only/perm_val_PS_95_thresh.xlsx", colNames = FALSE)
load("../data/RNA_seq_national/recip_only/rna_recip_cryo_PS_95.RData")
r_ps_95_beh <- read.xlsx("../data/RNAseq/recip_only/perm_val_beh_PS_95_thresh.xlsx")

summary_genes <- as.data.frame(t(c("Recip_PS", 0.95, length(r_ps_95_louis$X1),
                                   length(r_ps_95_cryo$X1), (ncol(rna_recip_cryo_PS_95)-1),
                                   length(r_ps_95_beh$sel_beh))))

r_ps_90_louis <- read.xlsx("../data/RNAseq/recip_only/perm_val_PS_90_thresh.xlsx", colNames = FALSE)
r_ps_90_cryo <- read.xlsx("../data/RNA_seq_national/recip_only/perm_val_PS_90_thresh.xlsx", colNames = FALSE)
load("../data/RNA_seq_national/recip_only/rna_recip_cryo_PS_90.RData")
r_ps_90_beh <- read.xlsx("../data/RNAseq/recip_only/perm_val_beh_PS_90_thresh.xlsx")

summary_genes <- rbind(summary_genes, 
                       as.data.frame(t(c("Recip_PS", 0.90, length(r_ps_90_louis$X1),
                                   length(r_ps_90_cryo$X1), (ncol(rna_recip_cryo_PS_90)-1),
                                   length(r_ps_90_beh$sel_beh)))))

d_ps_95_louis <- read.xlsx("../data/RNAseq/donors_only/perm_val_PS_95_thresh.xlsx", colNames = FALSE)
d_ps_95_cryo <- read.xlsx("../data/RNA_seq_national/donors_only/perm_val_PS_95_thresh.xlsx", colNames = FALSE)
load("../data/RNA_seq_national/donors_only/rna_donors_cryo_PS_95.RData")
d_ps_95_beh <- read.xlsx("../data/RNAseq/donors_only/perm_val_beh_PS_95_thresh.xlsx")

summary_genes <- rbind(summary_genes, 
                       as.data.frame(t(c("Donors_PS", 0.95, length(d_ps_95_louis$X1),
                                   length(d_ps_95_cryo$X1), ncol(rna_donors_cryo_PS_95)-1,
                                   length(d_ps_95_beh$sel_beh)))))

d_ps_90_louis <- read.xlsx("../data/RNAseq/donors_only/perm_val_PS_90_thresh.xlsx", colNames = FALSE)
d_ps_90_cryo <- read.xlsx("../data/RNA_seq_national/donors_only/perm_val_PS_90_thresh.xlsx", colNames = FALSE)
load("../data/RNA_seq_national/donors_only/rna_donors_cryo_PS_90.RData")
d_ps_90_beh <- read.xlsx("../data/RNAseq/donors_only/perm_val_beh_PS_90_thresh.xlsx")

summary_genes <- rbind(summary_genes, 
                       as.data.frame(t(c("Donors_PS", 0.90, length(d_ps_90_louis$X1),
                                   length(d_ps_90_cryo$X1), ncol(rna_donors_cryo_PS_90)-1,
                                   length(d_ps_90_beh$sel_beh)))))

dr_ps_95_louis <- read.xlsx("../data/RNAseq/donors_and_recip/perm_val_PS_95_thresh.xlsx", colNames = FALSE)
dr_ps_95_cryo <- read.xlsx("../data/RNA_seq_national/donors_and_recip/perm_val_PS_95_thresh.xlsx", colNames = FALSE)
load("../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_PS_95.RData")
dr_ps_95_beh <- read.xlsx("../data/RNAseq/donors_and_recip/perm_val_beh_PS_95_thresh.xlsx")

summary_genes <- rbind(summary_genes, 
                       as.data.frame(t(c("D-R_PS", 0.95, length(dr_ps_95_louis$X1),
                                   length(dr_ps_95_cryo$X1), ncol(rna_dr_cryo_PS_95)-1,
                                   length(dr_ps_95_beh$sel_beh)))))

dr_ps_90_louis <- read.xlsx("../data/RNAseq/donors_and_recip/perm_val_PS_90_thresh.xlsx", colNames = FALSE)
dr_ps_90_cryo <- read.xlsx("../data/RNA_seq_national/donors_and_recip/perm_val_PS_90_thresh.xlsx", colNames = FALSE)
load("../data/RNA_seq_national/donors_and_recip/rna_dr_cryo_PS_90.RData")
dr_ps_90_beh <- read.xlsx("../data/RNAseq/donors_and_recip/perm_val_beh_PS_90_thresh.xlsx")

summary_genes <- rbind(summary_genes, 
                       as.data.frame(t(c("D-R_PS", 0.90, length(dr_ps_90_louis$X1),
                                   length(dr_ps_90_cryo$X1), ncol(rna_dr_cryo_PS_90)-1,
                                   length(dr_ps_90_beh$sel_beh)))))

colnames(summary_genes) <- c("type", "threshold", "St Louis", "Cryostem", "both", "behaviour")

summary_genes %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover"))
```



