---
title: "BioGVHD project: integration of CyTOF, Metabolomics and RNA features in the St Louis cohort"
author: "Helena"
date: "06/01/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r imports, echo=FALSE, results='hide', warning = FALSE}
suppressPackageStartupMessages({
  library(tidyverse)
  library(openxlsx)
  library(BioGVHD)
  library(data.table)
  library(ggplot2)
  library(igraph)
})
options("scipen"=100)
```

## Focusing on the recipients

We can first focus on the features that we extracted as informative in the 
recipients, to see if, taken together, they would be more informative than 
when working on one data type alone.

### Tolerant versus non tolerant recipients

Since looking at differences between the three groups consisting of primary, 
secondary and non tolerant patients at the same time wasn't feasible, we divided
our problem in two distinct sub-problems. We first focus on the features that
seemed to play a role when trying to disinguish non tolerant from tolerant 
recipients.

We first need to gather the features that were extracted from the three data
sources:

- CyTOF features that were selected in both the St Louis and the Cryostem cohort, 
  with a threshold of 0.90. These features had the same behaviour in both cohorts.

```{r}
tmp <- readRDS("../data/cyto/log2/recip/pctgs_sel_ft_recip_TNT_90.RDS")
sel_beh <- read.xlsx("../data/cyto_national/log2/recip/perm_val_beh_TNT_90_thresh.xlsx")

ft_recip_cyto <- tmp %>% 
  select(sel_beh$features) 
```

- Metabolomics features that were selected in both the St Louis and the Cryostem cohort, 
  with a threshold of 0.90. These features had the same behaviour in both cohorts.

```{r}
tmp <- readRDS("../data/metabo/v2_012020/recip/pctgs_sel_ft_recip_TNT_90.RDS")
sel_beh <- read.xlsx("../data/metabo/v2_012020/recip/perm_val_beh_TNT_90_thresh.xlsx")

ft_recip_metabo <- tmp %>% 
  select(sel_beh$features) 
```

- Transcriptomics features that were selected in both the St Louis and the Cryostem cohort, 
  with a threshold of 0.90. These features had the same behaviour in both cohorts.

```{r}
load("../data/RNAseq/recip_only/rna_recip_louis_TNT_90.RData")
sel_beh <- read.xlsx("../data/RNAseq/recip_only/perm_val_beh_TNT_90_thresh.xlsx")

ft_recip_transcripto <- rna_recip_louis_TNT_90 %>% 
  select(sel_beh$sel_beh) 
```

I extract informations on the recipients group, age, gender, ... to be able to 
use this information later on.

```{r, echo = FALSE}
samp_rd <- read.xlsx("~/Documents/VIB/Projects/Integrative_Paris/Local_cohort/Data synthesis local cohort Saint-Louis 17012019.xlsx")
samp_rd$DOB <- as.Date(samp_rd$DOB, format = "%d.%m.%Y")
samp_rd$DOG <- as.Date(samp_rd$DOG, format = "%d.%m.%Y")
samp_rd$DATEOFSAMPLE <- as.Date(samp_rd$DATEOFSAMPLE, format = "%d.%m.%Y")
samp_rd <- samp_rd %>% 
  dplyr::filter(!is.na(METABONAME)) 
complete_couples <- samp_rd$COUPLENUMBER[which(duplicated(samp_rd$COUPLENUMBER))]

couple_info <- samp_rd %>% 
  dplyr::filter(COUPLENUMBER %in% complete_couples) %>% 
  arrange(COUPLENUMBER) %>% 
  column_to_rownames("METABONAME") %>% 
  rownames_to_column("rn") %>% 
  mutate(GROUP = as.factor(tolower(GROUP))) %>% 
  column_to_rownames("rn")
  
gender_donors <- couple_info$GENDER[c(TRUE,FALSE)]
gender_recip <- couple_info$GENDER[c(FALSE,TRUE)]
gender_comp <- paste0(gender_donors, gender_recip)

age_recip <- couple_info$DOG[c(FALSE,TRUE)] - couple_info$DOB[c(FALSE,TRUE)]

ifelse(grepl("R", rownames(couple_info)[1]), 
       cmv_donors <- couple_info$CMVStatus[c(FALSE,TRUE)],
       cmv_donors <- couple_info$CMVStatus[c(TRUE,FALSE)])

ifelse(grepl("R", rownames(couple_info)[1]), 
       cmv_recip <- couple_info$CMVStatus[c(TRUE,FALSE)],
       cmv_recip <- couple_info$CMVStatus[c(FALSE,TRUE)])

cmv_comp <- paste0(cmv_donors, cmv_recip)

couple_info <- couple_info %>% 
  rownames_to_column("rownames") %>% 
  mutate(gender_comp = rep(gender_comp, each = 2),
         age_recip = rep(age_recip, each = 2),
         cmv_comp = rep(cmv_comp, each = 2)) %>% 
  column_to_rownames("rownames")

head(couple_info[,c(1,2,3,5,6,55,56,57)])
```

I add a prefix ("cyt_", "rna_" or "met_") in front of the features names, and 
I then merge all features together:

```{r, echo = FALSE}
colnames(ft_recip_cyto) <- paste0("cyt_", colnames(ft_recip_cyto))
colnames(ft_recip_metabo) <- paste0("met_", colnames(ft_recip_metabo))

# I change the rownames in the metabo file so that they correspond to cryostem Ids
info_tmp <- couple_info[rownames(ft_recip_metabo),]
rownames(ft_recip_metabo) <- info_tmp$Id.Cryostem.R

colnames(ft_recip_transcripto) <- paste0("rna_", colnames(ft_recip_transcripto))

ft_integ <- ft_recip_cyto %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  inner_join(ft_recip_metabo %>% rownames_to_column("Id.Cryostem.R") , by = c("Id.Cryostem.R")) %>% 
  inner_join(ft_recip_transcripto %>% rownames_to_column("Id.Cryostem.R") , by = c("Id.Cryostem.R")) 
print(paste0("We have information on the CyTOF, metabolomics and transcriptomics data for ",
             nrow(ft_integ), " patients in the St Louis cohort."))
```

#### Principal component analysis

We can first generate a PCA on the data:
```{r, echo = FALSE}
mat4pca <- ft_integ %>% 
  select(-Id.Cryostem.R) %>%  
  apply(2, scale) %>% 
  as.matrix() 
rownames(mat4pca) <- ft_integ$Id.Cryostem.R
pca <- prcomp(mat4pca, scale. = TRUE)
```

The first principal component seems to hold much more information that the others
```{r}
plot(pca)
```

We can investigate how much of the principal components are driven by the 
different data types.
```{r, echo = FALSE}
pca_var <- pca$rotation %>% as.data.frame
pca_ft <- rep("x", nrow(pca_var))
pca_ft[grep("cyt_", rownames(pca_var))] <- "cyto"
pca_ft[grep("met_", rownames(pca_var))] <- "metabo"
pca_ft[grep("rna_", rownames(pca_var))] <- "rna"
pca_var <- pca_var %>% 
  rownames_to_column("rn") %>% 
  mutate(feature_orig = pca_ft)
```

The RNAseq data variability was captured much more than the other data types.
This is most probably due to the fact that we selected 278 genes, against
42 metabolites and 24 CyTOF features only.
```{r, echo = FALSE}
pc1 <- pca_var %>% 
  column_to_rownames("rn") %>% 
  group_by(feature_orig) %>% 
  summarise_all(sum)
pc1m <- pc1 %>% gather(PC, value, -feature_orig) %>% 
  mutate(label_ypos = round(value))

pc1m_sub <- pc1m[1:30,]

ggplot(data=pc1m_sub, aes(x=PC, y=value, fill=feature_orig)) +
  geom_bar(stat="identity")+
  # geom_text(aes(y=label_ypos, label=value), vjust=1.6, 
  #           color="white", size=3.5)+
  scale_fill_brewer(palette="Paired")+
  theme_minimal()
```

We can see how these principal components are correlated with tolerance. 
The first PC is very correlated with the tolerance. The PC4 (and slightly 
PC8 and PC10) also look correlated with tolerance.
```{r, echo = FALSE}
complete_patients <- rownames(mat4pca)
info_complete_patients <- couple_info %>%
  rownames_to_column("METABONAME") %>% 
  dplyr::filter(Id.Cryostem.R %in% complete_patients) %>% 
  arrange(Id.Cryostem.R) %>% 
  mutate(group = as.character(GROUP))
info_complete_patients$group[which(info_complete_patients$group != "non_tolerant")] <- "tolerant"
info_complete_patients$group <- as.factor(info_complete_patients$group)

pca_2plot <- left_join(ft_integ, info_complete_patients, by = c("Id.Cryostem.R")) %>% 
  mutate(PC1 = pca$x[,1],
         PC2 = pca$x[,2],
         PC3 = pca$x[,3],
         PC4 = pca$x[,4])

barplot(abs(cor(pca$x[,1:10], as.numeric(pca_2plot$group))[,1]))
```

We can then see how the PCs are correlated with the different informations 
that we have on the patients.

The first PC seems to be most correlated with group, cGVHD and ABO.incompatibility:

```{r}
totest <- pca_2plot[,c(348, 366:374, 376, 390, 400:403)]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,1], totest)), horiz = TRUE, las = 2, cex.names = 0.5, xlim = c(0,0.6))
```

The second PC seems to be slightly correlated with the disorder group:

```{r}
totest <- pca_2plot[,c(348, 366:374, 376, 390, 400:403)]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,2], totest)), horiz = TRUE, las = 2, cex.names = 0.5, xlim = c(0,0.6))
```

The third PC seems to be slightly correlated with aGVHD and ABO.incompatibility:

```{r}
totest <- pca_2plot[,c(348, 366:374, 376, 390, 400:403)]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,3], totest)), horiz = TRUE, las = 2, cex.names = 0.5, xlim = c(0,0.6))
```

The fourth PC seems to be most correlated with group and age_recip:

```{r}
totest <- pca_2plot[,c(348, 366:374, 376, 390, 400:403)]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,4], totest)), horiz = TRUE, las = 2, cex.names = 0.5, xlim = c(0,0.6))
```

Since the 1st and 4th PCs seem to be most correlated with tolerance, we visualise
the patients in these two dimensions.
We observe that these 2 components seem to separate the tolerant recipients 
(on the bottom-left) from the non tolerant recipients (on the top-right).
We can also see how the recipients are displayed in these two dimensions 
according to ABO incompatibility, recipients age and cGVHD.
```{r, echo = FALSE}
ggplot(pca_2plot, aes(x= PC1, y= PC4, col= group, label = Id.Cryostem.R)) + 
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Separation of tolerant vs non tolerant recipients")
ggplot(pca_2plot, aes(x= PC1, y= PC4, col= ABO.incompatibility, label = Id.Cryostem.R)) + 
  scale_color_manual(values = c("chartreuse2", "black", "gray48")) +
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Separation of recipients based on ABO.incompatibility")
ggplot(pca_2plot, aes(x= PC1, y= PC4, col= age_recip, label = Id.Cryostem.R)) +
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Gradient in the recipients based on age")
ggplot(pca_2plot, aes(x= PC1, y= PC4, col= as.factor(cGVHD), label = Id.Cryostem.R)) +
  scale_color_manual(values = c("black", "red")) +
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Separation of recipients based on cGVHD")
```





We can first look at the PC1, since it holds most of the data's variability:
We first focus on the features that are overexpressed in the tolerant recipients 
(= the features that are most expressed on the LEFT of the PC1). Therefore, the 
lower the feature value, the more it is expressed in tolerant recipients.
```{r}
plot(sort(pca$rotation[,1][which(pca$rotation[,1]<0)]))
overexp <- sort(pca$rotation[,1][which(pca$rotation[,1]<0)])
over <- as.data.frame(overexp)
orig <- strsplit(rownames(over), "_")
orig <- map(orig, function(x){x[1]}) %>% unlist
over$orig <- orig
ggplot(over, aes(x = orig, y = overexp, col = orig)) +
  geom_boxplot() +
  geom_jitter()
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 
```

We extract the "importance values" for these genes, metabolites and CyTOF features 
in an excel file. Here's a sample of the lowest features in the PC1 (under the
-0.6 threshold) :
```{r, echo = FALSE}
most_imp <- rownames(over)[which(over$overexp < -0.06)]
most_imp

underexp <- over %>% 
  rownames_to_column("feature") %>% 
  filter(overexp < 0) 
ft_names <- underexp$feature
ft_tmp <- strsplit(ft_names, "_")
ft_tmp2 <- lapply(ft_tmp, function(ft){
  ft[2]
}) %>% unlist 
underexp$feature <- ft_tmp2
underexp_otig_ordered <- underexp %>%
  arrange(orig)

write.xlsx(underexp, file = "~/Desktop/underexp_PC1.xlsx")
write.xlsx(underexp_otig_ordered, file = "~/Desktop/underexp_PC1_ordered_by_origin.xlsx")
```

We then focus on the features that are overexpressed in the non tolerant recipients 
(= the features that are most expressed on the RIGHT of the PC1). Therefore, the 
higher the feature value, the more it is expressed in non tolerant recipients.
```{r}
plot(sort(pca$rotation[,1][which(pca$rotation[,1]>0)]))
overexp <- sort(pca$rotation[,1][which(pca$rotation[,1]>0)])
over <- as.data.frame(overexp)
orig <- strsplit(rownames(over), "_")
orig <- map(orig, function(x){x[1]}) %>% unlist
over$orig <- orig
ggplot(over, aes(x = orig, y = overexp, col = orig)) +
  geom_boxplot() +
  geom_jitter()
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 
```

We extract the "importance values" for these genes, metabolites and CyTOF features 
in an excel file. Here's a sample of the highest features in the PC1 (over the
0.5 threshold) :
```{r}
most_imp <- over[which(over$overexp > 0.05),] %>% 
  rownames_to_column("feature") %>% 
  mutate(feature = gsub("^[a-z]+_", "", feature)) %>% 
  arrange(desc(overexp))
most_imp
```



## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
