---
title: "BioGVHD project: integration of CyTOF, Metabolomics and RNA features in the St Louis cohort"
author: "Helena"
date: "06/01/2020"
output:
  html_document:
    toc: true
    toc_depth: 4
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r imports, echo=FALSE, results='hide', warning = FALSE}
suppressPackageStartupMessages({
  library(tidyverse)
  library(openxlsx)
  library(BioGVHD)
  library(data.table)
  library(ggplot2)
  library(igraph)
  library(patchwork)
})
options("scipen"=100)
```

##Focusing on the recipients

We can first focus on the features that we extracted as informative in the 
recipients, to see if, taken together, they would be more informative than 
when working on one data type alone.

###Tolerant versus non tolerant recipients

Since looking at differences between the three groups consisting of primary, 
secondary and non tolerant patients at the same time wasn't feasible, we divided
our problem in two distinct sub-problems. We first focus on the features that
seemed to play a role when trying to disinguish non tolerant from tolerant 
recipients.

We first need to gather the features that were extracted from the three data
sources:

- CyTOF features that were selected in both the St Louis and the Cryostem cohort, 
  with a threshold of 0.90. These features had the same behaviour in both cohorts.

```{r, echo = FALSE}
tmp <- readRDS("../data/cyto/log2/recip/pctgs_sel_ft_recip_TNT_90.RDS")
sel_beh <- read.xlsx("../data/cyto_national/log2/recip/perm_val_beh_TNT_90_thresh.xlsx")

ft_recip_cyto <- tmp %>% 
  dplyr::select(sel_beh$features) 
```

- Metabolomics features that were selected in both the St Louis and the Cryostem cohort, 
  with a threshold of 0.90. These features had the same behaviour in both cohorts.

```{r, echo = FALSE}
tmp <- readRDS("../data/metabo/v2_012020/recip/pctgs_sel_ft_recip_TNT_90.RDS")
sel_beh <- read.xlsx("../data/metabo/v2_012020/recip/perm_val_beh_TNT_90_thresh.xlsx")

ft_recip_metabo <- tmp %>% 
  dplyr::select(sel_beh$features) 
```

- Transcriptomics features that were selected in both the St Louis and the Cryostem cohort, 
  with a threshold of 0.90. These features had the same behaviour in both cohorts.

```{r, echo = FALSE}
load("../data/RNAseq/recip_only/rna_recip_louis_TNT_90.RData")
sel_beh <- read.xlsx("../data/RNAseq/recip_only/perm_val_beh_TNT_90_thresh.xlsx")

ft_recip_transcripto <- rna_recip_louis_TNT_90 %>% 
  dplyr::select(sel_beh$sel_beh) 
```

I extract informations on the recipients group, age, gender, ... to be able to 
use this information later on.

```{r, echo = FALSE}
samp_rd <- read.xlsx("~/Documents/VIB/Projects/Integrative_Paris/Local_cohort/Data synthesis local cohort Saint-Louis 17012019.xlsx")
samp_rd$DOB <- as.Date(samp_rd$DOB, format = "%d.%m.%Y")
samp_rd$DOG <- as.Date(samp_rd$DOG, format = "%d.%m.%Y")
samp_rd$DATEOFSAMPLE <- as.Date(samp_rd$DATEOFSAMPLE, format = "%d.%m.%Y")
samp_rd <- samp_rd %>% 
  dplyr::filter(!is.na(METABONAME)) 
complete_couples <- samp_rd$COUPLENUMBER[which(duplicated(samp_rd$COUPLENUMBER))]

couple_info <- samp_rd %>% 
  dplyr::filter(COUPLENUMBER %in% complete_couples) %>% 
  arrange(COUPLENUMBER) %>% 
  column_to_rownames("METABONAME") %>% 
  rownames_to_column("rn") %>% 
  mutate(GROUP = as.factor(tolower(GROUP))) %>% 
  column_to_rownames("rn")
  
gender_donors <- couple_info$GENDER[c(TRUE,FALSE)]
gender_recip <- couple_info$GENDER[c(FALSE,TRUE)]
gender_comp <- paste0(gender_donors, gender_recip)

age_recip <- couple_info$DOG[c(FALSE,TRUE)] - couple_info$DOB[c(FALSE,TRUE)]

ifelse(grepl("R", rownames(couple_info)[1]), 
       cmv_donors <- couple_info$CMVStatus[c(FALSE,TRUE)],
       cmv_donors <- couple_info$CMVStatus[c(TRUE,FALSE)])

ifelse(grepl("R", rownames(couple_info)[1]), 
       cmv_recip <- couple_info$CMVStatus[c(TRUE,FALSE)],
       cmv_recip <- couple_info$CMVStatus[c(FALSE,TRUE)])

cmv_comp <- paste0(cmv_donors, cmv_recip)

couple_info <- couple_info %>% 
  rownames_to_column("rownames") %>% 
  mutate(gender_comp = rep(gender_comp, each = 2),
         age_recip = rep(age_recip, each = 2),
         cmv_comp = rep(cmv_comp, each = 2)) %>% 
  column_to_rownames("rownames")

head(couple_info[,c(1,2,3,5,6,55,56,57)])
```

I add a prefix ("cyt_", "rna_" or "met_") in front of the features names, and 
I then merge all features together:

```{r, echo = FALSE}
colnames(ft_recip_cyto) <- paste0("cyt_", colnames(ft_recip_cyto))
colnames(ft_recip_metabo) <- paste0("met_", colnames(ft_recip_metabo))

# I change the rownames in the metabo file so that they correspond to cryostem Ids
info_tmp <- couple_info[rownames(ft_recip_metabo),]
rownames(ft_recip_metabo) <- info_tmp$Id.Cryostem.R

colnames(ft_recip_transcripto) <- paste0("rna_", colnames(ft_recip_transcripto))

ft_integ <- ft_recip_cyto %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  inner_join(ft_recip_metabo %>% rownames_to_column("Id.Cryostem.R") , by = c("Id.Cryostem.R")) %>% 
  inner_join(ft_recip_transcripto %>% rownames_to_column("Id.Cryostem.R") , by = c("Id.Cryostem.R")) 
print(paste0("We have information on the CyTOF, metabolomics and transcriptomics data for ",
             nrow(ft_integ), " patients in the St Louis cohort."))

saveRDS(ft_integ, file = "../data/integ/cyt_met_rna_beh_features.RDS")
```

####Volcano plot analysis

We can do a volcano plot on the CyTOF features  metabolites and genes of the non
versus tolerant recipients:
```{r, echo = FALSE}
complete_patients <- ft_integ$Id.Cryostem.R
info_complete_patients <- couple_info %>%
  rownames_to_column("METABONAME") %>% 
  dplyr::filter(Id.Cryostem.R %in% complete_patients) %>% 
  arrange(Id.Cryostem.R) %>% 
  mutate(group = as.character(GROUP))
info_complete_patients$group[which(info_complete_patients$group != "non_tolerant")] <- "tolerant"
info_complete_patients$group <- as.factor(info_complete_patients$group)

mat4volcano <- ft_integ %>% 
  column_to_rownames("Id.Cryostem.R")
info_tmp <- info_complete_patients %>% 
  column_to_rownames("Id.Cryostem.R")
info_tmp <- info_tmp[rownames(mat4volcano),]
mat4volcano <- mat4volcano %>% 
  rownames_to_column("rn") %>% 
  mutate(group = info_tmp$group) %>% 
  dplyr::select(group, everything()) %>% 
  column_to_rownames("rn")
```

```{r, echo = FALSE}
res <- metabolomics::TwoGroup(mat4volcano)
# pdf("~/Desktop/volcanoplot.pdf", width = 24, height = 20)
metabolomics::VolcanoPlot(res$output[,4], res$output[,2], cexlab = 0.6)
# dev.off()

metabolomics::MetBoxPlots(mat4volcano, "rna_ADAMTS2",cols = c("red", "blue"),main = "rna_ADAMTS2")
metabolomics::MetBoxPlots(mat4volcano, "cyt_X23_DN..86.....CD8low..13...TSCM.like",cols = c("red", "blue"),main = "cyt_X23_DN..86.....CD8low..13...TSCM.like")
metabolomics::MetBoxPlots(mat4volcano, "met_androsterone.sulfate",cols = c("red", "blue"),main = "met_androsterone.sulfate")
```

#### Correlation analysis

```{r, eval = FALSE}
#pdf("~/Desktop/correlationplot_integration_recip_TNT.pdf", width = 16, height = 16)
gr <- plot_correlations(mat4volcano, compar = "TNT")
#dev.off()

```

#### Random forest modeling

To see if the features that we selected can help us to clssify the patients, we
apply a 5 fold crosse validation setting on the 23 patients for which we have 
CyTOF, metabolomics and transcriptomics information.

```{r}
set.seed(1)
cv_id <- list(c(1:5), c(6:10), c(11:15), c(16:20), c(21:23))
cv_accuracies <- lapply(seq_along(cv_id), function(idx){
  train_set <- mat4volcano[-cv_id[[idx]], c("group", colnames(ft_integ)[-1])]
  test_set <- mat4volcano[cv_id[[idx]], c(colnames(ft_integ)[-1])]
  true_labels <- mat4volcano[cv_id[[idx]], "group"]
  rf <- ranger::ranger(group~., train_set, num.trees = 5000,
                           importance = "impurity")
  pred <- predict(rf, test_set)
  acc <- ( length(which(pred$predictions == true_labels))/length(true_labels) )
})

accuracy <- mean(unlist(cv_accuracies))
accuracy
```


####Principal component analysis

We can first generate a PCA on the data:
```{r, echo = FALSE}
mat4pca <- ft_integ %>% 
  dplyr::select(-Id.Cryostem.R) %>%  
  apply(2, scale) %>% 
  as.matrix() 
rownames(mat4pca) <- ft_integ$Id.Cryostem.R
saveRDS(mat4pca, file ="../data/integ/mat4_pca_integ_louis.RDS")
pca <- prcomp(mat4pca)
```

The first principal component seems to hold much more information that the others
```{r, echo = FALSE}
plot(pca)
```

We can investigate how much of the principal components are driven by the 
different data types.
```{r, echo = FALSE}
pca_var <- pca$rotation %>% as.data.frame
pca_ft <- rep("x", nrow(pca_var))
pca_ft[grep("cyt_", rownames(pca_var))] <- "cyto"
pca_ft[grep("met_", rownames(pca_var))] <- "metabo"
pca_ft[grep("rna_", rownames(pca_var))] <- "rna"
pca_var <- pca_var %>% 
  rownames_to_column("rn") %>% 
  mutate(feature_orig = pca_ft)
```

The RNAseq data variability was captured much more than the other data types.
This is most probably due to the fact that we selected 278 genes, against
42 metabolites and 24 CyTOF features only.
```{r, echo = FALSE}
pc1 <- pca_var %>% 
  column_to_rownames("rn") %>% 
  group_by(feature_orig) %>% 
  summarise_all(sum)
pc1m <- pc1 %>% gather(PC, value, -feature_orig) %>% 
  mutate(label_ypos = round(value))

pc1m_sub <- pc1m[1:30,] %>% 
  mutate(PC = factor(PC, levels = c("PC1", "PC2", "PC3", "PC4", "PC5", "PC6", 
                                    "PC7", "PC8", "PC9", "PC10")))

ggplot(data=pc1m_sub, aes(x=PC, y=value, fill=feature_orig)) +
  geom_bar(stat="identity")+
  # geom_text(aes(y=label_ypos, label=value), vjust=1.6, 
  #           color="white", size=3.5)+
  #scale_fill_brewer(palette="Paired")+
  scale_fill_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  theme_minimal()
```

We can see how these principal components are correlated with tolerance. 
The first PC is very correlated with the tolerance. The PC4 (and slightly 
PC8 and PC10) also look correlated with tolerance.
```{r, echo = FALSE}
complete_patients <- rownames(mat4pca)
info_complete_patients <- couple_info %>%
  rownames_to_column("METABONAME") %>% 
  dplyr::filter(Id.Cryostem.R %in% complete_patients) %>% 
  arrange(Id.Cryostem.R) %>% 
  mutate(group = as.character(GROUP))
info_complete_patients$group[which(info_complete_patients$group != "non_tolerant")] <- "tolerant"
info_complete_patients$group <- as.factor(info_complete_patients$group)

pca_2plot <- left_join(ft_integ, info_complete_patients, by = c("Id.Cryostem.R")) %>% 
  mutate(PC1 = pca$x[,1],
         PC2 = pca$x[,2],
         PC3 = pca$x[,3],
         PC4 = pca$x[,4])

barplot(abs(cor(pca$x[,1:10], as.numeric(pca_2plot$group))[,1]), ylab = "Correlation with tolerance")
saveRDS(list(pca = pca, pca_2plot = pca_2plot), file = "../data/integ/pca_TNT_st_louis_recip.RDS")
```

We can then see how the PCs are correlated with the different informations 
that we have on the patients.

The first PC seems to be most correlated with group, cGVHD and ABO.incompatibility:

```{r, echo = FALSE}
totest <- pca_2plot[,c(348, 366:374, 376, 390, 400:403)]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,1], totest)), horiz = TRUE, las = 2, cex.names = 0.3, xlim = c(0,0.6))
```

The second PC seems to be slightly correlated with the disorder group:

```{r, echo = FALSE}
totest <- pca_2plot[,c(348, 366:374, 376, 390, 400:403)]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,2], totest)), horiz = TRUE, las = 2, cex.names = 0.3, xlim = c(0,0.6))
```

The third PC seems to be slightly correlated with aGVHD and ABO.incompatibility:

```{r, echo = FALSE}
totest <- pca_2plot[,c(348, 366:374, 376, 390, 400:403)]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,3], totest)), horiz = TRUE, las = 2, cex.names = 0.3, xlim = c(0,0.6))
```

The fourth PC seems to be most correlated with group and age_recip:

```{r, echo = FALSE}
totest <- pca_2plot[,c(348, 366:374, 376, 390, 400:403)]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,4], totest)), horiz = TRUE, las = 2, cex.names = 0.3, xlim = c(0,0.6))
```

Since the 1st and 4th PCs seem to be most correlated with tolerance, we visualise
the patients in these two dimensions.
We observe that these 2 components seem to separate the tolerant recipients 
(on the bottom-left) from the non tolerant recipients (on the top-right).
We can also see how the recipients are displayed in these two dimensions 
according to ABO incompatibility, recipients age and cGVHD.
```{r, echo = FALSE}
p1 <- ggplot(pca_2plot, aes(x= PC1, y= PC4, col= group, label = Id.Cryostem.R)) + 
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Separation of tolerant vs non tolerant recipients")
p2 <- ggplot(pca_2plot, aes(x= PC1, y= PC4, col= ABO.incompatibility, label = Id.Cryostem.R)) + 
  scale_color_manual(values = c("chartreuse2", "black", "gray48")) +
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Separation of recipients based on ABO.incompatibility")
p3 <- ggplot(pca_2plot, aes(x= PC1, y= PC4, col= age_recip, label = Id.Cryostem.R)) +
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Gradient in the recipients based on age")
p4 <- ggplot(pca_2plot, aes(x= PC1, y= PC4, col= as.factor(cGVHD), label = Id.Cryostem.R)) +
  scale_color_manual(values = c("black", "red")) +
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Separation of recipients based on cGVHD")
{p1 + p2}/{p3 + p4}
```

We can also map the cryostem patients on thiese PCs:
```{r}
other <- readRDS("~/Documents/VIB/Projects/Integrative_Paris/Integrative/outputs/data/integ_national/pca_2plot.RDS")
other4pca <- other %>% dplyr::select(colnames(mat4pca))
rownames(other4pca) <- other$Id.Cryostem.R

new_coords <- scale(other4pca, pca$center, pca$scale) %*% pca$rotation 

other_2plot <- other %>%
  dplyr::select(Id.Cryostem.R, group, age_recip, cGVHD) %>% 
  mutate(PC1 = new_coords[,1],
         PC4 = new_coords[,4])

all_2plot <- rbind(pca_2plot %>% 
                     dplyr::select(Id.Cryostem.R, group, age_recip, cGVHD, PC1, PC4) %>% 
                     mutate(cohort = rep("St_Louis", nrow(pca_2plot))),
                   other_2plot %>% 
                     mutate(cohort = rep("Cryostem", nrow(other_2plot))))

p1 <- ggplot(all_2plot, aes(x= PC1, y= PC4, col= group, label = Id.Cryostem.R)) + 
  geom_point(aes(shape = cohort, size = cohort)) +
  geom_text(aes(label=Id.Cryostem.R),hjust=-0.2, vjust=-0.2) +
  ggtitle("Separation of tolerant vs non tolerant recipients")
# p2 <- ggplot(pca_2plot, aes(x= PC1, y= PC4, col= ABO.incompatibility, label = Id.Cryostem.R)) + 
#   scale_color_manual(values = c("chartreuse2", "black", "gray48")) +
#   geom_point() +
#   geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
#   ggtitle("Separation of recipients based on ABO.incompatibility")
p3 <- ggplot(all_2plot, aes(x= PC1, y= PC4, col= age_recip, label = Id.Cryostem.R)) +
  geom_point(aes(shape = cohort, size = cohort)) +
  geom_text(aes(label=Id.Cryostem.R),hjust=-0.2, vjust=-0.2) +
  ggtitle("Gradient in the recipients based on age")
p4 <- ggplot(all_2plot, aes(x= PC1, y= PC4, col= as.factor(cGVHD), label = Id.Cryostem.R)) +
  scale_color_manual(values = c("black", "red")) +
  geom_point(aes(shape = cohort, size = cohort)) +
  geom_text(aes(label=Id.Cryostem.R),hjust=-0.2, vjust=-0.2) +
  ggtitle("Separation of recipients based on cGVHD")
{p1 +p3 + p4}
```

We can see the loadings of the different features for each PC
```{r}
pc <- "PC2"

pc_loadings <- pca$rotation[names(sort(abs(pca$rotation[,pc]))),pc]
#highest <- pc_loadings[(length(pc_loadings) - 100):length(pc_loadings)]

barplot(pc_loadings, las=2, cex.names=.5, main = paste0(pc, " loadings in comparison of prim vs sec tolerant recipients"))
```


##### Correlation between features and clinical variables

Correlation between the variables and the recipient age in the St Louis cohort:
```{r}
totest <- pca_2plot[,c(2:ncol(ft_integ))]
#totest <- pca_2plot[,c(348, 366:374, 376, 390, 400:403)]
#totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca_2plot[,401], totest)), horiz = TRUE, las = 2, cex.names = 0.3, xlim = c(0,0.6))
```

##### Features expressed in the first principal component

We can first look at the PC1, since it holds most of the data's variability:
We first focus on the features that are overexpressed in the tolerant recipients 
(= the features that are most expressed on the LEFT of the PC1). Therefore, the 
lower the feature value, the more it is expressed in tolerant recipients.
```{r, echo = FALSE}
#plot(sort(pca$rotation[,1][which(pca$rotation[,1]<0)]))
overexp <- sort(pca$rotation[,1][which(pca$rotation[,1]<0)])
over <- as.data.frame(overexp)
orig <- strsplit(rownames(over), "_")
orig <- map(orig, function(x){x[1]}) %>% unlist
over$orig <- orig
ggplot(over, aes(x = orig, y = overexp, col = orig)) +
  geom_boxplot() +
  scale_fill_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  geom_jitter()
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 
```

We extract the "importance values" for these genes, metabolites and CyTOF features 
in an excel file. Here's a sample of the lowest features in the PC1 (under the
-0.6 threshold) :
```{r, echo = FALSE}
most_imp <- over[which(over$overexp < -0.06),] %>% 
  rownames_to_column("feature") %>% 
  mutate(feature = gsub("^[a-z]+_", "", feature)) %>% 
  arrange(overexp)
most_imp

most_imp2 <- most_imp %>% 
  arrange(orig)

write.xlsx(most_imp, file = "../data/integ/tol_ft_PC1.xlsx")
write.xlsx(most_imp2, file = "../data/integ/tol_ft_PC1_ordered_by_origin.xlsx")
```

We then focus on the features that are overexpressed in the non tolerant recipients 
(= the features that are most expressed on the RIGHT of the PC1). Therefore, the 
higher the feature value, the more it is expressed in non tolerant recipients.
```{r, echo = FALSE}
#plot(sort(pca$rotation[,1][which(pca$rotation[,1]>0)]))
overexp <- sort(pca$rotation[,1][which(pca$rotation[,1]>0)])
over <- as.data.frame(overexp)
orig <- strsplit(rownames(over), "_")
orig <- map(orig, function(x){x[1]}) %>% unlist
over$orig <- orig
ggplot(over, aes(x = orig, y = overexp, col = orig)) +
  geom_boxplot() +
  scale_fill_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  geom_jitter()
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 
```

We extract the "importance values" for these genes, metabolites and CyTOF features 
in an excel file. Here's a sample of the highest features in the PC1 (over the
0.5 threshold) :
```{r, echo = FALSE}
most_imp <- over[which(over$overexp > 0.05),] %>% 
  rownames_to_column("feature") %>% 
  mutate(feature = gsub("^[a-z]+_", "", feature)) %>% 
  arrange(desc(overexp))
most_imp

most_imp2 <- most_imp %>% 
  arrange(orig)

write.xlsx(most_imp, file = "../data/integ/nontol_ft_PC1.xlsx")
write.xlsx(most_imp2, file = "../data/integ/nontol_ft_PC1_ordered_by_origin.xlsx")
```

##### Features expressed in the fourth prinipal component

We can first look at the PC1, since it holds most of the data's variability:
We first focus on the features that are overexpressed in the tolerant recipients 
(= the features that are most expressed on the LEFT of the PC1). Therefore, the 
lower the feature value, the more it is expressed in tolerant recipients.
```{r, echo = FALSE}
#plot(sort(pca$rotation[,4][which(pca$rotation[,4]<0)]))
overexp <- sort(pca$rotation[,4][which(pca$rotation[,4]<0)])
over <- as.data.frame(overexp)
orig <- strsplit(rownames(over), "_")
orig <- map(orig, function(x){x[1]}) %>% unlist
over$orig <- orig
ggplot(over, aes(x = orig, y = overexp, col = orig)) +
  geom_boxplot() +
  scale_fill_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  geom_jitter()
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 
```

We extract the "importance values" for these genes, metabolites and CyTOF features 
in an excel file. Here's a sample of the lowest features in the PC1 (under the
-0.6 threshold) :
```{r, echo = FALSE}
most_imp <- over[which(over$overexp < -0.1),] %>% 
  rownames_to_column("feature") %>% 
  mutate(feature = gsub("^[a-z]+_", "", feature)) %>% 
  arrange(overexp)
most_imp

most_imp2 <- most_imp %>% 
  arrange(orig)

write.xlsx(most_imp, file = "../data/integ/tol_ft_PC4.xlsx")
write.xlsx(most_imp2, file = "../data/integ/tol_ft_PC4_ordered_by_origin.xlsx")
```

We then focus on the features that are overexpressed in the non tolerant recipients 
(= the features that are most expressed on the TOP of the PC4). Therefore, the 
higher the feature value, the more it is expressed in non tolerant recipients.
```{r, echo = FALSE}
#plot(sort(pca$rotation[,4][which(pca$rotation[,4]>0)]))
overexp <- sort(pca$rotation[,4][which(pca$rotation[,4]>0)])
over <- as.data.frame(overexp)
orig <- strsplit(rownames(over), "_")
orig <- map(orig, function(x){x[1]}) %>% unlist
over$orig <- orig
ggplot(over, aes(x = orig, y = overexp, col = orig)) +
  geom_boxplot() +
  scale_fill_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  geom_jitter()
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 
```

We extract the "importance values" for these genes, metabolites and CyTOF features 
in an excel file. Here's a sample of the highest features in the PC1 (over the
0.5 threshold) :
```{r, echo = FALSE}
most_imp <- over[which(over$overexp > 0.1),] %>% 
  rownames_to_column("feature") %>% 
  mutate(feature = gsub("^[a-z]+_", "", feature)) %>% 
  arrange(desc(overexp))
most_imp

most_imp2 <- most_imp %>% 
  arrange(orig)

write.xlsx(most_imp, file = "../data/integ/non_tol_ft_PC4.xlsx")
write.xlsx(most_imp2, file = "../data/integ/non_tol_ft_PC4_ordered_by_origin.xlsx")
```

<!-- #### Correlations between features overexpressed in PC1 and PC4 -->

```{r, echo = FALSE}
# tolpc1 <- pca$rotation[,1][which(pca$rotation[,1]< -(0.05))]
# tolpc4 <- pca$rotation[,4][which(pca$rotation[,4]< -(0.05))]
# nontolpc1 <- pca$rotation[,1][which(pca$rotation[,1]>0.05)]
# nontolpc4 <- pca$rotation[,4][which(pca$rotation[,4]>0.05)]
# 
# ft_tol <- c(names(tolpc1), names(tolpc4))
# ft_nontol <- c(names(nontolpc1), names(nontolpc4))
```

```{r, echo = FALSE}
# plot_correlations(pca_2plot[,c("group", ft_tol)], compar = "TNT")
```

```{r, echo = FALSE}
# plot_correlations(pca_2plot[,c("group", ft_nontol)], compar = "TNT")
```

We can try to build a model that would predict a patients output (tolerant or not)
based on the PCs:
```{r, echo = FALSE}
data2model <- as.data.frame(pca$x) %>% mutate(group = pca_2plot$group)

set.seed(1) 

# Define training control
train.control <- caret::trainControl(method = "cv", number = 5)
# Train the model

model <- caret::train(group ~., data = data2model, method = "svmLinearWeights2",
               trControl = train.control)
# Summarize the results
print(model)
```

#### Compare the features found in the PC1 and PC4 to the MOFA features:

In MOFA, the factor one is most tolerance related, it consists mainly of CyTOF 
and Metabo features. In this MOFA factor, the recipients are nocely split into 
non tolerant (low loadings) to tolerant (high loadings on the first factor).
```{r}
cyto_MOFA <- read.csv("../data/integ/MOFA_res/MOFA_cyto_4F_weights.csv")
metabo_MOFA <- read.csv("../data/integ/MOFA_res/MOFA_meta_4F_weights.csv")
rna_MOFA <- read.csv("../data/integ/MOFA_res/MOFA_RNA_4F_weights.csv")
MOFA_res <- rbind(cyto_MOFA, metabo_MOFA, rna_MOFA) %>% 
  mutate(orig = c(rep("cyt", nrow(cyto_MOFA)),
                  rep("met", nrow(metabo_MOFA)),
                  rep("rna", nrow(rna_MOFA))))
```

PC1 loadings: (low loadings: tolerant ++, high loadings: non tolerant)
```{r}
# pc1_tol <- read.xlsx("../data/integ/tol_ft_PC1.xlsx")
overexp <- as.data.frame(pca$rotation[,c(1,4)]) %>% 
  rownames_to_column("X") %>% 
  magrittr::set_colnames(c("X", "PC1", "PC4"))
orig_split <- strsplit(overexp$X, "_")
orig <- map(orig_split, function(x){x[1]}) %>% unlist

overexp <- overexp %>% 
  mutate(orig = orig,
         X = gsub("^[a-z]+_", "", X))
```

We can then compare the results of the PCA (focusing on the PCs 1 and 4, which
were most associated with tolerance), to the 1st factor in MOFA ("LF1"):
The concordance between the cyto and metabo features is quite good. The MOFA LF1
didn't contain RNA features, which explains the uggly horizontal trend for this 
datatype.
```{r}
both <- overexp %>% 
  left_join(MOFA_res, by = c("X", "orig")) %>% 
  mutate(PC1 = -PC1,
         PC4 = -PC4)

ggplot(both, aes(x = PC1, y = LF1, col = orig)) +
  geom_point() +
  scale_fill_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  ggtitle("Concordance between MOFA and PCA results")

ggplot(both, aes(x = PC4, y = LF1, col = orig)) +
  geom_point() +
  scale_fill_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  ggtitle("Concordance between MOFA and PCA results")
```

I can make a heatmap of the most important genes along the PC1:
```{r}
pc1_rna_table <- overexp %>% 
  dplyr::filter(orig == "rna") %>% 
  mutate(PC1 = abs(PC1)) %>% 
  arrange(desc(PC1))
pc1_rna_features <- pc1_rna_table$X[1:20]

rna_table <- ft_integ[,paste0("rna_", "", pc1_rna_features)]
rownames(rna_table) <- ft_integ$Id.Cryostem.R
colnames(rna_table) <- gsub("rna_", "", colnames(rna_table))

info_table <- info_complete_patients %>% 
  mutate(cGVHD = as.factor(cGVHD)) %>% 
  column_to_rownames("Id.Cryostem.R") 
info_table <- info_table[rownames(rna_table),]

# arrange patients per PC1:
pca_table <- pca_2plot %>% 
  select(Id.Cryostem.R, PC1, PC4) %>% 
  arrange(PC1) 

info_table <- info_table[pca_table$Id.Cryostem.R,]
rna_table <- rna_table[pca_table$Id.Cryostem.R,]

pheatmap::pheatmap(t(rna_table), cluster_cols = FALSE,
                   annotation_col = info_table[,c("group", "cGVHD", "ABO.incompatibility")], 
                   annotation_colors = list(group = c(non_tolerant = "red", tolerant = "blue"),
                                            ABO.incompatibility = c(Major = "red", 
                                                                    Minor = "orange", 
                                                                    Compatible = "blue")))
```

I can make a heatmap of the most important metabolites along the PC1:
```{r}
pc1_met_table <- overexp %>% 
  dplyr::filter(orig == "met") %>% 
  mutate(PC1 = abs(PC1)) %>% 
  arrange(desc(PC1))
pc1_met_features <- pc1_met_table$X[1:20]

met_table <- ft_integ[,paste0("met_", "", pc1_met_features)]
rownames(met_table) <- ft_integ$Id.Cryostem.R
colnames(met_table) <- gsub("met_", "", colnames(met_table))

info_table <- info_complete_patients %>% 
  mutate(cGVHD = as.factor(cGVHD)) %>% 
  column_to_rownames("Id.Cryostem.R") 
info_table <- info_table[rownames(met_table),]

# arrange patients per PC1:
pca_table <- pca_2plot %>% 
  select(Id.Cryostem.R, PC1, PC4) %>% 
  arrange(PC1) 

info_table <- info_table[pca_table$Id.Cryostem.R,]
met_table <- met_table[pca_table$Id.Cryostem.R,]

pheatmap::pheatmap(t(met_table), cluster_cols = FALSE,
                   annotation_col = info_table[,c("group", "age_recip", "gender_comp")], 
                   annotation_colors = list(group = c(non_tolerant = "red", tolerant = "springgreen3"),
                                            gender_comp = c(FF = "hotpink", FM = "brown1",
                                                                    MF = "cornflowerblue", 
                                                                    MM = "springgreen2")))
```

I can make a heatmap of the most important CyTOF features along the PC1:
```{r}
pc1_cyt_table <- overexp %>% 
  dplyr::filter(orig == "cyt") %>% 
  mutate(PC1 = abs(PC1)) %>% 
  arrange(desc(PC1))
pc1_cyt_features <- pc1_cyt_table$X[1:20]

cyt_table <- ft_integ[,paste0("cyt_", "", pc1_cyt_features)]
rownames(cyt_table) <- ft_integ$Id.Cryostem.R
colnames(cyt_table) <- gsub("cyt_", "", colnames(cyt_table))

info_table <- info_complete_patients %>% 
  mutate(cGVHD = as.factor(cGVHD)) %>% 
  column_to_rownames("Id.Cryostem.R") 
info_table <- info_table[rownames(cyt_table),]

# arrange patients per PC1:
pca_table <- pca_2plot %>% 
  select(Id.Cryostem.R, PC1, PC4) %>% 
  arrange(PC1) 

info_table <- info_table[pca_table$Id.Cryostem.R,]
cyt_table <- cyt_table[pca_table$Id.Cryostem.R,]

pheatmap::pheatmap(t(cyt_table), cluster_cols = FALSE,
                   annotation_col = info_table[,c("group", "age_recip", "gender_comp")], 
                   annotation_colors = list(group = c(non_tolerant = "red", tolerant = "springgreen3"),
                                            gender_comp = c(FF = "hotpink", FM = "brown1",
                                                                    MF = "cornflowerblue", 
                                                                    MM = "springgreen2")))
```

I can make a heatmap of the most important genes along the PC4:
```{r}
pc4_rna_table <- overexp %>% 
  dplyr::filter(orig == "rna") %>% 
  mutate(PC4 = abs(PC4)) %>% 
  arrange(desc(PC4))
pc4_rna_features <- pc4_rna_table$X[1:20]

rna_table <- ft_integ[,paste0("rna_", "", pc4_rna_features)]
rownames(rna_table) <- ft_integ$Id.Cryostem.R
colnames(rna_table) <- gsub("rna_", "", colnames(rna_table))

info_table <- info_complete_patients %>% 
  mutate(cGVHD = as.factor(cGVHD)) %>% 
  column_to_rownames("Id.Cryostem.R") 
info_table <- info_table[rownames(rna_table),]

# arrange patients per PC1:
pca_table <- pca_2plot %>% 
  select(Id.Cryostem.R, PC1, PC4) %>% 
  arrange(PC4) 

info_table <- info_table[pca_table$Id.Cryostem.R,]
rna_table <- rna_table[pca_table$Id.Cryostem.R,]

pheatmap::pheatmap(t(rna_table), cluster_cols = FALSE,
                   annotation_col = info_table[,c("group", "age_recip")], 
                   annotation_colors = list(group = c(non_tolerant = "red", tolerant = "blue")))
```

I can make a heatmap of the most important metabolites along the PC4:
```{r}
pc4_met_table <- overexp %>% 
  dplyr::filter(orig == "met") %>% 
  mutate(PC4 = abs(PC4)) %>% 
  arrange(desc(PC4))
pc4_met_features <- pc4_met_table$X[1:20]

met_table <- ft_integ[,paste0("met_", "", pc4_met_features)]
rownames(met_table) <- ft_integ$Id.Cryostem.R
colnames(met_table) <- gsub("met_", "", colnames(met_table))

info_table <- info_complete_patients %>% 
  mutate(cGVHD = as.factor(cGVHD)) %>% 
  column_to_rownames("Id.Cryostem.R") 
info_table <- info_table[rownames(met_table),]

# arrange patients per PC1:
pca_table <- pca_2plot %>% 
  select(Id.Cryostem.R, PC1, PC4) %>% 
  arrange(PC4) 

info_table <- info_table[pca_table$Id.Cryostem.R,]
met_table <- met_table[pca_table$Id.Cryostem.R,]

pheatmap::pheatmap(t(met_table), cluster_cols = FALSE,
                   annotation_col = info_table[,c("group", "age_recip")],
                   annotation_colors = list(group = c(non_tolerant = "red", tolerant = "blue")))
```


#### Separate data type analyses

For comparison, we can see if we would have managed to separate the tolerant 
from non tolerant recipients as well without integrating the data:

##### CyTOF:

We perform PCA using only the 24 features that were selected as informative 
in both cohorts:

```{r, echo = FALSE}
cyt4pca <- ft_recip_cyto %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::filter(Id.Cryostem.R %in% ft_integ$Id.Cryostem.R) %>% 
  column_to_rownames("Id.Cryostem.R")
pca_cyto <- ade4::dudi.pca(cyt4pca, nf = 10, scannf = F)
```

The first comoponent seems to be capturing most of the data variability:
```{r, echo = FALSE}
plot(pca_cyto$eig, pch = 19)
```

```{r, echo = FALSE}
info_tmp <- couple_info %>% 
  rownames_to_column("rn") %>% 
  dplyr::filter(Id.Cryostem.R %in% rownames(cyt4pca)) %>%
  mutate(group = as.character(tolower(GROUP))) %>% 
  column_to_rownames("Id.Cryostem.R")
info_tmp <- info_tmp[rownames(cyt4pca),]
info_tmp$group[which(info_tmp$group != "non_tolerant")] <- "tolerant"
info_tmp$group <- as.factor(info_tmp$group)
```

We can see how these principal components are correlated with tolerance. 
The first PC is very correlated with the tolerance. The PC6 is the 2nd mostly
correlated with tolerance.
```{r, echo = FALSE}
barplot(abs(cor(pca_cyto$li[,1:10], as.numeric(info_tmp$group))[,1]))
```

The non tolerant recipients are slightly less well separated from the tolerant 
recipients than in the integrative PCA. 
```{r, echo = FALSE}
#information on the patients: 
info_pca <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R")
info_pca <- info_pca[rownames(cyt4pca),]
ade4::s.class(pca_cyto$li[,c(1,6)], info_pca$group, col=c("red", "blue"))
# ade4::s.class(pca_cyto$li[,c(1,6)], as.factor(info_pca$cGVHD), 
#               col = c("black", "red", "lightgreen", "green3"))
```

We can also have a look at the features that are driving these two components:
```{r, echo = FALSE}
keep1 <- order(abs(pca_cyto$c1$CS1), decreasing = T)[1:20]
# keep2 <- order(abs(pca_cyto$c1$CS2), decreasing = T)[1:5]
# tokeep <- unique(c(keep1, keep2))
tokeep <- keep1
ade4::s.arrow(pca_cyto$c1[tokeep,c(1,6)], boxes = F, clabel = 0.5)
```

We can try to build a model that would predict a patients output (tolerant or not)
based on the PCs:
```{r, echo = FALSE}
data2model <- as.data.frame(pca_cyto$li) %>% mutate(group = info_pca$group)

set.seed(1) 

# Define training control
train.control <- caret::trainControl(method = "cv", number = 5)
# Train the model

model <- caret::train(group ~., data = data2model, method = "svmLinearWeights2",
               trControl = train.control)
# Summarize the results
print(model)
```

The accuracy is quite lower than in the integrative model.

To see if the features that we selected can help us to clssify the patients, we
can also apply a 5 fold cross validation setting on the CyTOF features of the 
same 23 patients.

```{r}
set.seed(1)
info_rf <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R")
info_rf <- info_rf[rownames(cyt4pca),]
cyt4rf <- cbind(info_rf$group, cyt4pca)
colnames(cyt4rf)[1] <- "group"

cv_id <- list(c(1:5), c(6:10), c(11:15), c(16:20), c(21:23))
cv_accuracies <- lapply(seq_along(cv_id), function(idx){
  train_set <- cyt4rf[-cv_id[[idx]], ]
  test_set <- cyt4rf[cv_id[[idx]], -1]
  true_labels <- cyt4rf[cv_id[[idx]], "group"]
  rf <- ranger::ranger(group~., train_set, num.trees = 5000,
                           importance = "impurity")
  pred <- predict(rf, test_set)
  acc <- ( length(which(pred$predictions == true_labels))/length(true_labels) )
})

accuracy <- mean(unlist(cv_accuracies))
accuracy
```

##### Metabolomics:
```{r, echo = FALSE}
met4pca <- ft_recip_metabo %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::filter(Id.Cryostem.R %in% ft_integ$Id.Cryostem.R) %>% 
  column_to_rownames("Id.Cryostem.R")
pca_metabo <- ade4::dudi.pca(met4pca, nf = 10, scannf = F)
```

Variability captured by each component:
```{r, echo = FALSE}
plot(pca_metabo$eig, pch = 19)
```

```{r, echo = FALSE}
info_tmp <- couple_info %>% 
  rownames_to_column("rn") %>% 
  dplyr::filter(Id.Cryostem.R %in% rownames(met4pca)) %>%
  mutate(group = as.character(tolower(GROUP))) %>% 
  column_to_rownames("Id.Cryostem.R")
info_tmp <- info_tmp[rownames(met4pca),]
info_tmp$group[which(info_tmp$group != "non_tolerant")] <- "tolerant"
info_tmp$group <- as.factor(info_tmp$group)
```

We can see how these principal components are correlated with tolerance. 
The first PC is very correlated with the tolerance. The PC4 is the 2nd mostly
correlated with tolerance.
```{r, echo = FALSE}
barplot(abs(cor(pca_metabo$li[,1:10], as.numeric(info_tmp$group))[,1]))
```

The non tolerant recipients are slightly less well separated from the tolerant 
recipients than in the integrative PCA. 
```{r, echo = FALSE}
#information on the patients: 
info_pca <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R")
info_pca <- info_pca[rownames(met4pca),]
ade4::s.class(pca_metabo$li[,c(1,4)], info_pca$group, col=c("red", "blue"))
# ade4::s.class(pca_cyto$li[,c(1,6)], as.factor(info_pca$cGVHD), 
#               col = c("black", "red", "lightgreen", "green3"))
```

```{r, echo = FALSE}
keep1 <- order(abs(pca_metabo$c1$CS1), decreasing = T)[1:20]
# keep2 <- order(abs(pca_cyto$c1$CS2), decreasing = T)[1:5]
# tokeep <- unique(c(keep1, keep2))
tokeep <- keep1
ade4::s.arrow(pca_metabo$c1[tokeep, c(1,4)], boxes = F, clabel = 0.5)
```

We can try to build a model that would predict a patients output (tolerant or not)
based on the PCs:
```{r, echo = FALSE}
data2model <- as.data.frame(pca_metabo$li) %>% mutate(group = info_pca$group)

set.seed(1) 

# Define training control
train.control <- caret::trainControl(method = "cv", number = 5)
# Train the model

model <- caret::train(group ~., data = data2model, method = "svmLinearWeights2",
               trControl = train.control)
# Summarize the results
print(model)
```

Again, the accuracy of the model is slightly lower than the accuracy of the integrative model.

To see if the features that we selected can help us to clssify the patients, we
can also apply a 5 fold cross validation setting on the metabolites of the 
same 23 patients.

```{r}
set.seed(1)
info_rf <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R")
info_rf <- info_rf[rownames(met4pca),]
met4rf <- cbind(info_rf$group, met4pca)
colnames(met4rf)[1] <- "group"

cv_id <- list(c(1:5), c(6:10), c(11:15), c(16:20), c(21:23))
cv_accuracies <- lapply(seq_along(cv_id), function(idx){
  train_set <- met4rf[-cv_id[[idx]], ]
  test_set <- met4rf[cv_id[[idx]], -1]
  true_labels <- met4rf[cv_id[[idx]], "group"]
  rf <- ranger::ranger(group~., train_set, num.trees = 5000,
                           importance = "impurity")
  pred <- predict(rf, test_set)
  acc <- ( length(which(pred$predictions == true_labels))/length(true_labels) )
})

accuracy <- mean(unlist(cv_accuracies))
accuracy
```

##### Transcriptomics:
```{r, echo = FALSE}
rna4pca <- ft_recip_transcripto %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::filter(Id.Cryostem.R %in% ft_integ$Id.Cryostem.R) %>% 
  column_to_rownames("Id.Cryostem.R")
pca_transcripto <- ade4::dudi.pca(rna4pca, nf = 10, scannf = F)
```

Variability captured by each component:
```{r, echo = FALSE}
plot(pca_transcripto$eig, pch = 19)
```

```{r, echo = FALSE}
info_tmp <- couple_info %>% 
  rownames_to_column("rn") %>% 
  dplyr::filter(Id.Cryostem.R %in% rownames(rna4pca)) %>%
  mutate(group = as.character(tolower(GROUP))) %>% 
  column_to_rownames("Id.Cryostem.R")
info_tmp <- info_tmp[rownames(rna4pca),]
info_tmp$group[which(info_tmp$group != "non_tolerant")] <- "tolerant"
info_tmp$group <- as.factor(info_tmp$group)
```

We can see how these principal components are correlated with tolerance. 
The first PC is very correlated with the tolerance. The PC8 is the 2nd mostly
correlated with tolerance.
```{r, echo = FALSE}
barplot(abs(cor(pca_transcripto$li[,1:10], as.numeric(info_tmp$group))[,1]))
```

The non tolerant recipients are slightly less well separated from the tolerant 
recipients than in the integrative PCA. 
```{r, echo = FALSE}
#information on the patients: 
info_pca <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R")
info_pca <- info_pca[rownames(rna4pca),]
ade4::s.class(pca_transcripto$li[,c(1,8)], info_pca$group, col=c("red", "blue"))
# ade4::s.class(pca_cyto$li[,c(1,6)], as.factor(info_pca$cGVHD), 
#               col = c("black", "red", "lightgreen", "green3"))
```

```{r, echo = FALSE}
keep1 <- order(abs(pca_transcripto$c1$CS1), decreasing = T)[1:40]
# keep2 <- order(abs(pca_cyto$c1$CS2), decreasing = T)[1:5]
# tokeep <- unique(c(keep1, keep2))
tokeep <- keep1
ade4::s.arrow(pca_transcripto$c1[tokeep,c(1,8)], boxes = F, clabel = 0.5)
```

We can try to build a model that would predict a patients output (tolerant or not)
based on the PCs:
```{r, echo = FALSE}
data2model <- as.data.frame(pca_transcripto$li) %>% mutate(group = info_pca$group)

set.seed(1) 

# Define training control
train.control <- caret::trainControl(method = "cv", number = 5)
# Train the model

model <- caret::train(group ~., data = data2model, method = "svmLinearWeights2",
               trControl = train.control)
# Summarize the results
print(model)
```

It looks like, in our integrative approach, we managed to separate the recipients 
based on their tolerance better than when we look at the different data types
separately.

To see if the features that we selected can help us to clssify the patients, we
can also apply a 5 fold cross validation setting on the genes of the 
same 23 patients.

```{r}
set.seed(1)
info_rf <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R")
info_rf <- info_rf[rownames(rna4pca),]
rna4rf <- cbind(info_rf$group, rna4pca)
colnames(rna4rf)[1] <- "group"

cv_id <- list(c(1:5), c(6:10), c(11:15), c(16:20), c(21:23))
cv_accuracies <- lapply(seq_along(cv_id), function(idx){
  train_set <- rna4rf[-cv_id[[idx]], ]
  test_set <- rna4rf[cv_id[[idx]], -1]
  true_labels <- rna4rf[cv_id[[idx]], "group"]
  rf <- ranger::ranger(group~., train_set, num.trees = 5000,
                           importance = "impurity")
  pred <- predict(rf, test_set)
  acc <- ( length(which(pred$predictions == true_labels))/length(true_labels) )
})

accuracy <- mean(unlist(cv_accuracies))
accuracy
```

### Primary versus secondary tolerant recipients

We now focus on the features that seemed to play a role when trying to 
disinguish primary from secondary tolerant recipients.

We first need to gather the features that were extracted from the three data
sources:

- CyTOF features that were selected in both the St Louis and the Cryostem cohort, 
  with a threshold of 0.90. These features had the same behaviour in both cohorts.

```{r, echo = FALSE}
tmp <- readRDS("../data/cyto/log2/recip/pctgs_sel_ft_recip_PTST_90.RDS")
sel_beh <- read.xlsx("../data/cyto_national/log2/recip/perm_val_beh_PTST_90_thresh.xlsx")

ft_recip_cyto <- tmp %>% 
  dplyr::select(sel_beh$features) 
```

- Metabolomics features that were selected in both the St Louis and the Cryostem cohort, 
  with a threshold of 0.90. These features had the same behaviour in both cohorts.

```{r, echo = FALSE}
tmp <- readRDS("../data/metabo/v2_012020/recip/pctgs_sel_ft_recip_PTST_90.RDS")
sel_beh <- read.xlsx("../data/metabo_national/v2_012020/recip/perm_val_beh_PTST_90_thresh.xlsx")

ft_recip_metabo <- tmp %>% 
  dplyr::select(sel_beh$features) 
```

- Transcriptomics features that were selected in both the St Louis and the Cryostem cohort, 
  with a threshold of 0.90. These features had the same behaviour in both cohorts.

```{r, echo = FALSE}
load("../data/RNAseq/recip_only/rna_recip_louis_PS_90.RData")
sel_beh <- read.xlsx("../data/RNAseq/recip_only/perm_val_beh_PS_90_thresh.xlsx")

ft_recip_transcripto <- rna_recip_louis_PS_90 %>% 
  dplyr::select(sel_beh$sel_beh) 
```

I extract informations on the recipients group, age, gender, ... to be able to 
use this information later on.

```{r, echo = FALSE}
samp_rd <- read.xlsx("~/Documents/VIB/Projects/Integrative_Paris/Local_cohort/Data synthesis local cohort Saint-Louis 17012019.xlsx")
samp_rd$DOB <- as.Date(samp_rd$DOB, format = "%d.%m.%Y")
samp_rd$DOG <- as.Date(samp_rd$DOG, format = "%d.%m.%Y")
samp_rd$DATEOFSAMPLE <- as.Date(samp_rd$DATEOFSAMPLE, format = "%d.%m.%Y")
samp_rd <- samp_rd %>% 
  dplyr::filter(!is.na(METABONAME)) 
complete_couples <- samp_rd$COUPLENUMBER[which(duplicated(samp_rd$COUPLENUMBER))]

couple_info <- samp_rd %>% 
  dplyr::filter(COUPLENUMBER %in% complete_couples) %>% 
  arrange(COUPLENUMBER) %>% 
  column_to_rownames("METABONAME") %>% 
  rownames_to_column("rn") %>% 
  mutate(GROUP = as.factor(tolower(GROUP))) %>% 
  column_to_rownames("rn")
  
gender_donors <- couple_info$GENDER[c(TRUE,FALSE)]
gender_recip <- couple_info$GENDER[c(FALSE,TRUE)]
gender_comp <- paste0(gender_donors, gender_recip)

age_recip <- couple_info$DOG[c(FALSE,TRUE)] - couple_info$DOB[c(FALSE,TRUE)]

ifelse(grepl("R", rownames(couple_info)[1]), 
       cmv_donors <- couple_info$CMVStatus[c(FALSE,TRUE)],
       cmv_donors <- couple_info$CMVStatus[c(TRUE,FALSE)])

ifelse(grepl("R", rownames(couple_info)[1]), 
       cmv_recip <- couple_info$CMVStatus[c(TRUE,FALSE)],
       cmv_recip <- couple_info$CMVStatus[c(FALSE,TRUE)])

cmv_comp <- paste0(cmv_donors, cmv_recip)

couple_info <- couple_info %>% 
  rownames_to_column("rownames") %>% 
  mutate(gender_comp = rep(gender_comp, each = 2),
         age_recip = rep(age_recip, each = 2),
         cmv_comp = rep(cmv_comp, each = 2)) %>% 
  column_to_rownames("rownames")

head(couple_info[,c(1,2,3,5,6,55,56,57)])
```

I add a prefix ("cyt_", "rna_" or "met_") in front of the features names, and 
I then merge all features together:

```{r, echo = FALSE}
colnames(ft_recip_cyto) <- paste0("cyt_", colnames(ft_recip_cyto))
colnames(ft_recip_metabo) <- paste0("met_", colnames(ft_recip_metabo))

# I change the rownames in the metabo file so that they correspond to cryostem Ids
info_tmp <- couple_info[rownames(ft_recip_metabo),]
rownames(ft_recip_metabo) <- info_tmp$Id.Cryostem.R

colnames(ft_recip_transcripto) <- paste0("rna_", colnames(ft_recip_transcripto))

ft_integ <- ft_recip_cyto %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  inner_join(ft_recip_metabo %>% rownames_to_column("Id.Cryostem.R") , by = c("Id.Cryostem.R")) %>% 
  inner_join(ft_recip_transcripto %>% rownames_to_column("Id.Cryostem.R") , by = c("Id.Cryostem.R")) 
print(paste0("We have information on the CyTOF, metabolomics and transcriptomics data for ",
             nrow(ft_integ), " patients in the St Louis cohort."))

saveRDS(ft_integ, file = "../data/integ/cyt_met_rna_beh_PTST_features.RDS")
```

#### Volcano plot analysis

We can do a volcano plot on the CyTOF features  metabolites and genes of the non
versus tolerant recipients:
```{r, echo = FALSE}
complete_patients <- ft_integ$Id.Cryostem.R
info_complete_patients <- couple_info %>%
  rownames_to_column("METABONAME") %>% 
  dplyr::filter(Id.Cryostem.R %in% complete_patients) %>% 
  arrange(Id.Cryostem.R) %>% 
  mutate(group = as.character(GROUP))
info_complete_patients$group <- as.factor(info_complete_patients$group)

mat4volcano <- ft_integ %>% 
  column_to_rownames("Id.Cryostem.R")
info_tmp <- info_complete_patients %>% 
  column_to_rownames("Id.Cryostem.R")
info_tmp <- info_tmp[rownames(mat4volcano),]
mat4volcano <- mat4volcano %>% 
  rownames_to_column("rn") %>% 
  mutate(group = info_tmp$group) %>% 
  dplyr::select(group, everything()) %>% 
  column_to_rownames("rn")
```

```{r, echo = FALSE}
res <- metabolomics::TwoGroup(mat4volcano)
# pdf("~/Desktop/volcanoplot_integration_recip_PTST.pdf", width = 24, height = 20)
metabolomics::VolcanoPlot(res$output[,4], res$output[,2], cexlab = 0.6)
# dev.off()

metabolomics::MetBoxPlots(mat4volcano, "rna_KLRC2",cols = c("green", "blue"),main = "rna_KLRC2")
metabolomics::MetBoxPlots(mat4volcano, "rna_MAPT",cols = c("green", "blue"),main = "rna_MAPT")
metabolomics::MetBoxPlots(mat4volcano, "cyt_X.CD38._16_T.population.Naive.TSCM",cols = c("green", "blue"),main = "cyt_X.CD38._16_T.population.Naive.TSCM")
metabolomics::MetBoxPlots(mat4volcano, "rna_PRUNE2",cols = c("green", "blue"),main = "rna_PRUNE2")
```

#### Correlation analysis

```{r, echo = FALSE, eval = FALSE}
#pdf("~/Desktop/correlationplot_integration_recip_PTST.pdf", width = 16, height = 16)
gr <- plot_correlations(mat4volcano, compar = "PS")
#dev.off()

```

#### Principal component analysis

We can first generate a PCA on the data:
```{r, echo = FALSE}
mat4pca <- ft_integ %>% 
  dplyr::select(-Id.Cryostem.R) %>%  
  apply(2, scale) %>% 
  as.matrix() 
rownames(mat4pca) <- ft_integ$Id.Cryostem.R
pca <- prcomp(mat4pca, scale. = TRUE)
```

The first principal component seems to hold much more information that the others
```{r, echo = FALSE}
plot(pca)
```

We can investigate how much of the principal components are driven by the 
different data types.
```{r, echo = FALSE}
pca_var <- pca$rotation %>% as.data.frame
pca_ft <- rep("x", nrow(pca_var))
pca_ft[grep("cyt_", rownames(pca_var))] <- "cyto"
pca_ft[grep("met_", rownames(pca_var))] <- "metabo"
pca_ft[grep("rna_", rownames(pca_var))] <- "rna"
pca_var <- pca_var %>% 
  rownames_to_column("rn") %>% 
  mutate(feature_orig = pca_ft)
```

The RNAseq data variability was captured much more than the other data types.
This is most probably due to the fact that we selected 278 genes, against
42 metabolites and 24 CyTOF features only.
```{r, echo = FALSE}
pc1 <- pca_var %>% 
  column_to_rownames("rn") %>% 
  group_by(feature_orig) %>% 
  summarise_all(sum)
pc1m <- pc1 %>% gather(PC, value, -feature_orig) %>% 
  mutate(label_ypos = round(value))

pc1m_sub <- pc1m[1:30,]

ggplot(data=pc1m_sub, aes(x=PC, y=value, fill=feature_orig)) +
  geom_bar(stat="identity")+
  # geom_text(aes(y=label_ypos, label=value), vjust=1.6, 
  #           color="white", size=3.5)+
  # scale_fill_brewer(palette="Paired")+
  scale_fill_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  theme_minimal()
```

We can see how these principal components are correlated with tolerance. 
The first PC is very correlated with the tolerance. The PC2  also looks 
correlated with tolerance.
```{r, echo = FALSE}
complete_patients <- rownames(mat4pca)
info_complete_patients <- couple_info %>%
  rownames_to_column("METABONAME") %>% 
  dplyr::filter(Id.Cryostem.R %in% complete_patients) %>% 
  arrange(Id.Cryostem.R) %>% 
  mutate(group = as.character(GROUP))
info_complete_patients$group <- as.factor(info_complete_patients$group)

pca_2plot <- left_join(ft_integ, info_complete_patients, by = c("Id.Cryostem.R")) %>% 
  mutate(PC1 = pca$x[,1],
         PC2 = pca$x[,2],
         PC3 = pca$x[,3],
         PC4 = pca$x[,4])

barplot(abs(cor(pca$x[,1:10], as.numeric(pca_2plot$group))[,1]))
saveRDS(list(pca = pca, pca_2plot = pca_2plot), file = "../data/integ/pca_PTST_st_louis_recip.RDS")
```

We can then see how the PCs are correlated with the different informations 
that we have on the patients.

The first PC seems to be most correlated with group, SAL and blood incompatibility:

```{r, echo = FALSE}
# I replace missing values by 0.5 when the values are either 0 or 1
# I replace the missing values by the most represented one otherwise
pca_2plot$aGVHD[which(is.na(pca_2plot$aGVHD))] <- 0.5
pca_2plot$TBI[which(is.na(pca_2plot$TBI))] <- 0.5
pca_2plot$SAL[which(is.na(pca_2plot$SAL))] <- 0.5
pca_2plot$CONDITIONING_INTENSITY[which(is.na(pca_2plot$CONDITIONING_INTENSITY))] <- 
  names(which.max(table(pca_2plot$CONDITIONING_INTENSITY)))
pca_2plot$CONDITIONING_REGIMEN[which(is.na(pca_2plot$CONDITIONING_REGIMEN))] <- 
  names(which.max(table(pca_2plot$CONDITIONING_REGIMEN)))
pca_2plot$SOURCEOFGRAFT[which(is.na(pca_2plot$SOURCEOFGRAFT))] <- 
  names(which.max(table(pca_2plot$SOURCEOFGRAFT)))
pca_2plot$HEMATOLOGICDISORDER[which(is.na(pca_2plot$HEMATOLOGICDISORDER))] <- 
  names(which.max(table(pca_2plot$HEMATOLOGICDISORDER)))

totest <- pca_2plot[,c(32, 50:58, 60, 74, 84:87)]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,1], totest)), horiz = TRUE, las = 2, cex.names = 0.5, xlim = c(0,0.7))
```

The second PC seems to be most correlated with the source of graft, cGVHD and group:

```{r, echo = FALSE}
totest <- pca_2plot[,c(32, 50:58, 60, 74, 84:87)]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,2], totest)), horiz = TRUE, las = 2, cex.names = 0.5, xlim = c(0,0.6))
```

The third PC seems to be correlated with cmv:

```{r, echo = FALSE}
totest <- pca_2plot[,c(32, 50:58, 60, 74, 84:87)]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,3], totest)), horiz = TRUE, las = 2, cex.names = 0.5, xlim = c(0,0.6))
```

The fourth PC seems to be slightly correlated with the recipient's age:

```{r, echo = FALSE}
totest <- pca_2plot[,c(32, 50:58, 60, 74, 84:87)]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,4], totest)), horiz = TRUE, las = 2, cex.names = 0.5, xlim = c(0,0.6))
```

Since the 1st and 2nd PCs seem to be most correlated with tolerance, we visualise
the patients in these two dimensions.
We observe that these 2 components seem to separate the secondary tolerant recipients 
(on the bottom-left) from the primary tolerant recipients (on the top-right).
We can also see how the recipients are displayed in these two dimensions 
according to recipients age and aGVHD.
```{r, echo = FALSE}
ggplot(pca_2plot, aes(x= PC1, y= PC2, col= group, label = Id.Cryostem.R)) + 
  scale_color_manual(values = c("green", "blue")) +
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Separation of primary vs secondary tolerant recipients")
ggplot(pca_2plot, aes(x= PC1, y= PC2, col= SAL, label = Id.Cryostem.R)) +
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Separation of recipients based on SAL")
ggplot(pca_2plot, aes(x= PC1, y= PC2, col= SOURCEOFGRAFT, label = Id.Cryostem.R)) +
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Separation of recipients based on source of graft")
ggplot(pca_2plot, aes(x= PC1, y= PC2, col= as.factor(aGVHD), label = Id.Cryostem.R)) +
  scale_color_manual(values = c("black", "grey", "red")) +
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Separation of recipients based on aGVHD")
```



##### Features expressed in the first principal component

We can first look at the PC1, since it holds most of the data's variability:
We first focus on the features that are overexpressed in the primary tolerant recipients 
(= the features that are most expressed on the LEFT of the PC1). Therefore, the 
lower the feature value, the more it is expressed in primary tolerant recipients.
```{r, echo = FALSE}
#plot(sort(pca$rotation[,1][which(pca$rotation[,1]<0)]))
overexp <- sort(pca$rotation[,1][which(pca$rotation[,1]<0)])
over <- as.data.frame(overexp)
orig <- strsplit(rownames(over), "_")
orig <- map(orig, function(x){x[1]}) %>% unlist
over$orig <- orig
ggplot(over, aes(x = orig, y = overexp, col = orig)) +
  geom_boxplot() +
  scale_color_manual(values = c("mediumorchid4", "sandybrown")) +
  geom_jitter()
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 
```

We extract the "importance values" for these genes, metabolites and CyTOF features 
in an excel file. Here's a sample of the lowest features in the PC1 (under the
-0.2 threshold) :
```{r, echo = FALSE}
most_imp <- over[which(over$overexp < -0.02),] %>% 
  rownames_to_column("feature") %>% 
  mutate(feature = gsub("^[a-z]+_", "", feature)) %>% 
  arrange(overexp)
most_imp

most_imp2 <- most_imp %>% 
  arrange(orig)

write.xlsx(most_imp, file = "../data/integ/primtol_ft_PC1.xlsx")
write.xlsx(most_imp2, file = "../data/integ/primtol_ft_PC1_ordered_by_origin.xlsx")
```

We then focus on the features that are overexpressed in the secondary tolerant recipients 
(= the features that are most expressed on the RIGHT of the PC1). Therefore, the 
higher the feature value, the more it is expressed in secondary tolerant recipients.
```{r, echo = FALSE}
#plot(sort(pca$rotation[,1][which(pca$rotation[,1]>0)]))
overexp <- sort(pca$rotation[,1][which(pca$rotation[,1]>0)])
over <- as.data.frame(overexp)
orig <- strsplit(rownames(over), "_")
orig <- map(orig, function(x){x[1]}) %>% unlist
over$orig <- orig
ggplot(over, aes(x = orig, y = overexp, col = orig)) +
  geom_boxplot() +
  scale_fill_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  geom_jitter()
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 
```

We extract the "importance values" for these genes, metabolites and CyTOF features 
in an excel file. Here's a sample of the highest features in the PC1 (over the
0.05 threshold) :
```{r, echo = FALSE}
most_imp <- over[which(over$overexp > 0.05),] %>% 
  rownames_to_column("feature") %>% 
  mutate(feature = gsub("^[a-z]+_", "", feature)) %>% 
  arrange(desc(overexp))
most_imp

most_imp2 <- most_imp %>% 
  arrange(orig)

write.xlsx(most_imp, file = "../data/integ/sectol_ft_PC1.xlsx")
write.xlsx(most_imp2, file = "../data/integ/sectol_ft_PC1_ordered_by_origin.xlsx")
```

##### Features expressed in the third prinipal component

We first focus on the features that are overexpressed in the secondary tolerant recipients 
(= the features that are most expressed at the BOTTOM of the PC2). Therefore, the 
lower the feature value, the more it is expressed in secondary tolerant recipients.
```{r, echo = FALSE}
#plot(sort(pca$rotation[,4][which(pca$rotation[,4]<0)]))
overexp <- sort(pca$rotation[,2][which(pca$rotation[,2]<0)])
over <- as.data.frame(overexp)
orig <- strsplit(rownames(over), "_")
orig <- map(orig, function(x){x[1]}) %>% unlist
over$orig <- orig
ggplot(over, aes(x = orig, y = overexp, col = orig)) +
  geom_boxplot() +
  scale_fill_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  geom_jitter()
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 
```

We extract the "importance values" for these genes, metabolites and CyTOF features 
in an excel file. Here's a sample of the lowest features in the PC2 (under the
-0.3 threshold) :
```{r, echo = FALSE}
most_imp <- over[which(over$overexp < -0.3),] %>% 
  rownames_to_column("feature") %>% 
  mutate(feature = gsub("^[a-z]+_", "", feature)) %>% 
  arrange(overexp)
most_imp

most_imp2 <- most_imp %>% 
  arrange(orig)

write.xlsx(most_imp, file = "../data/integ/sectol_ft_PC3.xlsx")
write.xlsx(most_imp2, file = "../data/integ/sectol_ft_PC3_ordered_by_origin.xlsx")
```

We then focus on the features that are overexpressed in the primary tolerant recipients 
(= the features that are most expressed at the TOP of the PC2). Therefore, the 
higher the feature value, the more it is expressed in primary tolerant recipients.
```{r, echo = FALSE}
overexp <- sort(pca$rotation[,2][which(pca$rotation[,2]>0)])
over <- as.data.frame(overexp)
orig <- strsplit(rownames(over), "_")
orig <- map(orig, function(x){x[1]}) %>% unlist
over$orig <- orig
ggplot(over, aes(x = orig, y = overexp, col = orig)) +
  scale_fill_manual(values = c("mediumorchid4", "sandybrown")) +
  geom_boxplot() +
  geom_jitter()
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 
```

We extract the "importance values" for these genes, metabolites and CyTOF features 
in an excel file. Here's a sample of the highest features in the PC2 (over the
0.2 threshold) :
```{r, echo = FALSE}
most_imp <- over[which(over$overexp > 0.2),] %>% 
  rownames_to_column("feature") %>% 
  mutate(feature = gsub("^[a-z]+_", "", feature)) %>% 
  arrange(desc(overexp))
most_imp

most_imp2 <- most_imp %>% 
  arrange(orig)

write.xlsx(most_imp, file = "../data/integ/primtol_ft_PC3.xlsx")
write.xlsx(most_imp2, file = "../data/integ/primtol_ft_PC3_ordered_by_origin.xlsx")
```

We can try to build a model that would predict a patients output (tolerant or not)
based on the PCs:
```{r, echo = FALSE}
data2model <- as.data.frame(pca$x) %>% mutate(group = pca_2plot$group)

set.seed(1) 

# Define training control
train.control <- caret::trainControl(method = "cv", number = 5)
# Train the model

model <- caret::train(group ~., data = data2model, method = "svmLinearWeights2",
               trControl = train.control)
# Summarize the results
print(model)
```

#### Separate data type analyses

For comparison, we can see if we would have managed to separate the tolerant 
from non tolerant recipients as well without integrating the data:

##### CyTOF:

We apply PCA only on the 24 CyTOF features:
```{r, echo = FALSE}
cyt4pca <- ft_recip_cyto %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::filter(Id.Cryostem.R %in% ft_integ$Id.Cryostem.R) %>% 
  column_to_rownames("Id.Cryostem.R")
pca_cyto <- ade4::dudi.pca(cyt4pca, nf = 10, scannf = F)
```

Variability captured by each component:
```{r, echo = FALSE}
plot(pca_cyto$eig, pch = 19)
```

```{r, echo = FALSE}
info_tmp <- couple_info %>% 
  rownames_to_column("rn") %>% 
  dplyr::filter(Id.Cryostem.R %in% rownames(cyt4pca)) %>%
  mutate(group = as.character(tolower(GROUP))) %>% 
  column_to_rownames("Id.Cryostem.R")
info_tmp <- info_tmp[rownames(cyt4pca),]
info_tmp$group <- as.factor(info_tmp$group)
```

Since we have only two CyTOF features, the separation is not that good:

The secondary tolerant recipients are badly separated from the primary tolerant 
recipients compared to the integrative PCA. 
```{r, echo = FALSE}
#information on the patients: 
info_pca <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R")
info_pca <- info_pca[rownames(cyt4pca),]
ade4::s.class(pca_cyto$li[,c(1,2)], info_pca$group, col=c("green", "blue"))
# ade4::s.class(pca_cyto$li[,c(1,6)], as.factor(info_pca$cGVHD), 
#               col = c("black", "red", "lightgreen", "green3"))
```

```{r, echo = FALSE}
keep1 <- order(abs(pca_cyto$c1$CS1), decreasing = T)[1:2]
# keep2 <- order(abs(pca_cyto$c1$CS2), decreasing = T)[1:5]
# tokeep <- unique(c(keep1, keep2))
tokeep <- keep1
ade4::s.arrow(pca_cyto$c1[tokeep,], boxes = F, clabel = 0.5)
```

##### Metabolomics:
```{r, echo = FALSE}
met4pca <- ft_recip_metabo %>% 
  rownames_to_column("CyTOF.Name") %>% 
  dplyr::filter(CyTOF.Name %in% ft_integ$Id.Cryostem.R) %>% 
  column_to_rownames("CyTOF.Name")
pca_metabo <- ade4::dudi.pca(met4pca, nf = 10, scannf = F)
```

Variability captured by each component:
```{r, echo = FALSE}
plot(pca_metabo$eig, pch = 19)
```

```{r, echo = FALSE}
info_tmp <- couple_info %>% 
  rownames_to_column("rn") %>% 
  dplyr::filter(Id.Cryostem.R %in% rownames(met4pca)) %>%
  mutate(group = as.character(tolower(GROUP))) %>% 
  column_to_rownames("Id.Cryostem.R")
info_tmp <- info_tmp[rownames(met4pca),]
info_tmp$group <- as.factor(info_tmp$group)
```


The non tolerant recipients are slightly better separated from the tolerant 
recipients than in the integrative PCA. 
```{r, echo = FALSE}
#information on the patients: 
info_pca <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R")
info_pca <- info_pca[rownames(met4pca),]
ade4::s.class(pca_metabo$li[,c(1,2)], info_pca$group, col=c("green", "blue"))
# ade4::s.class(pca_cyto$li[,c(1,6)], as.factor(info_pca$cGVHD), 
#               col = c("black", "red", "lightgreen", "green3"))
```

```{r, echo = FALSE}
keep1 <- order(abs(pca_metabo$c1$CS1), decreasing = T)[1:2]
# keep2 <- order(abs(pca_cyto$c1$CS2), decreasing = T)[1:5]
# tokeep <- unique(c(keep1, keep2))
tokeep <- keep1
ade4::s.arrow(pca_metabo$c1[tokeep,], boxes = F, clabel = 0.5)
```

##### Transcriptomics:
```{r, echo = FALSE}
rna4pca <- ft_recip_transcripto %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::filter(Id.Cryostem.R %in% ft_integ$Id.Cryostem.R) %>% 
  column_to_rownames("Id.Cryostem.R")
pca_transcripto <- ade4::dudi.pca(rna4pca, nf = 10, scannf = F)
```

Variability captured by each component:
```{r, echo = FALSE}
plot(pca_transcripto$eig, pch = 19)
```

```{r, echo = FALSE}
info_tmp <- couple_info %>% 
  rownames_to_column("rn") %>% 
  dplyr::filter(Id.Cryostem.R %in% rownames(cyt4pca)) %>%
  mutate(group = as.character(tolower(GROUP))) %>% 
  column_to_rownames("Id.Cryostem.R")
info_tmp <- info_tmp[rownames(rna4pca),]
info_tmp$group <- as.factor(info_tmp$group)
```

We can see how these principal components are correlated with tolerance. 
The fourth PC is very correlated with the tolerance. The PC1 is the 2nd mostly
correlated with tolerance.
```{r, echo = FALSE}
barplot(abs(cor(pca_transcripto$li[,1:9], as.numeric(info_tmp$group))[,1]))
```

The non tolerant recipients are slightly better separated from the tolerant 
recipients than in the integrative PCA. 
```{r, echo = FALSE}
#information on the patients: 
info_pca <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R")
info_pca <- info_pca[rownames(rna4pca),]
ade4::s.class(pca_transcripto$li[,c(1,3)], info_pca$group, col=c("green", "blue"))
# ade4::s.class(pca_cyto$li[,c(1,6)], as.factor(info_pca$cGVHD), 
#               col = c("black", "red", "lightgreen", "green3"))
```

```{r, echo = FALSE}
keep1 <- order(abs(pca_transcripto$c1$CS1), decreasing = T)[1:9]
# keep2 <- order(abs(pca_cyto$c1$CS2), decreasing = T)[1:5]
# tokeep <- unique(c(keep1, keep2))
tokeep <- keep1
ade4::s.arrow(pca_transcripto$c1[tokeep,], boxes = F, clabel = 0.5)
```

We can try to build a model that would predict a patients output (tolerant or not)
based on the PCs:
```{r, echo = FALSE}
data2model <- as.data.frame(pca_transcripto$li) %>% mutate(group = info_pca$group)

set.seed(1) 

# Define training control
train.control <- caret::trainControl(method = "cv", number = 5)
# Train the model

model <- caret::train(group ~., data = data2model, method = "svmLinearWeights2",
               trControl = train.control)
# Summarize the results
print(model)
```








## Adding information on the couples to the recipients

We can then add the features that we extracted as informative in the 
couples, to see if, taken together, they would be more informative than 
when working on the integrated data of recipients alone.

### Tolerant versus non tolerant recipients

Since looking at differences between the three groups consisting of primary, 
secondary and non tolerant patients at the same time wasn't feasible, we divided
our problem in two distinct sub-problems. We first focus on the features that
seemed to play a role when trying to disinguish non tolerant from tolerant 
recipients and couples.

I extract informations on the recipients group, age, gender, ... to be able to 
use this information later on.

```{r, echo = FALSE}
samp_rd <- read.xlsx("~/Documents/VIB/Projects/Integrative_Paris/Local_cohort/Data synthesis local cohort Saint-Louis 17012019.xlsx")
samp_rd$DOB <- as.Date(samp_rd$DOB, format = "%d.%m.%Y")
samp_rd$DOG <- as.Date(samp_rd$DOG, format = "%d.%m.%Y")
samp_rd$DATEOFSAMPLE <- as.Date(samp_rd$DATEOFSAMPLE, format = "%d.%m.%Y")
samp_rd <- samp_rd %>% 
  dplyr::filter(!is.na(METABONAME)) 
complete_couples <- samp_rd$COUPLENUMBER[which(duplicated(samp_rd$COUPLENUMBER))]

couple_info <- samp_rd %>% 
  dplyr::filter(COUPLENUMBER %in% complete_couples) %>% 
  arrange(COUPLENUMBER) %>% 
  column_to_rownames("METABONAME") %>% 
  rownames_to_column("rn") %>% 
  mutate(GROUP = as.factor(tolower(GROUP))) %>% 
  column_to_rownames("rn")
  
gender_donors <- couple_info$GENDER[c(TRUE,FALSE)]
gender_recip <- couple_info$GENDER[c(FALSE,TRUE)]
gender_comp <- paste0(gender_donors, gender_recip)

age_recip <- couple_info$DOG[c(FALSE,TRUE)] - couple_info$DOB[c(FALSE,TRUE)]

ifelse(grepl("R", rownames(couple_info)[1]), 
       cmv_donors <- couple_info$CMVStatus[c(FALSE,TRUE)],
       cmv_donors <- couple_info$CMVStatus[c(TRUE,FALSE)])

ifelse(grepl("R", rownames(couple_info)[1]), 
       cmv_recip <- couple_info$CMVStatus[c(TRUE,FALSE)],
       cmv_recip <- couple_info$CMVStatus[c(FALSE,TRUE)])

cmv_comp <- paste0(cmv_donors, cmv_recip)

couple_info <- couple_info %>% 
  rownames_to_column("rownames") %>% 
  mutate(gender_comp = rep(gender_comp, each = 2),
         age_recip = rep(age_recip, each = 2),
         cmv_comp = rep(cmv_comp, each = 2)) %>% 
  column_to_rownames("rownames")

head(couple_info[,c(1,2,3,5,6,55,56,57)])
```


We then need to gather the features that were extracted from the three data
sources:

- CyTOF features that were selected in both the St Louis and the Cryostem cohort, 
  with a threshold of 0.90. These features had the same behaviour in both cohorts.

```{r, echo = FALSE}
tmp_r <- readRDS("../data/cyto/log2/recip/pctgs_sel_ft_recip_TNT_90.RDS")
sel_beh_r <- read.xlsx("../data/cyto_national/log2/recip/perm_val_beh_TNT_90_thresh.xlsx")
load("../data/cyto/rd/pctgs_sel_ft_couples_TNT_90.RData")
sel_beh_rd <- read.xlsx("../data/cyto/rd/perm_val_beh_TNT_90_thresh.xlsx")

tmp_r <- tmp_r %>% 
  select(sel_beh_r$features) %>% 
  rownames_to_column("Id.Cryostem.R") 

tmp_rd <- pctgs_sel_ft_couples_90 %>% 
  select(sel_beh_rd$sel_beh) %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  arrange(Id.Cryostem.R)

colnames(tmp_r)[-c(1)] <- paste0("cyt_R_", colnames(tmp_r)[-c(1)])
colnames(tmp_rd)[-c(1)] <- paste0("cyt_RD_", colnames(tmp_rd)[-c(1)])
  
ft_rdr_cyto <- tmp_r %>% 
  inner_join(tmp_rd, by = c("Id.Cryostem.R"))

```

- Metabolomics features that were selected in both the St Louis and the Cryostem cohort, 
  with a threshold of 0.90. These features had the same behaviour in both cohorts.

```{r, echo = FALSE}
tmp_r <- readRDS("../data/metabo/v2_012020/recip/pctgs_sel_ft_recip_TNT_90.RDS")
sel_beh_r <- read.xlsx("../data/metabo/v2_012020/recip/perm_val_beh_TNT_90_thresh.xlsx")
load("../data/metabo/r&d/metabo_ft_sel_rd_TNT_90.RData")
sel_beh_rd <- read.xlsx("../data/metabo/r&d/perm_val_beh_TNT_90_thresh.xlsx")

couple_info_tmp <- couple_info %>% 
  rownames_to_column("METABONAME") %>% 
  dplyr::filter(grepl("^R", METABONAME))

tmp_r <- tmp_r %>% 
  select(sel_beh_r$features) %>% 
  rownames_to_column("METABONAME") 

tmp_rd <- metabo_ft_sel_rd_TNT_90 %>% 
  rownames_to_column("COUPLENUMBER") %>% 
  mutate(COUPLENUMBER = as.numeric(gsub("couple_", "", COUPLENUMBER))) %>% 
  select(COUPLENUMBER, sel_beh_rd$sel_beh) %>% 
  inner_join(couple_info_tmp, by = "COUPLENUMBER") %>% 
  select(Id.Cryostem.R, METABONAME, sel_beh_rd$sel_beh)

colnames(tmp_r)[-c(1)] <- paste0("met_R_", colnames(tmp_r)[-c(1)])
colnames(tmp_rd)[-c(1,2)] <- paste0("met_RD_", colnames(tmp_rd)[-c(1,2)])
  
ft_rdr_metabo <- tmp_r %>% 
  inner_join(tmp_rd, by = c("METABONAME")) %>% 
  select(-METABONAME) %>% 
  select(Id.Cryostem.R, everything())
```

- Transcriptomics features that were selected in both the St Louis and the Cryostem cohort, 
  with a threshold of 0.90. These features had the same behaviour in both cohorts.

```{r, echo = FALSE}
load("../data/RNAseq/recip_only/rna_recip_louis_TNT_90.RData")
sel_beh_r <- read.xlsx("../data/RNAseq/recip_only/perm_val_beh_TNT_90_thresh.xlsx")
load("../data/RNAseq/donors_and_recip/rna_dr_louis_TNT_90.RData")
sel_beh_rd <- read.xlsx("../data/RNAseq/donors_and_recip/perm_val_beh_TNT_90_thresh.xlsx")

couple_info_tmp <- couple_info %>% 
  rownames_to_column("METABONAME") %>% 
  dplyr::filter(grepl("^R", METABONAME))

tmp_r <- rna_recip_louis_TNT_90 %>% 
  select(sel_beh_r$sel_beh) %>% 
  rownames_to_column("Id.Cryostem.R") 

tmp_rd <- rna_dr_louis_TNT_90 %>% 
  rownames_to_column("COUPLENUMBER") %>% 
  mutate(COUPLENUMBER = as.numeric(gsub("couple_", "", COUPLENUMBER))) %>% 
  select(COUPLENUMBER, sel_beh_rd$sel_beh) %>% 
  inner_join(couple_info_tmp, by = "COUPLENUMBER") %>% 
  select(Id.Cryostem.R, sel_beh_rd$sel_beh)

colnames(tmp_r)[-c(1)] <- paste0("rna_R_", colnames(tmp_r)[-c(1)])
colnames(tmp_rd)[-c(1)] <- gsub("subst_", "rna_RD_", colnames(tmp_rd)[-c(1)])

ft_rdr_transcripto <- tmp_r %>% 
  inner_join(tmp_rd, by = c("Id.Cryostem.R"))
```

I merge all features together:

```{r, echo = FALSE}
ft_integ <- ft_rdr_cyto %>% 
  inner_join(ft_rdr_metabo, by = c("Id.Cryostem.R")) %>% 
  inner_join(ft_rdr_transcripto, by = c("Id.Cryostem.R")) 
print(paste0("We have information on the CyTOF, metabolomics and transcriptomics data for ",
             nrow(ft_integ), " patients in the St Louis cohort."))
```

#### Principal component analysis

We can first generate a PCA on the data:
```{r, echo = FALSE}
mat4pca <- ft_integ %>% 
  select(-Id.Cryostem.R) %>%  
  apply(2, scale) %>% 
  as.matrix() 
rownames(mat4pca) <- ft_integ$Id.Cryostem.R
pca <- prcomp(mat4pca, scale. = TRUE)
```

The first principal component seems to hold much more information that the others
```{r, echo = FALSE}
plot(pca)
```

We can investigate how much of the principal components are driven by the 
different data types.
```{r, echo = FALSE}
pca_var <- pca$rotation %>% as.data.frame
pca_ft <- rep("x", nrow(pca_var))
pca_ft[grep("cyt_", rownames(pca_var))] <- "cyto"
pca_ft[grep("met_", rownames(pca_var))] <- "metabo"
pca_ft[grep("rna_", rownames(pca_var))] <- "rna"

pca_orig <- rep("x", nrow(pca_var))
pca_orig[grep("_R_", rownames(pca_var))] <- "R"
pca_orig[grep("_RD_", rownames(pca_var))] <- "RD"

pca_var <- pca_var %>% 
  rownames_to_column("rn") %>% 
  mutate(feature_orig = as.factor(pca_ft),
         feature_type = as.factor(pca_orig))
```

The RNAseq data variability was captured much more than the other data types.
This is most probably due to the fact that we selected 365 genes, against
45 metabolites and 49 CyTOF features only.
```{r, echo = FALSE}
pc1 <- pca_var %>% 
  column_to_rownames("rn") %>% 
  group_by(feature_orig) %>% 
  summarise_if(is.numeric, sum)
pc1m <- pc1 %>% gather(PC, value, -feature_orig) %>% 
  mutate(label_ypos = round(value))

pc1m_sub <- pc1m[1:30,] %>% 
  mutate(PC = factor(PC, levels = c("PC1", "PC2", "PC3", "PC4", "PC5", "PC6", 
                                    "PC7", "PC8", "PC9", "PC10")))

ggplot(data=pc1m_sub, aes(x=PC, y=value, fill=feature_orig)) +
  geom_bar(stat="identity")+
  # geom_text(aes(y=label_ypos, label=value), vjust=1.6, 
  #           color="white", size=3.5)+
  #scale_fill_brewer(palette="Paired")+
  scale_fill_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  theme_minimal()
```

We can also see how the features coming from recipients and couples are 
distributed in the PCs:
```{r, echo = FALSE}
pc1 <- pca_var %>% 
  column_to_rownames("rn") %>% 
  group_by(feature_type) %>% 
  summarise_if(is.numeric, sum)
pc1m <- pc1 %>% gather(PC, value, -feature_type) %>% 
  mutate(label_ypos = round(value))

pc1m_sub <- pc1m[1:30,] %>% 
  mutate(PC = factor(PC, levels = c("PC1", "PC2", "PC3", "PC4", "PC5", "PC6", 
                                    "PC7", "PC8", "PC9", "PC10")))

ggplot(data=pc1m_sub, aes(x=PC, y=value, fill=feature_type)) +
  geom_bar(stat="identity")+
  # geom_text(aes(y=label_ypos, label=value), vjust=1.6, 
  #           color="white", size=3.5)+
  scale_fill_brewer(palette="Paired")+
  #scale_fill_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  theme_minimal()
```

We can see how these principal components are correlated with tolerance. 
The first PC is very correlated with the tolerance. The PC4 (and slightly 
PC9 and PC10) also look correlated with tolerance.
```{r, echo = FALSE}
complete_patients <- rownames(mat4pca)
info_complete_patients <- couple_info %>%
  rownames_to_column("METABONAME") %>% 
  dplyr::filter(Id.Cryostem.R %in% complete_patients) %>% 
  arrange(Id.Cryostem.R) %>% 
  mutate(group = as.character(GROUP))
info_complete_patients$group[which(info_complete_patients$group != "non_tolerant")] <- "tolerant"
info_complete_patients$group <- as.factor(info_complete_patients$group)

pca_2plot <- left_join(ft_integ, info_complete_patients, by = c("Id.Cryostem.R")) %>% 
  mutate(PC1 = pca$x[,1],
         PC2 = pca$x[,2],
         PC3 = pca$x[,3],
         PC4 = pca$x[,4])

barplot(abs(cor(pca$x[,1:10], as.numeric(pca_2plot$group))[,1]))
```

We can then see how the PCs are correlated with the different informations 
that we have on the patients.

The first PC seems to be most correlated with group, cGVHD and ABO.incompatibility:

```{r, echo = FALSE}
totest <- pca_2plot[,c("group", "cmv_comp", "age_recip", "gender_comp", "cGVHD",
                       "aGVHD", "TBI", "SAL", "CONDITIONING_INTENSITY",
                       "CONDITIONING_REGIMEN", "SOURCEOFGRAFT", 
                       "DISORDER_GROUP", "HEMATOLOGICDISORDER", "ABO.incompatibility",
                       "GENDER")]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,1], totest)), horiz = TRUE, las = 2, cex.names = 0.5, xlim = c(0,0.6))
```

The second PC seems to be slightly correlated with the disorder group:

```{r, echo = FALSE}
totest <- pca_2plot[,c("group", "cmv_comp", "age_recip", "gender_comp", "cGVHD",
                       "aGVHD", "TBI", "SAL", "CONDITIONING_INTENSITY",
                       "CONDITIONING_REGIMEN", "SOURCEOFGRAFT", 
                       "DISORDER_GROUP", "HEMATOLOGICDISORDER", "ABO.incompatibility",
                       "GENDER")]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,2], totest)), horiz = TRUE, las = 2, cex.names = 0.5, xlim = c(0,0.6))
```

The third PC seems to be slightly correlated with GENDER:

```{r, echo = FALSE}
totest <- pca_2plot[,c("group", "cmv_comp", "age_recip", "gender_comp", "cGVHD",
                       "aGVHD", "TBI", "SAL", "CONDITIONING_INTENSITY",
                       "CONDITIONING_REGIMEN", "SOURCEOFGRAFT", 
                       "DISORDER_GROUP", "HEMATOLOGICDISORDER", "ABO.incompatibility",
                       "GENDER")]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,3], totest)), horiz = TRUE, las = 2, cex.names = 0.5, xlim = c(0,0.6))
```

The fourth PC seems to be most correlated with group and age_recip:

```{r, echo = FALSE}
totest <- pca_2plot[,c("group", "cmv_comp", "age_recip", "gender_comp", "cGVHD",
                       "aGVHD", "TBI", "SAL", "CONDITIONING_INTENSITY",
                       "CONDITIONING_REGIMEN", "SOURCEOFGRAFT", 
                       "DISORDER_GROUP", "HEMATOLOGICDISORDER", "ABO.incompatibility",
                       "GENDER")]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,4], totest)), horiz = TRUE, las = 2, cex.names = 0.5, xlim = c(0,0.6))
```

Since the 1st and 4th PCs seem to be most correlated with tolerance, we visualise
the patients in these two dimensions.
We observe that these 2 components seem to separate the tolerant recipients 
(on the bottom-left) from the non tolerant recipients (on the top-right).
We can also see how the recipients are displayed in these two dimensions 
according to ABO incompatibility, recipients age and cGVHD.
```{r, echo = FALSE}
ggplot(pca_2plot, aes(x= PC1, y= PC4, col= group, label = Id.Cryostem.R)) + 
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Separation of tolerant vs non tolerant recipients")
ggplot(pca_2plot, aes(x= PC1, y= PC4, col= ABO.incompatibility, label = Id.Cryostem.R)) + 
  scale_color_manual(values = c("chartreuse2", "black", "gray48")) +
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Separation of recipients based on ABO.incompatibility")
ggplot(pca_2plot, aes(x= PC1, y= PC4, col= age_recip, label = Id.Cryostem.R)) +
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Gradient in the recipients based on age")
ggplot(pca_2plot, aes(x= PC1, y= PC4, col= as.factor(cGVHD), label = Id.Cryostem.R)) +
  scale_color_manual(values = c("black", "red")) +
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Separation of recipients based on cGVHD")
```



##### Features expressed in the first principal component

We can first look at the PC1, since it holds most of the data's variability:
We first focus on the features that are overexpressed in the tolerant recipients 
(= the features that are most expressed on the LEFT of the PC1). Therefore, the 
lower the feature value, the more it is expressed in tolerant recipients.
```{r, echo = FALSE}
#plot(sort(pca$rotation[,1][which(pca$rotation[,1]<0)]))
overexp <- sort(pca$rotation[,1][which(pca$rotation[,1]<0)])
over <- as.data.frame(overexp)
orig <- strsplit(rownames(over), "_")
orig <- map(orig, function(x){x[1]}) %>% unlist
over$orig <- orig

type <- rep("x", nrow(over))
type[grepl("_R_", rownames(over))] <- "R"
type[grepl("_RD_", rownames(over))] <- "RD"

ggplot(over, aes(x = orig, y = overexp, col = orig)) +
  scale_color_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  geom_boxplot() +
  geom_jitter()
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 

ggplot(over, aes(x = orig, y = overexp)) +
  geom_boxplot() +
  geom_jitter(aes(x = orig, y = overexp, col = type), data = over)
```

We extract the "importance values" for these genes, metabolites and CyTOF features 
in an excel file. Here's a sample of the lowest features in the PC1 (under the
-0.6 threshold) :
```{r, echo = FALSE}
most_imp <- over[which(over$overexp < -0.06),] %>% 
  rownames_to_column("feature") %>% 
  mutate(feature = gsub("^[a-z]+_", "", feature)) %>% 
  arrange(overexp)
most_imp

most_imp2 <- most_imp %>% 
  arrange(orig)

write.xlsx(most_imp, file = "../data/integ/RDR/tol_ft_PC1.xlsx")
write.xlsx(most_imp2, file = "../data/integ/RDR/tol_ft_PC1_ordered_by_origin.xlsx")
```

We then focus on the features that are overexpressed in the non tolerant recipients 
(= the features that are most expressed on the RIGHT of the PC1). Therefore, the 
higher the feature value, the more it is expressed in non tolerant recipients.
```{r, echo = FALSE}
#plot(sort(pca$rotation[,1][which(pca$rotation[,1]>0)]))
overexp <- sort(pca$rotation[,1][which(pca$rotation[,1]>0)])
over <- as.data.frame(overexp)
orig <- strsplit(rownames(over), "_")
orig <- map(orig, function(x){x[1]}) %>% unlist
over$orig <- orig

type <- rep("x", nrow(over))
type[grepl("_R_", rownames(over))] <- "R"
type[grepl("_RD_", rownames(over))] <- "RD"
over$type <- type

ggplot(over, aes(x = orig, y = overexp, col = orig)) +
  scale_color_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  geom_boxplot() +
  geom_jitter()
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 

ggplot(over, aes(x = orig, y = overexp)) +
  geom_boxplot() +
  geom_jitter(aes(x = orig, y = overexp, col = type), data = over)
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 
```

We extract the "importance values" for these genes, metabolites and CyTOF features 
in an excel file. Here's a sample of the highest features in the PC1 (over the
0.5 threshold) :
```{r, echo = FALSE}
most_imp <- over[which(over$overexp > 0.05),] %>% 
  rownames_to_column("feature") %>% 
  mutate(feature = gsub("^[a-z]+_", "", feature)) %>% 
  arrange(desc(overexp))
most_imp

most_imp2 <- most_imp %>% 
  arrange(orig)

write.xlsx(most_imp, file = "../data/integ/RDR/nontol_ft_PC1.xlsx")
write.xlsx(most_imp2, file = "../data/integ/RDR/nontol_ft_PC1_ordered_by_origin.xlsx")
```

##### Features expressed in the fourth prinipal component

We can first look at the PC1, since it holds most of the data's variability:
We first focus on the features that are overexpressed in the tolerant recipients 
(= the features that are most expressed on the LEFT of the PC1). Therefore, the 
lower the feature value, the more it is expressed in tolerant recipients.
```{r, echo = FALSE}
#plot(sort(pca$rotation[,4][which(pca$rotation[,4]<0)]))
overexp <- sort(pca$rotation[,4][which(pca$rotation[,4]<0)])
over <- as.data.frame(overexp)
orig <- strsplit(rownames(over), "_")
orig <- map(orig, function(x){x[1]}) %>% unlist
over$orig <- orig

type <- rep("x", nrow(over))
type[grepl("_R_", rownames(over))] <- "R"
type[grepl("_RD_", rownames(over))] <- "RD"
over$type <- type

ggplot(over, aes(x = orig, y = overexp, col = orig)) +
  scale_color_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  geom_boxplot() +
  geom_jitter()
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 

ggplot(over, aes(x = orig, y = overexp)) +
  geom_boxplot() +
  geom_jitter(aes(x = orig, y = overexp, col = type), data = over)

# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 
```

We extract the "importance values" for these genes, metabolites and CyTOF features 
in an excel file. Here's a sample of the lowest features in the PC1 (under the
-0.6 threshold) :
```{r, echo = FALSE}
most_imp <- over[which(over$overexp < -0.1),] %>% 
  rownames_to_column("feature") %>% 
  mutate(feature = gsub("^[a-z]+_", "", feature)) %>% 
  arrange(overexp)
most_imp

most_imp2 <- most_imp %>% 
  arrange(orig)

write.xlsx(most_imp, file = "../data/integ/RDR/tol_ft_PC4.xlsx")
write.xlsx(most_imp2, file = "../data/integ/RDR/tol_ft_PC4_ordered_by_origin.xlsx")
```

We then focus on the features that are overexpressed in the non tolerant recipients 
(= the features that are most expressed on the RIGHT of the PC1). Therefore, the 
higher the feature value, the more it is expressed in non tolerant recipients.
```{r, echo = FALSE}
#plot(sort(pca$rotation[,4][which(pca$rotation[,4]>0)]))
overexp <- sort(pca$rotation[,4][which(pca$rotation[,4]>0)])
over <- as.data.frame(overexp)
orig <- strsplit(rownames(over), "_")
orig <- map(orig, function(x){x[1]}) %>% unlist
over$orig <- orig

type <- rep("x", nrow(over))
type[grepl("_R_", rownames(over))] <- "R"
type[grepl("_RD_", rownames(over))] <- "RD"
over$type <- type

ggplot(over, aes(x = orig, y = overexp, col = orig)) +
  scale_color_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  geom_boxplot() +
  geom_jitter()
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 

ggplot(over, aes(x = orig, y = overexp)) +
  geom_boxplot() +
  geom_jitter(aes(x = orig, y = overexp, col = type), data = over)

# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 
```

We extract the "importance values" for these genes, metabolites and CyTOF features 
in an excel file. Here's a sample of the highest features in the PC1 (over the
0.5 threshold) :
```{r, echo = FALSE}
most_imp <- over[which(over$overexp > 0.1),] %>% 
  rownames_to_column("feature") %>% 
  mutate(feature = gsub("^[a-z]+_", "", feature)) %>% 
  arrange(desc(overexp))
most_imp

most_imp2 <- most_imp %>% 
  arrange(orig)

write.xlsx(most_imp, file = "../data/integ/RDR/non_tol_ft_PC4.xlsx")
write.xlsx(most_imp2, file = "../data/integ/RDR/non_tol_ft_PC4_ordered_by_origin.xlsx")
```

We can try to build a model that would predict a patients output (tolerant or not)
based on the PCs:
```{r, echo = FALSE}
data2model <- as.data.frame(pca$x) %>% mutate(group = pca_2plot$group)

set.seed(1) 

# Define training control
train.control <- caret::trainControl(method = "cv", number = 5)
# Train the model

model <- caret::train(group ~., data = data2model, method = "svmLinearWeights2",
               trControl = train.control)
# Summarize the results
print(model)
```

The results haven't been improved by the addition of the couple's features.

#### Random forest modeling

To see if the features that we selected can help us to clssify the patients, we
apply a 5 fold cross validation setting on the 23 patients for which we have 
CyTOF, metabolomics and transcriptomics information.

```{r}
set.seed(1)
cv_id <- list(c(1:5), c(6:10), c(11:15), c(16:20), c(21:23))
cv_accuracies <- lapply(seq_along(cv_id), function(idx){
  train_set <- pca_2plot[-cv_id[[idx]], c("group", colnames(ft_integ)[-1])]
  test_set <- pca_2plot[cv_id[[idx]], c(colnames(ft_integ)[-1])]
  true_labels <- pca_2plot[cv_id[[idx]], "group"]
  rf <- ranger::ranger(group~., train_set, num.trees = 5000,
                           importance = "impurity")
  pred <- predict(rf, test_set)
  acc <- ( length(which(pred$predictions == true_labels))/length(true_labels) )
})

accuracy <- mean(unlist(cv_accuracies))
accuracy
```

#### Separate data type analyses

For comparison, we can see if we would have managed to separate the tolerant 
from non tolerant recipients and couples as well without integrating the data:

##### CyTOF:

We perform PCA using only the 49 features that were selected as informative 
in the recipients and couples in both cohorts:

```{r, echo = FALSE}
cyt4pca <- ft_rdr_cyto %>% 
  dplyr::filter(Id.Cryostem.R %in% ft_integ$Id.Cryostem.R) %>% 
  column_to_rownames("Id.Cryostem.R")
pca_cyto <- ade4::dudi.pca(cyt4pca, nf = 10, scannf = F)
```

The first comoponent seems to be capturing most of the data variability:
```{r, echo = FALSE}
plot(pca_cyto$eig, pch = 19)
```

```{r, echo = FALSE}
info_tmp <- couple_info %>% 
  rownames_to_column("rn") %>% 
  dplyr::filter(Id.Cryostem.R %in% rownames(cyt4pca)) %>%
  mutate(group = as.character(tolower(GROUP))) %>% 
  column_to_rownames("Id.Cryostem.R")
info_tmp <- info_tmp[rownames(cyt4pca),]
info_tmp$group[which(info_tmp$group != "non_tolerant")] <- "tolerant"
info_tmp$group <- as.factor(info_tmp$group)
```

We can see how these principal components are correlated with tolerance. 
The first PC is very correlated with the tolerance. The PC9 is the 2nd mostly
correlated with tolerance.
```{r, echo = FALSE}
barplot(abs(cor(pca_cyto$li[,1:10], as.numeric(info_tmp$group))[,1]))
```

The non tolerant recipients are slightly better separated from the tolerant 
recipients than in the integrative PCA. 
```{r, echo = FALSE}
#information on the patients: 
info_pca <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R")
info_pca <- info_pca[rownames(cyt4pca),]
ade4::s.class(pca_cyto$li[,c(1,9)], info_pca$group, col=c("red", "blue"))
# ade4::s.class(pca_cyto$li[,c(1,6)], as.factor(info_pca$cGVHD), 
#               col = c("black", "red", "lightgreen", "green3"))
```

We can also have a look at the features that are driving these two components:
```{r, echo = FALSE}
keep1 <- order(abs(pca_cyto$c1$CS1), decreasing = T)[1:20]
# keep2 <- order(abs(pca_cyto$c1$CS2), decreasing = T)[1:5]
# tokeep <- unique(c(keep1, keep2))
tokeep <- keep1
ade4::s.arrow(pca_cyto$c1[tokeep,c(1,9)], boxes = F, clabel = 0.5)
```

We can try to build a model that would predict a patients output (tolerant or not)
based on the PCs:
```{r, echo = FALSE}
data2model <- as.data.frame(pca_cyto$li) %>% mutate(group = info_pca$group)

set.seed(1) 

# Define training control
train.control <- caret::trainControl(method = "cv", number = 5)
# Train the model

model <- caret::train(group ~., data = data2model, method = "svmLinearWeights2",
               trControl = train.control)
# Summarize the results
print(model)
```

The accuracy is quite lower than in the integrative model.

To see if the features that we selected can help us to clssify the patients, we
can also apply a 5 fold cross validation setting on the CyTOF features of the 
same 23 patients.

```{r}
set.seed(1)
info_rf <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R")
info_rf <- info_rf[rownames(cyt4pca),]
cyt4rf <- cbind(info_rf$group, cyt4pca)
colnames(cyt4rf)[1] <- "group"

cv_id <- list(c(1:5), c(6:10), c(11:15), c(16:20), c(21:23))
cv_accuracies <- lapply(seq_along(cv_id), function(idx){
  train_set <- cyt4rf[-cv_id[[idx]], ]
  test_set <- cyt4rf[cv_id[[idx]], -1]
  true_labels <- cyt4rf[cv_id[[idx]], "group"]
  rf <- ranger::ranger(group~., train_set, num.trees = 5000,
                           importance = "impurity")
  pred <- predict(rf, test_set)
  acc <- ( length(which(pred$predictions == true_labels))/length(true_labels) )
})

accuracy <- mean(unlist(cv_accuracies))
accuracy
```

##### Metabolomics:
```{r, echo = FALSE}
met4pca <- ft_rdr_metabo %>% 
  dplyr::filter(Id.Cryostem.R %in% ft_integ$Id.Cryostem.R) %>% 
  column_to_rownames("Id.Cryostem.R")
pca_metabo <- ade4::dudi.pca(met4pca, nf = 10, scannf = F)
```

Variability captured by each component:
```{r, echo = FALSE}
plot(pca_metabo$eig, pch = 19)
```

```{r, echo = FALSE}
info_tmp <- couple_info %>% 
  rownames_to_column("rn") %>% 
  dplyr::filter(Id.Cryostem.R %in% rownames(met4pca)) %>%
  mutate(group = as.character(tolower(GROUP))) %>% 
  column_to_rownames("Id.Cryostem.R")
info_tmp <- info_tmp[rownames(met4pca),]
info_tmp$group[which(info_tmp$group != "non_tolerant")] <- "tolerant"
info_tmp$group <- as.factor(info_tmp$group)
```

We can see how these principal components are correlated with tolerance. 
The first PC is very correlated with the tolerance. The PC4 is the 2nd mostly
correlated with tolerance.
```{r, echo = FALSE}
barplot(abs(cor(pca_metabo$li[,1:10], as.numeric(info_tmp$group))[,1]))
```

The non tolerant recipients are slightly less well separated from the tolerant 
recipients than in the integrative PCA. 
```{r, echo = FALSE}
#information on the patients: 
info_pca <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R")
info_pca <- info_pca[rownames(met4pca),]
ade4::s.class(pca_metabo$li[,c(1,4)], info_pca$group, col=c("red", "blue"))
# ade4::s.class(pca_cyto$li[,c(1,6)], as.factor(info_pca$cGVHD), 
#               col = c("black", "red", "lightgreen", "green3"))
```

```{r, echo = FALSE}
keep1 <- order(abs(pca_metabo$c1$CS1), decreasing = T)[1:20]
# keep2 <- order(abs(pca_cyto$c1$CS2), decreasing = T)[1:5]
# tokeep <- unique(c(keep1, keep2))
tokeep <- keep1
ade4::s.arrow(pca_metabo$c1[tokeep, c(1,4)], boxes = F, clabel = 0.5)
```

We can try to build a model that would predict a patients output (tolerant or not)
based on the PCs:
```{r, echo = FALSE}
data2model <- as.data.frame(pca_metabo$li) %>% mutate(group = info_pca$group)

set.seed(1) 

# Define training control
train.control <- caret::trainControl(method = "cv", number = 5)
# Train the model

model <- caret::train(group ~., data = data2model, method = "svmLinearWeights2",
               trControl = train.control)
# Summarize the results
print(model)
```

Again, the accuracy of the model is slightly lower than the accuracy of the integrative model.

To see if the features that we selected can help us to clssify the patients, we
can also apply a 5 fold cross validation setting on the CyTOF features of the 
same 23 patients.

```{r}
set.seed(1)
info_rf <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R")
info_rf <- info_rf[rownames(met4pca),]
met4rf <- cbind(info_rf$group, met4pca)
colnames(met4rf)[1] <- "group"

cv_id <- list(c(1:5), c(6:10), c(11:15), c(16:20), c(21:23))
cv_accuracies <- lapply(seq_along(cv_id), function(idx){
  train_set <- met4rf[-cv_id[[idx]], ]
  test_set <- met4rf[cv_id[[idx]], -1]
  true_labels <- met4rf[cv_id[[idx]], "group"]
  rf <- ranger::ranger(group~., train_set, num.trees = 5000,
                           importance = "impurity")
  pred <- predict(rf, test_set)
  acc <- ( length(which(pred$predictions == true_labels))/length(true_labels) )
})

accuracy <- mean(unlist(cv_accuracies))
accuracy
```

##### Transcriptomics:
```{r, echo = FALSE}
rna4pca <- ft_rdr_transcripto %>%
  dplyr::filter(Id.Cryostem.R %in% ft_integ$Id.Cryostem.R) %>% 
  column_to_rownames("Id.Cryostem.R")
pca_transcripto <- ade4::dudi.pca(rna4pca, nf = 10, scannf = F)
```

Variability captured by each component:
```{r, echo = FALSE}
plot(pca_transcripto$eig, pch = 19)
```

```{r, echo = FALSE}
info_tmp <- couple_info %>% 
  rownames_to_column("rn") %>% 
  dplyr::filter(Id.Cryostem.R %in% rownames(rna4pca)) %>%
  mutate(group = as.character(tolower(GROUP))) %>% 
  column_to_rownames("Id.Cryostem.R")
info_tmp <- info_tmp[rownames(rna4pca),]
info_tmp$group[which(info_tmp$group != "non_tolerant")] <- "tolerant"
info_tmp$group <- as.factor(info_tmp$group)
```

We can see how these principal components are correlated with tolerance. 
The first PC is very correlated with the tolerance. The PC8 is the 2nd mostly
correlated with tolerance.
```{r, echo = FALSE}
barplot(abs(cor(pca_transcripto$li[,1:10], as.numeric(info_tmp$group))[,1]))
```

The non tolerant recipients are slightly less well separated from the tolerant 
recipients than in the integrative PCA. 
```{r, echo = FALSE}
#information on the patients: 
info_pca <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R")
info_pca <- info_pca[rownames(rna4pca),]
ade4::s.class(pca_transcripto$li[,c(1,8)], info_pca$group, col=c("red", "blue"))
# ade4::s.class(pca_cyto$li[,c(1,6)], as.factor(info_pca$cGVHD), 
#               col = c("black", "red", "lightgreen", "green3"))
```

```{r, echo = FALSE}
keep1 <- order(abs(pca_transcripto$c1$CS1), decreasing = T)[1:20]
# keep2 <- order(abs(pca_cyto$c1$CS2), decreasing = T)[1:5]
# tokeep <- unique(c(keep1, keep2))
tokeep <- keep1
ade4::s.arrow(pca_transcripto$c1[tokeep,c(1,8)], boxes = F, clabel = 0.5)
```

We can try to build a model that would predict a patients output (tolerant or not)
based on the PCs:
```{r, echo = FALSE}
data2model <- as.data.frame(pca_transcripto$li) %>% mutate(group = info_pca$group)

set.seed(1) 

# Define training control
train.control <- caret::trainControl(method = "cv", number = 5)
# Train the model

model <- caret::train(group ~., data = data2model, method = "svmLinearWeights2",
               trControl = train.control)
# Summarize the results
print(model)
```

It looks like, in our integrative approach, we managed to separate the recipients 
based on their tolerance better than when we look at the different data types
separately.

To see if the features that we selected can help us to clssify the patients, we
can also apply a 5 fold cross validation setting on the CyTOF features of the 
same 23 patients.

```{r}
set.seed(1)
info_rf <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R")
info_rf <- info_rf[rownames(rna4pca),]
rna4rf <- cbind(info_rf$group, rna4pca)
colnames(rna4rf)[1] <- "group"

cv_id <- list(c(1:5), c(6:10), c(11:15), c(16:20), c(21:23))
cv_accuracies <- lapply(seq_along(cv_id), function(idx){
  train_set <- rna4rf[-cv_id[[idx]], ]
  test_set <- rna4rf[cv_id[[idx]], -1]
  true_labels <- rna4rf[cv_id[[idx]], "group"]
  rf <- ranger::ranger(group~., train_set, num.trees = 5000,
                           importance = "impurity")
  pred <- predict(rf, test_set)
  acc <- ( length(which(pred$predictions == true_labels))/length(true_labels) )
})

accuracy <- mean(unlist(cv_accuracies))
accuracy
```

### Looking at differences between primary and secondary recipients + couples

Since looking at differences between the three groups consisting of primary, 
secondary and non tolerant patients at the same time wasn't feasible, we divided
our problem in two distinct sub-problems. We now focus on the features that
seemed to play a role when trying to disinguish primary tolerant from secondary 
tolerant recipients and couples.

I extract informations on the recipients group, age, gender, ... to be able to 
use this information later on.

```{r, echo = FALSE}
samp_rd <- read.xlsx("~/Documents/VIB/Projects/Integrative_Paris/Local_cohort/Data synthesis local cohort Saint-Louis 17012019.xlsx")
samp_rd$DOB <- as.Date(samp_rd$DOB, format = "%d.%m.%Y")
samp_rd$DOG <- as.Date(samp_rd$DOG, format = "%d.%m.%Y")
samp_rd$DATEOFSAMPLE <- as.Date(samp_rd$DATEOFSAMPLE, format = "%d.%m.%Y")
# samp_rd <- samp_rd %>% 
#   dplyr::filter(!is.na(METABONAME)) 
complete_couples <- samp_rd$COUPLENUMBER[which(duplicated(samp_rd$COUPLENUMBER))]

couple_info <- samp_rd %>% 
  dplyr::filter(COUPLENUMBER %in% complete_couples) %>% 
  arrange(COUPLENUMBER) %>% 
  mutate(GROUP = as.factor(tolower(GROUP))) %>% 
  column_to_rownames("Id.Cryostem.R")
  
gender_donors <- couple_info$GENDER[c(TRUE,FALSE)]
gender_recip <- couple_info$GENDER[c(FALSE,TRUE)]
gender_comp <- paste0(gender_donors, gender_recip)

age_recip <- couple_info$DOG[c(FALSE,TRUE)] - couple_info$DOB[c(FALSE,TRUE)]

ifelse(grepl("R", rownames(couple_info)[1]), 
       cmv_donors <- couple_info$CMVStatus[c(FALSE,TRUE)],
       cmv_donors <- couple_info$CMVStatus[c(TRUE,FALSE)])

ifelse(grepl("R", rownames(couple_info)[1]), 
       cmv_recip <- couple_info$CMVStatus[c(TRUE,FALSE)],
       cmv_recip <- couple_info$CMVStatus[c(FALSE,TRUE)])

cmv_comp <- paste0(cmv_donors, cmv_recip)

couple_info <- couple_info %>% 
  rownames_to_column("rownames") %>% 
  mutate(gender_comp = rep(gender_comp, each = 2),
         age_recip = rep(age_recip, each = 2),
         cmv_comp = rep(cmv_comp, each = 2)) %>% 
  column_to_rownames("rownames")

head(couple_info[,c(1,2,3,5,6,55,56,57)])
```


We then need to gather the features that were extracted from the three data
sources:

- CyTOF features that were selected in both the St Louis and the Cryostem cohort, 
  with a threshold of 0.90. These features had the same behaviour in both cohorts.

```{r, echo = FALSE}
tmp_r <- readRDS("../data/cyto/log2/recip/pctgs_sel_ft_recip_PTST_90.RDS")
sel_beh_r <- read.xlsx("../data/cyto_national/log2/recip/perm_val_beh_PTST_90_thresh.xlsx")
load("../data/cyto/rd/pctgs_sel_ft_couples_PTST_90.RData")
sel_beh_rd <- read.xlsx("../data/cyto/rd/perm_val_beh_PS_90_thresh.xlsx")

tmp_r <- tmp_r %>% 
  select(sel_beh_r$features) %>% 
  rownames_to_column("Id.Cryostem.R") 

tmp_info <- couple_info %>% rownames_to_column("rn") %>% 
  dplyr::filter(grepl("R", rn)) %>% 
  distinct(COUPLENUMBER, .keep_all = TRUE) %>% 
  column_to_rownames("COUPLENUMBER")
tmp_info <- tmp_info[pctgs_sel_ft_couples_90$COUPLENUMBER,]

tmp_rd <- pctgs_sel_ft_couples_90 %>% 
  select(sel_beh_rd$sel_beh) %>% 
  mutate(Id.Cryostem.R = tmp_info$rn) %>% 
  arrange(Id.Cryostem.R)

colnames(tmp_r)[-c(1)] <- paste0("cyt_R_", colnames(tmp_r)[-c(1)])
colnames(tmp_rd)[-c(2)] <- paste0("cyt_RD_", colnames(tmp_rd)[-c(2)])
  
ft_rdr_cyto <- tmp_r %>% 
  inner_join(tmp_rd, by = c("Id.Cryostem.R"))

```

- Metabolomics features that were selected in both the St Louis and the Cryostem cohort, 
  with a threshold of 0.90. These features had the same behaviour in both cohorts.

```{r, echo = FALSE}
tmp_r <- readRDS("../data/metabo/v2_012020/recip/pctgs_sel_ft_recip_PTST_90.RDS")
sel_beh_r <- read.xlsx("../data/metabo/v2_012020/recip/perm_val_beh_PTST_90_thresh.xlsx")
load("../data/metabo/r&d/metabo_ft_sel_rd_PS_90.RData")
sel_beh_rd <- read.xlsx("../data/metabo/r&d/perm_val_beh_PS_90_thresh.xlsx")

couple_info_tmp <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::filter(grepl("^R", METABONAME))

tmp_r <- tmp_r %>% 
  select(sel_beh_r$features) %>% 
  rownames_to_column("METABONAME") 

tmp_rd <- metabo_ft_sel_rd_PS_90 %>% 
  rownames_to_column("COUPLENUMBER") %>% 
  mutate(COUPLENUMBER = as.numeric(gsub("couple_", "", COUPLENUMBER))) %>% 
  select(COUPLENUMBER, sel_beh_rd$sel_beh) %>% 
  inner_join(couple_info_tmp, by = "COUPLENUMBER") %>% 
  select(Id.Cryostem.R, METABONAME, sel_beh_rd$sel_beh)

colnames(tmp_r)[-c(1)] <- paste0("met_R_", colnames(tmp_r)[-c(1)])
colnames(tmp_rd)[-c(1,2)] <- paste0("met_RD_", colnames(tmp_rd)[-c(1,2)])
  
ft_rdr_metabo <- tmp_r %>% 
  inner_join(tmp_rd, by = c("METABONAME")) %>% 
  select(-METABONAME) %>% 
  select(Id.Cryostem.R, everything())
```

- Transcriptomics features that were selected in both the St Louis and the Cryostem cohort, 
  with a threshold of 0.90. These features had the same behaviour in both cohorts.

```{r, echo = FALSE}
load("../data/RNAseq/recip_only/rna_recip_louis_PS_90.RData")
sel_beh_r <- read.xlsx("../data/RNAseq/recip_only/perm_val_beh_PS_90_thresh.xlsx")
load("../data/RNAseq/donors_and_recip/rna_dr_louis_PS_90.RData")
sel_beh_rd <- read.xlsx("../data/RNAseq/donors_and_recip/perm_val_beh_PS_90_thresh.xlsx")

couple_info_tmp <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::filter(grepl("^R", METABONAME))

tmp_r <- rna_recip_louis_PS_90 %>% 
  select(sel_beh_r$sel_beh) %>% 
  rownames_to_column("Id.Cryostem.R") 

tmp_rd <- rna_dr_louis_PS_90 %>% 
  rownames_to_column("COUPLENUMBER") %>% 
  mutate(COUPLENUMBER = as.numeric(gsub("couple_", "", COUPLENUMBER))) %>% 
  select(COUPLENUMBER, sel_beh_rd$sel_beh) %>% 
  inner_join(couple_info_tmp, by = "COUPLENUMBER") %>% 
  select(Id.Cryostem.R, sel_beh_rd$sel_beh)

colnames(tmp_r)[-c(1)] <- paste0("rna_R_", colnames(tmp_r)[-c(1)])
colnames(tmp_rd)[-c(1)] <- gsub("subst_", "rna_RD_", colnames(tmp_rd)[-c(1)])

ft_rdr_transcripto <- tmp_r %>% 
  inner_join(tmp_rd, by = c("Id.Cryostem.R"))
```

I merge all features together:

```{r, echo = FALSE}
ft_integ <- ft_rdr_cyto %>% 
  inner_join(ft_rdr_metabo, by = c("Id.Cryostem.R")) %>% 
  inner_join(ft_rdr_transcripto, by = c("Id.Cryostem.R")) 
print(paste0("We have information on the CyTOF, metabolomics and transcriptomics data for ",
             nrow(ft_integ), " patients in the St Louis cohort."))
```

#### Principal component analysis

We can first generate a PCA on the data:
```{r, echo = FALSE}
mat4pca <- ft_integ %>% 
  select(-Id.Cryostem.R) %>%  
  apply(2, scale) %>% 
  as.matrix() 
rownames(mat4pca) <- ft_integ$Id.Cryostem.R
pca <- prcomp(mat4pca, scale. = TRUE)
```

The first principal component seems to hold much more information that the others
```{r, echo = FALSE}
plot(pca)
```

We can investigate how much of the principal components are driven by the 
different data types.
```{r, echo = FALSE}
pca_var <- pca$rotation %>% as.data.frame
pca_ft <- rep("x", nrow(pca_var))
pca_ft[grep("cyt_", rownames(pca_var))] <- "cyto"
pca_ft[grep("met_", rownames(pca_var))] <- "metabo"
pca_ft[grep("rna_", rownames(pca_var))] <- "rna"

pca_orig <- rep("x", nrow(pca_var))
pca_orig[grep("_R_", rownames(pca_var))] <- "R"
pca_orig[grep("_RD_", rownames(pca_var))] <- "RD"

pca_var <- pca_var %>% 
  rownames_to_column("rn") %>% 
  mutate(feature_orig = as.factor(pca_ft),
         feature_type = as.factor(pca_orig))
```

The RNAseq data variability was captured much more than the other data types.
This is most probably due to the fact that we selected 365 genes, against
45 metabolites and 49 CyTOF features only.
```{r, echo = FALSE}
pc1 <- pca_var %>% 
  column_to_rownames("rn") %>% 
  group_by(feature_orig) %>% 
  summarise_if(is.numeric, sum)
pc1m <- pc1 %>% gather(PC, value, -feature_orig) %>% 
  mutate(label_ypos = round(value))

pc1m_sub <- pc1m[1:30,] %>% 
  mutate(PC = factor(PC, levels = c("PC1", "PC2", "PC3", "PC4", "PC5", "PC6", 
                                    "PC7", "PC8", "PC9", "PC10")))

ggplot(data=pc1m_sub, aes(x=PC, y=value, fill=feature_orig)) +
  geom_bar(stat="identity")+
  # geom_text(aes(y=label_ypos, label=value), vjust=1.6, 
  #           color="white", size=3.5)+
  #scale_fill_brewer(palette="Paired")+
  scale_fill_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  theme_minimal()
```

We can also see how the features coming from recipients and couples are 
distributed in the PCs:
```{r, echo = FALSE}
pc1 <- pca_var %>% 
  column_to_rownames("rn") %>% 
  group_by(feature_type) %>% 
  summarise_if(is.numeric, sum)
pc1m <- pc1 %>% gather(PC, value, -feature_type) %>% 
  mutate(label_ypos = round(value))

pc1m_sub <- pc1m[1:30,] %>% 
  mutate(PC = factor(PC, levels = c("PC1", "PC2", "PC3", "PC4", "PC5", "PC6", 
                                    "PC7", "PC8", "PC9", "PC10")))

ggplot(data=pc1m_sub, aes(x=PC, y=value, fill=feature_type)) +
  geom_bar(stat="identity")+
  # geom_text(aes(y=label_ypos, label=value), vjust=1.6, 
  #           color="white", size=3.5)+
  scale_fill_brewer(palette="Paired")+
  #scale_fill_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  theme_minimal()
```

We can see how these principal components are correlated with tolerance. 
The first PC is very correlated with the tolerance. The PC4 (and slightly 
PC9 and PC10) also look correlated with tolerance.
```{r, echo = FALSE}
complete_patients <- rownames(mat4pca)
info_complete_patients <- couple_info %>%
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::filter(Id.Cryostem.R %in% complete_patients) %>% 
  arrange(Id.Cryostem.R) %>% 
  mutate(group = as.character(GROUP)) %>% 
  dplyr::filter(group != "non_tolerant")

info_complete_patients$group <- as.factor(info_complete_patients$group)

pca_2plot <- left_join(ft_integ, info_complete_patients, by = c("Id.Cryostem.R")) %>% 
  mutate(PC1 = pca$x[,1],
         PC2 = pca$x[,2],
         PC3 = pca$x[,3],
         PC4 = pca$x[,4])

barplot(abs(cor(pca$x[,1:10], as.numeric(pca_2plot$group))[,1]))
```

We can then see how the PCs are correlated with the different informations 
that we have on the patients.

The first PC seems to be most correlated with group, and SAL:

```{r, echo = FALSE}
totest <- pca_2plot[,c("group", "cmv_comp", "age_recip", "gender_comp", "cGVHD",
                       "aGVHD", "TBI", "SAL", "CONDITIONING_INTENSITY",
                       "CONDITIONING_REGIMEN", "SOURCEOFGRAFT", 
                       "DISORDER_GROUP", "HEMATOLOGICDISORDER", "ABO.incompatibility",
                       "GENDER")]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,1], totest)), horiz = TRUE, las = 2, cex.names = 0.5, xlim = c(0,0.6))
```

The second PC seems to be correlated with the hematologic disorder group:

```{r, echo = FALSE}
totest <- pca_2plot[,c("group", "cmv_comp", "age_recip", "gender_comp", "cGVHD",
                       "aGVHD", "TBI", "SAL", "CONDITIONING_INTENSITY",
                       "CONDITIONING_REGIMEN", "SOURCEOFGRAFT", 
                       "DISORDER_GROUP", "HEMATOLOGICDISORDER", "ABO.incompatibility",
                       "GENDER")]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,2], totest)), horiz = TRUE, las = 2, cex.names = 0.5, xlim = c(0,0.6))
```

The third PC seems to be slightly correlated with cmv compatibility:

```{r, echo = FALSE}
totest <- pca_2plot[,c("group", "cmv_comp", "age_recip", "gender_comp", "cGVHD",
                       "aGVHD", "TBI", "SAL", "CONDITIONING_INTENSITY",
                       "CONDITIONING_REGIMEN", "SOURCEOFGRAFT", 
                       "DISORDER_GROUP", "HEMATOLOGICDISORDER", "ABO.incompatibility",
                       "GENDER")]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,3], totest)), horiz = TRUE, las = 2, cex.names = 0.5, xlim = c(0,0.6))
```

The fourth PC seems to be most correlated with age_recip:

```{r, echo = FALSE}
totest <- pca_2plot[,c("group", "cmv_comp", "age_recip", "gender_comp", "cGVHD",
                       "aGVHD", "TBI", "SAL", "CONDITIONING_INTENSITY",
                       "CONDITIONING_REGIMEN", "SOURCEOFGRAFT", 
                       "DISORDER_GROUP", "HEMATOLOGICDISORDER", "ABO.incompatibility",
                       "GENDER")]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,4], totest)), horiz = TRUE, las = 2, cex.names = 0.5, xlim = c(0,0.6))
```


Since the 1st PC seems to be most correlated with tolerance, we visualise
the patients in the two dimensions that carry most variability: the PCs 1 and 2.
We observe that these 2 components seem to separate the tolerant recipients 
(on the bottom-left) from the non tolerant recipients (on the top-right).
We can also see how the recipients are displayed in these two dimensions 
according to ABO incompatibility, recipients age and cGVHD.
```{r, echo = FALSE}
ggplot(pca_2plot, aes(x= PC1, y= PC4, col= group, label = Id.Cryostem.R)) + 
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  scale_color_manual(values = c("green", "blue")) +
  ggtitle("Separation of tolerant vs non tolerant recipients")
ggplot(pca_2plot, aes(x= PC1, y= PC4, col= SAL, label = Id.Cryostem.R)) + 
  scale_color_manual(values = c("chartreuse2", "black", "gray48")) +
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Separation of recipients based on SAL")
ggplot(pca_2plot, aes(x= PC1, y= PC4, col= HEMATOLOGICDISORDER, label = Id.Cryostem.R)) + 
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Separation of recipients based on HEMATOLOGICDISORDER")
ggplot(pca_2plot, aes(x= PC1, y= PC4, col= age_recip, label = Id.Cryostem.R)) +
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Gradient in the recipients based on age")
ggplot(pca_2plot, aes(x= PC1, y= PC4, col= as.factor(cGVHD), label = Id.Cryostem.R)) +
  scale_color_manual(values = c("black", "red")) +
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Separation of recipients based on cGVHD")
```

##### Features expressed in the first principal component

Since most information related to tolerance is carried in the PC1, we look into it:
We first focus on the features that are overexpressed in the primary tolerant recipients 
(= the features that are most expressed on the LEFT of the PC1). Therefore, the 
lower the feature value, the more it is expressed in primary tolerant recipients.
```{r, echo = FALSE}
#plot(sort(pca$rotation[,1][which(pca$rotation[,1]<0)]))
overexp <- sort(pca$rotation[,1][which(pca$rotation[,1]<0)])
over <- as.data.frame(overexp)
orig <- strsplit(rownames(over), "_")
orig <- map(orig, function(x){x[1]}) %>% unlist
over$orig <- orig

type <- rep("x", nrow(over))
type[grepl("_R_", rownames(over))] <- "R"
type[grepl("_RD_", rownames(over))] <- "RD"

ggplot(over, aes(x = orig, y = overexp, col = orig)) +
  scale_color_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  geom_boxplot() +
  geom_jitter()
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 

ggplot(over, aes(x = orig, y = overexp)) +
  geom_boxplot() +
  geom_jitter(aes(x = orig, y = overexp, col = type), data = over)
```

We extract the "importance values" for these genes, metabolites and CyTOF features 
in an excel file
```{r, echo = FALSE}
most_imp <- over %>% 
  rownames_to_column("feature") %>% 
  mutate(feature = gsub("^[a-z]+_", "", feature)) %>% 
  arrange(overexp)
most_imp

most_imp2 <- most_imp %>% 
  arrange(orig)

write.xlsx(most_imp, file = "../data/integ/RDR/prim_tol_ft_PC1.xlsx")
write.xlsx(most_imp2, file = "../data/integ/RDR/prim_tol_ft_PC1_ordered_by_origin.xlsx")
```

We then focus on the features that are overexpressed in the secondary tolerant recipients 
(= the features that are most expressed on the RIGHT of the PC1). Therefore, the 
higher the feature value, the more it is expressed in secondary tolerant recipients.
```{r, echo = FALSE}
#plot(sort(pca$rotation[,1][which(pca$rotation[,1]>0)]))
overexp <- sort(pca$rotation[,1][which(pca$rotation[,1]>0)])
over <- as.data.frame(overexp)
orig <- strsplit(rownames(over), "_")
orig <- map(orig, function(x){x[1]}) %>% unlist
over$orig <- orig

type <- rep("x", nrow(over))
type[grepl("_R_", rownames(over))] <- "R"
type[grepl("_RD_", rownames(over))] <- "RD"
over$type <- type

ggplot(over, aes(x = orig, y = overexp, col = orig)) +
  scale_color_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  geom_boxplot() +
  geom_jitter()
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 

ggplot(over, aes(x = orig, y = overexp)) +
  geom_boxplot() +
  geom_jitter(aes(x = orig, y = overexp, col = type), data = over) 
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 
```

We extract the "importance values" for these genes, metabolites and CyTOF features 
in an excel file. Here's a sample of the highest features in the PC1 (over the
0.5 threshold) :
```{r, echo = FALSE}
most_imp <- over %>% 
  rownames_to_column("feature") %>% 
  mutate(feature = gsub("^[a-z]+_", "", feature)) %>% 
  arrange(desc(overexp))
most_imp

most_imp2 <- most_imp %>% 
  arrange(orig)

write.xlsx(most_imp, file = "../data/integ/RDR/sectol_ft_PC1.xlsx")
write.xlsx(most_imp2, file = "../data/integ/RDR/sectol_ft_PC1_ordered_by_origin.xlsx")
```







## Adding information on the donors to the recipients

We can then add the features that we extracted as informative in the 
donors, to see if, taken together, they would be more informative than 
when working on the integrated data of recipients alone.

### Tolerant versus non tolerant recipients + donors

Since looking at differences between the three groups consisting of primary, 
secondary and non tolerant patients at the same time wasn't feasible, we divided
our problem in two distinct sub-problems. We first focus on the features that
seemed to play a role when trying to disinguish non tolerant from tolerant 
recipients and donors.

I extract informations on the recipients group, age, gender, ... to be able to 
use this information later on.

```{r, echo = FALSE}
samp_rd <- read.xlsx("~/Documents/VIB/Projects/Integrative_Paris/Local_cohort/Data synthesis local cohort Saint-Louis 17012019.xlsx")
samp_rd$DOB <- as.Date(samp_rd$DOB, format = "%d.%m.%Y")
samp_rd$DOG <- as.Date(samp_rd$DOG, format = "%d.%m.%Y")
samp_rd$DATEOFSAMPLE <- as.Date(samp_rd$DATEOFSAMPLE, format = "%d.%m.%Y")
samp_rd <- samp_rd %>% 
  dplyr::filter(!is.na(METABONAME)) 
complete_couples <- samp_rd$COUPLENUMBER[which(duplicated(samp_rd$COUPLENUMBER))]

couple_info <- samp_rd %>% 
  dplyr::filter(COUPLENUMBER %in% complete_couples) %>% 
  arrange(COUPLENUMBER) %>% 
  column_to_rownames("METABONAME") %>% 
  rownames_to_column("rn") %>% 
  mutate(GROUP = as.factor(tolower(GROUP))) %>% 
  column_to_rownames("rn")
  
gender_donors <- couple_info$GENDER[c(TRUE,FALSE)]
gender_recip <- couple_info$GENDER[c(FALSE,TRUE)]
gender_comp <- paste0(gender_donors, gender_recip)

age_recip <- couple_info$DOG[c(FALSE,TRUE)] - couple_info$DOB[c(FALSE,TRUE)]

ifelse(grepl("R", rownames(couple_info)[1]), 
       cmv_donors <- couple_info$CMVStatus[c(FALSE,TRUE)],
       cmv_donors <- couple_info$CMVStatus[c(TRUE,FALSE)])

ifelse(grepl("R", rownames(couple_info)[1]), 
       cmv_recip <- couple_info$CMVStatus[c(TRUE,FALSE)],
       cmv_recip <- couple_info$CMVStatus[c(FALSE,TRUE)])

cmv_comp <- paste0(cmv_donors, cmv_recip)

couple_info <- couple_info %>% 
  rownames_to_column("rownames") %>% 
  mutate(gender_comp = rep(gender_comp, each = 2),
         age_recip = rep(age_recip, each = 2),
         cmv_comp = rep(cmv_comp, each = 2)) %>% 
  column_to_rownames("rownames")

head(couple_info[,c(1,2,3,5,6,55,56,57)])
```


We then need to gather the features that were extracted from the three data
sources:

- CyTOF features that were selected in both the St Louis and the Cryostem cohort, 
  with a threshold of 0.90. These features had the same behaviour in both cohorts.

```{r, echo = FALSE}
tmp_r <- readRDS("../data/cyto/log2/recip/pctgs_sel_ft_recip_TNT_90.RDS")
sel_beh_r <- read.xlsx("../data/cyto_national/log2/recip/perm_val_beh_TNT_90_thresh.xlsx")
tmp_d <- readRDS("../data/cyto/log2/donors/pctgs_sel_ft_donors_TNT_90.RDS") 
sel_beh_d <- read.xlsx("../data/cyto_national/log2/donors/perm_val_beh_TNT_90_thresh.xlsx")

couple_info_D <- couple_info %>% 
  dplyr::filter(grepl("^D", Id.Cryostem.R)) %>% 
  arrange(COUPLENUMBER)

couple_info_R <- couple_info %>% 
  dplyr::filter(grepl("^R", Id.Cryostem.R)) %>% 
  arrange(COUPLENUMBER)

couple_info_tmp <- couple_info_D %>% 
  mutate(recip_names = couple_info_R$Id.Cryostem.R)

tmp_r <- tmp_r %>% 
  select(sel_beh_r$features) %>% 
  rownames_to_column("Id.Cryostem.R") 

tmp_d <- tmp_d %>%  
  rownames_to_column("Id.Cryostem.R") %>% 
  inner_join(couple_info_tmp, by = "Id.Cryostem.R") %>% 
  select(recip_names, sel_beh_d$features)

colnames(tmp_d)[1] <- "Id.Cryostem.R"

colnames(tmp_r)[-c(1)] <- paste0("cyt_R_", colnames(tmp_r)[-c(1)])
colnames(tmp_d)[-c(1)] <- paste0("cyt_D_", colnames(tmp_d)[-c(1)])
  
ft_rdr_cyto <- tmp_r %>% 
  inner_join(tmp_d, by = c("Id.Cryostem.R"))

```

- Metabolomics features that were selected in both the St Louis and the Cryostem cohort, 
  with a threshold of 0.90. These features had the same behaviour in both cohorts.

```{r, echo = FALSE}
tmp_r <- readRDS("../data/metabo/v2_012020/recip/pctgs_sel_ft_recip_TNT_90.RDS")
sel_beh_r <- read.xlsx("../data/metabo/v2_012020/recip/perm_val_beh_TNT_90_thresh.xlsx")
tmp_d <- readRDS("../data/metabo/v2_012020/donors/pctgs_sel_ft_donors_TNT_90.RDS") 
sel_beh_d <- read.xlsx("../data/metabo/v2_012020/donors/perm_val_beh_TNT_90_thresh.xlsx")

couple_info_D <- couple_info %>% 
  rownames_to_column("METABONAME") %>% 
  dplyr::filter(grepl("^D", Id.Cryostem.R)) %>% 
  arrange(COUPLENUMBER)

couple_info_R <- couple_info %>% 
  rownames_to_column("METABONAME") %>% 
  dplyr::filter(grepl("^R", Id.Cryostem.R)) %>% 
  arrange(COUPLENUMBER)

couple_info_tmp <- couple_info_D %>% 
  mutate(recip_names = couple_info_R$Id.Cryostem.R)

tmp_d <- tmp_d %>%  
  rownames_to_column("METABONAME") %>% 
  inner_join(couple_info_tmp, by = "METABONAME") %>% 
  select(recip_names, sel_beh_d$features)

colnames(tmp_d)[1] <- "Id.Cryostem.R"

couple_info_tmp <- couple_info_R %>% 
  mutate(recip_names = couple_info_R$Id.Cryostem.R)

tmp_r <- tmp_r %>%
  rownames_to_column("METABONAME") %>% 
  inner_join(couple_info_tmp, by = "METABONAME") %>% 
  select(recip_names, sel_beh_r$features)

colnames(tmp_r)[1] <- "Id.Cryostem.R"

colnames(tmp_r)[-c(1)] <- paste0("met_R_", colnames(tmp_r)[-c(1)])
colnames(tmp_d)[-c(1)] <- paste0("met_D_", colnames(tmp_d)[-c(1)])
  
ft_rdr_metabo <- tmp_r %>% 
  inner_join(tmp_d, by = c("Id.Cryostem.R"))
```

- Transcriptomics features that were selected in both the St Louis and the Cryostem cohort, 
  with a threshold of 0.90. These features had the same behaviour in both cohorts.

```{r, echo = FALSE}
load("../data/RNAseq/recip_only/rna_recip_louis_TNT_90.RData")
sel_beh_r <- read.xlsx("../data/RNAseq/recip_only/perm_val_beh_TNT_90_thresh.xlsx")
load("../data/RNAseq/donors_only/rna_donor_louis_TNT_90.RData")
sel_beh_d <- read.xlsx("../data/RNAseq/donors_only/perm_val_beh_TNT_90_thresh.xlsx")

couple_info_D <- couple_info %>% 
  dplyr::filter(grepl("^D", Id.Cryostem.R)) %>% 
  arrange(COUPLENUMBER)

couple_info_R <- couple_info %>% 
  dplyr::filter(grepl("^R", Id.Cryostem.R)) %>% 
  arrange(COUPLENUMBER)

couple_info_tmp <- couple_info_D %>% 
  mutate(recip_names = couple_info_R$Id.Cryostem.R)

tmp_r <- rna_recip_louis_TNT_90 %>% 
  select(sel_beh_r$sel_beh) %>% 
  rownames_to_column("Id.Cryostem.R") 

tmp_d <- rna_donor_louis_TNT_90 %>%  
  rownames_to_column("Id.Cryostem.R") %>% 
  inner_join(couple_info_tmp, by = "Id.Cryostem.R") %>% 
  select(recip_names, sel_beh_d$sel_beh)

colnames(tmp_d)[1] <- "Id.Cryostem.R"

colnames(tmp_r)[-c(1)] <- paste0("rna_R_", colnames(tmp_r)[-c(1)])
colnames(tmp_d)[-c(1)] <- paste0("rna_D_", colnames(tmp_d)[-c(1)])
  
ft_rdr_transcripto <- tmp_r %>% 
  inner_join(tmp_d, by = c("Id.Cryostem.R"))
```

I merge all features together:

```{r, echo = FALSE}
ft_integ <- ft_rdr_cyto %>% 
  inner_join(ft_rdr_metabo, by = c("Id.Cryostem.R")) %>% 
  inner_join(ft_rdr_transcripto, by = c("Id.Cryostem.R")) 
print(paste0("We have information on the CyTOF, metabolomics and transcriptomics data for ",
             nrow(ft_integ), " patients in the St Louis cohort."))
```

#### Principal component analysis

We can first generate a PCA on the data:
```{r, echo = FALSE}
mat4pca <- ft_integ %>% 
  select(-Id.Cryostem.R) %>%  
  apply(2, scale) %>% 
  as.matrix() 
rownames(mat4pca) <- ft_integ$Id.Cryostem.R
pca <- prcomp(mat4pca, scale. = TRUE)
```

The first principal component seems to hold much more information than the others
```{r, echo = FALSE}
plot(pca)
```

We can investigate how much of the principal components are driven by the 
different data types.
```{r, echo = FALSE}
pca_var <- pca$rotation %>% as.data.frame
pca_ft <- rep("x", nrow(pca_var))
pca_ft[grep("cyt_", rownames(pca_var))] <- "cyto"
pca_ft[grep("met_", rownames(pca_var))] <- "metabo"
pca_ft[grep("rna_", rownames(pca_var))] <- "rna"

pca_orig <- rep("x", nrow(pca_var))
pca_orig[grep("_R_", rownames(pca_var))] <- "R"
pca_orig[grep("_D_", rownames(pca_var))] <- "D"

pca_var <- pca_var %>% 
  rownames_to_column("rn") %>% 
  mutate(feature_orig = as.factor(pca_ft),
         feature_type = as.factor(pca_orig))
```

We can see the data origin (CyTOF, metabolomics or transcriptomics) 
of the variables in each PC.
```{r, echo = FALSE}
pc1 <- pca_var %>% 
  column_to_rownames("rn") %>% 
  group_by(feature_orig) %>% 
  summarise_if(is.numeric, sum)
pc1m <- pc1 %>% gather(PC, value, -feature_orig) %>% 
  mutate(label_ypos = round(value))

pc1m_sub <- pc1m[1:30,] %>% 
  mutate(PC = factor(PC, levels = c("PC1", "PC2", "PC3", "PC4", "PC5", "PC6", 
                                    "PC7", "PC8", "PC9", "PC10")))

ggplot(data=pc1m_sub, aes(x=PC, y=value, fill=feature_orig)) +
  geom_bar(stat="identity")+
  # geom_text(aes(y=label_ypos, label=value), vjust=1.6, 
  #           color="white", size=3.5)+
  #scale_fill_brewer(palette="Paired")+
  scale_fill_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  theme_minimal()
```

We can also see how the features coming from recipients and donors are 
distributed in the PCs:
```{r, echo = FALSE}
pc1 <- pca_var %>% 
  column_to_rownames("rn") %>% 
  group_by(feature_type) %>% 
  summarise_if(is.numeric, sum)
pc1m <- pc1 %>% gather(PC, value, -feature_type) %>% 
  mutate(label_ypos = round(value))

pc1m_sub <- pc1m[1:30,] %>% 
  mutate(PC = factor(PC, levels = c("PC1", "PC2", "PC3", "PC4", "PC5", "PC6", 
                                    "PC7", "PC8", "PC9", "PC10")))

ggplot(data=pc1m_sub, aes(x=PC, y=value, fill=feature_type)) +
  geom_bar(stat="identity")+
  # geom_text(aes(y=label_ypos, label=value), vjust=1.6, 
  #           color="white", size=3.5)+
  scale_fill_brewer(palette="Paired")+
  #scale_fill_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  theme_minimal()
```

We can see how these principal components are correlated with tolerance. 
The first PC is very correlated with the tolerance. The PC4 (and slightly 
PC8 and PC10) also look correlated with tolerance.
```{r, echo = FALSE}
complete_patients <- rownames(mat4pca)
info_complete_patients <- couple_info %>%
  rownames_to_column("METABONAME") %>% 
  dplyr::filter(Id.Cryostem.R %in% complete_patients) %>% 
  arrange(Id.Cryostem.R) %>% 
  mutate(group = as.character(GROUP))
info_complete_patients$group[which(info_complete_patients$group != "non_tolerant")] <- "tolerant"
info_complete_patients$group <- as.factor(info_complete_patients$group)

pca_2plot <- left_join(ft_integ, info_complete_patients, by = c("Id.Cryostem.R")) %>% 
  mutate(PC1 = pca$x[,1],
         PC2 = pca$x[,2],
         PC3 = pca$x[,3],
         PC4 = pca$x[,4])

barplot(abs(cor(pca$x[,1:10], as.numeric(pca_2plot$group))[,1]))
saveRDS(list(pca = pca, pca_2plot = pca_2plot), file = "../data/integ/pca_TNT_st_louis_recip_donors.RDS")
```

We can then see how the PCs are correlated with the different informations 
that we have on the patients.

The first PC seems to be most correlated with group, cGVHD and ABO.incompatibility:

```{r, echo = FALSE}
totest <- pca_2plot[,c("group", "cmv_comp", "age_recip", "gender_comp", "cGVHD",
                       "aGVHD", "TBI", "SAL", "CONDITIONING_INTENSITY",
                       "CONDITIONING_REGIMEN", "SOURCEOFGRAFT", 
                       "DISORDER_GROUP", "HEMATOLOGICDISORDER", "ABO.incompatibility",
                       "GENDER")]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,1], totest)), horiz = TRUE, las = 2, cex.names = 0.5, xlim = c(0,0.6))
```

The second PC seems to be slightly correlated with the disorder group:

```{r, echo = FALSE}
totest <- pca_2plot[,c("group", "cmv_comp", "age_recip", "gender_comp", "cGVHD",
                       "aGVHD", "TBI", "SAL", "CONDITIONING_INTENSITY",
                       "CONDITIONING_REGIMEN", "SOURCEOFGRAFT", 
                       "DISORDER_GROUP", "HEMATOLOGICDISORDER", "ABO.incompatibility",
                       "GENDER")]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,2], totest)), horiz = TRUE, las = 2, cex.names = 0.5, xlim = c(0,0.6))
```

The third PC seems to be slightly correlated with aGVHD:

```{r, echo = FALSE}
totest <- pca_2plot[,c("group", "cmv_comp", "age_recip", "gender_comp", "cGVHD",
                       "aGVHD", "TBI", "SAL", "CONDITIONING_INTENSITY",
                       "CONDITIONING_REGIMEN", "SOURCEOFGRAFT", 
                       "DISORDER_GROUP", "HEMATOLOGICDISORDER", "ABO.incompatibility",
                       "GENDER")]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,3], totest)), horiz = TRUE, las = 2, cex.names = 0.5, xlim = c(0,0.6))
```

The fourth PC seems to be most correlated with group and age_recip:

```{r, echo = FALSE}
totest <- pca_2plot[,c("group", "cmv_comp", "age_recip", "gender_comp", "cGVHD",
                       "aGVHD", "TBI", "SAL", "CONDITIONING_INTENSITY",
                       "CONDITIONING_REGIMEN", "SOURCEOFGRAFT", 
                       "DISORDER_GROUP", "HEMATOLOGICDISORDER", "ABO.incompatibility",
                       "GENDER")]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,4], totest)), horiz = TRUE, las = 2, cex.names = 0.5, xlim = c(0,0.6))
```

Since the 1st and 4th PCs seem to be most correlated with tolerance, we visualise
the patients in these two dimensions.
We observe that these 2 components seem to separate the tolerant recipients 
(on the bottom-left) from the non tolerant recipients (on the top-right).
We can also see how the recipients are displayed in these two dimensions 
according to ABO incompatibility, recipients age and cGVHD.
```{r, echo = FALSE}
ggplot(pca_2plot, aes(x= PC1, y= PC4, col= group, label = Id.Cryostem.R)) + 
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Separation of tolerant vs non tolerant recipients")
ggplot(pca_2plot, aes(x= PC1, y= PC4, col= ABO.incompatibility, label = Id.Cryostem.R)) + 
  scale_color_manual(values = c("chartreuse2", "black", "gray48")) +
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Separation of recipients based on ABO.incompatibility")
ggplot(pca_2plot, aes(x= PC1, y= PC4, col= age_recip, label = Id.Cryostem.R)) +
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Gradient in the recipients based on age")
ggplot(pca_2plot, aes(x= PC1, y= PC4, col= as.factor(cGVHD), label = Id.Cryostem.R)) +
  scale_color_manual(values = c("black", "red")) +
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Separation of recipients based on cGVHD")
```



##### Features expressed in the first principal component

We can first look at the PC1, since it holds most of the data's variability:
We first focus on the features that are overexpressed in the tolerant recipients 
(= the features that are most expressed on the LEFT of the PC1). Therefore, the 
lower the feature value, the more it is expressed in tolerant recipients.
```{r, echo = FALSE}
#plot(sort(pca$rotation[,1][which(pca$rotation[,1]>0)]))
overexp <- sort(pca$rotation[,1][which(pca$rotation[,1]<0)])
over <- as.data.frame(overexp)
orig <- strsplit(rownames(over), "_")
orig <- map(orig, function(x){x[1]}) %>% unlist
over$orig <- orig

type <- rep("x", nrow(over))
type[grepl("_R_", rownames(over))] <- "R"
type[grepl("_D_", rownames(over))] <- "D"
over$type <- type

ggplot(over, aes(x = orig, y = overexp, col = orig)) +
  scale_color_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  geom_boxplot() +
  geom_jitter()
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 

ggplot(over, aes(x = orig, y = overexp)) +
  geom_boxplot() +
  geom_jitter(aes(x = orig, y = overexp, col = type), data = over) 
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 
```

We extract the "importance values" for these genes, metabolites and CyTOF features 
in an excel file. Here's a sample of the lowest features in the PC1 (under the
-0.6 threshold) :
```{r, echo = FALSE}
most_imp <- over[which(over$overexp < -0.06),] %>% 
  rownames_to_column("feature") %>% 
  mutate(feature = gsub("^[a-z]+_", "", feature)) %>% 
  arrange(overexp)
most_imp

most_imp2 <- most_imp %>% 
  arrange(orig)

write.xlsx(most_imp, file = "../data/integ/RD/tol_ft_PC1.xlsx")
write.xlsx(most_imp2, file = "../data/integ/RD/tol_ft_PC1_ordered_by_origin.xlsx")
```

We then focus on the features that are overexpressed in the non tolerant recipients 
(= the features that are most expressed on the RIGHT of the PC1). Therefore, the 
higher the feature value, the more it is expressed in non tolerant recipients.
```{r, echo = FALSE}
#plot(sort(pca$rotation[,1][which(pca$rotation[,1]>0)]))
overexp <- sort(pca$rotation[,1][which(pca$rotation[,1]>0)])
over <- as.data.frame(overexp)
orig <- strsplit(rownames(over), "_")
orig <- map(orig, function(x){x[1]}) %>% unlist
over$orig <- orig

type <- rep("x", nrow(over))
type[grepl("_R_", rownames(over))] <- "R"
type[grepl("_D_", rownames(over))] <- "D"
over$type <- type

ggplot(over, aes(x = orig, y = overexp, col = orig)) +
  scale_color_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  geom_boxplot() +
  geom_jitter()
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 

ggplot(over, aes(x = orig, y = overexp)) +
  geom_boxplot() +
  geom_jitter(aes(x = orig, y = overexp, col = type), data = over) 
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 
```

We extract the "importance values" for these genes, metabolites and CyTOF features 
in an excel file. Here's a sample of the highest features in the PC1 (over the
0.5 threshold) :
```{r, echo = FALSE}
most_imp <- over[which(over$overexp > 0.05),] %>% 
  rownames_to_column("feature") %>% 
  mutate(feature = gsub("^[a-z]+_", "", feature)) %>% 
  arrange(desc(overexp))
most_imp

most_imp2 <- most_imp %>% 
  arrange(orig)

write.xlsx(most_imp, file = "../data/integ/RD/nontol_ft_PC1.xlsx")
write.xlsx(most_imp2, file = "../data/integ/RD/nontol_ft_PC1_ordered_by_origin.xlsx")
```

##### Features expressed in the fourth principal component

We can first look at the PC1, since it holds most of the data's variability:
We first focus on the features that are overexpressed in the tolerant recipients 
(= the features that are most expressed on the LEFT of the PC1). Therefore, the 
lower the feature value, the more it is expressed in tolerant recipients.
```{r, echo = FALSE}
#plot(sort(pca$rotation[,1][which(pca$rotation[,1]>0)]))
overexp <- sort(pca$rotation[,4][which(pca$rotation[,4]<0)])
over <- as.data.frame(overexp)
orig <- strsplit(rownames(over), "_")
orig <- map(orig, function(x){x[1]}) %>% unlist
over$orig <- orig

type <- rep("x", nrow(over))
type[grepl("_R_", rownames(over))] <- "R"
type[grepl("_D_", rownames(over))] <- "D"
over$type <- type

ggplot(over, aes(x = orig, y = overexp, col = orig)) +
  scale_color_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  geom_boxplot() +
  geom_jitter()
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 

ggplot(over, aes(x = orig, y = overexp)) +
  geom_boxplot() +
  geom_jitter(aes(x = orig, y = overexp, col = type), data = over) 
```

We extract the "importance values" for these genes, metabolites and CyTOF features 
in an excel file. Here's a sample of the lowest features in the PC1 (under the
-0.6 threshold) :
```{r, echo = FALSE}
most_imp <- over[which(over$overexp < -0.1),] %>% 
  rownames_to_column("feature") %>% 
  mutate(feature = gsub("^[a-z]+_", "", feature)) %>% 
  arrange(overexp)
most_imp

most_imp2 <- most_imp %>% 
  arrange(orig)

write.xlsx(most_imp, file = "../data/integ/RD/tol_ft_PC4.xlsx")
write.xlsx(most_imp2, file = "../data/integ/RD/tol_ft_PC4_ordered_by_origin.xlsx")
```

We then focus on the features that are overexpressed in the non tolerant recipients 
(= the features that are most expressed on the RIGHT of the PC1). Therefore, the 
higher the feature value, the more it is expressed in non tolerant recipients.
```{r, echo = FALSE}
overexp <- sort(pca$rotation[,4][which(pca$rotation[,4]>0)])
over <- as.data.frame(overexp)
orig <- strsplit(rownames(over), "_")
orig <- map(orig, function(x){x[1]}) %>% unlist
over$orig <- orig

type <- rep("x", nrow(over))
type[grepl("_R_", rownames(over))] <- "R"
type[grepl("_D_", rownames(over))] <- "D"
over$type <- type

ggplot(over, aes(x = orig, y = overexp, col = orig)) +
  scale_color_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  geom_boxplot() +
  geom_jitter()

ggplot(over, aes(x = orig, y = overexp)) +
  geom_boxplot() +
  geom_jitter(aes(x = orig, y = overexp, col = type), data = over) 
```

We extract the "importance values" for these genes, metabolites and CyTOF features 
in an excel file. Here's a sample of the highest features in the PC1 (over the
0.5 threshold) :
```{r, echo = FALSE}
most_imp <- over[which(over$overexp > 0.1),] %>% 
  rownames_to_column("feature") %>% 
  mutate(feature = gsub("^[a-z]+_", "", feature)) %>% 
  arrange(desc(overexp))
most_imp

most_imp2 <- most_imp %>% 
  arrange(orig)

write.xlsx(most_imp, file = "../data/integ/RD/non_tol_ft_PC4.xlsx")
write.xlsx(most_imp2, file = "../data/integ/RD/non_tol_ft_PC4_ordered_by_origin.xlsx")
```

We can try to build a model that would predict a patients output (tolerant or not)
based on the PCs:
```{r, echo = FALSE}
data2model <- as.data.frame(pca$x) %>% mutate(group = pca_2plot$group)

set.seed(1) 

# Define training control
train.control <- caret::trainControl(method = "cv", number = 5)
# Train the model

model <- caret::train(group ~., data = data2model, method = "svmLinearWeights2",
               trControl = train.control)
# Summarize the results
print(model)
```

The results haven't been improved by the addition of the couple's features.


#### Correlations between features

We can now see how the variables that we extracted from recipients and donors
are coorelated. Correlations between variables are represented by edges in 
the graph. We kept only correlations that were common between the St Louis 
and the Cryostem cohort.

```{r, echo = FALSE}
pval_threshold <- 0.01
compar <- "TNT"

# St Louis
gr_medians_l <- pca_2plot %>%
  select(c(Id.Cryostem.R, group, colnames(ft_integ))) %>% 
  column_to_rownames("Id.Cryostem.R") %>% 
  group_by(group) %>% 
  summarize_if(is.numeric, median) %>% 
  as.data.frame()
rownames(gr_medians_l) <- gr_medians_l$group

tmp_l <- ft_integ %>% 
  column_to_rownames("Id.Cryostem.R") %>% 
  apply(2, scale)

# correlation St Louis
ft_df <- tmp_l %>% 
  as.data.frame() %>% 
  mutate(group = pca_2plot$group) %>% 
  magrittr::set_rownames(ft_integ$Id.Cryostem.R)

correlations <- ft_df %>%
  dplyr::select(-group) %>%
  as.matrix %>%
  cor %>%
  as.data.frame %>%
  rownames_to_column(var = 'var1') %>%
  gather(var2, value, -var1) %>%
  dplyr::filter(var1 < var2)
  
# cor.tests to define which correlations are significant
cor_sel <- correlations[which(correlations[,3] > 0),]
pvals <- rep(1, nrow(cor_sel))
for(i in 1: nrow(cor_sel)){
  var1 <- ft_df[,cor_sel[i,1]]
  var2 <- ft_df[,cor_sel[i,2]]
  pvals[i] <- cor.test(var1, var2, method = "spearman", exact = FALSE)$p.value
}
  
cor_sel <- cor_sel[which(pvals < pval_threshold),]

  
  
# Cryostem
cryo_for_cor <- readRDS("../data/integ_national/data_for_correlation.RDS")
gr_c <- tolower(as.character(cryo_for_cor$mat_withinfo$GROUP))
gr_c[which(gr_c != "non_tolerant")] <- "tolerant"

gr_medians_c <- cryo_for_cor$ft_integ %>%
  column_to_rownames("Id.Cryostem.R") %>% 
  mutate(group = as.factor(gr_c)) %>% 
  group_by(group) %>% 
  summarize_if(is.numeric, median) %>% 
  as.data.frame()
rownames(gr_medians_c) <- gr_medians_c$group

tmp_c <- cryo_for_cor$ft_integ %>% 
  column_to_rownames("Id.Cryostem.R") %>% 
  apply(2, scale)

# Correlations Cryostem
other <- tmp_c %>% 
  as.data.frame() %>% 
  mutate(group = gr_c) %>% 
  magrittr::set_rownames(cryo_for_cor$ft_integ$Id.Cryostem.R)

correlations_other <- other %>%
  dplyr::select(-group) %>%
  as.matrix %>%
  cor %>%
  as.data.frame %>%
  rownames_to_column(var = 'var1') %>%
  gather(var2, value, -var1) %>%
  dplyr::filter(var1 < var2)
  
# cor.tests to define which correlations are significant
cor_sel_other <- correlations_other[which(correlations_other[,3] > 0),]
pvals_other <- rep(1, nrow(cor_sel_other))
for(i in 1: nrow(cor_sel_other)){
  var1 <- other[,cor_sel_other[i,1]]
  var2 <- other[,cor_sel_other[i,2]]
  pvals_other[i] <- cor.test(var1, var2, method = "spearman", exact = FALSE)$p.value
}
  
cor_sel_other <- cor_sel_other[which(pvals_other < pval_threshold),]



  
# Both cohorts
correlated_both <- dplyr::intersect(cor_sel[,1:2], cor_sel_other[,1:2])
cor <- rep(0, nrow(correlated_both))
for(i in 1:nrow(correlated_both)){
  cor1 <- cor_sel[which((cor_sel[,1] == correlated_both[i,1]) &
                          (cor_sel[,2] == correlated_both[i,2])), 3]
  cor2 <- cor_sel_other[which((cor_sel_other[,1] == correlated_both[i,1]) &
                                (cor_sel_other[,2] == correlated_both[i,2])), 3]
  cor[i] <- mean(cor1, cor2)
}
correlated_both$value <- cor

gr <- graph_from_data_frame(correlated_both, directed = FALSE,
                            vertices=colnames(ft_df)[-which(colnames(ft_df)=="group")])
nodes <- get.vertex.attribute(gr)
  
if(compar == "TNT"){
  color_nodes <- rep("#FF000080", length(nodes$name))
  
  gr_medians <- ft_df %>%
    group_by(group) %>%
    summarize_if(is.numeric, median) %>%
    as.data.frame()
  rownames(gr_medians) <- gr_medians$group
  
  medians <- gr_medians[,nodes$name]
  for (gene in seq_along(colnames(medians))){
    if (medians["tolerant", gene] == max(medians[, gene])){
      color_nodes[gene] <- "#0000FF80"
    }
  }
} else if(compar == "PS"){
  color_nodes <- rep("#0000FF80", length(nodes$name))
  
  gr_medians <- ft_df %>%
    group_by(group) %>%
    summarize_if(is.numeric, median) %>%
    as.data.frame()
  rownames(gr_medians) <- gr_medians$group
  
  medians <- gr_medians[,nodes$name]
  for (gene in seq_along(colnames(medians))){
    if (medians["primary_tolerant", gene] == max(medians[, gene])){
      color_nodes[gene] <- "#00FF0080"
    }
  }
}

color_frame <- rep("palegreen3", length(nodes$name))
color_frame[grep("met_",nodes$name)] <- "mediumorchid4"
color_frame[grep("rna_",nodes$name)] <- "sandybrown"

set.seed(42)
plot(gr, vertex.size = 7, vertex.label.cex = .5, edge.width = (E(gr)$value - .7) * 20,
     vertex.color = color_frame, vertex.frame.color= color_nodes)
legend(1, 1, legend=c(round(min(E(gr)$value), digits = 3), round(max(E(gr)$value), digits = 3)),
         col=c("grey", "grey"),
         lwd=c((min(E(gr)$value) - round(min(E(gr)$value), digits = 2)) * 10,
               (max(E(gr)$value) - round(min(E(gr)$value), digits = 2)) * 10),
         cex=0.8,
         title="Edge weight", text.font=4, bg='white')
```

```{r, echo = FALSE}
gr_components <- decompose(gr, mode = "strong", max.comps = NA,
                 min.vertices = 0)
comp_sizes <- sapply(gr_components, diameter)

list_components <- lapply(which(comp_sizes != 0), function(i){
  vertex_names <- vertex_attr(gr_components[[i]], index = V(gr_components[[i]]))[[1]]
  vertex_names
})

write.xlsx(list_components, file = "../data/integ_2_cohorts/graph_components_TNT_recipients_and_donors.xlsx")
write.xlsx(correlated_both, file = "../data/integ_2_cohorts/table_correlations_2cohorts_recipients_and_donors_TNT.xlsx")
```

#### Random forest modeling

To see if the features that we selected can help us to clssify the patients, we
apply a 5 fold cross validation setting on the 23 patients for which we have 
CyTOF, metabolomics and transcriptomics information.

```{r}
set.seed(1)
cv_id <- list(c(1:5), c(6:10), c(11:15), c(16:20), c(21:23))
cv_accuracies <- lapply(seq_along(cv_id), function(idx){
  train_set <- pca_2plot[-cv_id[[idx]], c("group", colnames(ft_integ)[-1])]
  test_set <- pca_2plot[cv_id[[idx]], c(colnames(ft_integ)[-1])]
  true_labels <- pca_2plot[cv_id[[idx]], "group"]
  rf <- ranger::ranger(group~., train_set, num.trees = 5000,
                           importance = "impurity")
  pred <- predict(rf, test_set)
  acc <- ( length(which(pred$predictions == true_labels))/length(true_labels) )
})

accuracy <- mean(unlist(cv_accuracies))
accuracy
```

#### Separate data type analyses

For comparison, we can see if we would have managed to separate the tolerant 
from non tolerant recipients and couples as well without integrating the data:

##### CyTOF:

We perform PCA using only the 49 features that were selected as informative 
in the recipients and couples in both cohorts:

```{r, echo = FALSE}
cyt4pca <- ft_rdr_cyto %>% 
  dplyr::filter(Id.Cryostem.R %in% ft_integ$Id.Cryostem.R) %>% 
  column_to_rownames("Id.Cryostem.R")
pca_cyto <- ade4::dudi.pca(cyt4pca, nf = 10, scannf = F)
```

The first comoponent seems to be capturing most of the data variability:
```{r, echo = FALSE}
plot(pca_cyto$eig, pch = 19)
```

```{r, echo = FALSE}
info_tmp <- couple_info %>% 
  rownames_to_column("rn") %>% 
  dplyr::filter(Id.Cryostem.R %in% rownames(cyt4pca)) %>%
  mutate(group = as.character(tolower(GROUP))) %>% 
  column_to_rownames("Id.Cryostem.R")
info_tmp <- info_tmp[rownames(cyt4pca),]
info_tmp$group[which(info_tmp$group != "non_tolerant")] <- "tolerant"
info_tmp$group <- as.factor(info_tmp$group)
```

We can see how these principal components are correlated with tolerance. 
The first PC is very correlated with the tolerance. The PC10 is the 2nd mostly
correlated with tolerance.
```{r, echo = FALSE}
barplot(abs(cor(pca_cyto$li[,1:10], as.numeric(info_tmp$group))[,1]))
```

The non tolerant recipients are slightly better separated from the tolerant 
recipients than in the integrative PCA. 
```{r, echo = FALSE}
#information on the patients: 
info_pca <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R")
info_pca <- info_pca[rownames(cyt4pca),]
ade4::s.class(pca_cyto$li[,c(1,10)], info_pca$group, col=c("red", "blue"))
# ade4::s.class(pca_cyto$li[,c(1,6)], as.factor(info_pca$cGVHD), 
#               col = c("black", "red", "lightgreen", "green3"))
```

We can also have a look at the features that are driving these two components:
```{r, echo = FALSE}
keep1 <- order(abs(pca_cyto$c1$CS1), decreasing = T)[1:20]
# keep2 <- order(abs(pca_cyto$c1$CS2), decreasing = T)[1:5]
# tokeep <- unique(c(keep1, keep2))
tokeep <- keep1
ade4::s.arrow(pca_cyto$c1[tokeep,c(1,10)], boxes = F, clabel = 0.5)
```

We can try to build a model that would predict a patients output (tolerant or not)
based on the PCs:
```{r, echo = FALSE}
data2model <- as.data.frame(pca_cyto$li) %>% mutate(group = info_pca$group)

set.seed(1) 

# Define training control
train.control <- caret::trainControl(method = "cv", number = 5)
# Train the model

model <- caret::train(group ~., data = data2model, method = "svmLinearWeights2",
               trControl = train.control)
# Summarize the results
print(model)
```

The accuracy is quite lower than in the integrative model and hasn't improved 
with the addition of features from the donors.

To see if the features that we selected can help us to clssify the patients, we
can also apply a 5 fold cross validation setting on the CyTOF features of the 
same 23 patients.

```{r}
set.seed(1)
info_rf <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R")
info_rf <- info_rf[rownames(cyt4pca),]
cyt4rf <- cbind(info_rf$group, cyt4pca)
colnames(cyt4rf)[1] <- "group"

cv_id <- list(c(1:5), c(6:10), c(11:15), c(16:20), c(21:23))
cv_accuracies <- lapply(seq_along(cv_id), function(idx){
  train_set <- cyt4rf[-cv_id[[idx]], ]
  test_set <- cyt4rf[cv_id[[idx]], -1]
  true_labels <- cyt4rf[cv_id[[idx]], "group"]
  rf <- ranger::ranger(group~., train_set, num.trees = 5000,
                           importance = "impurity")
  pred <- predict(rf, test_set)
  acc <- ( length(which(pred$predictions == true_labels))/length(true_labels) )
})

accuracy <- mean(unlist(cv_accuracies))
accuracy
```

##### Metabolomics:
```{r, echo = FALSE}
met4pca <- ft_rdr_metabo %>% 
  dplyr::filter(Id.Cryostem.R %in% ft_integ$Id.Cryostem.R) %>% 
  column_to_rownames("Id.Cryostem.R")
pca_metabo <- ade4::dudi.pca(met4pca, nf = 10, scannf = F)
```

Variability captured by each component:
```{r, echo = FALSE}
plot(pca_metabo$eig, pch = 19)
```

```{r, echo = FALSE}
info_tmp <- couple_info %>% 
  rownames_to_column("rn") %>% 
  dplyr::filter(Id.Cryostem.R %in% rownames(met4pca)) %>%
  mutate(group = as.character(tolower(GROUP))) %>% 
  column_to_rownames("Id.Cryostem.R")
info_tmp <- info_tmp[rownames(met4pca),]
info_tmp$group[which(info_tmp$group != "non_tolerant")] <- "tolerant"
info_tmp$group <- as.factor(info_tmp$group)
```

We can see how these principal components are correlated with tolerance. 
The first PC is very correlated with the tolerance. The PC5 is the 2nd mostly
correlated with tolerance.
```{r, echo = FALSE}
barplot(abs(cor(pca_metabo$li[,1:10], as.numeric(info_tmp$group))[,1]))
```

The non tolerant recipients are slightly less well separated from the tolerant 
recipients than in the integrative PCA. 
```{r, echo = FALSE}
#information on the patients: 
info_pca <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R")
info_pca <- info_pca[rownames(met4pca),]
ade4::s.class(pca_metabo$li[,c(1,5)], info_pca$group, col=c("red", "blue"))
# ade4::s.class(pca_cyto$li[,c(1,6)], as.factor(info_pca$cGVHD), 
#               col = c("black", "red", "lightgreen", "green3"))
```

```{r, echo = FALSE}
keep1 <- order(abs(pca_metabo$c1$CS1), decreasing = T)[1:20]
# keep2 <- order(abs(pca_cyto$c1$CS2), decreasing = T)[1:5]
# tokeep <- unique(c(keep1, keep2))
tokeep <- keep1
ade4::s.arrow(pca_metabo$c1[tokeep, c(1,5)], boxes = F, clabel = 0.5)
```

We can try to build a model that would predict a patients output (tolerant or not)
based on the PCs:
```{r, echo = FALSE}
data2model <- as.data.frame(pca_metabo$li) %>% mutate(group = info_pca$group)

set.seed(1) 

# Define training control
train.control <- caret::trainControl(method = "cv", number = 5)
# Train the model

model <- caret::train(group ~., data = data2model, method = "svmLinearWeights2",
               trControl = train.control)
# Summarize the results
print(model)
```

Again, the accuracy of the model is slightly lower than the accuracy of the integrative model.

To see if the features that we selected can help us to clssify the patients, we
can also apply a 5 fold cross validation setting on the CyTOF features of the 
same 23 patients.

```{r}
set.seed(1)
info_rf <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R")
info_rf <- info_rf[rownames(met4pca),]
met4rf <- cbind(info_rf$group, met4pca)
colnames(met4rf)[1] <- "group"

cv_id <- list(c(1:5), c(6:10), c(11:15), c(16:20), c(21:23))
cv_accuracies <- lapply(seq_along(cv_id), function(idx){
  train_set <- met4rf[-cv_id[[idx]], ]
  test_set <- met4rf[cv_id[[idx]], -1]
  true_labels <- met4rf[cv_id[[idx]], "group"]
  rf <- ranger::ranger(group~., train_set, num.trees = 5000,
                           importance = "impurity")
  pred <- predict(rf, test_set)
  acc <- ( length(which(pred$predictions == true_labels))/length(true_labels) )
})

accuracy <- mean(unlist(cv_accuracies))
accuracy
```

##### Transcriptomics:
```{r, echo = FALSE}
rna4pca <- ft_rdr_transcripto %>%
  dplyr::filter(Id.Cryostem.R %in% ft_integ$Id.Cryostem.R) %>% 
  column_to_rownames("Id.Cryostem.R")
pca_transcripto <- ade4::dudi.pca(rna4pca, nf = 10, scannf = F)
```

Variability captured by each component:
```{r, echo = FALSE}
plot(pca_transcripto$eig, pch = 19)
```

```{r, echo = FALSE}
info_tmp <- couple_info %>% 
  rownames_to_column("rn") %>% 
  dplyr::filter(Id.Cryostem.R %in% rownames(rna4pca)) %>%
  mutate(group = as.character(tolower(GROUP))) %>% 
  column_to_rownames("Id.Cryostem.R")
info_tmp <- info_tmp[rownames(rna4pca),]
info_tmp$group[which(info_tmp$group != "non_tolerant")] <- "tolerant"
info_tmp$group <- as.factor(info_tmp$group)
```

We can see how these principal components are correlated with tolerance. 
The first PC is very correlated with the tolerance. The PC8 is the 2nd mostly
correlated with tolerance.
```{r, echo = FALSE}
barplot(abs(cor(pca_transcripto$li[,1:10], as.numeric(info_tmp$group))[,1]))
```

The non tolerant recipients are slightly less well separated from the tolerant 
recipients than in the integrative PCA. 
```{r, echo = FALSE}
#information on the patients: 
info_pca <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R")
info_pca <- info_pca[rownames(rna4pca),]
ade4::s.class(pca_transcripto$li[,c(1,8)], info_pca$group, col=c("red", "blue"))
# ade4::s.class(pca_cyto$li[,c(1,6)], as.factor(info_pca$cGVHD), 
#               col = c("black", "red", "lightgreen", "green3"))
```

```{r, echo = FALSE}
keep1 <- order(abs(pca_transcripto$c1$CS1), decreasing = T)[1:20]
# keep2 <- order(abs(pca_cyto$c1$CS2), decreasing = T)[1:5]
# tokeep <- unique(c(keep1, keep2))
tokeep <- keep1
ade4::s.arrow(pca_transcripto$c1[tokeep,c(1,8)], boxes = F, clabel = 0.5)
```

We can try to build a model that would predict a patients output (tolerant or not)
based on the PCs:
```{r, echo = FALSE}
data2model <- as.data.frame(pca_transcripto$li) %>% mutate(group = info_pca$group)

set.seed(1) 

# Define training control
train.control <- caret::trainControl(method = "cv", number = 5)
# Train the model

model <- caret::train(group ~., data = data2model, method = "svmLinearWeights2",
               trControl = train.control)
# Summarize the results
print(model)
```

It looks like, in our integrative approach, we managed to separate the recipients 
based on their tolerance better than when we look at the different data types
separately.

To see if the features that we selected can help us to clssify the patients, we
can also apply a 5 fold cross validation setting on the CyTOF features of the 
same 23 patients.

```{r}
set.seed(1)
info_rf <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R")
info_rf <- info_rf[rownames(rna4pca),]
rna4rf <- cbind(info_rf$group, rna4pca)
colnames(rna4rf)[1] <- "group"

cv_id <- list(c(1:5), c(6:10), c(11:15), c(16:20), c(21:23))
cv_accuracies <- lapply(seq_along(cv_id), function(idx){
  train_set <- rna4rf[-cv_id[[idx]], ]
  test_set <- rna4rf[cv_id[[idx]], -1]
  true_labels <- rna4rf[cv_id[[idx]], "group"]
  rf <- ranger::ranger(group~., train_set, num.trees = 5000,
                           importance = "impurity")
  pred <- predict(rf, test_set)
  acc <- ( length(which(pred$predictions == true_labels))/length(true_labels) )
})

accuracy <- mean(unlist(cv_accuracies))
accuracy
```



### Primary tolerant versus secondary tolerant recipients + donors

Since looking at differences between the three groups consisting of primary, 
secondary and non tolerant patients at the same time wasn't feasible, we divided
our problem in two distinct sub-problems. We now focus on the features that
seemed to play a role when trying to disinguish primary tolerant from secondary
tolerant recipients.

I extract informations on the recipients group, age, gender, ... to be able to 
use this information later on.

```{r, echo = FALSE}
samp_rd <- read.xlsx("~/Documents/VIB/Projects/Integrative_Paris/Local_cohort/Data synthesis local cohort Saint-Louis 17012019.xlsx")
samp_rd$DOB <- as.Date(samp_rd$DOB, format = "%d.%m.%Y")
samp_rd$DOG <- as.Date(samp_rd$DOG, format = "%d.%m.%Y")
samp_rd$DATEOFSAMPLE <- as.Date(samp_rd$DATEOFSAMPLE, format = "%d.%m.%Y")
# samp_rd <- samp_rd %>% 
#   dplyr::filter(!is.na(METABONAME)) 
complete_couples <- samp_rd$COUPLENUMBER[which(duplicated(samp_rd$COUPLENUMBER))]

couple_info <- samp_rd %>% 
  dplyr::filter(COUPLENUMBER %in% complete_couples) %>% 
  arrange(COUPLENUMBER) %>% 
  mutate(GROUP = as.factor(tolower(GROUP))) %>% 
  column_to_rownames("Id.Cryostem.R")
  
gender_donors <- couple_info$GENDER[c(TRUE,FALSE)]
gender_recip <- couple_info$GENDER[c(FALSE,TRUE)]
gender_comp <- paste0(gender_donors, gender_recip)

age_recip <- couple_info$DOG[c(FALSE,TRUE)] - couple_info$DOB[c(FALSE,TRUE)]

ifelse(grepl("R", rownames(couple_info)[1]), 
       cmv_donors <- couple_info$CMVStatus[c(FALSE,TRUE)],
       cmv_donors <- couple_info$CMVStatus[c(TRUE,FALSE)])

ifelse(grepl("R", rownames(couple_info)[1]), 
       cmv_recip <- couple_info$CMVStatus[c(TRUE,FALSE)],
       cmv_recip <- couple_info$CMVStatus[c(FALSE,TRUE)])

cmv_comp <- paste0(cmv_donors, cmv_recip)

couple_info <- couple_info %>% 
  rownames_to_column("rownames") %>% 
  mutate(gender_comp = rep(gender_comp, each = 2),
         age_recip = rep(age_recip, each = 2),
         cmv_comp = rep(cmv_comp, each = 2)) %>% 
  column_to_rownames("rownames")

head(couple_info[,c(1,2,3,5,6,55,56,57)])
```


We then need to gather the features that were extracted from the three data
sources:

- CyTOF features that were selected in both the St Louis and the Cryostem cohort, 
  with a threshold of 0.90. These features had the same behaviour in both cohorts.

```{r, echo = FALSE}
tmp_r <- readRDS("../data/cyto/log2/recip/pctgs_sel_ft_recip_PTST_90.RDS")
sel_beh_r <- read.xlsx("../data/cyto_national/log2/recip/perm_val_beh_PTST_90_thresh.xlsx")
tmp_d <- readRDS("../data/cyto/log2/donors/pctgs_sel_ft_donors_PTST_90.RDS") 
sel_beh_d <- read.xlsx("../data/cyto_national/log2/donors/perm_val_beh_PTST_90_thresh.xlsx")

couple_info_D <- couple_info %>%
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::filter(grepl("^D", Id.Cryostem.R)) %>% 
  arrange(COUPLENUMBER)

couple_info_R <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::filter(grepl("^R", Id.Cryostem.R)) %>% 
  arrange(COUPLENUMBER)

couple_info_tmp <- couple_info_D %>% 
  mutate(recip_names = couple_info_R$Id.Cryostem.R)

tmp_r <- tmp_r %>% 
  select(sel_beh_r$features) %>% 
  rownames_to_column("Id.Cryostem.R") 

tmp_d <- tmp_d %>%  
  rownames_to_column("Id.Cryostem.R") %>% 
  inner_join(couple_info_tmp, by = "Id.Cryostem.R") %>% 
  select(recip_names, sel_beh_d$features)

colnames(tmp_d)[1] <- "Id.Cryostem.R"

colnames(tmp_r)[-c(1)] <- paste0("cyt_R_", colnames(tmp_r)[-c(1)])
colnames(tmp_d)[-c(1)] <- paste0("cyt_D_", colnames(tmp_d)[-c(1)])
  
ft_rdr_cyto <- tmp_r %>% 
  inner_join(tmp_d, by = c("Id.Cryostem.R"))

```

- Metabolomics features that were selected in both the St Louis and the Cryostem cohort, 
  with a threshold of 0.90. These features had the same behaviour in both cohorts.

```{r, echo = FALSE}
tmp_r <- readRDS("../data/metabo/v2_012020/recip/pctgs_sel_ft_recip_PTST_90.RDS")
sel_beh_r <- read.xlsx("../data/metabo/v2_012020/recip/perm_val_beh_PTST_90_thresh.xlsx")
tmp_d <- readRDS("../data/metabo/v2_012020/donors/pctgs_sel_ft_donors_PTST_90.RDS") 
sel_beh_d <- read.xlsx("../data/metabo/v2_012020/donors/perm_val_beh_PTST_90_thresh.xlsx")

couple_info_D <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::filter(grepl("^D", Id.Cryostem.R)) %>% 
  arrange(COUPLENUMBER)

couple_info_R <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::filter(grepl("^R", Id.Cryostem.R)) %>% 
  arrange(COUPLENUMBER)

couple_info_tmp <- couple_info_D %>% 
  mutate(recip_names = couple_info_R$Id.Cryostem.R)

tmp_d <- tmp_d %>%  
  rownames_to_column("METABONAME") %>% 
  inner_join(couple_info_tmp, by = "METABONAME") %>% 
  select(recip_names, sel_beh_d$features)

colnames(tmp_d)[1] <- "Id.Cryostem.R"

couple_info_tmp <- couple_info_R %>% 
  mutate(recip_names = couple_info_R$Id.Cryostem.R)

tmp_r <- tmp_r %>%
  rownames_to_column("METABONAME") %>% 
  inner_join(couple_info_tmp, by = "METABONAME") %>% 
  select(recip_names, sel_beh_r$features)

colnames(tmp_r)[1] <- "Id.Cryostem.R"

colnames(tmp_r)[-c(1)] <- paste0("met_R_", colnames(tmp_r)[-c(1)])
colnames(tmp_d)[-c(1)] <- paste0("met_D_", colnames(tmp_d)[-c(1)])
  
ft_rdr_metabo <- tmp_r %>% 
  inner_join(tmp_d, by = c("Id.Cryostem.R"))
```

- Transcriptomics features that were selected in both the St Louis and the Cryostem cohort, 
  with a threshold of 0.90. These features had the same behaviour in both cohorts.

```{r, echo = FALSE}
load("../data/RNAseq/recip_only/rna_recip_louis_PS_90.RData")
sel_beh_r <- read.xlsx("../data/RNAseq/recip_only/perm_val_beh_PS_90_thresh.xlsx")
rna_donor_louis_PS_90 <- readRDS("../data/RNAseq/donors_only/rna_donor_louis_PS_90.RDS")
sel_beh_d <- read.xlsx("../data/RNAseq/donors_only/perm_val_beh_PS_90_thresh.xlsx")

couple_info_D <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::filter(grepl("^D", Id.Cryostem.R)) %>% 
  arrange(COUPLENUMBER)

couple_info_R <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::filter(grepl("^R", Id.Cryostem.R)) %>% 
  arrange(COUPLENUMBER)

couple_info_tmp <- couple_info_D %>% 
  mutate(recip_names = couple_info_R$Id.Cryostem.R)

tmp_r <- rna_recip_louis_PS_90 %>% 
  select(sel_beh_r$sel_beh) %>% 
  rownames_to_column("Id.Cryostem.R") 

tmp_d <- rna_donor_louis_PS_90 %>%  
  rownames_to_column("Id.Cryostem.R") %>% 
  inner_join(couple_info_tmp, by = "Id.Cryostem.R") %>% 
  select(recip_names, sel_beh_d$sel_beh)

colnames(tmp_d)[1] <- "Id.Cryostem.R"

colnames(tmp_r)[-c(1)] <- paste0("rna_R_", colnames(tmp_r)[-c(1)])
colnames(tmp_d)[-c(1)] <- paste0("rna_D_", colnames(tmp_d)[-c(1)])
  
ft_rdr_transcripto <- tmp_r %>% 
  inner_join(tmp_d, by = c("Id.Cryostem.R"))
```

I merge all features together:

```{r, echo = FALSE}
ft_integ <- ft_rdr_cyto %>% 
  inner_join(ft_rdr_metabo, by = c("Id.Cryostem.R")) %>% 
  inner_join(ft_rdr_transcripto, by = c("Id.Cryostem.R")) 
print(paste0("We have information on the CyTOF, metabolomics and transcriptomics data for ",
             nrow(ft_integ), " patients in the St Louis cohort."))
```

#### Principal component analysis

We can first generate a PCA on the data:
```{r, echo = FALSE}
mat4pca <- ft_integ %>% 
  select(-Id.Cryostem.R) %>%  
  apply(2, scale) %>% 
  as.matrix() 
rownames(mat4pca) <- ft_integ$Id.Cryostem.R
pca <- prcomp(mat4pca, scale. = TRUE)
```

The first principal component seems to hold much more information that the others
```{r, echo = FALSE}
plot(pca)
```

We can investigate how much of the principal components are driven by the 
different data types.
```{r, echo = FALSE}
pca_var <- pca$rotation %>% as.data.frame
pca_ft <- rep("x", nrow(pca_var))
pca_ft[grep("cyt_", rownames(pca_var))] <- "cyto"
pca_ft[grep("met_", rownames(pca_var))] <- "metabo"
pca_ft[grep("rna_", rownames(pca_var))] <- "rna"

pca_orig <- rep("x", nrow(pca_var))
pca_orig[grep("_R_", rownames(pca_var))] <- "R"
pca_orig[grep("_D_", rownames(pca_var))] <- "D"

pca_var <- pca_var %>% 
  rownames_to_column("rn") %>% 
  mutate(feature_orig = as.factor(pca_ft),
         feature_type = as.factor(pca_orig))
```

We can see the data origin (CyTOF, metabolomics or transcriptomics) 
of the variables in each PC.
```{r, echo = FALSE}
pc1 <- pca_var %>% 
  column_to_rownames("rn") %>% 
  group_by(feature_orig) %>% 
  summarise_if(is.numeric, sum)
pc1m <- pc1 %>% gather(PC, value, -feature_orig) %>% 
  mutate(label_ypos = round(value))

pc1m_sub <- pc1m[1:30,] %>% 
  mutate(PC = factor(PC, levels = c("PC1", "PC2", "PC3", "PC4", "PC5", "PC6", 
                                    "PC7", "PC8", "PC9", "PC10")))

ggplot(data=pc1m_sub, aes(x=PC, y=value, fill=feature_orig)) +
  geom_bar(stat="identity")+
  # geom_text(aes(y=label_ypos, label=value), vjust=1.6, 
  #           color="white", size=3.5)+
  #scale_fill_brewer(palette="Paired")+
  scale_fill_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  theme_minimal()
```

We can also see how the features coming from recipients and donors are 
distributed in the PCs:
```{r, echo = FALSE}
pc1 <- pca_var %>% 
  column_to_rownames("rn") %>% 
  group_by(feature_type) %>% 
  summarise_if(is.numeric, sum)
pc1m <- pc1 %>% gather(PC, value, -feature_type) %>% 
  mutate(label_ypos = round(value))

pc1m_sub <- pc1m[1:30,] %>% 
  mutate(PC = factor(PC, levels = c("PC1", "PC2", "PC3", "PC4", "PC5", "PC6", 
                                    "PC7", "PC8", "PC9", "PC10")))

ggplot(data=pc1m_sub, aes(x=PC, y=value, fill=feature_type)) +
  geom_bar(stat="identity")+
  # geom_text(aes(y=label_ypos, label=value), vjust=1.6, 
  #           color="white", size=3.5)+
  scale_fill_brewer(palette="Paired")+
  #scale_fill_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  theme_minimal()
```

We can see how these principal components are correlated with tolerance. 
The first PC is very correlated with the tolerance. The PC4 (and slightly 
PC8 and PC10) also look correlated with tolerance.
```{r, echo = FALSE}
complete_patients <- rownames(mat4pca)
info_complete_patients <- couple_info %>%
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::filter(Id.Cryostem.R %in% complete_patients) %>% 
  arrange(Id.Cryostem.R) %>% 
  mutate(group = as.character(GROUP))
# info_complete_patients$group[which(info_complete_patients$group != "non_tolerant")] <- "tolerant"
info_complete_patients$group <- as.factor(as.character(info_complete_patients$group))

pca_2plot <- left_join(ft_integ, info_complete_patients, by = c("Id.Cryostem.R")) %>% 
  mutate(PC1 = pca$x[,1],
         PC2 = pca$x[,2],
         PC3 = pca$x[,3],
         PC4 = pca$x[,4])

barplot(abs(cor(pca$x[,1:10], as.numeric(pca_2plot$group))[,1]))
saveRDS(list(pca = pca, pca_2plot = pca_2plot), file = "../data/integ/pca_PTST_st_louis_recip_donors.RDS")
```

We can then see how the PCs are correlated with the different informations 
that we have on the patients.

The first PC seems to be most correlated with group, SAL, cGVHD and ABO.incompatibility:

```{r, echo = FALSE}
totest <- pca_2plot[,c("group", "cmv_comp", "age_recip", "gender_comp", "cGVHD",
                       "aGVHD", "TBI", "SAL", "CONDITIONING_INTENSITY",
                       "CONDITIONING_REGIMEN", "SOURCEOFGRAFT", 
                       "DISORDER_GROUP", "HEMATOLOGICDISORDER", "ABO.incompatibility",
                       "GENDER")]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,1], totest)), horiz = TRUE, las = 2, cex.names = 0.5, xlim = c(0,0.6))
```

The second PC seems to be slightly correlated with the hematologic disorder,
source of graft and tolerance group:

```{r, echo = FALSE}
totest <- pca_2plot[,c("group", "cmv_comp", "age_recip", "gender_comp", "cGVHD",
                       "aGVHD", "TBI", "SAL", "CONDITIONING_INTENSITY",
                       "CONDITIONING_REGIMEN", "SOURCEOFGRAFT", 
                       "DISORDER_GROUP", "HEMATOLOGICDISORDER", "ABO.incompatibility",
                       "GENDER")]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,2], totest)), horiz = TRUE, las = 2, cex.names = 0.5, xlim = c(0,0.6))
```

The third PC seems to be slightly correlated with cmv compatibility between donors and recipients:

```{r, echo = FALSE}
totest <- pca_2plot[,c("group", "cmv_comp", "age_recip", "gender_comp", "cGVHD",
                       "aGVHD", "TBI", "SAL", "CONDITIONING_INTENSITY",
                       "CONDITIONING_REGIMEN", "SOURCEOFGRAFT", 
                       "DISORDER_GROUP", "HEMATOLOGICDISORDER", "ABO.incompatibility",
                       "GENDER")]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,3], totest)), horiz = TRUE, las = 2, cex.names = 0.5, xlim = c(0,0.6))
```

The fourth PC seems to be most correlated with gender compatibility between donors and recipients:

```{r, echo = FALSE}
totest <- pca_2plot[,c("group", "cmv_comp", "age_recip", "gender_comp", "cGVHD",
                       "aGVHD", "TBI", "SAL", "CONDITIONING_INTENSITY",
                       "CONDITIONING_REGIMEN", "SOURCEOFGRAFT", 
                       "DISORDER_GROUP", "HEMATOLOGICDISORDER", "ABO.incompatibility",
                       "GENDER")]
totest <- apply(totest, 2, function(x){as.numeric(as.factor(x))})
barplot(abs(cor(pca$x[,4], totest)), horiz = TRUE, las = 2, cex.names = 0.5, xlim = c(0,0.6))
```

Since the 1st and 2nd PCs seem to be most correlated with tolerance, we visualise
the patients in these two dimensions.
We observe that these 2 components seem to separate the primary tolerant 
recipients (on the left) from the secondary tolerant recipients (on the top-right).
We can also see how the recipients are displayed in these two dimensions 
according to ABO incompatibility, recipients age and cGVHD.
```{r, echo = FALSE}
ggplot(pca_2plot, aes(x= PC1, y= PC2, col= group, label = Id.Cryostem.R)) + 
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  scale_color_manual(values = c("green", "blue")) +
  ggtitle("Separation of tolerant vs non tolerant recipients")
ggplot(pca_2plot, aes(x= PC1, y= PC2, col= ABO.incompatibility, label = Id.Cryostem.R)) + 
  scale_color_manual(values = c("chartreuse2", "black", "gray48")) +
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Separation of recipients based on ABO.incompatibility")
ggplot(pca_2plot, aes(x= PC1, y= PC2, col= SAL, label = Id.Cryostem.R)) +
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Separation of recipients based on SAL")
ggplot(pca_2plot, aes(x= PC1, y= PC2, col= as.factor(SOURCEOFGRAFT), label = Id.Cryostem.R)) +
  scale_color_manual(values = c("black", "red")) +
  geom_point() +
  geom_text(aes(label=Id.Cryostem.R),hjust=0, vjust=0) +
  ggtitle("Separation of recipients based on source of graft")
```



##### Features expressed in the first prinipal component

We can first look at the PC1, since it holds most of the data's variability:
We first focus on the features that are overexpressed in the primary tolerant recipients 
(= the features that are most expressed on the LEFT of the PC1). Therefore, the 
lower the feature value, the more it is expressed in primary tolerant recipients.
```{r, echo = FALSE}
#plot(sort(pca$rotation[,1][which(pca$rotation[,1]>0)]))
overexp <- sort(pca$rotation[,1][which(pca$rotation[,1]<0)])
over <- as.data.frame(overexp)
orig <- strsplit(rownames(over), "_")
orig <- map(orig, function(x){x[1]}) %>% unlist
over$orig <- orig

type <- rep("x", nrow(over))
type[grepl("_R_", rownames(over))] <- "R"
type[grepl("_D_", rownames(over))] <- "D"
over$type <- type

ggplot(over, aes(x = orig, y = overexp, col = orig)) +
  scale_color_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  geom_boxplot() +
  geom_jitter()
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 

ggplot(over, aes(x = orig, y = overexp)) +
  geom_boxplot() +
  geom_jitter(aes(x = orig, y = overexp, col = type), data = over) 
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 
```

We extract the "importance values" for these genes, metabolites and CyTOF features 
in an excel file. Here's a sample of the lowest features in the PC1 (under the
-0.6 threshold) :
```{r, echo = FALSE}
most_imp <- over %>% 
  rownames_to_column("feature") %>% 
  mutate(feature = gsub("^[a-z]+_", "", feature)) %>% 
  arrange(overexp)
most_imp

most_imp2 <- most_imp %>% 
  arrange(orig)

write.xlsx(most_imp, file = "../data/integ/RD/primtol_ft_PC1.xlsx")
write.xlsx(most_imp2, file = "../data/integ/RD/primtol_ft_PC1_ordered_by_origin.xlsx")
```

We then focus on the features that are overexpressed in the secondary tolerant recipients 
(= the features that are most expressed on the RIGHT of the PC1). Therefore, the 
higher the feature value, the more it is expressed in secondary tolerant recipients.
```{r, echo = FALSE}
#plot(sort(pca$rotation[,1][which(pca$rotation[,1]>0)]))
overexp <- sort(pca$rotation[,1][which(pca$rotation[,1]>0)])
over <- as.data.frame(overexp)
orig <- strsplit(rownames(over), "_")
orig <- map(orig, function(x){x[1]}) %>% unlist
over$orig <- orig

type <- rep("x", nrow(over))
type[grepl("_R_", rownames(over))] <- "R"
type[grepl("_D_", rownames(over))] <- "D"
over$type <- type

ggplot(over, aes(x = orig, y = overexp, col = orig)) +
  scale_color_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  geom_boxplot() +
  geom_jitter()
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 

ggplot(over, aes(x = orig, y = overexp)) +
  geom_boxplot() +
  geom_jitter(aes(x = orig, y = overexp, col = type), data = over) 
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 
```

We extract the "importance values" for these genes, metabolites and CyTOF features 
in an excel file. Here's a sample of the highest features in the PC1 (over the
0.5 threshold) :
```{r, echo = FALSE}
most_imp <- over %>% 
  rownames_to_column("feature") %>% 
  mutate(feature = gsub("^[a-z]+_", "", feature)) %>% 
  arrange(desc(overexp))
most_imp

most_imp2 <- most_imp %>% 
  arrange(orig)

write.xlsx(most_imp, file = "../data/integ/RD/sectol_ft_PC1.xlsx")
write.xlsx(most_imp2, file = "../data/integ/RD/sectol_ft_PC1_ordered_by_origin.xlsx")
```

##### Features expressed in the second principal component

We first focus on the features that are overexpressed in the primary tolerant recipients 
(= the features that are most expressed at the bottom of the PC2). Therefore, the 
lower the feature value, the more it is expressed in primary tolerant recipients.
```{r, echo = FALSE}
#plot(sort(pca$rotation[,1][which(pca$rotation[,1]>0)]))
overexp <- sort(pca$rotation[,2][which(pca$rotation[,2]<0)])
over <- as.data.frame(overexp)
orig <- strsplit(rownames(over), "_")
orig <- map(orig, function(x){x[1]}) %>% unlist
over$orig <- orig

type <- rep("x", nrow(over))
type[grepl("_R_", rownames(over))] <- "R"
type[grepl("_D_", rownames(over))] <- "D"
over$type <- type

ggplot(over, aes(x = orig, y = overexp, col = orig)) +
  scale_color_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  geom_boxplot() +
  geom_jitter()
# group_by feature origin (in pca_var), do box plots of importance per feature orig
# also look at these features, per feature orig/ or per imprtance in PC1? 

ggplot(over, aes(x = orig, y = overexp)) +
  geom_boxplot() +
  geom_jitter(aes(x = orig, y = overexp, col = type), data = over) 
```

We extract the "importance values" for these genes, metabolites and CyTOF features 
in an excel file. Here's a sample of the lowest features in the PC1 (under the
-0.6 threshold) :
```{r, echo = FALSE}
most_imp <- over %>% 
  rownames_to_column("feature") %>% 
  mutate(feature = gsub("^[a-z]+_", "", feature)) %>% 
  arrange(overexp)
most_imp

most_imp2 <- most_imp %>% 
  arrange(orig)

write.xlsx(most_imp, file = "../data/integ/RD/primtol_ft_PC2.xlsx")
write.xlsx(most_imp2, file = "../data/integ/RD/primtol_ft_PC2_ordered_by_origin.xlsx")
```

We then focus on the features that are overexpressed in the secondary tolerant recipients 
(= the features that are most expressed on the top of the PC2). Therefore, the 
higher the feature value, the more it is expressed in secondary tolerant recipients.
```{r, echo = FALSE}
overexp <- sort(pca$rotation[,2][which(pca$rotation[,2]>0)])
over <- as.data.frame(overexp)
orig <- strsplit(rownames(over), "_")
orig <- map(orig, function(x){x[1]}) %>% unlist
over$orig <- orig

type <- rep("x", nrow(over))
type[grepl("_R_", rownames(over))] <- "R"
type[grepl("_D_", rownames(over))] <- "D"
over$type <- type

ggplot(over, aes(x = orig, y = overexp, col = orig)) +
  scale_color_manual(values = c("palegreen3", "mediumorchid4", "sandybrown")) +
  geom_boxplot() +
  geom_jitter()

ggplot(over, aes(x = orig, y = overexp)) +
  geom_boxplot() +
  geom_jitter(aes(x = orig, y = overexp, col = type), data = over) 
```

We extract the "importance values" for these genes, metabolites and CyTOF features 
in an excel file. Here's a sample of the highest features in the PC1 (over the
0.5 threshold) :
```{r, echo = FALSE}
most_imp <- over %>% 
  rownames_to_column("feature") %>% 
  mutate(feature = gsub("^[a-z]+_", "", feature)) %>% 
  arrange(desc(overexp))
most_imp

most_imp2 <- most_imp %>% 
  arrange(orig)

write.xlsx(most_imp, file = "../data/integ/RD/sec_tol_ft_PC2.xlsx")
write.xlsx(most_imp2, file = "../data/integ/RD/sec_tol_ft_PC2_ordered_by_origin.xlsx")
```

<!-- We can try to build a model that would predict a patients output (tolerant or not) -->
<!-- based on the PCs: -->
<!-- ```{r, echo = FALSE} -->
<!-- data2model <- as.data.frame(pca$x) %>% mutate(group = pca_2plot$group) -->

<!-- set.seed(1)  -->

<!-- # Define training control -->
<!-- train.control <- caret::trainControl(method = "cv", number = 5) -->
<!-- # Train the model -->

<!-- model <- caret::train(group ~., data = data2model, method = "svmLinearWeights2", -->
<!--                trControl = train.control) -->
<!-- # Summarize the results -->
<!-- print(model) -->
<!-- ``` -->

<!-- The results haven't been improved by the addition of the couple's features. -->

<!-- #### Random forest modeling -->

<!-- To see if the features that we selected can help us to classify the patients, we -->
<!-- apply a 5 fold cross validation setting on the 10 patients for which we have  -->
<!-- CyTOF, metabolomics and transcriptomics information. -->

<!-- ```{r} -->
<!-- set.seed(1) -->
<!-- cv_id <- list(c(1:5), c(6:10), c(11:15), c(16:20), c(21:23)) -->
<!-- cv_accuracies <- lapply(seq_along(cv_id), function(idx){ -->
<!--   train_set <- pca_2plot[-cv_id[[idx]], c("group", colnames(ft_integ)[-1])] -->
<!--   test_set <- pca_2plot[cv_id[[idx]], c(colnames(ft_integ)[-1])] -->
<!--   true_labels <- pca_2plot[cv_id[[idx]], "group"] -->
<!--   rf <- ranger::ranger(group~., train_set, num.trees = 5000, -->
<!--                            importance = "impurity") -->
<!--   pred <- predict(rf, test_set) -->
<!--   acc <- ( length(which(pred$predictions == true_labels))/length(true_labels) ) -->
<!-- }) -->

<!-- accuracy <- mean(unlist(cv_accuracies)) -->
<!-- accuracy -->
<!-- ``` -->

#### Correlations between features

We can now see how the variables that we extracted from recipients and donors
are coorelated. Correlations between variables are represented by edges in 
the graph. We kept only correlations that were common between the St Louis 
and the Cryostem cohort.

```{r, echo = FALSE}
pval_threshold <- 0.01
compar <- "PS"

# St Louis
gr_medians_l <- pca_2plot %>%
  select(c(Id.Cryostem.R, group, colnames(ft_integ))) %>% 
  column_to_rownames("Id.Cryostem.R") %>% 
  group_by(group) %>% 
  summarize_if(is.numeric, median) %>% 
  as.data.frame()
rownames(gr_medians_l) <- gr_medians_l$group

tmp_l <- ft_integ %>% 
  column_to_rownames("Id.Cryostem.R") %>% 
  apply(2, scale)

# correlation St Louis
ft_df <- tmp_l %>% 
  as.data.frame() %>% 
  mutate(group = pca_2plot$group) %>% 
  magrittr::set_rownames(ft_integ$Id.Cryostem.R)

correlations <- ft_df %>%
  dplyr::select(-group) %>%
  as.matrix %>%
  cor %>%
  as.data.frame %>%
  rownames_to_column(var = 'var1') %>%
  gather(var2, value, -var1) %>%
  dplyr::filter(var1 < var2)
  
# cor.tests to define which correlations are significant
cor_sel <- correlations[which(correlations[,3] > 0),]
pvals <- rep(1, nrow(cor_sel))
for(i in 1: nrow(cor_sel)){
  var1 <- ft_df[,cor_sel[i,1]]
  var2 <- ft_df[,cor_sel[i,2]]
  pvals[i] <- cor.test(var1, var2, method = "spearman", exact = FALSE)$p.value
}
  
cor_sel <- cor_sel[which(pvals < pval_threshold),]

  
  
# Cryostem
cryo_for_cor <- readRDS("../data/integ_national/data_for_correlation_R_D_PTST.RDS")
gr_c <- tolower(as.character(cryo_for_cor$mat_withinfo$GROUP))

gr_medians_c <- cryo_for_cor$ft_integ %>%
  column_to_rownames("Id.Cryostem.R") %>% 
  mutate(group = as.factor(gr_c)) %>% 
  group_by(group) %>% 
  summarize_if(is.numeric, median) %>% 
  as.data.frame()
rownames(gr_medians_c) <- gr_medians_c$group

tmp_c <- cryo_for_cor$ft_integ %>% 
  column_to_rownames("Id.Cryostem.R") %>% 
  apply(2, scale)

# Correlations Cryostem
other <- tmp_c %>% 
  as.data.frame() %>% 
  mutate(group = gr_c) %>% 
  magrittr::set_rownames(cryo_for_cor$ft_integ$Id.Cryostem.R)

correlations_other <- other %>%
  dplyr::select(-group) %>%
  as.matrix %>%
  cor %>%
  as.data.frame %>%
  rownames_to_column(var = 'var1') %>%
  gather(var2, value, -var1) %>%
  dplyr::filter(var1 < var2)
  
# cor.tests to define which correlations are significant
cor_sel_other <- correlations_other[which(correlations_other[,3] > 0),]
pvals_other <- rep(1, nrow(cor_sel_other))
for(i in 1: nrow(cor_sel_other)){
  var1 <- other[,cor_sel_other[i,1]]
  var2 <- other[,cor_sel_other[i,2]]
  pvals_other[i] <- cor.test(var1, var2, method = "spearman", exact = FALSE)$p.value
}
  
cor_sel_other <- cor_sel_other[which(pvals_other < pval_threshold),]



  
# Both cohorts
correlated_both <- dplyr::intersect(cor_sel[,1:2], cor_sel_other[,1:2])
cor <- rep(0, nrow(correlated_both))
for(i in 1:nrow(correlated_both)){
  cor1 <- cor_sel[which((cor_sel[,1] == correlated_both[i,1]) &
                          (cor_sel[,2] == correlated_both[i,2])), 3]
  cor2 <- cor_sel_other[which((cor_sel_other[,1] == correlated_both[i,1]) &
                                (cor_sel_other[,2] == correlated_both[i,2])), 3]
  cor[i] <- mean(cor1, cor2)
}
correlated_both$value <- cor

gr <- graph_from_data_frame(correlated_both, directed = FALSE,
                            vertices=colnames(ft_df)[-which(colnames(ft_df)=="group")])
nodes <- get.vertex.attribute(gr)
  
if(compar == "TNT"){
  color_nodes <- rep("#FF000080", length(nodes$name))
  
  gr_medians <- ft_df %>%
    group_by(group) %>%
    summarize_if(is.numeric, median) %>%
    as.data.frame()
  rownames(gr_medians) <- gr_medians$group
  
  medians <- gr_medians[,nodes$name]
  for (gene in seq_along(colnames(medians))){
    if (medians["tolerant", gene] == max(medians[, gene])){
      color_nodes[gene] <- "#0000FF80"
    }
  }
} else if(compar == "PS"){
  color_nodes <- rep("#0000FF80", length(nodes$name))
  
  gr_medians <- ft_df %>%
    group_by(group) %>%
    summarize_if(is.numeric, median) %>%
    as.data.frame()
  rownames(gr_medians) <- gr_medians$group
  
  medians <- gr_medians[,nodes$name]
  for (gene in seq_along(colnames(medians))){
    if (medians["primary_tolerant", gene] == max(medians[, gene])){
      color_nodes[gene] <- "#00FF0080"
    }
  }
}

color_frame <- rep("palegreen3", length(nodes$name))
color_frame[grep("met_",nodes$name)] <- "mediumorchid4"
color_frame[grep("rna_",nodes$name)] <- "sandybrown"

set.seed(42)
plot(gr, vertex.size = 7, vertex.label.cex = .5, edge.width = (E(gr)$value - .7) * 20,
     vertex.color = color_frame, vertex.frame.color= color_nodes)
legend(1, 1, legend=c(round(min(E(gr)$value), digits = 3), round(max(E(gr)$value), digits = 3)),
         col=c("grey", "grey"),
         lwd=c((min(E(gr)$value) - round(min(E(gr)$value), digits = 2)) * 10,
               (max(E(gr)$value) - round(min(E(gr)$value), digits = 2)) * 10),
         cex=0.8,
         title="Edge weight", text.font=4, bg='white')
```

```{r, echo = FALSE}
gr_components <- decompose(gr, mode = "strong", max.comps = NA,
                 min.vertices = 0)
comp_sizes <- sapply(gr_components, diameter)

list_components <- lapply(which(comp_sizes != 0), function(i){
  vertex_names <- vertex_attr(gr_components[[i]], index = V(gr_components[[i]]))[[1]]
  vertex_names
})

write.xlsx(list_components, file = "../data/integ_2_cohorts/graph_components_PTST_recipients_and_donors.xlsx")
write.xlsx(correlated_both, file = "../data/integ_2_cohorts/table_correlations_2cohorts_recipients_and_donors_PTST.xlsx")
```

#### Separate data type analyses

For comparison, we can see if we would have managed to separate the tolerant 
from non tolerant recipients and couples as well without integrating the data:

##### CyTOF:

We perform PCA using only the 49 features that were selected as informative 
in the recipients and couples in both cohorts:

```{r, echo = FALSE}
cyt4pca <- ft_rdr_cyto %>% 
  dplyr::filter(Id.Cryostem.R %in% ft_integ$Id.Cryostem.R) %>% 
  column_to_rownames("Id.Cryostem.R")
pca_cyto <- ade4::dudi.pca(cyt4pca, nf = 10, scannf = F)
```

The first comoponent seems to be capturing most of the data variability:
```{r, echo = FALSE}
plot(pca_cyto$eig, pch = 19)
```

```{r, echo = FALSE}
info_tmp <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::filter(Id.Cryostem.R %in% rownames(cyt4pca)) %>%
  mutate(group = as.character(tolower(GROUP))) %>% 
  column_to_rownames("Id.Cryostem.R")
info_tmp <- info_tmp[rownames(cyt4pca),]
info_tmp$group <- as.factor(info_tmp$group)
```

We can see how these principal components are correlated with tolerance. 
The first PC is very correlated with the tolerance. The PC4 is the 2nd mostly
correlated with tolerance.
```{r, echo = FALSE}
barplot(abs(cor(pca_cyto$li[,1:5], as.numeric(info_tmp$group))[,1]))
```

The primary and secondary patients are well separatd in this PCA. 
```{r, echo = FALSE}
#information on the patients: 
info_pca <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R")
info_pca <- info_pca[rownames(cyt4pca),]
ade4::s.class(pca_cyto$li[,c(1,4)], info_pca$group, col=c("green", "blue"))
# ade4::s.class(pca_cyto$li[,c(1,6)], as.factor(info_pca$cGVHD), 
#               col = c("black", "red", "lightgreen", "green3"))
```

We can also have a look at the features that are driving these two components:
```{r, echo = FALSE}
keep1 <- order(abs(pca_cyto$c1$CS1), decreasing = T)[1:5]
# keep2 <- order(abs(pca_cyto$c1$CS2), decreasing = T)[1:5]
# tokeep <- unique(c(keep1, keep2))
tokeep <- keep1
ade4::s.arrow(pca_cyto$c1[tokeep,c(1,4)], boxes = F, clabel = 0.5)
```

<!-- We can try to build a model that would predict a patients output (tolerant or not) -->
<!-- based on the PCs: -->
<!-- ```{r, echo = FALSE} -->
<!-- data2model <- as.data.frame(pca_cyto$li) %>% mutate(group = info_pca$group) -->

<!-- set.seed(1)  -->

<!-- # Define training control -->
<!-- train.control <- caret::trainControl(method = "cv", number = 5) -->
<!-- # Train the model -->

<!-- model <- caret::train(group ~., data = data2model, method = "svmLinearWeights2", -->
<!--                trControl = train.control) -->
<!-- # Summarize the results -->
<!-- print(model) -->
<!-- ``` -->

<!-- The accuracy is quite lower than in the integrative model and hasn't improved  -->
<!-- with the addition of features from the donors. -->

<!-- To see if the features that we selected can help us to clssify the patients, we -->
<!-- can also apply a 5 fold cross validation setting on the CyTOF features of the  -->
<!-- same 23 patients. -->

<!-- ```{r} -->
<!-- set.seed(1) -->
<!-- info_rf <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R") -->
<!-- info_rf <- info_rf[rownames(cyt4pca),] -->
<!-- cyt4rf <- cbind(info_rf$group, cyt4pca) -->
<!-- colnames(cyt4rf)[1] <- "group" -->

<!-- cv_id <- list(c(1:5), c(6:10), c(11:15), c(16:20), c(21:23)) -->
<!-- cv_accuracies <- lapply(seq_along(cv_id), function(idx){ -->
<!--   train_set <- cyt4rf[-cv_id[[idx]], ] -->
<!--   test_set <- cyt4rf[cv_id[[idx]], -1] -->
<!--   true_labels <- cyt4rf[cv_id[[idx]], "group"] -->
<!--   rf <- ranger::ranger(group~., train_set, num.trees = 5000, -->
<!--                            importance = "impurity") -->
<!--   pred <- predict(rf, test_set) -->
<!--   acc <- ( length(which(pred$predictions == true_labels))/length(true_labels) ) -->
<!-- }) -->

<!-- accuracy <- mean(unlist(cv_accuracies)) -->
<!-- accuracy -->
<!-- ``` -->

##### Metabolomics:
```{r, echo = FALSE}
met4pca <- ft_rdr_metabo %>% 
  dplyr::filter(Id.Cryostem.R %in% ft_integ$Id.Cryostem.R) %>% 
  column_to_rownames("Id.Cryostem.R")
pca_metabo <- ade4::dudi.pca(met4pca, nf = 10, scannf = F)
```

Variability captured by each component:
```{r, echo = FALSE}
plot(pca_metabo$eig, pch = 19)
```

```{r, echo = FALSE}
info_tmp <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::filter(Id.Cryostem.R %in% rownames(met4pca)) %>%
  mutate(group = as.character(tolower(GROUP))) %>% 
  column_to_rownames("Id.Cryostem.R")
info_tmp <- info_tmp[rownames(met4pca),]
info_tmp$group <- as.factor(info_tmp$group)
```

We can see how these principal components are correlated with tolerance. 
The first PC is very correlated with the tolerance. The PC2 is the 2nd mostly
correlated with tolerance.
```{r, echo = FALSE}
barplot(abs(cor(pca_metabo$li[,1:3], as.numeric(info_tmp$group))[,1]))
```

The primary and secondary tolerant patients are less well separated in this PCA. 
```{r, echo = FALSE}
#information on the patients: 
info_pca <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R")
info_pca <- info_pca[rownames(met4pca),]
ade4::s.class(pca_metabo$li[,c(1,2)], info_pca$group, col=c("green", "blue"))
# ade4::s.class(pca_cyto$li[,c(1,6)], as.factor(info_pca$cGVHD), 
#               col = c("black", "red", "lightgreen", "green3"))
```

```{r, echo = FALSE}
keep1 <- order(abs(pca_metabo$c1$CS1), decreasing = T)[1:3]
# keep2 <- order(abs(pca_cyto$c1$CS2), decreasing = T)[1:5]
# tokeep <- unique(c(keep1, keep2))
tokeep <- keep1
ade4::s.arrow(pca_metabo$c1[tokeep, c(1,2)], boxes = F, clabel = 0.5)
```

<!-- We can try to build a model that would predict a patients output (tolerant or not) -->
<!-- based on the PCs: -->
<!-- ```{r, echo = FALSE} -->
<!-- data2model <- as.data.frame(pca_metabo$li) %>% mutate(group = info_pca$group) -->

<!-- set.seed(1)  -->

<!-- # Define training control -->
<!-- train.control <- caret::trainControl(method = "cv", number = 5) -->
<!-- # Train the model -->

<!-- model <- caret::train(group ~., data = data2model, method = "svmLinearWeights2", -->
<!--                trControl = train.control) -->
<!-- # Summarize the results -->
<!-- print(model) -->
<!-- ``` -->

<!-- Again, the accuracy of the model is slightly lower than the accuracy of the integrative model. -->

<!-- To see if the features that we selected can help us to clssify the patients, we -->
<!-- can also apply a 5 fold cross validation setting on the CyTOF features of the  -->
<!-- same 23 patients. -->

<!-- ```{r} -->
<!-- set.seed(1) -->
<!-- info_rf <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R") -->
<!-- info_rf <- info_rf[rownames(met4pca),] -->
<!-- met4rf <- cbind(info_rf$group, met4pca) -->
<!-- colnames(met4rf)[1] <- "group" -->

<!-- cv_id <- list(c(1:5), c(6:10), c(11:15), c(16:20), c(21:23)) -->
<!-- cv_accuracies <- lapply(seq_along(cv_id), function(idx){ -->
<!--   train_set <- met4rf[-cv_id[[idx]], ] -->
<!--   test_set <- met4rf[cv_id[[idx]], -1] -->
<!--   true_labels <- met4rf[cv_id[[idx]], "group"] -->
<!--   rf <- ranger::ranger(group~., train_set, num.trees = 5000, -->
<!--                            importance = "impurity") -->
<!--   pred <- predict(rf, test_set) -->
<!--   acc <- ( length(which(pred$predictions == true_labels))/length(true_labels) ) -->
<!-- }) -->

<!-- accuracy <- mean(unlist(cv_accuracies)) -->
<!-- accuracy -->
<!-- ``` -->

##### Transcriptomics:
```{r, echo = FALSE}
rna4pca <- ft_rdr_transcripto %>%
  dplyr::filter(Id.Cryostem.R %in% ft_integ$Id.Cryostem.R) %>% 
  column_to_rownames("Id.Cryostem.R")
pca_transcripto <- ade4::dudi.pca(rna4pca, nf = 10, scannf = F)
```

Variability captured by each component:
```{r, echo = FALSE}
plot(pca_transcripto$eig, pch = 19)
```

```{r, echo = FALSE}
info_tmp <- couple_info %>% 
  rownames_to_column("Id.Cryostem.R") %>% 
  dplyr::filter(Id.Cryostem.R %in% rownames(rna4pca)) %>%
  mutate(group = as.character(tolower(GROUP))) %>% 
  column_to_rownames("Id.Cryostem.R")
info_tmp <- info_tmp[rownames(rna4pca),]
info_tmp$group <- as.factor(info_tmp$group)
```

We can see how these principal components are correlated with tolerance. 
The first PC is very correlated with the tolerance. The PC5 is the 2nd mostly
correlated with tolerance.
```{r, echo = FALSE}
barplot(abs(cor(pca_transcripto$li[,1:9], as.numeric(info_tmp$group))[,1]))
```

The primary tolerant patients are perfectly separated from the 
secondary tolerant patients in this PCA. 
```{r, echo = FALSE}
#information on the patients: 
info_pca <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R")
info_pca <- info_pca[rownames(rna4pca),]
ade4::s.class(pca_transcripto$li[,c(1,5)], info_pca$group, col=c("green", "blue"))
# ade4::s.class(pca_cyto$li[,c(1,6)], as.factor(info_pca$cGVHD), 
#               col = c("black", "red", "lightgreen", "green3"))
```

```{r, echo = FALSE}
keep1 <- order(abs(pca_transcripto$c1$CS1), decreasing = T)[1:9]
# keep2 <- order(abs(pca_cyto$c1$CS2), decreasing = T)[1:5]
# tokeep <- unique(c(keep1, keep2))
tokeep <- keep1
ade4::s.arrow(pca_transcripto$c1[tokeep,c(1,5)], boxes = F, clabel = 0.5)
```

<!-- We can try to build a model that would predict a patients output (tolerant or not) -->
<!-- based on the PCs: -->
<!-- ```{r, echo = FALSE} -->
<!-- data2model <- as.data.frame(pca_transcripto$li) %>% mutate(group = info_pca$group) -->

<!-- set.seed(1)  -->

<!-- # Define training control -->
<!-- train.control <- caret::trainControl(method = "cv", number = 5) -->
<!-- # Train the model -->

<!-- model <- caret::train(group ~., data = data2model, method = "svmLinearWeights2", -->
<!--                trControl = train.control) -->
<!-- # Summarize the results -->
<!-- print(model) -->
<!-- ``` -->

<!-- It looks like, in our integrative approach, we managed to separate the recipients  -->
<!-- based on their tolerance better than when we look at the different data types -->
<!-- separately. -->

<!-- To see if the features that we selected can help us to clssify the patients, we -->
<!-- can also apply a 5 fold cross validation setting on the CyTOF features of the  -->
<!-- same 23 patients. -->

<!-- ```{r} -->
<!-- set.seed(1) -->
<!-- info_rf <- info_complete_patients %>% column_to_rownames("Id.Cryostem.R") -->
<!-- info_rf <- info_rf[rownames(rna4pca),] -->
<!-- rna4rf <- cbind(info_rf$group, rna4pca) -->
<!-- colnames(rna4rf)[1] <- "group" -->

<!-- cv_id <- list(c(1:5), c(6:10), c(11:15), c(16:20), c(21:23)) -->
<!-- cv_accuracies <- lapply(seq_along(cv_id), function(idx){ -->
<!--   train_set <- rna4rf[-cv_id[[idx]], ] -->
<!--   test_set <- rna4rf[cv_id[[idx]], -1] -->
<!--   true_labels <- rna4rf[cv_id[[idx]], "group"] -->
<!--   rf <- ranger::ranger(group~., train_set, num.trees = 5000, -->
<!--                            importance = "impurity") -->
<!--   pred <- predict(rf, test_set) -->
<!--   acc <- ( length(which(pred$predictions == true_labels))/length(true_labels) ) -->
<!-- }) -->

<!-- accuracy <- mean(unlist(cv_accuracies)) -->
<!-- accuracy -->
<!-- ``` -->
